

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API Documentation &mdash; GenomeKit 6.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=b6ed0530"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Genomes" href="genomes.html" />
    <link rel="prev" title="Anchors" href="anchors.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            GenomeKit
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="anchors.html">Anchors</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#interval">Interval</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome_kit.Interval"><code class="docutils literal notranslate"><span class="pre">Interval</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.__init__"><code class="docutils literal notranslate"><span class="pre">Interval.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.chromosome"><code class="docutils literal notranslate"><span class="pre">Interval.chromosome</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.strand"><code class="docutils literal notranslate"><span class="pre">Interval.strand</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.start"><code class="docutils literal notranslate"><span class="pre">Interval.start</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.end"><code class="docutils literal notranslate"><span class="pre">Interval.end</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.end5"><code class="docutils literal notranslate"><span class="pre">Interval.end5</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.end3"><code class="docutils literal notranslate"><span class="pre">Interval.end3</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.reference_genome"><code class="docutils literal notranslate"><span class="pre">Interval.reference_genome</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.anchor"><code class="docutils literal notranslate"><span class="pre">Interval.anchor</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.anchor_offset"><code class="docutils literal notranslate"><span class="pre">Interval.anchor_offset</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.shift"><code class="docutils literal notranslate"><span class="pre">Interval.shift()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.expand"><code class="docutils literal notranslate"><span class="pre">Interval.expand()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.intersect"><code class="docutils literal notranslate"><span class="pre">Interval.intersect()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.subtract"><code class="docutils literal notranslate"><span class="pre">Interval.subtract()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.is_positive_strand"><code class="docutils literal notranslate"><span class="pre">Interval.is_positive_strand()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.as_positive_strand"><code class="docutils literal notranslate"><span class="pre">Interval.as_positive_strand()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.as_negative_strand"><code class="docutils literal notranslate"><span class="pre">Interval.as_negative_strand()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.as_opposite_strand"><code class="docutils literal notranslate"><span class="pre">Interval.as_opposite_strand()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.upstream_of"><code class="docutils literal notranslate"><span class="pre">Interval.upstream_of()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.dnstream_of"><code class="docutils literal notranslate"><span class="pre">Interval.dnstream_of()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.contains"><code class="docutils literal notranslate"><span class="pre">Interval.contains()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.within"><code class="docutils literal notranslate"><span class="pre">Interval.within()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.overlaps"><code class="docutils literal notranslate"><span class="pre">Interval.overlaps()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.from_dna0_coord"><code class="docutils literal notranslate"><span class="pre">Interval.from_dna0_coord()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.from_dna0"><code class="docutils literal notranslate"><span class="pre">Interval.from_dna0()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.from_dna1"><code class="docutils literal notranslate"><span class="pre">Interval.from_dna1()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.from_rna1"><code class="docutils literal notranslate"><span class="pre">Interval.from_rna1()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.spanning"><code class="docutils literal notranslate"><span class="pre">Interval.spanning()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.midpoint"><code class="docutils literal notranslate"><span class="pre">Interval.midpoint</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.distance"><code class="docutils literal notranslate"><span class="pre">Interval.distance()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.as_dna0"><code class="docutils literal notranslate"><span class="pre">Interval.as_dna0()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.as_dna1"><code class="docutils literal notranslate"><span class="pre">Interval.as_dna1()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.as_rna1"><code class="docutils literal notranslate"><span class="pre">Interval.as_rna1()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.as_ucsc"><code class="docutils literal notranslate"><span class="pre">Interval.as_ucsc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.with_anchor"><code class="docutils literal notranslate"><span class="pre">Interval.with_anchor()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.__repr__"><code class="docutils literal notranslate"><span class="pre">Interval.__repr__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Interval.__str__"><code class="docutils literal notranslate"><span class="pre">Interval.__str__()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#variant">Variant</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome_kit.Variant"><code class="docutils literal notranslate"><span class="pre">Variant</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.__init__"><code class="docutils literal notranslate"><span class="pre">Variant.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.interval"><code class="docutils literal notranslate"><span class="pre">Variant.interval</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.position"><code class="docutils literal notranslate"><span class="pre">Variant.position</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.ref"><code class="docutils literal notranslate"><span class="pre">Variant.ref</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.alt"><code class="docutils literal notranslate"><span class="pre">Variant.alt</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.as_variant_string"><code class="docutils literal notranslate"><span class="pre">Variant.as_variant_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.__repr__"><code class="docutils literal notranslate"><span class="pre">Variant.__repr__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.__str__"><code class="docutils literal notranslate"><span class="pre">Variant.__str__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.from_string"><code class="docutils literal notranslate"><span class="pre">Variant.from_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Variant.spanning"><code class="docutils literal notranslate"><span class="pre">Variant.spanning()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#genome">Genome</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome_kit.Genome"><code class="docutils literal notranslate"><span class="pre">Genome</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.__init__"><code class="docutils literal notranslate"><span class="pre">Genome.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.__reduce__"><code class="docutils literal notranslate"><span class="pre">Genome.__reduce__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.__new__"><code class="docutils literal notranslate"><span class="pre">Genome.__new__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.interval"><code class="docutils literal notranslate"><span class="pre">Genome.interval()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.variant"><code class="docutils literal notranslate"><span class="pre">Genome.variant()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.dna"><code class="docutils literal notranslate"><span class="pre">Genome.dna()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.variant_dna"><code class="docutils literal notranslate"><span class="pre">Genome.variant_dna()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.annotation"><code class="docutils literal notranslate"><span class="pre">Genome.annotation</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.genes"><code class="docutils literal notranslate"><span class="pre">Genome.genes</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.transcripts"><code class="docutils literal notranslate"><span class="pre">Genome.transcripts</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.exons"><code class="docutils literal notranslate"><span class="pre">Genome.exons</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.introns"><code class="docutils literal notranslate"><span class="pre">Genome.introns</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.cdss"><code class="docutils literal notranslate"><span class="pre">Genome.cdss</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.utr5s"><code class="docutils literal notranslate"><span class="pre">Genome.utr5s</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.utr3s"><code class="docutils literal notranslate"><span class="pre">Genome.utr3s</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.find_motif"><code class="docutils literal notranslate"><span class="pre">Genome.find_motif()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.reference_genome"><code class="docutils literal notranslate"><span class="pre">Genome.reference_genome</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.chromosomes"><code class="docutils literal notranslate"><span class="pre">Genome.chromosomes</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.chromosome_size"><code class="docutils literal notranslate"><span class="pre">Genome.chromosome_size()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.config"><code class="docutils literal notranslate"><span class="pre">Genome.config</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.data_dir"><code class="docutils literal notranslate"><span class="pre">Genome.data_dir</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.__repr__"><code class="docutils literal notranslate"><span class="pre">Genome.__repr__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.appris_transcripts"><code class="docutils literal notranslate"><span class="pre">Genome.appris_transcripts()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.appris_principality"><code class="docutils literal notranslate"><span class="pre">Genome.appris_principality()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Genome.__weakref__"><code class="docutils literal notranslate"><span class="pre">Genome.__weakref__</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#variantgenome">VariantGenome</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome_kit.VariantGenome"><code class="docutils literal notranslate"><span class="pre">VariantGenome</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.VariantGenome.__init__"><code class="docutils literal notranslate"><span class="pre">VariantGenome.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.VariantGenome.dna"><code class="docutils literal notranslate"><span class="pre">VariantGenome.dna()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.VariantGenome.find_motif"><code class="docutils literal notranslate"><span class="pre">VariantGenome.find_motif()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.VariantGenome.__repr__"><code class="docutils literal notranslate"><span class="pre">VariantGenome.__repr__()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tracks">Tracks</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genomedna">GenomeDNA</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.GenomeDNA"><code class="docutils literal notranslate"><span class="pre">GenomeDNA</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#genometrack">GenomeTrack</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.GenomeTrack"><code class="docutils literal notranslate"><span class="pre">GenomeTrack</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#genometrackbuilder">GenomeTrackBuilder</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.GenomeTrackBuilder"><code class="docutils literal notranslate"><span class="pre">GenomeTrackBuilder</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#annotations">Annotations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genomeannotation">GenomeAnnotation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.GenomeAnnotation"><code class="docutils literal notranslate"><span class="pre">GenomeAnnotation</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#gene">Gene</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Gene"><code class="docutils literal notranslate"><span class="pre">Gene</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#genetable">GeneTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.GeneTable"><code class="docutils literal notranslate"><span class="pre">GeneTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#transcript">Transcript</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Transcript"><code class="docutils literal notranslate"><span class="pre">Transcript</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#transcripttable">TranscriptTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.TranscriptTable"><code class="docutils literal notranslate"><span class="pre">TranscriptTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#exon">Exon</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Exon"><code class="docutils literal notranslate"><span class="pre">Exon</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#exontable">ExonTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.ExonTable"><code class="docutils literal notranslate"><span class="pre">ExonTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#intron">Intron</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Intron"><code class="docutils literal notranslate"><span class="pre">Intron</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#introntable">IntronTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.IntronTable"><code class="docutils literal notranslate"><span class="pre">IntronTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#cds">Cds</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Cds"><code class="docutils literal notranslate"><span class="pre">Cds</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#cdstable">CdsTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.CdsTable"><code class="docutils literal notranslate"><span class="pre">CdsTable</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#read-alignments">Read Alignments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#readalignments">ReadAlignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.ReadAlignments"><code class="docutils literal notranslate"><span class="pre">ReadAlignments</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#junction">Junction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Junction"><code class="docutils literal notranslate"><span class="pre">Junction</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#junctiontable">JunctionTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.JunctionTable"><code class="docutils literal notranslate"><span class="pre">JunctionTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#alignment">Alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.Alignment"><code class="docutils literal notranslate"><span class="pre">Alignment</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#alignmenttable">AlignmentTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.AlignmentTable"><code class="docutils literal notranslate"><span class="pre">AlignmentTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#alignmentmatch">AlignmentMatch</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.AlignmentMatch"><code class="docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#alignmentmatchtable">AlignmentMatchTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.AlignmentMatchTable"><code class="docutils literal notranslate"><span class="pre">AlignmentMatchTable</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#junction-read-alignments">Junction Read Alignments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#jreadalignments">JReadAlignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.JReadAlignments"><code class="docutils literal notranslate"><span class="pre">JReadAlignments</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#junctionreadalignmentstable">JunctionReadAlignmentsTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.JunctionReadAlignmentsTable"><code class="docutils literal notranslate"><span class="pre">JunctionReadAlignmentsTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#junctionreadalignments">JunctionReadAlignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.JunctionReadAlignments"><code class="docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#junctionreadalignment">JunctionReadAlignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.JunctionReadAlignment"><code class="docutils literal notranslate"><span class="pre">JunctionReadAlignment</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#readdistributions">ReadDistributions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.ReadDistributions"><code class="docutils literal notranslate"><span class="pre">ReadDistributions</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#junctionreaddistributiontable">JunctionReadDistributionTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.JunctionReadDistributionTable"><code class="docutils literal notranslate"><span class="pre">JunctionReadDistributionTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#junctionreaddistribution">JunctionReadDistribution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.JunctionReadDistribution"><code class="docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#junctionreadcount">JunctionReadCount</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.JunctionReadCount"><code class="docutils literal notranslate"><span class="pre">JunctionReadCount</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#vcf-tables">VCF Tables</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#vcftable">VCFTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.VCFTable"><code class="docutils literal notranslate"><span class="pre">VCFTable</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#vcfvariant">VCFVariant</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.VCFVariant"><code class="docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#varianttable">VariantTable</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.VariantTable"><code class="docutils literal notranslate"><span class="pre">VariantTable</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#data-access">Data Access</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome-kit-datamanager">genome_kit.DataManager</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.DataManager"><code class="docutils literal notranslate"><span class="pre">DataManager</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#internals">Internals</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-genome_kit.gk_data">genome_kit.gk_data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.gk_data.get_file"><code class="docutils literal notranslate"><span class="pre">get_file()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.gk_data.upload_file"><code class="docutils literal notranslate"><span class="pre">upload_file()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.gk_data.wget"><code class="docutils literal notranslate"><span class="pre">wget()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#genome-kit-data-manager">genome_kit.data_manager</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit.data_manager.ProgressPercentage"><code class="docutils literal notranslate"><span class="pre">ProgressPercentage</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#genome-kit-twobitutils">genome_kit._twobitutils</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit._twobitutils.write2bit"><code class="docutils literal notranslate"><span class="pre">write2bit()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#genome-kit-apply-variants">genome_kit._apply_variants</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit._apply_variants.apply_variants"><code class="docutils literal notranslate"><span class="pre">apply_variants()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#genome-kit-cxx-util">genome_kit._cxx_util</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit._cxx_util.mock"><code class="docutils literal notranslate"><span class="pre">mock()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit._cxx_util.mock_result"><code class="docutils literal notranslate"><span class="pre">mock_result()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit._cxx_util.mock_unreachable"><code class="docutils literal notranslate"><span class="pre">mock_unreachable()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome_kit._cxx_util.strip_mock_bases"><code class="docutils literal notranslate"><span class="pre">strip_mock_bases()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="genomes.html">Genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="develop.html">Developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GenomeKit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API Documentation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="api-documentation">
<span id="api"></span><h1>API Documentation<a class="headerlink" href="#api-documentation" title="Link to this heading"></a></h1>
<section id="interval">
<h2>Interval<a class="headerlink" href="#interval" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Interval">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Interval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anchor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anchor_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval" title="Link to this definition"></a></dt>
<dd><p>A genomic interval.</p>
<p>An Interval represents a contiguous stranded genomic interval, identified
by a chromosome, strand, start position, end position, and reference genome.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anchor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anchor_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize an interval from a DNA0 position.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The chromosome name (‘chr1’, …).</p></li>
<li><p><strong>strand</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">chr</span></code>) – The strand (“+” or “-“).</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The start position of the interval (inclusive, 0-based).</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The end position of the interval (exclusive, 0-based).</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome in which the interval is defined,
for example “hg19”. If a genome object is given, its reference
genome will be used.</p></li>
<li><p><strong>anchor</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a> | None, optional) – The anchor is only relevant when extracting DNA from a
<cite>VariantGenome</cite> and specifies how length-changing variants are
applied. Conceptually, the anchor defines a fixed point when the
sequence changes length. See (TODO: link) for the details.
The anchor can be any position in the interval. If anchor is
<cite>None</cite>, then length-changing variants will change the length of the
interval. Anchor can be a string: <cite>5p</cite> or <cite>3p</cite> sets the anchor to
the respective end of the interval, respecting the sense strand.
<cite>start</cite> or <cite>end</cite> set the anchor to the start or end of the interval
on the forward strand. When anchor is <cite>center</cite> then the anchor is
set to the middle position of the interval (if the interval has odd
length, the position is rounded towards the 5p end of the interval,
respecting the sense strand). If anchor is an integer or an empty
interval, the anchor is set to that position.</p></li>
<li><p><strong>anchor_offset</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, optional) – The anchor offset is a shift that’s applied to the anchor only on
the variant genome. The purpose is to align a position on the
reference genome with a position <em>inside</em> an insertion. The anchor
itself would only allow to align with the beginning or end of the
insertion.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.chromosome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome</span></span><a class="headerlink" href="#genome_kit.Interval.chromosome" title="Link to this definition"></a></dt>
<dd><p>The chromosome name.</p>
<p>A shorthand property <cite>chrom</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The name of the chromosome this interval is defined on  (e.g. “chr1”).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.strand">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#genome_kit.Interval.strand" title="Link to this definition"></a></dt>
<dd><p>The strand (“+” or “-“).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The strand this interval is defined on.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">chr</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.start">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">start</span></span><a class="headerlink" href="#genome_kit.Interval.start" title="Link to this definition"></a></dt>
<dd><p>The integer position of the start (inclusive, 0-based).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The start position of this interval within the reference genome.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.end">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">end</span></span><a class="headerlink" href="#genome_kit.Interval.end" title="Link to this definition"></a></dt>
<dd><p>The integer position of the end (exclusive, 0-based).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The end position of the this interval within the reference genome.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.end5">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">end5</span></span><a class="headerlink" href="#genome_kit.Interval.end5" title="Link to this definition"></a></dt>
<dd><p>The 5p end of this interval, which is a length-0 interval.</p>
<p>Not to be confused with the <cite>end</cite> attribute, which</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The length-0 interval at the 5p end of the interval, i.e. the gap preceding
the most upstream base in the interval.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.end3">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">end3</span></span><a class="headerlink" href="#genome_kit.Interval.end3" title="Link to this definition"></a></dt>
<dd><p>The 3p end of this interval, which is a length-0 interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The length-0 interval at the 3p end of the interval, i.e. the gap succeeding
the most downstream base in the interval.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.reference_genome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_genome</span></span><a class="headerlink" href="#genome_kit.Interval.reference_genome" title="Link to this definition"></a></dt>
<dd><p>The reference genome (assembly name in UCSC format).</p>
<p>A shorthand property <cite>refg</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The name of the reference genome this interval is defined in.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.anchor">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">anchor</span></span><a class="headerlink" href="#genome_kit.Interval.anchor" title="Link to this definition"></a></dt>
<dd><p>The anchor position.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The anchor position associated with this interval, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.anchor_offset">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">anchor_offset</span></span><a class="headerlink" href="#genome_kit.Interval.anchor_offset" title="Link to this definition"></a></dt>
<dd><p>The anchor offset.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The anchor offset associated with this interval, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.shift">
<span class="sig-name descname"><span class="pre">shift</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amount</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.shift"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.shift" title="Link to this definition"></a></dt>
<dd><p>The interval shifted upstream/downstream.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>amount</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The amount by which to shift the start and end positions.
A positive amount shifts downstream according to the current strand,
and a negative amount shifts the position upstream.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The shifted interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.expand">
<span class="sig-name descname"><span class="pre">expand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">upstream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnstream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.expand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.expand" title="Link to this definition"></a></dt>
<dd><p>The interval expanded upstream/downstream.</p>
<p>Can also be called with a single argument, in which case
upstream and dnstream take on the same value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>upstream</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of upstream positions to add to the interval.</p></li>
<li><p><strong>dnstream</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of downstream positions to add to the interval.
If not specified, defaults to <cite>upstream</cite> (i.e. symmetric).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An interval containing <cite>upstream</cite> + <cite>downstream</cite> + 1 positions.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.intersect">
<span class="sig-name descname"><span class="pre">intersect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.intersect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.intersect" title="Link to this definition"></a></dt>
<dd><p>The interval intersecting both intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – interval to intersect with this interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The intersecting interval (possibly empty) or <cite>None</cite> if the
intervals are disjoint</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.subtract">
<span class="sig-name descname"><span class="pre">subtract</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><span class="pre">Interval</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><span class="pre">Interval</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.subtract"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.subtract" title="Link to this definition"></a></dt>
<dd><p>A list of intervals, representing this interval with the its
intersection of another interval removed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> – interval to subtract its intersection from this interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The intervals after subtraction. If the intervals were disjoint,
returns [self].</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Sequence[<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval">Interval</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.is_positive_strand">
<span class="sig-name descname"><span class="pre">is_positive_strand</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.is_positive_strand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.is_positive_strand" title="Link to this definition"></a></dt>
<dd><p>Test if this interval is on the positive strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>True if <cite>strand</cite> is equal to ‘+’</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.as_positive_strand">
<span class="sig-name descname"><span class="pre">as_positive_strand</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.as_positive_strand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.as_positive_strand" title="Link to this definition"></a></dt>
<dd><p>The same interval on the positive strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The same interval, but on the positive (‘+’) strand.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.as_negative_strand">
<span class="sig-name descname"><span class="pre">as_negative_strand</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.as_negative_strand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.as_negative_strand" title="Link to this definition"></a></dt>
<dd><p>The same interval on the negative strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The same interval, but on the negative (‘-’) strand.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.as_opposite_strand">
<span class="sig-name descname"><span class="pre">as_opposite_strand</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.as_opposite_strand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.as_opposite_strand" title="Link to this definition"></a></dt>
<dd><p>The same interval on the opposite strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The same interval, but on the opposite strand.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.upstream_of">
<span class="sig-name descname"><span class="pre">upstream_of</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.upstream_of"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.upstream_of" title="Link to this definition"></a></dt>
<dd><p>Test if this interval is strictly upstream of <cite>other</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Any coordinate or interval subclass defined on the same
reference genome and strand as this interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if this interval is strictly upstream of <cite>other</cite>, with no overlap.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.dnstream_of">
<span class="sig-name descname"><span class="pre">dnstream_of</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.dnstream_of"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.dnstream_of" title="Link to this definition"></a></dt>
<dd><p>Test if this interval is strictly downstream of <cite>other</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Any coordinate or interval subclass defined on the same
reference genome and strand as this interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if this interval is strictly downstream of <cite>other</cite>, with no overlap.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.contains">
<span class="sig-name descname"><span class="pre">contains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.contains"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.contains" title="Link to this definition"></a></dt>
<dd><p>Test if <cite>other</cite> lies within the extents of this interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Any coordinate or interval subclass defined on the same
reference genome and strand as this interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if <cite>other</cite> lies entirely within the extents of this interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.within">
<span class="sig-name descname"><span class="pre">within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.within" title="Link to this definition"></a></dt>
<dd><p>Test if this interval lies within the extents of <cite>other</cite>.</p>
<p>Use of <code class="docutils literal notranslate"><span class="pre">interval</span> <span class="pre">in</span> <span class="pre">other</span></code> expressions is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Any coordinate or interval subclass defined on the same
reference genome and strand as this interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if this interval lies within the extents of <cite>other</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.overlaps">
<span class="sig-name descname"><span class="pre">overlaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.overlaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.overlaps" title="Link to this definition"></a></dt>
<dd><p>Test if this interval overlaps the extents of <cite>other</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Any coordinate or interval subclass defined on the same
reference genome and strand as this interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if this interval overlaps the extents of <cite>other</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.from_dna0_coord">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_dna0_coord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.from_dna0_coord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.from_dna0_coord" title="Link to this definition"></a></dt>
<dd><p>Create an interval spanning a single DNA0 position.</p>
<p>Equivalent to
<code class="docutils literal notranslate"><span class="pre">Interval(chromosome,</span> <span class="pre">strand,</span> <span class="pre">position,</span> <span class="pre">position+1,</span> <span class="pre">reference_genome)</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The chromosome name (e.g. “chr1”).</p></li>
<li><p><strong>strand</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The strand (“+” or “-“).</p></li>
<li><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The position the interval should span (0-based).</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome in which the interval is defined,
for example “hg19”. If a genome object is given, its reference
genome will be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The resulting length-1 interval spanning <cite>position</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.from_dna0">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_dna0</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.from_dna0"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.from_dna0" title="Link to this definition"></a></dt>
<dd><p>Create an interval from DNA0 positions.</p>
<p>Equivalent to calling <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The chromosome name (e.g. “chr1”).</p></li>
<li><p><strong>strand</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The strand (“+” or “-“).</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The start position of the interval (inclusive, 0-based).</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The end position of the interval (exclusive, 0-based).</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome in which the interval is defined,
for example “hg19”. If a genome object is given, its reference
genome will be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The resulting interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.from_dna1">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_dna1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.from_dna1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.from_dna1" title="Link to this definition"></a></dt>
<dd><p>Create an interval from DNA1 positions.</p>
<p>Positions are automatically converted to DNA0 as (start-1, end).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The chromosome name (e.g. “chr1”).</p></li>
<li><p><strong>strand</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The strand (“+” or “-“).</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The start position of the interval (inclusive, 1-based).</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The end position of the interval (inclusive, 1-based).</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome in which the interval is defined,
for example “hg19”. If a genome object is given, its reference
genome will be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The resulting interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.from_rna1">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_rna1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.from_rna1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.from_rna1" title="Link to this definition"></a></dt>
<dd><p>Create an interval from RNA1 positions.</p>
<p>The start/end positions are automatically converted to DNA0 as
<cite>min(abs(start),abs(end))-1</cite> and <cite>max(abs(start), abs(end)))</cite> respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The chromosome name (e.g. “chr1”).</p></li>
<li><p><strong>strand</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The strand (“+” or “-“).</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The start position of the interval (signed, inclusive, 1-based).</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The end position of the interval (signed, inclusive, 1-based).</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome in which the interval is defined,
for example “hg19”. If a genome object is given, its reference
genome will be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The resulting interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.spanning">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">spanning</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.spanning"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.spanning" title="Link to this definition"></a></dt>
<dd><p>Create the smallest interval spanning the given pair of existing intervals,
or raise an exception if given intervals not on the same chromosome, strand,
and reference genome. The given intervals are not required to be overlapping,
and may both be of zero length.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">250</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Interval</span><span class="o">.</span><span class="n">spanning</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
<span class="go">Interval(&quot;chr1&quot;, &quot;+&quot;, 100, 250, &quot;hg19&quot;)</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>This method is not anchor aware, since that requires a list of variants for context.
See <a class="reference internal" href="#genome_kit.Variant.spanning" title="genome_kit.Variant.spanning"><code class="xref py py-func docutils literal notranslate"><span class="pre">spanning()</span></code></a> for that use case.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval1</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – an interval</p></li>
<li><p><strong>interval2</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – another interval</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The resulting spanning interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Interval.midpoint">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">midpoint</span></span><a class="headerlink" href="#genome_kit.Interval.midpoint" title="Link to this definition"></a></dt>
<dd><p>Identify the midpoint of an interval.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">midpoint</span>
<span class="go">Interval(&quot;chr1&quot;, &quot;+&quot;, 125, 125, &quot;hg19&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">151</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">midpoint</span>
<span class="go">Interval(&quot;chr1&quot;, &quot;+&quot;, 125, 126, &quot;hg19&quot;)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanning the midpoint of the interval.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.distance">
<span class="sig-name descname"><span class="pre">distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'midpoint'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.distance" title="Link to this definition"></a></dt>
<dd><p>The distance this interval and another.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">interval_a</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval_b</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">300</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval_c</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">400</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval_a</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">interval_b</span><span class="p">)</span>
<span class="go">125.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval_a</span><span class="o">.</span><span class="n">distance</span><span class="p">([</span><span class="n">interval_b</span><span class="p">,</span> <span class="n">interval_c</span><span class="p">])</span>
<span class="go">[125.0, 325.0]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a> or <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>) – an interval or list of intervals</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – determines which (point) attribute of this and the other interval to
use when computing the distance.
Must be one of <cite>midpoint</cite>, <cite>end5</cite>, or <cite>end3</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Distance to the interval. If other is of type Interval, return distance as a float;
otherwise, a list of distances.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#float" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a> or <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.as_dna0">
<span class="sig-name descname"><span class="pre">as_dna0</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.as_dna0"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.as_dna0" title="Link to this definition"></a></dt>
<dd><p>Returns (start, end) positions using DNA0 convention.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The positions <code class="docutils literal notranslate"><span class="pre">(start,</span> <span class="pre">end)</span></code> using DNA0 convention.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#tuple" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.as_dna1">
<span class="sig-name descname"><span class="pre">as_dna1</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.as_dna1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.as_dna1" title="Link to this definition"></a></dt>
<dd><p>Returns (start, end) positions using DNA1 convention.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The positions <code class="docutils literal notranslate"><span class="pre">(start,</span> <span class="pre">end)</span></code> using DNA1 convention.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#tuple" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.as_rna1">
<span class="sig-name descname"><span class="pre">as_rna1</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.as_rna1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.as_rna1" title="Link to this definition"></a></dt>
<dd><p>Returns (start, end) positions using RNA1 convention.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The positions <code class="docutils literal notranslate"><span class="pre">(start,</span> <span class="pre">end)</span></code> using RNA1 convention.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#tuple" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.as_ucsc">
<span class="sig-name descname"><span class="pre">as_ucsc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.as_ucsc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.as_ucsc" title="Link to this definition"></a></dt>
<dd><p>Returns this interval as a UCSC browser coordinate.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A string in UCSC browser convention, e.g. <code class="docutils literal notranslate"><span class="pre">&quot;chr5:500201-5002010&quot;</span></code>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.with_anchor">
<span class="sig-name descname"><span class="pre">with_anchor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anchor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anchor_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.with_anchor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.with_anchor" title="Link to this definition"></a></dt>
<dd><p>Returns an anchored version of this interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An anchored version of this interval.
The previous anchor, if any, is ignored.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Interval.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/interval.html#Interval.__str__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Interval.__str__" title="Link to this definition"></a></dt>
<dd><p>Return str(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="variant">
<h2>Variant<a class="headerlink" href="#variant" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Variant">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Variant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#Variant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Variant" title="Link to this definition"></a></dt>
<dd><p>A variant.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Variant.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#Variant.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Variant.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize a variant from a DNA0 position.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The chromosome name (‘chr1’, …).</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The start of the reference genome spanned by <cite>ref</cite> (inclusive, 0-based).</p></li>
<li><p><strong>ref</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The reference allele.</p></li>
<li><p><strong>alt</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The alternate allele.</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome. See <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Variant.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.Variant.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this variant in the reference genome.</p>
<p>Note that <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this variant’s <cite>ref</cite> sequence in the reference genome.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Variant.position">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">position</span></span><a class="headerlink" href="#genome_kit.Variant.position" title="Link to this definition"></a></dt>
<dd><p>The integer position of the start (inclusive, 0-based).</p>
<p>This property is a synonym for <cite>start</cite> and is provided for
historical reasons.  For consistency with other interval-based
objects, prefer the <cite>start</cite> property.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The start position of this interval within the reference genome.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Variant.ref">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ref</span></span><a class="headerlink" href="#genome_kit.Variant.ref" title="Link to this definition"></a></dt>
<dd><p>The reference allele as a DNA string.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The sequence of the reference allele that is altered by this variant.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Variant.alt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">alt</span></span><a class="headerlink" href="#genome_kit.Variant.alt" title="Link to this definition"></a></dt>
<dd><p>The alternate allele as a DNA string.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The sequence of the alternate allele that this variant represents.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Variant.as_variant_string">
<span class="sig-name descname"><span class="pre">as_variant_string</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#Variant.as_variant_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Variant.as_variant_string" title="Link to this definition"></a></dt>
<dd><p>Returns a string representation of the variant.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#genome_kit.Variant.from_string" title="genome_kit.Variant.from_string"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_string()</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The variant in the format <code class="docutils literal notranslate"><span class="pre">&quot;chromosome:start:ref:alt&quot;</span></code>, where
start is in DNA1 coordinates.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Variant.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#Variant.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Variant.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Variant.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genome_kit.Variant.__str__" title="Link to this definition"></a></dt>
<dd><p>Returns a string representation of the variant.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#genome_kit.Variant.from_string" title="genome_kit.Variant.from_string"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_string()</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The variant in the format <code class="docutils literal notranslate"><span class="pre">&quot;chromosome:start:ref:alt&quot;</span></code>, where
start is in DNA1 coordinates.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Variant.from_string">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#Variant.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Variant.from_string" title="Link to this definition"></a></dt>
<dd><p>Initialize a variant object from a variant string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – A variant in the format <code class="docutils literal notranslate"><span class="pre">&quot;chromosome:start:ref:alt&quot;</span></code> where start is in DNA1 coordinates.</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome that <cite>variant</cite> diverges from.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#genome_kit.Variant.as_variant_string" title="genome_kit.Variant.as_variant_string"><code class="xref py py-meth docutils literal notranslate"><span class="pre">as_variant_string()</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Variant.spanning">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">spanning</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variants</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#Variant.spanning"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Variant.spanning" title="Link to this definition"></a></dt>
<dd><p>Extends <a class="reference internal" href="#genome_kit.Interval.spanning" title="genome_kit.Interval.spanning"><code class="xref py py-meth docutils literal notranslate"><span class="pre">spanning()</span></code></a> such that it is aware
of variants and anchors.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genome_kit</span> <span class="kn">import</span> <span class="n">Genome</span><span class="p">,</span> <span class="n">Variant</span><span class="p">,</span> <span class="n">Interval</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hg19</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;hg19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">v</span> <span class="o">=</span> <span class="n">Variant</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="mi">10</span> <span class="o">*</span> <span class="s2">&quot;N&quot;</span><span class="p">,</span> <span class="n">hg19</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="n">hg19</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">250</span><span class="p">,</span> <span class="n">hg19</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Interval</span><span class="o">.</span><span class="n">spanning</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
<span class="go">Interval(&quot;chr1&quot;, &quot;+&quot;, 100, 250, &quot;hg19&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Variant</span><span class="o">.</span><span class="n">spanning</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="p">[</span><span class="n">v</span><span class="p">])</span>
<span class="go">Interval(&quot;chr1&quot;, &quot;+&quot;, 100, 260, &quot;hg19&quot;, 100)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – an interval</p></li>
<li><p><strong>y</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – another interval</p></li>
<li><p><strong>variants</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> or None, optional) – variants to apply to <cite>x</cite> and <cite>y</cite>; variants cannot overlap.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The resulting spanning interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="genome">
<h2>Genome<a class="headerlink" href="#genome" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Genome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">config</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome" title="Link to this definition"></a></dt>
<dd><p>Initialize a reference genome and its corresponding resources (DNA, annotations, etc).</p>
<p>A genome object represents a reference genome and resources available for it,
such as DNA and annotations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;gencode.v19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exon</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">exons</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>                                <span class="c1"># 1st annotated exon.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span>  <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">dna</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>                               <span class="c1"># Extract DNA.</span>
</pre></div>
</div>
<p>The initializer argument specifies a versioned resource:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Access resources that are hg19</span>
<span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;hg19&quot;</span><span class="p">)</span>

<span class="c1"># GENCODE v29 implies hg38</span>
<span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;gencode.v29&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The currently recognized strings depend on the data made available by data_manager.
Some common ones are:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reference_genomes</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;hg19&quot;</span><span class="p">,</span>
    <span class="s2">&quot;hg19.p13.plusMT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;hg38&quot;</span><span class="p">,</span>
    <span class="s2">&quot;hg38.p12&quot;</span><span class="p">,</span>
    <span class="s2">&quot;hg38.p13&quot;</span><span class="p">,</span>
    <span class="s2">&quot;hg38.p14&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mm10.p6&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mm39&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rn6&quot;</span><span class="p">,</span>
    <span class="s2">&quot;macFas5&quot;</span><span class="p">,</span>
    <span class="s2">&quot;susScr11&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">annotations</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;gencode.v19&quot;</span><span class="p">,</span>                  <span class="c1"># implies hg19.p13.plusMT</span>
    <span class="s2">&quot;gencode.v29&quot;</span><span class="p">,</span>                  <span class="c1"># implies hg38.p12</span>
    <span class="s2">&quot;gencode.v29.basic&quot;</span><span class="p">,</span>            <span class="c1"># implies hg38.p12</span>
    <span class="s2">&quot;gencode.v29lift37&quot;</span><span class="p">,</span>            <span class="c1"># implies hg19</span>
    <span class="s2">&quot;gencode.v29lift37.basic&quot;</span><span class="p">,</span>      <span class="c1"># implies hg19</span>
    <span class="s2">&quot;gencode.v41&quot;</span><span class="p">,</span>                  <span class="c1"># implies hg38.p13</span>
    <span class="s2">&quot;gencode.v41.basic&quot;</span><span class="p">,</span>            <span class="c1"># implies hg38.p13</span>
    <span class="s2">&quot;gencode.v41lift37&quot;</span><span class="p">,</span>            <span class="c1"># implies hg19</span>
    <span class="s2">&quot;gencode.v41lift37.basic&quot;</span><span class="p">,</span>      <span class="c1"># implies hg19</span>
    <span class="s2">&quot;ucsc_refseq.2017-06-25&quot;</span><span class="p">,</span>       <span class="c1"># implies hg19</span>
    <span class="s2">&quot;ncbi_refseq.v105.20190906&quot;</span><span class="p">,</span>    <span class="c1"># implies hg19.p13.plusMT</span>
    <span class="s2">&quot;ncbi_refseq.v109&quot;</span><span class="p">,</span>             <span class="c1"># implies hg38.p12</span>
    <span class="s2">&quot;ncbi_refseq.v110&quot;</span><span class="p">,</span>             <span class="c1"># implies hg38.p14</span>
    <span class="s2">&quot;gencode.vM19&quot;</span><span class="p">,</span>                 <span class="c1"># implies mm10.p6</span>
    <span class="s2">&quot;gencode.vM19.basic&quot;</span><span class="p">,</span>           <span class="c1"># implies mm10.p6</span>
    <span class="s2">&quot;gencode.vM30&quot;</span><span class="p">,</span>                 <span class="c1"># implies mm39</span>
    <span class="s2">&quot;gencode.vM30.basic&quot;</span><span class="p">,</span>           <span class="c1"># implies mm39</span>
    <span class="s2">&quot;gencode.vM31&quot;</span><span class="p">,</span>                 <span class="c1"># implies mm39</span>
    <span class="s2">&quot;gencode.vM31.basic&quot;</span><span class="p">,</span>           <span class="c1"># implies mm39</span>
    <span class="s2">&quot;ncbi_refseq.m39.v109&quot;</span><span class="p">,</span>         <span class="c1"># implies mm39</span>
    <span class="s2">&quot;ensembl.Rnor_6.0.88&quot;</span><span class="p">,</span>          <span class="c1"># implies rn6</span>
    <span class="s2">&quot;ncbi_refseq.Macfas_5.0.v101&quot;</span><span class="p">,</span>  <span class="c1"># implies macFas5</span>
    <span class="s2">&quot;ensembl.Macfas_5.0.95&quot;</span><span class="p">,</span>        <span class="c1"># implies macFas5</span>
    <span class="s2">&quot;ensembl.Sscrofa11.1.98&quot;</span><span class="p">,</span>       <span class="c1"># implies susScr11</span>
<span class="p">]</span>
</pre></div>
</div>
<p>If a resource is requested after object creation, but the version
needed cannot be disambiguated from <cite>config</cite>, an exception is raised.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>config</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Identifies a versioned resource that this genome object is
expected to serve.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">config</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.__reduce__">
<span class="sig-name descname"><span class="pre">__reduce__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.__reduce__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.__reduce__" title="Link to this definition"></a></dt>
<dd><p>Helper for pickle.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.__new__">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">__new__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.__new__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.__new__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.interval">
<span class="sig-name descname"><span class="pre">interval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.interval"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.interval" title="Link to this definition"></a></dt>
<dd><p>Shorthand method for creating an Interval with the same refg as the genome object called on.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.variant">
<span class="sig-name descname"><span class="pre">variant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_string</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.variant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.variant" title="Link to this definition"></a></dt>
<dd><p>Shorthand method for creating a Variant with the same refg as the genome object called on.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.dna">
<span class="sig-name descname"><span class="pre">dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_outside_chromosome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.dna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.dna" title="Link to this definition"></a></dt>
<dd><p>Extract DNA from a reference genome.</p>
<p>If <cite>interval</cite> is on the negative strand, the reverse complement sequence is returned.
If <cite>interval</cite> is outside the range of the chromosome, an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IndexError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IndexError</span></code></a> is raised.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p></li>
<li><p><strong>allow_outside_chromosome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – If False, does not allow the interval to be outside the range of the chromosome.
Attempting to pass an interval outside the chromosome range will raise an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IndexError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IndexError</span></code></a>.
If not specified, defaults to True. Any base outside the chromosome range will be returned as an ‘N’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The DNA sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.variant_dna">
<span class="sig-name descname"><span class="pre">variant_dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variants</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_outside_chromosome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.variant_dna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.variant_dna" title="Link to this definition"></a></dt>
<dd><p>Extract DNA from a genome with specific variants applied to it.</p>
<p>Follows the same rules as <a class="reference internal" href="#genome_kit.VariantGenome.dna" title="genome_kit.VariantGenome.dna"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dna()</span></code></a>.
Intended for applying sets of variants that may overlap <cite>interval</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p></li>
<li><p><strong>variants</strong> (<a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of (<a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a>)) – Either a single variant or a list of variants. See <a class="reference internal" href="#genome_kit.VariantGenome" title="genome_kit.VariantGenome"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantGenome</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The variant DNA sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.annotation">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotation</span></span><a class="headerlink" href="#genome_kit.Genome.annotation" title="Link to this definition"></a></dt>
<dd><p>Access to annotations.</p>
<p>A shorthand property <cite>anno</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that provides access to annotated genes, transcripts, and exons.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.GenomeAnnotation" title="genome_kit.GenomeAnnotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomeAnnotation</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genes</span></span><a class="headerlink" href="#genome_kit.Genome.genes" title="Link to this definition"></a></dt>
<dd><p>Access to annotated genes.</p>
<p>This is shorthand for <cite>annotation.genes</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated genes.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.GeneTable" title="genome_kit.GeneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.transcripts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">transcripts</span></span><a class="headerlink" href="#genome_kit.Genome.transcripts" title="Link to this definition"></a></dt>
<dd><p>Access to annotated transcripts.</p>
<p>This is shorthand for <cite>annotation.transcripts</cite>.</p>
<p>A shorthand property <cite>trans</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated transcripts.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.TranscriptTable" title="genome_kit.TranscriptTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">TranscriptTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.exons">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exons</span></span><a class="headerlink" href="#genome_kit.Genome.exons" title="Link to this definition"></a></dt>
<dd><p>Access to annotated exons.</p>
<p>This is shorthand for <cite>annotation.exons</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated exons.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.ExonTable" title="genome_kit.ExonTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ExonTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.introns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">introns</span></span><a class="headerlink" href="#genome_kit.Genome.introns" title="Link to this definition"></a></dt>
<dd><p>Access to annotated introns.</p>
<p>This is shorthand for <cite>annotation.introns</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated introns.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.IntronTable" title="genome_kit.IntronTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IntronTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.cdss">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cdss</span></span><a class="headerlink" href="#genome_kit.Genome.cdss" title="Link to this definition"></a></dt>
<dd><p>Access to annotated coding sequences (CDSs).</p>
<p>This is shorthand for <cite>annotation.cdss</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated coding sequences.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.CdsTable" title="genome_kit.CdsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CdsTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.utr5s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">utr5s</span></span><a class="headerlink" href="#genome_kit.Genome.utr5s" title="Link to this definition"></a></dt>
<dd><p>Access to annotated UTR5 sequences.</p>
<p>This is shorthand for <cite>annotation.utr5s</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated coding sequences.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Utr5Table</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.utr3s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">utr3s</span></span><a class="headerlink" href="#genome_kit.Genome.utr3s" title="Link to this definition"></a></dt>
<dd><p>Access to annotated UTR3 sequences.</p>
<p>This is shorthand for <cite>annotation.utr3s</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated coding sequences.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Utr3Table</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.find_motif">
<span class="sig-name descname"><span class="pre">find_motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">match_position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">find_overlapping_motifs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.find_motif"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.find_motif" title="Link to this definition"></a></dt>
<dd><p>Find a genomic motif in an interval on a <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – A genomic interval to search for the motif.</p></li>
<li><p><strong>motif</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – A genomic motif to search for.</p></li>
<li><p><strong>match_position</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Very often you want to find a position after a motif or in the middle
of a motif. For example, when you are searching for a <code class="docutils literal notranslate"><span class="pre">AG</span></code> core
acceptor site motif you want to match the 3p end of the <code class="docutils literal notranslate"><span class="pre">AG</span></code> motif to
return the position of the putative splice site.
To do this you can specify a match position as string with the values
<code class="docutils literal notranslate"><span class="pre">&quot;3p&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;5p&quot;</span></code>, or you can pass an integer that indicates a position
in the motif with <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">match_position</span> <span class="pre">&lt;=</span> <span class="pre">len(motif)</span></code>.</p></li>
<li><p><strong>find_overlapping_motifs</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to return matches for overlapping motifs. The default is to only
find non-overlapping motifs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of all matches (length-0 intervals) found in the query interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">genome37</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="s1">&#39;hg19&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span> <span class="o">=</span> <span class="n">genome37</span><span class="o">.</span><span class="n">interval</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="mi">40000</span><span class="p">,</span> <span class="mi">40080</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">motif</span> <span class="o">=</span> <span class="s1">&#39;AA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome37</span><span class="o">.</span><span class="n">find_motif</span><span class="p">(</span><span class="n">interval</span><span class="p">,</span> <span class="n">motif</span><span class="p">)</span>  
<span class="go">[Interval(&quot;chr1&quot;, &quot;+&quot;, 40031, 40031, &quot;hg19&quot;),</span>
<span class="go"> Interval(&quot;chr1&quot;, &quot;+&quot;, 40061, 40061, &quot;hg19&quot;)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.reference_genome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_genome</span></span><a class="headerlink" href="#genome_kit.Genome.reference_genome" title="Link to this definition"></a></dt>
<dd><p>The name of the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The name of the reference genome that any resources map to, e.g. “hg19” or “hg38”.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.chromosomes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosomes</span></span><a class="headerlink" href="#genome_kit.Genome.chromosomes" title="Link to this definition"></a></dt>
<dd><p>A list of valid chromosome names for this genome.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.chromosome_size">
<span class="sig-name descname"><span class="pre">chromosome_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.chromosome_size"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.chromosome_size" title="Link to this definition"></a></dt>
<dd><p>Returns the size of the given chromosome, in bp.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The size of the given chromosome, as number of basepairs.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.config">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">config</span></span><a class="headerlink" href="#genome_kit.Genome.config" title="Link to this definition"></a></dt>
<dd><p>The configuration strings of this genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The config used to initialize this object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Genome.data_dir">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">data_dir</span></span><a class="headerlink" href="#genome_kit.Genome.data_dir" title="Link to this definition"></a></dt>
<dd><p>The data_dir from which resources are opened.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to where resource files are being opened.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.appris_transcripts">
<span class="sig-name descname"><span class="pre">appris_transcripts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.appris_transcripts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.appris_transcripts" title="Link to this definition"></a></dt>
<dd><p>Returns a list of transcript objects ordered by our custom APPRIS index.
If a gene is provided, returns a list of transcripts for that gene ordered by decreasing priority.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gene</strong> (<a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, optional) – Restrict to a specific gene. If using <cite>str</cite>, use the geneID.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of transcripts that have APPRIS scores available. If a gene is provided as input, a list
of transcripts for that gene in the order from most to least principle; so the first transcript should
be used as the canonical.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ApprisNotAvailableError:</strong> – APPRIS data is not available for this genome’s annotations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Genome.appris_principality">
<span class="sig-name descname"><span class="pre">appris_principality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transcript</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_string</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome.html#Genome.appris_principality"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Genome.appris_principality" title="Link to this definition"></a></dt>
<dd><p>Returns the APPRIS principality index (isoform specific number) for this transcript.
An APPRIS principality index is an integer from 0..6 where 0 is considered most prinicpal and 6 considered least
principal.
In the original APPRIS database, the 7 principality levels are reported using strings, in the following order:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">appris_index_names</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;PRINCIPAL:1&quot;</span><span class="p">,</span>     <span class="c1"># index 0</span>
    <span class="s2">&quot;PRINCIPAL:2&quot;</span><span class="p">,</span>     <span class="c1"># index 1</span>
    <span class="s2">&quot;PRINCIPAL:3&quot;</span><span class="p">,</span>     <span class="c1"># index 2</span>
    <span class="s2">&quot;PRINCIPAL:4&quot;</span><span class="p">,</span>     <span class="c1"># index 3</span>
    <span class="s2">&quot;PRINCIPAL:5&quot;</span><span class="p">,</span>     <span class="c1"># index 4</span>
    <span class="s2">&quot;ALTERNATIVE:1&quot;</span><span class="p">,</span>   <span class="c1"># index 5</span>
    <span class="s2">&quot;ALTERNATIVE:2&quot;</span><span class="p">,</span>   <span class="c1"># index 6</span>
<span class="p">]</span>
</pre></div>
</div>
<p>See the <a class="reference external" href="http://appris.bioinfo.cnio.es/#/help/database/report">APPRIS database documentation</a> for details.</p>
<p>The APPRIS version defaults to <cite>2017_05.v22</cite> most annotations; see _build_appris.REMOTE_FILES for the rest.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>transcript</strong> (<a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – A transcript to query.</p></li>
<li><p><strong>as_string</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, optional) – Toggle True if the APPRIS principality string (e.g. <cite>“PRINCIPAL:1”</cite>) is desired.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The APPRIS principality index [0..6] of this transcript, or the string representation if
<cite>as_string=True</cite></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ApprisNotAvailableError:</strong> – APPRIS data is not available for this genome’s annotations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.Genome.__weakref__">
<span class="sig-name descname"><span class="pre">__weakref__</span></span><a class="headerlink" href="#genome_kit.Genome.__weakref__" title="Link to this definition"></a></dt>
<dd><p>list of weak references to the object (if defined)</p>
</dd></dl>

</dd></dl>

</section>
<section id="variantgenome">
<h2>VariantGenome<a class="headerlink" href="#variantgenome" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.VariantGenome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">VariantGenome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variants</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant_genome.html#VariantGenome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantGenome" title="Link to this definition"></a></dt>
<dd><p>A reference genome with a number of variants applied to it.</p>
<p>A variant genome behaves like a <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a> object,
object, except any requests for DNA sequence (or quantities that depend
on DNA sequence) will be returned with the variants already applied.
If the variants change the length of the sequence, then the
interval’s <cite>anchor</cite> and <cite>anchor_offset</cite> determine how the query interval
will be aligned to the variant genome’s coordinate system.</p>
<p>TODO: see explanation of anchors.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantGenome.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variants</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant_genome.html#VariantGenome.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantGenome.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize a variant genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome to which variants are applied.</p></li>
<li><p><strong>variants</strong> (<a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of (<a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a>)) – Either a single variant or a list of variants. Aligned-normalization of
variants is not yet implemented, so the user is currently
responsible for not passing variants that overlap. The variants must be provided
as a <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> object. String variants are no longer allowed,
please see <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> for more information. Since we assume that
variants never overlap one another this means that all variant
positions are with respect to the reference genome.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantGenome.dna">
<span class="sig-name descname"><span class="pre">dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_outside_chromosome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant_genome.html#VariantGenome.dna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantGenome.dna" title="Link to this definition"></a></dt>
<dd><p>Extract variant DNA for an anchored interval or coordinate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – If <cite>interval</cite> is on the negative strand, the reverse complement
sequence is returned. The <cite>interval.anchor</cite> and <cite>interval.anchor_offset</cite>
attributes determine the policy of how length-changing variants are applied.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The variant DNA sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantGenome.find_motif">
<span class="sig-name descname"><span class="pre">find_motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">match_position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">find_overlapping_motifs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant_genome.html#VariantGenome.find_motif"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantGenome.find_motif" title="Link to this definition"></a></dt>
<dd><p>Find a genomic motif in an interval on a <a class="reference internal" href="#genome_kit.VariantGenome" title="genome_kit.VariantGenome"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantGenome</span></code></a>.</p>
<p>This method finds all occurrences of a genomic motif on a VariantGenome. It can
handle all kinds of variants including indels and complex variants. When a motif
is found within an insertion, it will set the <code class="docutils literal notranslate"><span class="pre">anchor_offset</span></code> attribute on
the returned <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – A genomic interval to search for the motif.</p></li>
<li><p><strong>motif</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – A genomic motif to search for.</p></li>
<li><p><strong>match_position</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Very often you want to find a position after a motif or in the middle
of a motif. For example, when you are searching for a <code class="docutils literal notranslate"><span class="pre">AG</span></code> core
acceptor site motif you want to match the 3p end of the <code class="docutils literal notranslate"><span class="pre">AG</span></code> motif to
return the position of the putative splice site.
To do this you can specify a match position as string with the values
<code class="docutils literal notranslate"><span class="pre">&quot;3p&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;5p&quot;</span></code>, or you can pass an integer that indicates a position
in the motif with <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">match_position</span> <span class="pre">&lt;=</span> <span class="pre">len(motif)</span></code>.</p></li>
<li><p><strong>find_overlapping_motifs</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to return matches for overlapping motifs. The default is to only
find non-overlapping motifs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of all matches (length-0 intervals) found in the query interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>py:class`list` of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genome_kit</span> <span class="kn">import</span> <span class="n">Genome</span><span class="p">,</span> <span class="n">Variant</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="s1">&#39;hg19&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variant</span> <span class="o">=</span> <span class="n">Variant</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s2">&quot;chr1:40033:A:G&quot;</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variant_genome</span> <span class="o">=</span> <span class="n">VariantGenome</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">variant</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">interval</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="mi">40000</span><span class="p">,</span> <span class="mi">40080</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">motif</span> <span class="o">=</span> <span class="s1">&#39;TAG&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variant_genome</span><span class="o">.</span><span class="n">find_motif</span><span class="p">(</span><span class="n">interval</span><span class="p">,</span> <span class="n">motif</span><span class="p">)</span>
<span class="go">[Interval(&quot;chr1&quot;, &quot;+&quot;, 40030, 40030, &quot;hg19&quot;, 40030)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantGenome.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant_genome.html#VariantGenome.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantGenome.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="tracks">
<h2>Tracks<a class="headerlink" href="#tracks" title="Link to this heading"></a></h2>
<section id="genomedna">
<h3>GenomeDNA<a class="headerlink" href="#genomedna" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.GenomeDNA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GenomeDNA</span></span><a class="reference internal" href="_modules/genome_kit/genome_dna.html#GenomeDNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeDNA" title="Link to this definition"></a></dt>
<dd><p>Access to the DNA sequence of a reference genome.</p>
<p>This object should be accessed through the <cite>dna</cite> property of
<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genome_kit</span> <span class="kn">import</span> <span class="n">Genome</span><span class="p">,</span> <span class="n">Interval</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s2">&quot;chr5&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="mi">50000</span><span class="p">,</span> <span class="mi">50010</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;hg19&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span><span class="o">.</span><span class="n">dna</span><span class="p">(</span><span class="n">interval</span><span class="p">)</span>
<span class="go">&#39;TAAACCACAT&#39;</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeDNA.__call__">
<span class="sig-name descname"><span class="pre">__call__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_outside_chromosome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_dna.html#GenomeDNA.__call__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeDNA.__call__" title="Link to this definition"></a></dt>
<dd><p>Extract DNA from a reference genome.</p>
<p>If <cite>interval</cite> is on the negative strand, the reverse complement sequence is returned.
If <cite>interval</cite> is outside the range of the chromosome, an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IndexError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IndexError</span></code></a> is raised.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p></li>
<li><p><strong>allow_outside_chromosome</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – If False, does not allow the interval to be outside the range of the chromosome.
Attempting to pass an interval outside the chromosome range will raise an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IndexError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IndexError</span></code></a>.
If not specified, defaults to True. Any base outside the chromosome range will be returned as an ‘N’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The DNA sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeDNA.reference_genome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_genome</span></span><a class="headerlink" href="#genome_kit.GenomeDNA.reference_genome" title="Link to this definition"></a></dt>
<dd><p>The name of the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The name of the reference genome from which DNA will be extracted, e.g. “hg38”.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeDNA.filename">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filename</span></span><a class="headerlink" href="#genome_kit.GenomeDNA.filename" title="Link to this definition"></a></dt>
<dd><p>The path to the file from which DNA is extracted.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the file, e.g. “/data/hg19.2bit”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeDNA.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_dna.html#GenomeDNA.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeDNA.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="genometrack">
<h3>GenomeTrack<a class="headerlink" href="#genometrack" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GenomeTrack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrack"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrack" title="Link to this definition"></a></dt>
<dd><p>Access to a genomic data track.</p>
<p>A <a class="reference internal" href="#genome_kit.GenomeTrack" title="genome_kit.GenomeTrack"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomeTrack</span></code></a> provides a read-only view of a <cite>.gtrack</cite> file.
A track can be easily opened and queried:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="s1">&#39;chr17&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="mi">41246881</span><span class="p">,</span> <span class="mi">41246886</span><span class="p">,</span> <span class="s1">&#39;hg19&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">track</span> <span class="o">=</span> <span class="n">GenomeTrack</span><span class="p">(</span><span class="s2">&quot;hg19.phastcons_mammal.f4.gtrack&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">track</span><span class="p">(</span><span class="n">interval</span><span class="p">)</span>
<span class="go">array([[ 1.   ],</span>
<span class="go">       [ 0.857],</span>
<span class="go">       [ 0.571],</span>
<span class="go">       [ 0.357],</span>
<span class="go">       [ 0.286]], dtype=float16)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">track</span><span class="p">(</span><span class="n">interval</span><span class="o">.</span><span class="n">as_opposite_strand</span><span class="p">())</span>
<span class="go">array([[ 0.286],</span>
<span class="go">       [ 0.357],</span>
<span class="go">       [ 0.571],</span>
<span class="go">       [ 0.857],</span>
<span class="go">       [ 1.   ]], dtype=float16)</span>
</pre></div>
</div>
<p>You can also open a track with a specific scope:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">GenomeTrack</span><span class="p">(</span><span class="s2">&quot;foo.gtrack&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">track</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">values</span> <span class="o">=</span> <span class="n">track</span><span class="p">(</span><span class="n">interval</span><span class="p">)</span>
</pre></div>
</div>
<p>Tracks can be encoded in formats called <cite>etypes</cite>.
The available etypes trade representation power
for storage size and query speed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">etypes</span> <span class="o">=</span> <span class="p">[</span>  <span class="c1"># DECODABLE TYPES          RANGE                ENCODED TYPE</span>
    <span class="s2">&quot;m0&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, uint8, bool  {1}                  0-bit (mask)</span>
    <span class="s2">&quot;u1&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, uint8, bool  {0..1}               1-bit</span>
    <span class="s2">&quot;u2&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, uint8        {0..3}               2-bit</span>
    <span class="s2">&quot;u3&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, uint8        {0..7}               3-bit</span>
    <span class="s2">&quot;u4&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, uint8        {0..15}              4-bit</span>
    <span class="s2">&quot;u5&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, uint8        {0..31}              5-bit</span>
    <span class="s2">&quot;u6&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, uint8        {0..63}              6-bit</span>
    <span class="s2">&quot;u8&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, uint8        {0..255}             8-bit</span>
    <span class="s2">&quot;i8&quot;</span><span class="p">,</span>   <span class="c1"># float16/32, int8         {-128..127}          8-bit</span>
    <span class="s2">&quot;f2&quot;</span><span class="p">,</span>   <span class="c1"># float16/32               [-65504,65504]       2-bit dictionary</span>
    <span class="s2">&quot;f3&quot;</span><span class="p">,</span>   <span class="c1"># float16/32               [-65504,65504]       3-bit dictionary</span>
    <span class="s2">&quot;f4&quot;</span><span class="p">,</span>   <span class="c1"># float16/32               [-65504,65504]       4-bit dictionary</span>
    <span class="s2">&quot;f5&quot;</span><span class="p">,</span>   <span class="c1"># float16/32               [-65504,65504]       5-bit dictionary</span>
    <span class="s2">&quot;f6&quot;</span><span class="p">,</span>   <span class="c1"># float16/32               [-65504,65504]       6-bit dictionary</span>
    <span class="s2">&quot;f8&quot;</span><span class="p">,</span>   <span class="c1"># float16/32               [-65504,65504]       8-bit dictionary</span>
    <span class="s2">&quot;f16&quot;</span><span class="p">,</span>  <span class="c1"># float16/32               [-65504,65504]       16-bit float</span>
    <span class="s2">&quot;f32&quot;</span><span class="p">,</span>  <span class="c1"># float16/32               [-1.2e-38, 3.4e38]   32-bit float</span>
<span class="p">]</span>
</pre></div>
</div>
<p>The default dictionary for <cite>f2..8</cite> types is the range [0,1] computed
as <code class="docutils literal notranslate"><span class="pre">np.linspace(0.0,</span> <span class="pre">1.0,</span> <span class="pre">2**num_bits,</span> <span class="pre">dtype=np.float16)</span></code>.</p>
<p>Decoded types are called <cite>dtypes</cite>, and are indeed represented by
numpy dtypes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dtypes</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">,</span>    <span class="c1"># {False, True}</span>
    <span class="n">np</span><span class="o">.</span><span class="n">uint8</span><span class="p">,</span>    <span class="c1"># {0..255}</span>
    <span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">,</span>     <span class="c1"># {-128..127}</span>
    <span class="n">np</span><span class="o">.</span><span class="n">float16</span><span class="p">,</span>  <span class="c1"># [-65504, 65504] IEEE 754 half-precision</span>
    <span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">,</span>  <span class="c1"># [-1.2e-38, 3.4e38]</span>
<span class="p">]</span>
</pre></div>
</div>
<p>New tracks can be can created using <a class="reference internal" href="#genome_kit.GenomeTrackBuilder" title="genome_kit.GenomeTrackBuilder"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomeTrackBuilder</span></code></a>.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrack.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrack.__init__" title="Link to this definition"></a></dt>
<dd><p>Open a .gtrack file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>infile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The .gtrack file to open.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.__call__">
<span class="sig-name descname"><span class="pre">__call__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrack.__call__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrack.__call__" title="Link to this definition"></a></dt>
<dd><p>Extract data from a genomic track.</p>
<p>Data is always returned in sense-strand order, meaning it is
always sensitive to the strand of the interval,
even on an unstranded track.</p>
<p>You can optionally specify the dtype that a track should be
decoded to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The stranded query interval.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#type" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">type</span></code></a>) – Optional. The numpy dtype of the resulting array. Ignored if
<code class="docutils literal notranslate"><span class="pre">out</span></code> is provided.</p></li>
<li><p><strong>out</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>) – Optional. A numpy array to hold the result. If provided, it
overrides <code class="docutils literal notranslate"><span class="pre">dtype</span></code> and it must have a shape that the inputs
broadcast to. If not provided or None, a freshly-allocated array
is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The decoded track data with <code class="docutils literal notranslate"><span class="pre">len(interval)</span></code> rows and <code class="docutils literal notranslate"><span class="pre">dim</span></code> columns.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.dim">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dim</span></span><a class="headerlink" href="#genome_kit.GenomeTrack.dim" title="Link to this definition"></a></dt>
<dd><p>The dimensionality of the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of columns in each returned array.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.resolution">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">resolution</span></span><a class="headerlink" href="#genome_kit.GenomeTrack.resolution" title="Link to this definition"></a></dt>
<dd><p>The resolution of the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of genomic positions spanned by each datum.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.GenomeTrack.stranded" title="Link to this definition"></a></dt>
<dd><p>The strandedness of the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether the negative strand stores data unique from the positive strand.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.etype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">etype</span></span><a class="headerlink" href="#genome_kit.GenomeTrack.etype" title="Link to this definition"></a></dt>
<dd><p>The encoding type of the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A string identifying the encoding type.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.dtype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#genome_kit.GenomeTrack.dtype" title="Link to this definition"></a></dt>
<dd><p>The default dtype for the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The numpy dtype that this track will decode to by default.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#type" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">type</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.reference_genome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_genome</span></span><a class="headerlink" href="#genome_kit.GenomeTrack.reference_genome" title="Link to this definition"></a></dt>
<dd><p>The name of the reference genome this track is mapped to.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The name of the reference genome, e.g. “hg38”.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.filename">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filename</span></span><a class="headerlink" href="#genome_kit.GenomeTrack.filename" title="Link to this definition"></a></dt>
<dd><p>The path to <cite>.gtrack</cite> file from which data is extracted.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the file, e.g. “/data/phastcons.gtrack”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrack.intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrack.intervals" title="Link to this definition"></a></dt>
<dd><p>The list of intervals covered by this track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The list of intervals with data in this track.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.histogram">
<span class="sig-name descname"><span class="pre">histogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separate_dims</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrack.histogram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrack.histogram" title="Link to this definition"></a></dt>
<dd><p>Generate a histogram of the values in the track.</p>
<p>Returns a dictionary where each key is a float appearing
at least once in the track, and each value is a count of the
number of times it occurred. NaN values are omitted.</p>
<p>If <cite>region</cite> is a <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>, computes
a genome-wide histogram.
If <cite>region</cite> is an interval or list of intervals, computes the
histogram over that region only.</p>
<p>If <cite>separate_dims</cite> is true, returns a list of
<a class="reference internal" href="#genome_kit.GenomeTrack.dim" title="genome_kit.GenomeTrack.dim"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dim</span></code></a> histograms,
one for each track dimension.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a> | <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a> |                 <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Specific interval(s) upon which to compute
the histogram.</p></li>
<li><p><strong>separate_dims</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Optional. Whether to generate a single histogram over all
dimensions, or to return a separate histogram for each.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary <code class="docutils literal notranslate"><span class="pre">{</span> <span class="pre">value</span> <span class="pre">:</span> <span class="pre">count</span> <span class="pre">}</span></code> where the <cite>value</cite>
appearing in the track <cite>count</cite> times. Zero</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#dict" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#dict" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrack.close"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrack.close" title="Link to this definition"></a></dt>
<dd><p>Close the file handle.</p>
<p>Note that if you use a <cite>with</cite> statement the file handle is automatically closed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Open the file</span>
<span class="k">with</span> <span class="n">GenomeTrack</span><span class="p">(</span><span class="s1">&#39;foo.gtrack&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">track</span><span class="p">:</span>
    <span class="o">...</span>
<span class="c1"># &lt;-- File is now closed</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.gtrack_version">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gtrack_version</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrack.gtrack_version"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrack.gtrack_version" title="Link to this definition"></a></dt>
<dd><p>The <cite>gtrack</cite> file format version that this build supports.</p>
<p>Attempting to open a <cite>gtrack</cite> file of a mismatched version will raise an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IOError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IOError</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The version number.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrack.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrack.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrack.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="genometrackbuilder">
<h3>GenomeTrackBuilder<a class="headerlink" href="#genometrackbuilder" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GenomeTrackBuilder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">etype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder" title="Link to this definition"></a></dt>
<dd><p>Builds a .gtrack file.</p>
<p>Creating a .gtrack file with <a class="reference internal" href="#genome_kit.GenomeTrackBuilder" title="genome_kit.GenomeTrackBuilder"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomeTrackBuilder</span></code></a>
generally involves four steps:</p>
<ol class="arabic simple">
<li><p>Open the file for writing.</p></li>
<li><p>Set further configuration options.</p></li>
<li><p>Set data on intervals, either as numpy arrays or from disk.</p></li>
<li><p>Call <a class="reference internal" href="#genome_kit.GenomeTrackBuilder.finalize" title="genome_kit.GenomeTrackBuilder.finalize"><code class="xref py py-meth docutils literal notranslate"><span class="pre">finalize()</span></code></a>.</p></li>
</ol>
<p>For example, assuming we had a list of genomic intervals and
corresponding data arrays we wanted to fill them with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. Open a .gtrack 8-bit encoding of floats in range [0, 1]</span>
<span class="n">track</span> <span class="o">=</span> <span class="n">GenomeTrackBuilder</span><span class="p">(</span><span class="s2">&quot;foo.gtrack&quot;</span><span class="p">,</span> <span class="s2">&quot;f8&quot;</span><span class="p">,</span> <span class="s2">&quot;single_stranded&quot;</span><span class="p">,</span> <span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;hg19&quot;</span><span class="p">))</span>

<span class="c1"># 2. Configure the track further</span>
<span class="n">track</span><span class="o">.</span><span class="n">set_default_value</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>          <span class="c1"># Fill gaps in the data with 0</span>

<span class="c1"># 3. Set data for each interval</span>
<span class="k">for</span> <span class="n">interval</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">entries</span><span class="p">:</span>
    <span class="n">track</span><span class="o">.</span><span class="n">set_data</span><span class="p">(</span><span class="n">interval</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>

<span class="c1"># 4. Finish writing the file</span>
<span class="n">track</span><span class="o">.</span><span class="n">finalize</span><span class="p">()</span>
</pre></div>
</div>
<p>Tracks can have any dimension (the <cite>dim</cite> argument) or resolution
(the <cite>resolution</cite> argument).
Resolution allows track data to be specified and stored at coarser
than 1bp resolution.</p>
<p>For example, a track with <cite>resolution=5</cite> can
load a WIG file with step/span of 5bp:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fixedStep</span> <span class="n">chrom</span><span class="o">=</span><span class="n">chr1</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span> <span class="n">step</span><span class="o">=</span><span class="mi">5</span> <span class="n">span</span><span class="o">=</span><span class="mi">5</span>
<span class="mf">0.57</span>
<span class="mf">0.23</span>
<span class="o">...</span>
</pre></div>
</div>
<p>The resulting track is equivalent to the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span>    <span class="mi">1</span>    <span class="mi">2</span>    <span class="mi">3</span>    <span class="mi">4</span>    <span class="mi">5</span>    <span class="mi">6</span>    <span class="mi">7</span>    <span class="mi">8</span>    <span class="mi">9</span>    <span class="o">...</span>  <span class="o">&lt;--</span> <span class="n">position</span>
<span class="mf">0.57</span> <span class="mf">0.57</span> <span class="mf">0.57</span> <span class="mf">0.57</span> <span class="mf">0.57</span> <span class="mf">0.23</span> <span class="mf">0.23</span> <span class="mf">0.23</span> <span class="mf">0.23</span> <span class="mf">0.23</span> <span class="o">...</span>  <span class="o">&lt;--</span> <span class="n">decoded</span> <span class="n">value</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">etype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.__init__" title="Link to this definition"></a></dt>
<dd><p>Open a .gtrack for writing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The .gtrack file to create.</p></li>
<li><p><strong>etype</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The encoding format.
See <a class="reference internal" href="#genome_kit.GenomeTrack" title="genome_kit.GenomeTrack"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomeTrack</span></code></a> for details.</p></li>
<li><p><strong>strandedness</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – <p>Determines the order by which the data is applied.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;single_stranded&quot;</span></code>: both strands share the same data. The data is applied in Interval coordinate (reference strand) order.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;strand_unaware&quot;</span></code>: ignores the Interval strand, data is applied in Interval coordinate (reference strand) order.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;strand_aware&quot;</span></code>: data is applied from 5” end to 3” end (sense strand order).</p></li>
</ul>
</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome for the track data to build.</p></li>
<li><p><strong>dim</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The number of dimensions in the resulting track.</p></li>
<li><p><strong>resolution</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The resolution of the track data, in genomic positions.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_default_value">
<span class="sig-name descname"><span class="pre">set_default_value</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_default_value"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_default_value" title="Link to this definition"></a></dt>
<dd><p>Set the fill value for gaps in the track.</p>
<p>The default value does not need to be encodable.
For example, a default value of -1.0 for a dictionary that
does not contain -1.0 is fine.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>value</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#float" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – The value to fill gaps with.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_sparsity">
<span class="sig-name descname"><span class="pre">set_sparsity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_run</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">48</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_sparsity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_sparsity" title="Link to this definition"></a></dt>
<dd><p>Enable automatic sparsification of data intervals.</p>
<p>Reduces file size of tracks that are dominated by <cite>default_value</cite>,
with non-default values appearing at sparse intervals.
<em>Highly recommended for variableStep WIG files.</em></p>
<p>The default behaviour of a track is to encode all data
passed to <a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_data" title="genome_kit.GenomeTrackBuilder.set_data"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_data()</span></code></a>.
After calling <a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_sparsity" title="genome_kit.GenomeTrackBuilder.set_sparsity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_sparsity()</span></code></a>,
the encoder will analyze the data arrays passed in and avoid
encoding runs of <cite>default_value</cite>.</p>
<p>Specifically, if there exists a run of at least <cite>min_run</cite>
values such that each <code class="docutils literal notranslate"><span class="pre">abs(data[i]</span> <span class="pre">-</span> <span class="pre">default_value)</span> <span class="pre">&lt;=</span> <span class="pre">min_delta</span></code>,
then that entire run will be treated as <cite>default_value</cite> and
not explicitly stored in the file.</p>
<p>Note that, in order to exclude a run from the track, a 12-byte entry
must be addd to the track index, so setting <cite>min_run</cite> to very small values
can actually increase file size and will definitely increase query time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>min_run</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The minimum length of a run before it’s excluded.</p></li>
<li><p><strong>min_delta</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#float" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – Optional. The minimum delta for a value to be considered distinct
from <cite>default_value</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_clamping">
<span class="sig-name descname"><span class="pre">set_clamping</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_clamping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_clamping" title="Link to this definition"></a></dt>
<dd><p>Clamp all values to the encodable range.</p>
<p>By default, a value outside the encodable range will trigger
an error, to notify the user that his/her data may not be encoded
properly by the current format.</p>
<p>Calling this method shows intent that values should be clamped
to the encodable range. Currently allowed only for f-type encodings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_dict">
<span class="sig-name descname"><span class="pre">set_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_dict" title="Link to this definition"></a></dt>
<dd><p>Set the dictionary for etypes <cite>f2..8</cite>.</p>
<p>The dictionary must be a float16/32 numpy array.
For etype f&lt;n&gt;, there must be <cite>2**n</cite> entries in the array.
Numerical values must be in non-decreasing order.
The last entry in <cite>dict</cite> can be <code class="xref py py-data docutils literal notranslate"><span class="pre">np.nan</span></code>
if you wish to be able to encode NaN at specific positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dict</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">np.ndarray</span></code>) – The array of values comprising the dictionary.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_transform">
<span class="sig-name descname"><span class="pre">set_transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_transform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_transform" title="Link to this definition"></a></dt>
<dd><p>Transform data from range <cite>[a, b]</cite> to range <cite>[c, d]</cite>.</p>
<p>This specifies that any time data is about to be encoded, it
will undergo an affine transformation from numeric range <cite>[a, b]</cite>
to range <cite>[c, d]</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#float" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – The source range.</p></li>
<li><p><strong>b</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#float" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – The source range.</p></li>
<li><p><strong>c</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#float" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – The target range.</p></li>
<li><p><strong>d</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#float" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – The target range.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_restriction">
<span class="sig-name descname"><span class="pre">set_restriction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">restriction</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_restriction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_restriction" title="Link to this definition"></a></dt>
<dd><p>Restrict data to a specific interval.</p>
<p>The restriction interval acts as a allowed region, so that all
data outside that region is either ignored or cropped away by
<a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_data" title="genome_kit.GenomeTrackBuilder.set_data"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_data()</span></code></a>.</p>
<p>The restriction interval’s strand is ignored, i.e. the restriction
always allows data on either strand.</p>
<p>This method is useful for making small version of full-sized data pipeline,
for the sake of creating unit tests or iterating faster during development.</p>
<p>If the restriction interval is not aligned with the track resolution, some
then the restriction interval will be expanded until it is aligned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>restriction</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The interval on which to keep data.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_data">
<span class="sig-name descname"><span class="pre">set_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_data" title="Link to this definition"></a></dt>
<dd><p>Set the track data on a specific interval.</p>
<p>If the interval overlaps a previously-specified interval,
an error is raised.</p>
<p>The <cite>dtype</cite> of the data array must be compatible with the
<cite>etype</cite> of the track being built.</p>
<p>Normally, <cite>data</cite> must be an NxM array where N is
the length of <cite>interval</cite> and M is <cite>dim</cite>.</p>
<p>If track resolution R is greater than 1, then <cite>data</cite>
must be (N/R)xM where N is the length of <cite>interval</cite>.
The start and end of <cite>interval</cite> must be aligned to R.
In other words, data is supplied (and stored) at
a coarsened resolution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The interval to fill with data.</p></li>
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">np.ndarray</span></code>) – The data array, ordered according to the strandedness argument passed to the
builder’s constructor.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_data_from_wig">
<span class="sig-name descname"><span class="pre">set_data_from_wig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos_strand_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neg_strand_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_data_from_wig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_data_from_wig" title="Link to this definition"></a></dt>
<dd><p>Load all data from a .wig or .wig.gz file.</p>
<p>Loads all data from a fixedStep or variableStep WIG file.</p>
<p>If the track is stranded, then the negative strand
track data will be loaded from the second file.</p>
<p>The number of data columns in the WIG must match the track dimension.
The span of the data must match the track resolution, with the exception
of the very last datum on a chromosome.</p>
<p>Adjacent WIG intervals will automatically be merged into a single GTRACK
interval, with different data encoded for each spanned position.
Consider calling <a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_sparsity" title="genome_kit.GenomeTrackBuilder.set_sparsity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_sparsity()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pos_strand_file</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Path to the .wig file for positive strand.</p></li>
<li><p><strong>neg_strand_file</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Optional. Path to the .wig file for negative strand, if stranded.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_data_from_bedgraph">
<span class="sig-name descname"><span class="pre">set_data_from_bedgraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos_strand_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neg_strand_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_data_from_bedgraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_data_from_bedgraph" title="Link to this definition"></a></dt>
<dd><p>Load all data from a .bedgraph or .bedgraph.gz file.</p>
<p>Loads all data from a BEDGRAPH file.
A direct analogue of <a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_data_from_wig" title="genome_kit.GenomeTrackBuilder.set_data_from_wig"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_data_from_wig()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pos_strand_file</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Path to the .bedgraph file for positive strand.</p></li>
<li><p><strong>neg_strand_file</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Optional. Path to the .bedgraph file for negative strand, if stranded.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.set_data_from_bed">
<span class="sig-name descname"><span class="pre">set_data_from_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bedfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">categories</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.set_data_from_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.set_data_from_bed" title="Link to this definition"></a></dt>
<dd><p>Load all data from a .bed or .bed.gz file.</p>
<p>Loads each interval from a BED file and incorporates it into the track.</p>
<p>If <cite>categories</cite> is not specified, then the value contained in the <cite>score</cite>
column of the BED file will be stored in the corresponding interval.</p>
<p>If <cite>categories</cite> is specified, then the <cite>name</cite> column of the BED file
determines the value stored on each interval. The interval’s <cite>name</cite> will
be used to find an index in <cite>categories</cite>, and that index will be stored.</p>
<p>The strandedness of the track must match the BED file (i.e. an unstranded
track expects the <cite>strand</cite> column to contain <code class="docutils literal notranslate"><span class="pre">'.'</span></code>).
The track dimension must be 1.
The start and end of each interval must match the track resolution,
with the exception of the very last position of a chromosome.</p>
<p>Adjacent BED intervals will automatically be merged into a single contiguous
GTRACK interval.
In a GTRACK file, the data value of an interval is repeated for the size of the
interval, unlike in a BED file where it’s specified only once per interval.
This can result in large GTRACK files, especially when at 1bp resolution (the default).
Consider using <a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_default_value" title="genome_kit.GenomeTrackBuilder.set_default_value"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_default_value()</span></code></a>
and <a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_sparsity" title="genome_kit.GenomeTrackBuilder.set_sparsity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_sparsity()</span></code></a> to manage file size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bedfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Path to the .bed file.</p></li>
<li><p><strong>categories</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Optional. List of categories to match with the <cite>name</cite> column
in the BED file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.flush"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.flush" title="Link to this definition"></a></dt>
<dd><p>Flush the currently accumulated data to disk.</p>
<p>This function can free up memory while building a track,
but should only be called immediately after an entire chromosome
has had its data set. Once this is called all chromosomes
that currently have <em>any</em> data on them become fixed.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.finalize">
<span class="sig-name descname"><span class="pre">finalize</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.finalize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.finalize" title="Link to this definition"></a></dt>
<dd><p>Finish writing the file and close it.</p>
<p>This function must be called last, to complete creation of the GTRACK file.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.dim">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dim</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.dim" title="Link to this definition"></a></dt>
<dd><p>The dimensionality of the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of columns in each returned array.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.resolution">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">resolution</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.resolution" title="Link to this definition"></a></dt>
<dd><p>The resolution of the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of genomic positions spanned by each datum.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.stranded" title="Link to this definition"></a></dt>
<dd><p>The strandedness of the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether the negative strand stores data unique from the positive strand.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.etype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">etype</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.etype" title="Link to this definition"></a></dt>
<dd><p>The encoding type of the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A string identifying the encoding type.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.dtype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.dtype" title="Link to this definition"></a></dt>
<dd><p>The default dtype for the track.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The numpy dtype that this track will decode to by default.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#type" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">type</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.reference_genome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_genome</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.reference_genome" title="Link to this definition"></a></dt>
<dd><p>The name of the reference genome the track will be mapped to.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The name of the reference genome, e.g. “hg38”.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.filename">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filename</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.filename" title="Link to this definition"></a></dt>
<dd><p>The path to <cite>.gtrack</cite> file being written to.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the file, e.g. “/data/phastcons.gtrack”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.data_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">data_size</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.data_size" title="Link to this definition"></a></dt>
<dd><p>The number of positions spanned by data in the GTRACK file.</p>
<p>If sparsity is disabled, the data size is the number of positions
spanned by data via
<a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_data" title="genome_kit.GenomeTrackBuilder.set_data"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_data()</span></code></a>.</p>
<p>Enabling sparsity may decrease the data size.</p>
<p>Using resolution &gt; 1bp will also decrease the data size, since data is
encoded into a coarse coordinate system, not at full genomic resolution.</p>
<p>The number of bits that the data occupies in the GTRACK file is
approximately <cite>data_size * dim * bits_per_datum</cite>, but the exact
computation is non-trivial, so portion of GTRACK file dedicated
to data may be higher or lower than this rough estimate.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of data spanned by encoded data.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.index_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">index_size</span></span><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.index_size" title="Link to this definition"></a></dt>
<dd><p>The number of separate intervals in the GTRACK index.</p>
<p>If sparsity is disabled, the index size is the number of times
<a class="reference internal" href="#genome_kit.GenomeTrackBuilder.set_data" title="genome_kit.GenomeTrackBuilder.set_data"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_data()</span></code></a> was called.</p>
<p>Enabling sparsity may increase the index size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of unique intervals in the index.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeTrackBuilder.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_track.html#GenomeTrackBuilder.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeTrackBuilder.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="annotations">
<h2>Annotations<a class="headerlink" href="#annotations" title="Link to this heading"></a></h2>
<section id="genomeannotation">
<h3>GenomeAnnotation<a class="headerlink" href="#genomeannotation" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GenomeAnnotation</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GenomeAnnotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeAnnotation" title="Link to this definition"></a></dt>
<dd><p>Annotations for a reference genome.</p>
<p>This object should not be directly created.
Instead, create a <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;gencode.v19&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a> object also exposes shortcut
attributes, for convenience:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>See <a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a> for a list of annotations
available, e.g. <code class="docutils literal notranslate"><span class="pre">gencode.v29</span></code>, <code class="docutils literal notranslate"><span class="pre">ucsc_refseq.2017-06-25</span></code>, etc.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>UCSC RefSeq can contain multiple versions of the same gene/transcript.</strong>
These tend to be in ambiguously mapped regions. In GenomeKit, they’re
retained as separate entries in the gene and transcript tables, with
distinct intervals but otherwise identical.</p>
<p>For example, in GENCODE v19 the SMN1/2 genes are located on chromosome
5 at 70220767-70249769 and 69345349-69374349 respectively.
In UCSC RefSeq, SMN1 is duplicated at both 70220767-70248838 and
69345349-69373418, so there are two gene entries named SMN1, each
with their own version of transcripts NM_000344, NM_001297715, NM_022874.
Similarly, SMN2 is duplicated at both 70220767-70248842 and 69345349-69373422.</p>
</div>
<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genes</span></span><a class="headerlink" href="#genome_kit.GenomeAnnotation.genes" title="Link to this definition"></a></dt>
<dd><p>An indexed table of annotated genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated genes.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.GeneTable" title="genome_kit.GeneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.transcripts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">transcripts</span></span><a class="headerlink" href="#genome_kit.GenomeAnnotation.transcripts" title="Link to this definition"></a></dt>
<dd><p>An indexed table of annotated transcripts.</p>
<p>A shorthand property <cite>trans</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated transcripts.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.TranscriptTable" title="genome_kit.TranscriptTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">TranscriptTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.exons">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exons</span></span><a class="headerlink" href="#genome_kit.GenomeAnnotation.exons" title="Link to this definition"></a></dt>
<dd><p>An indexed table of annotated exons.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated exons.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.ExonTable" title="genome_kit.ExonTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ExonTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.introns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">introns</span></span><a class="headerlink" href="#genome_kit.GenomeAnnotation.introns" title="Link to this definition"></a></dt>
<dd><p>An indexed table of annotated introns.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated introns.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.IntronTable" title="genome_kit.IntronTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IntronTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.cdss">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cdss</span></span><a class="headerlink" href="#genome_kit.GenomeAnnotation.cdss" title="Link to this definition"></a></dt>
<dd><p>An indexed table of annotated coding sequences (CDSs) within exons.</p>
<p>This is shorthand for <cite>annotation.cdss</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated coding sequences.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.CdsTable" title="genome_kit.CdsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CdsTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.utr5s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">utr5s</span></span><a class="headerlink" href="#genome_kit.GenomeAnnotation.utr5s" title="Link to this definition"></a></dt>
<dd><p>An indexed table of UTR5s within exons.</p>
<p>This is shorthand for <cite>annotation.utr5s</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated coding sequences.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">UtrTable</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.utr3s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">utr3s</span></span><a class="headerlink" href="#genome_kit.GenomeAnnotation.utr3s" title="Link to this definition"></a></dt>
<dd><p>An indexed table of UTR3s within exons.</p>
<p>This is shorthand for <cite>annotation.utr3s</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over annotated coding sequences.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">UtrTable</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.filename">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filename</span></span><a class="headerlink" href="#genome_kit.GenomeAnnotation.filename" title="Link to this definition"></a></dt>
<dd><p>The path to the file from which annotations are retrieved.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the file, e.g. “/data/gencode.v19.annotation.dganno”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.binary_version">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">binary_version</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GenomeAnnotation.binary_version"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeAnnotation.binary_version" title="Link to this definition"></a></dt>
<dd><p>The binary file format version that this build supports.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The version number.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.build_gencode">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_gencode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GenomeAnnotation.build_gencode"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeAnnotation.build_gencode" title="Link to this definition"></a></dt>
<dd><p>Build a binary version of GENCODE/Ensembl annotations from a GFF3 file.</p>
<p>Currently supports conversion from <cite>gff3[.gz]</cite>.</p>
<p>In addition to the versioned <cite>dganno</cite> file, a <cite>cfg</cite> file will also be created.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the original GFF3 file.</p></li>
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the destination <cite>dganno</cite> file.</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference assembly used to annotate <code class="docutils literal notranslate"><span class="pre">infile</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of built files created from the GFF3.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.build_ncbi_refseq">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_ncbi_refseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GenomeAnnotation.build_ncbi_refseq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeAnnotation.build_ncbi_refseq" title="Link to this definition"></a></dt>
<dd><p>Build a binary version of NCBI annotations from a GFF3 file.</p>
<p>Currently supports conversion from <cite>gff3[.gz]</cite>.</p>
<p>In addition to the versioned <cite>dganno</cite> file, a <cite>cfg</cite> file will also be created.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the original GFF3 file.</p></li>
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the destination <cite>dganno</cite> file.</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference assembly used to annotate <code class="docutils literal notranslate"><span class="pre">infile</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of built files created from the GFF3.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.build_ucsc_refseq">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_ucsc_refseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ucsc_db_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GenomeAnnotation.build_ucsc_refseq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeAnnotation.build_ucsc_refseq" title="Link to this definition"></a></dt>
<dd><p>Build a binary version of UCSC RefSeq annotations.</p>
<p>The <cite>ucsc_db_dir</cite> should be a directory that already contains
the following files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">refGene.txt[.gz]</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">refLink.txt[.gz]</span></code></p></li>
</ul>
<p>These files will be parsed and used to build a <cite>dganno</cite> file.</p>
<p>In addition to the versioned <cite>dganno</cite> file, a <cite>cfg</cite> file will also be created.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ucsc_db_dir</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to a directory with the UCSC RefSeq source files.</p></li>
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the destination <cite>dganno</cite> file.</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference assembly used  <code class="docutils literal notranslate"><span class="pre">ucsc_db_dir</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of built files created from the UCSC RefSeq files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GenomeAnnotation.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GenomeAnnotation.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GenomeAnnotation.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="gene">
<h3>Gene<a class="headerlink" href="#gene" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Gene">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Gene</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Gene" title="Link to this definition"></a></dt>
<dd><p>An annotated gene.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Gene.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Gene.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Gene.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Gene.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.Gene.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this gene.</p>
<p>Note that <a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this gene.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Gene.id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#genome_kit.Gene.id" title="Link to this definition"></a></dt>
<dd><p>The ID of this gene.</p>
<p>Returns either an Ensembl ID (for GENCODE) or a Entrez ID (for RefSeq)
depending on which annotation is being represented.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The ID of this transcript, e.g. “ENSG00000115275.11” or “7841”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Gene.name">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#genome_kit.Gene.name" title="Link to this definition"></a></dt>
<dd><p>The name of this gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The name of this gene, e.g. “BRCA1”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Gene.type">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#genome_kit.Gene.type" title="Link to this definition"></a></dt>
<dd><p>The type of this gene.</p>
<p>This represents the <cite>gene_type</cite> field in most GFF3 annotations.</p>
<p>For GENCODE/Ensembl/NCBI RefSeq, the result can be any standard <a class="reference external" href="https://www.gencodegenes.org/gencode_biotypes.html">GENCODE biotype</a></p>
<p>For UCSC RefSeq, the result is “protein_coding” if any transcript is
protein coding, and otherwise “non_coding”.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The type of this gene, e.g. “protein_coding” or “non_coding”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Gene.level">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">level</span></span><a class="headerlink" href="#genome_kit.Gene.level" title="Link to this definition"></a></dt>
<dd><p>The evidence level of this gene.</p>
<p>This represents the <cite>level</cite> field in most GFF3 annotations,
either 1, 2, or 3.</p>
<p>Available for <cite>gencode</cite> and <cite>ensembl</cite> only.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The evidence level of this gene.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Gene.transcripts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">transcripts</span></span><a class="headerlink" href="#genome_kit.Gene.transcripts" title="Link to this definition"></a></dt>
<dd><p>The transcripts for this gene.</p>
<p>A shorthand property <cite>trans</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of the list of annotated transcripts for this gene.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Gene.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Gene.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Gene.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="genetable">
<h3>GeneTable<a class="headerlink" href="#genetable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.GeneTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GeneTable</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all genes that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All genes that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all genes that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All genes that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all genes that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All genes that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all genes that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All genes that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all genes that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All genes that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all genes that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All genes that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all genes that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All genes that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.GeneTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.GeneTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Lookup a gene by index or ID string.</p>
<p>If <cite>index</cite> is a string the gene with matching <code class="docutils literal notranslate"><span class="pre">id</span></code> is returned (by linear search):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="s1">&#39;ENSG00000115275&#39;</span><span class="p">]</span>  <span class="c1"># MOGS gene</span>
</pre></div>
</div>
<p>If the ID string has a version suffix (<code class="docutils literal notranslate"><span class="pre">'ENSG00000115275.11'</span></code>) then the match must be exact.</p>
<p>One can also iterate over all genes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The results will be ordered as in the source file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The integer index or ID string of the requested gene.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The gene object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.first_by_name">
<span class="sig-name descname"><span class="pre">first_by_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#genome_kit.Gene" title="genome_kit.genome_annotation.Gene"><span class="pre">Gene</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><span class="pre">None</span></a></span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.first_by_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.first_by_name" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The first gene with the specified name or None if none matched.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.GeneTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#GeneTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.GeneTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="transcript">
<h3>Transcript<a class="headerlink" href="#transcript" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Transcript">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Transcript</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Transcript"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Transcript" title="Link to this definition"></a></dt>
<dd><p>An annotated transcript.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Transcript.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Transcript.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Transcript.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.Transcript.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this transcript.</p>
<p>Note that <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#genome_kit.Transcript.id" title="Link to this definition"></a></dt>
<dd><p>The ID of this transcript.</p>
<p>Returns either an Ensembl ID or a RefSeq ID depending on which
annotation is being represented.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The ID of this transcript, e.g. “ENST00000233616.8” or “NM_006302”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.tsl">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">tsl</span></span><a class="headerlink" href="#genome_kit.Transcript.tsl" title="Link to this definition"></a></dt>
<dd><p>The support level of this transcript.</p>
<ul class="simple">
<li><p>tsl1: all splice junctions supported by at least one non-suspect mRNA</p></li>
<li><p>tsl2: best supporting mRNA is flagged as suspect or support from multiple ESTs</p></li>
<li><p>tsl3: only support is from a single EST</p></li>
<li><p>tsl4: best supporting EST is flagged as suspect</p></li>
<li><p>tsl5: no single transcript supports the model structure</p></li>
</ul>
<p>See <a class="reference external" href="http://www.ensembl.org/Help/Glossary?id=492">Ensemble transcript support level</a>.</p>
<p>Available for <cite>gencode</cite> only.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The support level of this transcript, where values 1..5 and 6 correspond to
GENCODE identifiers [tsl1, tsl2, tsl3, tsl4, tsl5, tslNA]</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.type">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#genome_kit.Transcript.type" title="Link to this definition"></a></dt>
<dd><p>The type of this transcript.</p>
<p>This represents the <cite>transcript_type</cite> field in most GFF3 annotations,
which may differ from the parent gene’s <cite>type</cite>.</p>
<p>For GENCODE/Ensembl/NCBI RefSeq, the result can be any standard <a class="reference external" href="https://www.gencodegenes.org/gencode_biotypes.html">GENCODE biotype</a></p>
<p>For UCSC RefSeq, the result is either “protein_coding” or “non_coding”.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The type of this transcript, e.g. “protein_coding” or “non_coding”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.level">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">level</span></span><a class="headerlink" href="#genome_kit.Transcript.level" title="Link to this definition"></a></dt>
<dd><p>The evidence level of this transcript.</p>
<p>This represents the <cite>level</cite> field in most GFF3 annotations,
either 1, 2, or 3.</p>
<p>Available for <cite>gencode</cite> and <cite>ensembl</cite> only.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The evidence level of this transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.ccds_id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ccds_id</span></span><a class="headerlink" href="#genome_kit.Transcript.ccds_id" title="Link to this definition"></a></dt>
<dd><p>The CCDSID of this transcript’s coding sequence, if applicable.</p>
<p>Available for <cite>gencode</cite> only.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The CCDSID of this transcript, e.g. “CCDS53759.1”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.protein_id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">protein_id</span></span><a class="headerlink" href="#genome_kit.Transcript.protein_id" title="Link to this definition"></a></dt>
<dd><p>The protein ID of this transcript, if applicable.</p>
<p>Available for <cite>gencode</cite>, <cite>ensembl</cite>, <cite>ucsc_refseq</cite>, and <cite>ncbi_refseq</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The protein ID of this transcript, e.g. “ENSP00000233616.4” or “NP_006293”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.product">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">product</span></span><a class="headerlink" href="#genome_kit.Transcript.product" title="Link to this definition"></a></dt>
<dd><p>A description of the transcript’s product.</p>
<p>Available for <cite>ucsc_refseq</cite> and <cite>ncbi_refseq</cite> only.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The product of this transcript, e.g. “breast cancer type 1 susceptibility protein isoform 2”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.gene">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gene</span></span><a class="headerlink" href="#genome_kit.Transcript.gene" title="Link to this definition"></a></dt>
<dd><p>The parent gene of this transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The parent gene for this transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.exons">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exons</span></span><a class="headerlink" href="#genome_kit.Transcript.exons" title="Link to this definition"></a></dt>
<dd><p>The exons for this transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of the list of annotated exons for this transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.introns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">introns</span></span><a class="headerlink" href="#genome_kit.Transcript.introns" title="Link to this definition"></a></dt>
<dd><p>The introns for this transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of the list of annotated introns for this transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.cdss">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cdss</span></span><a class="headerlink" href="#genome_kit.Transcript.cdss" title="Link to this definition"></a></dt>
<dd><p>The coding sequences (CDSs) for this transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of the list of annotated coding sequences for this transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.utr5s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">utr5s</span></span><a class="headerlink" href="#genome_kit.Transcript.utr5s" title="Link to this definition"></a></dt>
<dd><p>The UTR5 segments for this transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of the list of annotated UTR5 segments for this transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <code class="xref py py-class docutils literal notranslate"><span class="pre">Utr</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Transcript.utr3s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">utr3s</span></span><a class="headerlink" href="#genome_kit.Transcript.utr3s" title="Link to this definition"></a></dt>
<dd><p>The UTR3 segments for this transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of the list of annotated UTR3 segments for this transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <code class="xref py py-class docutils literal notranslate"><span class="pre">Utr</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Transcript.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Transcript.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Transcript.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="transcripttable">
<h3>TranscriptTable<a class="headerlink" href="#transcripttable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TranscriptTable</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all transcripts that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tran</span> <span class="k">for</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">transcripts</span> <span class="k">if</span> <span class="n">tran</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All transcripts that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all transcripts that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tran</span> <span class="k">for</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">transcripts</span> <span class="k">if</span> <span class="n">tran</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All transcripts that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all transcripts that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tran</span> <span class="k">for</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">transcripts</span> <span class="k">if</span> <span class="n">tran</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All transcripts that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all transcripts that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tran</span> <span class="k">for</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">transcripts</span> <span class="k">if</span> <span class="n">tran</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All transcripts that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all transcripts that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tran</span> <span class="k">for</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">transcripts</span> <span class="k">if</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All transcripts that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all transcripts that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tran</span> <span class="k">for</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">transcripts</span> <span class="k">if</span> <span class="n">tran</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All transcripts that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all transcripts that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tran</span> <span class="k">for</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">transcripts</span> <span class="k">if</span> <span class="n">tran</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All transcripts that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.TranscriptTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Lookup a transcript by index or ID string.</p>
<p>If <cite>index</cite> is a string the transcript with matching <code class="docutils literal notranslate"><span class="pre">id</span></code> is returned (by linear search):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tran</span> <span class="o">=</span> <span class="n">transcripts</span><span class="p">[</span><span class="s1">&#39;ENST00000233616&#39;</span><span class="p">]</span>  <span class="c1"># MOGS transcript</span>
</pre></div>
</div>
<p>If the ID string has a version suffix (<code class="docutils literal notranslate"><span class="pre">'ENST00000233616.8'</span></code>) then the match must be exact.</p>
<p>One can also iterate over all transcripts:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">tran</span> <span class="ow">in</span> <span class="n">transcripts</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The results will be ordered as in the source file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The integer index or ID string of the requested transcript.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The transcript object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.TranscriptTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#TranscriptTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.TranscriptTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="exon">
<h3>Exon<a class="headerlink" href="#exon" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Exon">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Exon</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Exon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Exon" title="Link to this definition"></a></dt>
<dd><p>An annotated exon.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<p>Each <cite>Exon</cite> object is equal only to itself. To build a set of unique exonic
intervals, use the <a class="reference internal" href="#genome_kit.Exon.interval" title="genome_kit.Exon.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">interval</span></code></a> attribute:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genome_kit</span> <span class="kn">import</span> <span class="n">Genome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="s1">&#39;gencode.v19&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exonic_intervals</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">exon</span><span class="o">.</span><span class="n">interval</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">exons</span><span class="p">])</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Exon.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Exon.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Exon.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.Exon.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this exon.</p>
<p>Note that <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this exon.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#genome_kit.Exon.id" title="Link to this definition"></a></dt>
<dd><p>The ID of this exon.</p>
<p>Available for <cite>gencode</cite> and <cite>ensembl</cite> only.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The ID of this exon, e.g. “ENSE00000846737.3”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.index">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">index</span></span><a class="headerlink" href="#genome_kit.Exon.index" title="Link to this definition"></a></dt>
<dd><p>The index of this exon within the transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The index of the exon number within the parent transcript (0-based).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.transcript">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">transcript</span></span><a class="headerlink" href="#genome_kit.Exon.transcript" title="Link to this definition"></a></dt>
<dd><p>The parent transcript of this exon.</p>
<p>A shorthand property <cite>tran</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The parent transcript of this exon.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.cds">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cds</span></span><a class="headerlink" href="#genome_kit.Exon.cds" title="Link to this definition"></a></dt>
<dd><p>The CDS on this exon, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The CDS contained within this exon, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.utr5">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">utr5</span></span><a class="headerlink" href="#genome_kit.Exon.utr5" title="Link to this definition"></a></dt>
<dd><p>The UTR5 on this exon, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The UTR5 contained within this exon, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Utr5</span></code> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.utr3">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">utr3</span></span><a class="headerlink" href="#genome_kit.Exon.utr3" title="Link to this definition"></a></dt>
<dd><p>The UTR3 on this exon, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The UTR3 contained within this exon, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Utr3</span></code> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.next_exon">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">next_exon</span></span><a class="headerlink" href="#genome_kit.Exon.next_exon" title="Link to this definition"></a></dt>
<dd><p>The next exon in the transcript, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The next annotated exon on the parent transcript, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.prev_exon">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prev_exon</span></span><a class="headerlink" href="#genome_kit.Exon.prev_exon" title="Link to this definition"></a></dt>
<dd><p>The previous exon in the transcript, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The previous annotated exon on the parent transcript, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.prev_intron">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prev_intron</span></span><a class="headerlink" href="#genome_kit.Exon.prev_intron" title="Link to this definition"></a></dt>
<dd><p>Returns the previous intron in the transcript, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The previous annotated intron on the parent transcript, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Exon.next_intron">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">next_intron</span></span><a class="headerlink" href="#genome_kit.Exon.next_intron" title="Link to this definition"></a></dt>
<dd><p>Returns the next intron in the transcript, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The next annotated intron on the parent transcript, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Exon.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Exon.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Exon.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="exontable">
<h3>ExonTable<a class="headerlink" href="#exontable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.ExonTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ExonTable</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all exons that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">exon</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span> <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All exons that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all exons that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">exon</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span> <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All exons that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all exons that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">exon</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span> <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All exons that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all exons that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">exon</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span> <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All exons that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all exons that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">exon</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span> <span class="k">if</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All exons that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all exons that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">exon</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span> <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All exons that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all exons that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">exon</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span> <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All exons that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.ExonTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.ExonTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Lookup an exon by index.</p>
<p>Allows iteration over all exons:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The results will be ordered as in the source file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested exon.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The exon object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ExonTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#ExonTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ExonTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="intron">
<h3>Intron<a class="headerlink" href="#intron" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Intron">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Intron</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Intron"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Intron" title="Link to this definition"></a></dt>
<dd><p>An annotated intron.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<p>Each <cite>Intron</cite> object is equal only to itself. To build a set of unique intronic
intervals, use the <a class="reference internal" href="#genome_kit.Intron.interval" title="genome_kit.Intron.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">interval</span></code></a> attribute:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genome_kit</span> <span class="kn">import</span> <span class="n">Genome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="s1">&#39;gencode.v19&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intronic_intervals</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">intron</span><span class="o">.</span><span class="n">interval</span> <span class="k">for</span> <span class="n">intron</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">introns</span><span class="p">])</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Intron.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Intron.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Intron.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Intron.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.Intron.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this intron.</p>
<p>Note that <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this intron.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Intron.index">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">index</span></span><a class="headerlink" href="#genome_kit.Intron.index" title="Link to this definition"></a></dt>
<dd><p>The index of the intron within the transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The index of the intron within the parent transcript (0-based).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Intron.transcript">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">transcript</span></span><a class="headerlink" href="#genome_kit.Intron.transcript" title="Link to this definition"></a></dt>
<dd><p>The parent transcript of this intron.</p>
<p>A shorthand property <cite>tran</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The parent transcript of this intron.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Intron.next_intron">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">next_intron</span></span><a class="headerlink" href="#genome_kit.Intron.next_intron" title="Link to this definition"></a></dt>
<dd><p>The next intron in the transcript, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The next annotated intron on the parent transcript, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Intron.prev_intron">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prev_intron</span></span><a class="headerlink" href="#genome_kit.Intron.prev_intron" title="Link to this definition"></a></dt>
<dd><p>The previous intron in the transcript, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The previous annotated intron on the parent transcript, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Intron.prev_exon">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prev_exon</span></span><a class="headerlink" href="#genome_kit.Intron.prev_exon" title="Link to this definition"></a></dt>
<dd><p>Returns the previous exon in the transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The previous annotated exon on the parent transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Intron.next_exon">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">next_exon</span></span><a class="headerlink" href="#genome_kit.Intron.next_exon" title="Link to this definition"></a></dt>
<dd><p>Returns the next exon in the transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The next annotated exon on the parent transcript.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Intron.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Intron.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Intron.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="introntable">
<h3>IntronTable<a class="headerlink" href="#introntable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.IntronTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">IntronTable</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all introns that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">intr</span> <span class="k">for</span> <span class="n">intr</span> <span class="ow">in</span> <span class="n">introns</span> <span class="k">if</span> <span class="n">intr</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All introns that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all introns that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">intr</span> <span class="k">for</span> <span class="n">intr</span> <span class="ow">in</span> <span class="n">introns</span> <span class="k">if</span> <span class="n">intr</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All introns that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all introns that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">intr</span> <span class="k">for</span> <span class="n">intr</span> <span class="ow">in</span> <span class="n">introns</span> <span class="k">if</span> <span class="n">intr</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All introns that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all introns that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">intr</span> <span class="k">for</span> <span class="n">intr</span> <span class="ow">in</span> <span class="n">introns</span> <span class="k">if</span> <span class="n">intr</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All introns that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all introns that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">intr</span> <span class="k">for</span> <span class="n">intr</span> <span class="ow">in</span> <span class="n">introns</span> <span class="k">if</span> <span class="n">intr</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All introns that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all introns that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">intr</span> <span class="k">for</span> <span class="n">intr</span> <span class="ow">in</span> <span class="n">introns</span> <span class="k">if</span> <span class="n">intr</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All introns that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all introns that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">intr</span> <span class="k">for</span> <span class="n">intr</span> <span class="ow">in</span> <span class="n">introns</span> <span class="k">if</span> <span class="n">intr</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All introns that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.IntronTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.IntronTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Lookup an intron by index.</p>
<p>Allows iteration over all introns:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">intron</span> <span class="ow">in</span> <span class="n">introns</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The results will be ordered as in the source file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested intron.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The intron object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Intron" title="genome_kit.Intron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Intron</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.IntronTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#IntronTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.IntronTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="cds">
<h3>Cds<a class="headerlink" href="#cds" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Cds">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Cds</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Cds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Cds" title="Link to this definition"></a></dt>
<dd><p>An annotated coding sequence (CDS) element within an exon.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Cds.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Cds.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Cds.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Cds.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.Cds.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this CDS.</p>
<p>Note that <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this CDS.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Cds.phase">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">phase</span></span><a class="headerlink" href="#genome_kit.Cds.phase" title="Link to this definition"></a></dt>
<dd><p>The phase of this CDS’s start position.</p>
<p>The <cite>phase</cite> is the number of bases that should be trimmed from the 5’ end of the
CDS element to reach the start of next codon:</p>
<ul class="simple">
<li><p>phase 0 means the 5p end is in-frame</p></li>
<li><p>phase 1 means the 5p end is 1 bp away from being in-frame.</p></li>
<li><p>phase 2 means the 5p end is 2 bp away from being in-frame.</p></li>
</ul>
<p>See the <cite>GFF3 documentation &lt;http://gmod.org/wiki/GFF3&gt;</cite> for a precise explanation.</p>
<p>The <cite>frame</cite> attribute in UCSC RefSeq is not available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The phase this CDS, either 0, 1, or 2.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Cds.exon">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exon</span></span><a class="headerlink" href="#genome_kit.Cds.exon" title="Link to this definition"></a></dt>
<dd><p>The parent exon of this CDS.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The parent exon of this CDS.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Exon" title="genome_kit.Exon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Exon</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Cds.transcript">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">transcript</span></span><a class="headerlink" href="#genome_kit.Cds.transcript" title="Link to this definition"></a></dt>
<dd><p>The parent transcript of this CDS.</p>
<p>A shorthand property <cite>tran</cite> is also available.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The parent transcript of this CDS.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Transcript" title="genome_kit.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">Transcript</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Cds.next_cds">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">next_cds</span></span><a class="headerlink" href="#genome_kit.Cds.next_cds" title="Link to this definition"></a></dt>
<dd><p>The next CDS in the transcript, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The next annotated CDS on the parent transcript, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Cds.prev_cds">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prev_cds</span></span><a class="headerlink" href="#genome_kit.Cds.prev_cds" title="Link to this definition"></a></dt>
<dd><p>The previous CDS in the transcript, if any.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The previous annotated CDS on the parent transcript, if any.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Cds.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#Cds.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Cds.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="cdstable">
<h3>CdsTable<a class="headerlink" href="#cdstable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.CdsTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">CdsTable</span></span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all CDSs that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cds</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span> <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All coding sequences (CDSs) that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all CDSs that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cds</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span> <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All coding sequences (CDSs) that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all CDSs that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cds</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span> <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All coding sequences (CDSs) that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all CDSs that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cds</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span> <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All coding sequences (CDSs) that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all CDSs that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cds</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span> <span class="k">if</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All coding sequences (CDSs) that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all CDSs that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cds</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span> <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All coding sequences (CDSs) that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all CDSs that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cds</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span> <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All coding sequences (CDSs) that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.CdsTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.CdsTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Lookup a coding sequence element by index.</p>
<p>Allows iteration over all coding sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested CDS.</p></li>
<li><p><strong>file.</strong> (<em>The results will be ordered as in the source</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The CDS identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Cds" title="genome_kit.Cds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cds</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.CdsTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/genome_annotation.html#CdsTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.CdsTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="read-alignments">
<h2>Read Alignments<a class="headerlink" href="#read-alignments" title="Link to this heading"></a></h2>
<section id="readalignments">
<h3>ReadAlignments<a class="headerlink" href="#readalignments" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ReadAlignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#ReadAlignments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadAlignments" title="Link to this definition"></a></dt>
<dd><p>Access to read alignments.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#ReadAlignments.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadAlignments.__init__" title="Link to this definition"></a></dt>
<dd><p>Open a .ralign file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>infile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The .ralign file to open.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IOError" title="(in Python v3.7)"><strong>IOError</strong></a> – Occurs when <cite>infile</cite> refers to an incompatible ralign file for this version of <code class="docutils literal notranslate"><span class="pre">genome_kit</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.junctions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">junctions</span></span><a class="headerlink" href="#genome_kit.ReadAlignments.junctions" title="Link to this definition"></a></dt>
<dd><p>Access to junctions inferred from read alignments</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A table with optimized methods to query over all junctions by position.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.JunctionTable" title="genome_kit.JunctionTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">alignments</span></span><a class="headerlink" href="#genome_kit.ReadAlignments.alignments" title="Link to this definition"></a></dt>
<dd><p>Access to read alignments, i.e. a row in BAM file</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A table with optimized methods to query over all read alignments by position.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.AlignmentTable" title="genome_kit.AlignmentTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.matches">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matches</span></span><a class="headerlink" href="#genome_kit.ReadAlignments.matches" title="Link to this definition"></a></dt>
<dd><p>Access to aligned regions of read alignments</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A table with optimized methods to query over all aligned regions by position.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.AlignmentMatchTable" title="genome_kit.AlignmentMatchTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatchTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.variants">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">variants</span></span><a class="headerlink" href="#genome_kit.ReadAlignments.variants" title="Link to this definition"></a></dt>
<dd><p>Access to variants observed by read alignments</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A table where each element (each row) is a variant that was observed
in at least one read alignment.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.VariantTable" title="genome_kit.VariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.filename">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filename</span></span><a class="headerlink" href="#genome_kit.ReadAlignments.filename" title="Link to this definition"></a></dt>
<dd><p>The path to the file from which read alignments are retrieved.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the file, e.g. “/mnt/data/sample000132.ralign”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#ReadAlignments.close"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadAlignments.close" title="Link to this definition"></a></dt>
<dd><p>Close the file handle.</p>
<p>Note that if you use a <cite>with</cite> statement the file handle is automatically closed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Open the file</span>
<span class="k">with</span> <span class="n">ReadAlignments</span><span class="p">(</span><span class="s1">&#39;my_file.ralign&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">raligns</span><span class="p">:</span>
    <span class="o">...</span>

<span class="c1"># &lt;-- File is now closed</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.ralign_version">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ralign_version</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#ReadAlignments.ralign_version"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadAlignments.ralign_version" title="Link to this definition"></a></dt>
<dd><p>The <cite>ralign</cite> file format version that this build supports.</p>
<p>Attempting to open an <cite>ralign</cite> file of a mismatched version will raise an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IOError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IOError</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The version number.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.build_ralign">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_ralign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_duplicates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">library_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'U'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#ReadAlignments.build_ralign"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadAlignments.build_ralign" title="Link to this definition"></a></dt>
<dd><p>Build an <cite>ralign</cite> file from one or more <cite>SAM</cite> files.</p>
<p>If multiple input files are specified, their junctions and alignments will be pooled.</p>
<p>Once an <cite>ralign</cite> file is created, it can be opened by creating
a <a class="reference internal" href="#genome_kit.ReadAlignments" title="genome_kit.ReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadAlignments</span></code></a> object.</p>
<p><strong>Alignments on the reference genome.</strong>
All junction, alignment, match, and variant intervals in the resulting file
are on the reference genome.</p>
<p>For example, an alignment starting at reference genome position 2
and with CIGAR string <code class="docutils literal notranslate"><span class="pre">1M1D1M2N1M</span></code>
is matched to the reference genome as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">4</span> <span class="mi">5</span> <span class="mi">6</span> <span class="mi">7</span> <span class="mi">8</span> <span class="mi">9</span>     <span class="o">&lt;--</span> <span class="n">reference</span> <span class="n">genome</span> <span class="n">coordinates</span>
    <span class="n">M</span> <span class="n">D</span> <span class="n">M</span> <span class="o">-</span> <span class="o">-</span> <span class="n">M</span> <span class="n">M</span>       <span class="o">&lt;--</span> <span class="n">M</span> <span class="o">=</span> <span class="n">match</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="n">deletion</span>
</pre></div>
</div>
<p>This results in one alignment (2:9), two matches (2:5, 7:9),
one junction (5:7), and one variant (3:4).</p>
<p>Similarly for insertions, with CIGAR string <code class="docutils literal notranslate"><span class="pre">1M1I1M2N1M</span></code> aligns as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">4</span> <span class="mi">5</span> <span class="mi">6</span> <span class="mi">7</span> <span class="mi">8</span> <span class="mi">9</span>
    <span class="n">M</span> <span class="n">M</span> <span class="o">-</span> <span class="o">-</span> <span class="n">M</span> <span class="n">M</span>
     <span class="o">^</span>
     <span class="n">I</span>                  <span class="o">&lt;--</span> <span class="n">I</span> <span class="o">=</span> <span class="n">insertion</span> <span class="n">between</span>
</pre></div>
</div>
<p>This results in one alignment (2:8), two matches (2:4, 6:8),
one junction (4:6), and one variant (3:3).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the destination <cite>rdist</cite> file.</p></li>
<li><p><strong>infiles</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <code class="xref py py-class docutils literal notranslate"><span class="pre">file</span></code>) – The paths to the source <cite>SAM</cite> files, or a reference to <code class="docutils literal notranslate"><span class="pre">sys.stdin</span></code>.
Streaming lines from stdin is useful for reading directly from BAM
via <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">view</span> <span class="pre">-h</span></code>. Be sure to use <code class="docutils literal notranslate"><span class="pre">-h</span></code> so that
the reference genome can be inferred from the header lines.</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome of the data in <code class="docutils literal notranslate"><span class="pre">infiles</span></code>.
If the genome contains annotations, junction strands will be
inferred and validated against the introns annotations.</p></li>
<li><p><strong>exclude</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>, optional) – Junctions within these intervals will be excluded.</p></li>
<li><p><strong>allow</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>, optional) – Only junctions within these intervals will be included,
so long as they are not excluded.</p></li>
<li><p><strong>include_duplicates</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, optional) – Includes duplicates when <cite>True</cite>. Defaults to <cite>False</cite>. See
<a class="reference external" href="https://github.com/samtools/hts-specs/blob/master/SAMv1.pdf">https://github.com/samtools/hts-specs/blob/master/SAMv1.pdf</a>
subsection 1.4.2</p></li>
<li><p><strong>library_format</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, optional) – Specifies to infer junction strands via the library format string.
“SF” implies the first read is the sense strand, whereas “SR”
implies the second read is the sense strand. (the default is ‘U’,
which represents unstranded [no strand inference]).
See <a class="reference external" href="https://salmon.readthedocs.io/en/stable/library_type.html">https://salmon.readthedocs.io/en/stable/library_type.html</a>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.pileup">
<span class="sig-name descname"><span class="pre">pileup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=&lt;class</span> <span class="pre">'numpy.int32'&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#ReadAlignments.pileup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadAlignments.pileup" title="Link to this definition"></a></dt>
<dd><p>Returns the DNA sequence pileup for body reads in this <cite>ReadAlignments</cite>
as a 5-track. The pileup is variant-aware, so the track counts
the alternate sequences, instead of the reference, on any alignment
matches with variants.</p>
<p>Currently, only acgtACGT substitutions and deletions variants are
supported.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – A positive interval to query over all the read alignments.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dtype</span></code></a>, optional) – The data type for the returned counts.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An array of <code class="docutils literal notranslate"><span class="pre">shape</span></code> (len(<cite>interval</cite>), 5) where the second
dimension is indexed via <code class="docutils literal notranslate"><span class="pre">PILEUP_A</span></code>, <code class="docutils literal notranslate"><span class="pre">PILEUP_C</span></code>, <code class="docutils literal notranslate"><span class="pre">PILEUP_G</span></code>,
<code class="docutils literal notranslate"><span class="pre">PILEUP_T</span></code>, and <code class="docutils literal notranslate"><span class="pre">PILEUP_DEL</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadAlignments.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#ReadAlignments.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadAlignments.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="junction">
<h3>Junction<a class="headerlink" href="#junction" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Junction">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Junction</span></span><a class="reference internal" href="_modules/genome_kit/ralign.html#Junction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Junction" title="Link to this definition"></a></dt>
<dd><p>A Junction represents a gap in a read alignment when compared to the reference genome.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Junction.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#Junction.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Junction.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Junction.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.Junction.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this junction.</p>
<p>Note that <a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this junction.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Junction.alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">alignments</span></span><a class="headerlink" href="#genome_kit.Junction.alignments" title="Link to this definition"></a></dt>
<dd><p>The read alignments across this junction.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The read alignments across this junction.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#tuple" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a> of <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Junction.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#Junction.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Junction.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="junctiontable">
<h3>JunctionTable<a class="headerlink" href="#junctiontable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.JunctionTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">JunctionTable</span></span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junction that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.JunctionTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Lookup a junction by index</p>
<p>Allows iteration over all junctions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested junction.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The junction object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Junction" title="genome_kit.Junction"><code class="xref py py-class docutils literal notranslate"><span class="pre">Junction</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#JunctionTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="alignment">
<h3>Alignment<a class="headerlink" href="#alignment" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.Alignment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Alignment</span></span><a class="reference internal" href="_modules/genome_kit/ralign.html#Alignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Alignment" title="Link to this definition"></a></dt>
<dd><p>A read alignment</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Alignment.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#Alignment.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Alignment.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Alignment.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.Alignment.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this read alignment.</p>
<p>Note that <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this read alignment.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.Alignment.matches">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matches</span></span><a class="headerlink" href="#genome_kit.Alignment.matches" title="Link to this definition"></a></dt>
<dd><p>The matching regions of this read alignment</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A list of matching regions</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.Alignment.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#Alignment.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.Alignment.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="alignmenttable">
<h3>AlignmentTable<a class="headerlink" href="#alignmenttable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">AlignmentTable</span></span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all read alignments that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">read</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All read alignments that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all read alignments that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">read</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All read alignments that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all read alignments that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">read</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All read alignments that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all read alignments that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">read</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All read alignments that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all read alignments that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">read</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span> <span class="k">if</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All read alignment that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all read alignments that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">read</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All read alignments that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all read alignments that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">read</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All read alignments that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.AlignmentTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Lookup a read alignment by index</p>
<p>Allows iteration over all read alignments:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested read alignment.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The read alignment object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Alignment" title="genome_kit.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Alignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="alignmentmatch">
<h3>AlignmentMatch<a class="headerlink" href="#alignmentmatch" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatch">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">AlignmentMatch</span></span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatch" title="Link to this definition"></a></dt>
<dd><p>A matching region a read alignment</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatch.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatch.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatch.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatch.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.AlignmentMatch.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by the matching region of a read alignment.</p>
<p>Note that <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this alignment match.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatch.variants">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">variants</span></span><a class="headerlink" href="#genome_kit.AlignmentMatch.variants" title="Link to this definition"></a></dt>
<dd><p>A list of variants observed in this matching region of an alignment.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A tuple of variants that this matching region of an alignment contains</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#tuple" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatch.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatch.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatch.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="alignmentmatchtable">
<h3>AlignmentMatchTable<a class="headerlink" href="#alignmentmatchtable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">AlignmentMatchTable</span></span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all alignment matches that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">match</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">match</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All alignment matches that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all alignment matches that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">match</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">match</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All alignment matches that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all alignment matches that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">match</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">match</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All alignment matches that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all alignment matches that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">match</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">match</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All alignments matches that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all alignment matches that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">match</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All alignment matches that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all alignment matches that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">match</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">match</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All alignment matches that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all alignment matches that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">match</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">match</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All alignment matches that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.AlignmentMatchTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Lookup an alignment match by index</p>
<p>Allows iteration over all alignment matches:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested alignment match.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The alignment match object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.AlignmentMatch" title="genome_kit.AlignmentMatch"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentMatch</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.AlignmentMatchTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/ralign.html#AlignmentMatchTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.AlignmentMatchTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="junction-read-alignments">
<h2>Junction Read Alignments<a class="headerlink" href="#junction-read-alignments" title="Link to this heading"></a></h2>
<section id="jreadalignments">
<h3>JReadAlignments<a class="headerlink" href="#jreadalignments" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">JReadAlignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JReadAlignments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JReadAlignments" title="Link to this definition"></a></dt>
<dd><p>Access to read alignments.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JReadAlignments.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JReadAlignments.__init__" title="Link to this definition"></a></dt>
<dd><p>Open a .jralign file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>infile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The .jralign file to open.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments.junctions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">junctions</span></span><a class="headerlink" href="#genome_kit.JReadAlignments.junctions" title="Link to this definition"></a></dt>
<dd><p>Access to read alignments that are split across junctions.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over junctions, where
each junction provides a read distribution.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.JunctionReadAlignmentsTable" title="genome_kit.JunctionReadAlignmentsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignmentsTable</span></code></a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The returned table. junctions, and child objects should not be accessed
after this JReadAlignments has been closed via <a class="reference internal" href="#genome_kit.JReadAlignments.close" title="genome_kit.JReadAlignments.close"><code class="xref py py-meth docutils literal notranslate"><span class="pre">close()</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments.variants">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">variants</span></span><a class="headerlink" href="#genome_kit.JReadAlignments.variants" title="Link to this definition"></a></dt>
<dd><p>Access to variants observed over all read alignments</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A table where each element (each row) is a variant that was observed
in at least one read alignment.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.VariantTable" title="genome_kit.VariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantTable</span></code></a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The returned table and variants should not be accessed after this
JReadAlignments has been closed via <a class="reference internal" href="#genome_kit.JReadAlignments.close" title="genome_kit.JReadAlignments.close"><code class="xref py py-meth docutils literal notranslate"><span class="pre">close()</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments.filename">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filename</span></span><a class="headerlink" href="#genome_kit.JReadAlignments.filename" title="Link to this definition"></a></dt>
<dd><p>The path to the file from which read alignments are retrieved.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the file, e.g. “/mnt/data/sample000132.jralign”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JReadAlignments.close"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JReadAlignments.close" title="Link to this definition"></a></dt>
<dd><p>Close the file handle.</p>
<p>Note that if you use a <cite>with</cite> statement the file handle is automatically closed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Open the file</span>
<span class="k">with</span> <span class="n">JReadAlignments</span><span class="p">(</span><span class="s1">&#39;my_file.jralign&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">jraligns</span><span class="p">:</span>
    <span class="o">...</span>

<span class="c1"># &lt;-- File is now closed</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>Closing this JReadAlignments will invalidate all tables and objects previously accessed via
<a class="reference internal" href="#genome_kit.JReadAlignments.junctions" title="genome_kit.JReadAlignments.junctions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">junctions()</span></code></a> and <a class="reference internal" href="#genome_kit.JunctionReadAlignment.variants" title="genome_kit.JunctionReadAlignment.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments.jralign_version">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">jralign_version</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JReadAlignments.jralign_version"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JReadAlignments.jralign_version" title="Link to this definition"></a></dt>
<dd><p>The <cite>jralign</cite> file format version that this build supports.</p>
<p>Attempting to open an <cite>jralign</cite> file of a mismatched version will raise an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IOError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IOError</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The version number.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments.build_jralign">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_jralign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_reads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overhang</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_variants</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_duplicates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">library_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'U'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang_error</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'error'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JReadAlignments.build_jralign"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JReadAlignments.build_jralign" title="Link to this definition"></a></dt>
<dd><p>Build an <cite>jralign</cite> file from one or more <cite>SAM</cite> files.</p>
<p>If multiple input files are specified, their junctions and read alignments will be pooled.</p>
<p>Once an <cite>jralign</cite> file is created, it can be opened by creating
a <a class="reference internal" href="#genome_kit.JReadAlignments" title="genome_kit.JReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JReadAlignments</span></code></a> object.</p>
<p>Below are some details on how SAM alignments are processed, which are
important for understanding how a CIGAR strings containing insertions/deletions
(<cite>5M8D5M</cite>) or multiple splits (<cite>5M100N10M50N5M</cite>) are processed:</p>
<ul>
<li><p><strong>Multi-split alignments.</strong>
If a read maps across multiple junctions, its CIGAR string will
contain multiple <code class="docutils literal notranslate"><span class="pre">N</span></code> blocks. The read will then contribute a
separate (left, right) alignment to <em>each</em> junction that it spans.</p>
<p>For example, consider <code class="docutils literal notranslate"><span class="pre">1S2M2N1M1N1M</span></code> where the first match is
aligned to start position 2:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">4</span> <span class="mi">5</span> <span class="mi">6</span> <span class="mi">7</span> <span class="mi">8</span> <span class="mi">9</span>
  <span class="n">S</span> <span class="n">M</span> <span class="n">M</span> <span class="o">-</span> <span class="o">-</span> <span class="n">M</span> <span class="o">-</span> <span class="n">M</span>
</pre></div>
</div>
<p>This read will contribute a separate alignment to each junction.
The <cite>left</cite> and <cite>right</cite> overhang is the sum of all alignment matches
before/after that particular junction.
In the above example, the resulting alignments would be
be as if two separate reads were aligned as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">4</span> <span class="mi">5</span> <span class="mi">6</span> <span class="mi">7</span> <span class="mi">8</span> <span class="mi">9</span>
  <span class="n">S</span> <span class="n">M</span> <span class="n">M</span> <span class="o">-</span> <span class="o">-</span> <span class="n">M</span> <span class="n">M</span>        <span class="o">&lt;--</span> <span class="n">alignment</span> <span class="k">for</span> <span class="mi">1</span><span class="n">st</span> <span class="n">junction</span>
      <span class="n">S</span> <span class="n">M</span> <span class="n">M</span> <span class="n">M</span> <span class="o">-</span> <span class="n">M</span>      <span class="o">&lt;--</span> <span class="n">alignment</span> <span class="k">for</span> <span class="mi">2</span><span class="n">nd</span> <span class="n">junction</span>
</pre></div>
</div>
<p>In this scenario, the read would not actually match the reference genome
in the above alignments, but this scheme facilitates counting of relative
junction usage (assuming usage of the two junctions is independent)
and also is allows positional bootstrap to be applied correctly
without modification.</p>
</li>
<li><p><strong>Overhang filter.</strong>
The <code class="docutils literal notranslate"><span class="pre">min_overhang</span></code> filter is applied on a per-junction bases.
For example, if <code class="docutils literal notranslate"><span class="pre">min_overhang=5</span></code> then a read mapped across
two junctions <code class="docutils literal notranslate"><span class="pre">20M100N10M100N3M</span></code> will contribute an alignment
only to the first, because the second has a right overhang of 3.</p></li>
<li><p><strong>Counting insertions/deletions.</strong>
Only alignment matches are counted. Other CIGAR types (e.g.,
insertions, deletions, clippings) are not relevant when filtering
for alignment quality.</p>
<p>For example, a read starting at reference genome position 2
and with CIGAR string <code class="docutils literal notranslate"><span class="pre">1M1D1M2N1M</span></code>
is matched to the reference genome as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">4</span> <span class="mi">5</span> <span class="mi">6</span> <span class="mi">7</span> <span class="mi">8</span> <span class="mi">9</span>     <span class="o">&lt;--</span> <span class="n">reference</span> <span class="n">genome</span> <span class="n">coordinates</span>
    <span class="n">M</span> <span class="n">D</span> <span class="n">M</span> <span class="o">-</span> <span class="o">-</span> <span class="n">M</span>         <span class="o">&lt;--</span> <span class="n">M</span> <span class="o">=</span> <span class="n">match</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="n">deletion</span>
</pre></div>
</div>
<p>so the start of the junction is 2+1+1+1 = 5, but the
the left overhang is only 2, because the deleted base
doesn’t exist in the private genome.</p>
<p>Similarly for insertions, with CIGAR string <code class="docutils literal notranslate"><span class="pre">1M1I1M2N1M</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>   <span class="mi">0</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">4</span> <span class="mi">5</span> <span class="mi">6</span> <span class="mi">7</span> <span class="mi">8</span> <span class="mi">9</span>
       <span class="n">M</span> <span class="n">M</span> <span class="o">-</span> <span class="o">-</span> <span class="n">M</span>
        <span class="o">^</span>
        <span class="n">I</span>                <span class="o">&lt;--</span> <span class="n">I</span> <span class="o">=</span> <span class="n">insertion</span> <span class="n">between</span>

<span class="n">the</span> <span class="n">start</span> <span class="n">of</span> <span class="n">the</span> <span class="n">junction</span> <span class="ow">is</span> <span class="mi">2</span><span class="o">+</span><span class="mi">1</span><span class="o">+</span><span class="mi">1</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">but</span> <span class="n">the</span> <span class="n">left</span>
<span class="n">overhang</span> <span class="ow">is</span> <span class="n">actually</span> <span class="mi">2</span><span class="p">,</span> <span class="n">because</span> <span class="n">the</span> <span class="n">inserted</span> <span class="n">base</span> <span class="n">exists</span>
<span class="n">only</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">private</span> <span class="n">genome</span><span class="o">.</span>
</pre></div>
</div>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the destination <cite>rdist</cite> file.</p></li>
<li><p><strong>infiles</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <code class="xref py py-class docutils literal notranslate"><span class="pre">file</span></code>) – The paths to the source <cite>SAM</cite> files, or a reference to <code class="docutils literal notranslate"><span class="pre">sys.stdin</span></code>.
Streaming lines from stdin is useful for reading directly from BAM
via <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">view</span> <span class="pre">-h</span></code>. Be sure to use <code class="docutils literal notranslate"><span class="pre">-h</span></code> so that
the reference genome can be inferred from the header lines.</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome of the data in <code class="docutils literal notranslate"><span class="pre">infiles</span></code>.
If the genome contains annotations, junction strands will be
inferred and validated against the introns annotations.</p></li>
<li><p><strong>min_reads</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The minimum number of reads across a junction to
be considered for inclusion in the output file.</p></li>
<li><p><strong>min_overhang</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The minimum overhang for a read to be included.</p></li>
<li><p><strong>exclude</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Optional. Junctions within these intervals will be excluded.</p></li>
<li><p><strong>allow</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Optional. Only junctions within these intervals will be included,
so long as they are not excluded.</p></li>
<li><p><strong>include_variants</strong> (<em>:py:class`bool`</em>) – Optional. True if variants are to be included in the output file.
Default will exclude all variants.</p></li>
<li><p><strong>include_duplicates</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, optional) – Includes duplicates when <cite>True</cite>. Defaults to <cite>False</cite>. See
<a class="reference external" href="https://github.com/samtools/hts-specs/blob/master/SAMv1.pdf">https://github.com/samtools/hts-specs/blob/master/SAMv1.pdf</a>
subsection 1.4.2</p></li>
<li><p><strong>library_format</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, optional) – Specifies to infer junction strands via the library format string.
“SF” implies the first read is the sense strand, whereas “SR”
implies the second read is the sense strand. (the default is ‘U’,
which represents unstranded [no strand inference]).
See <a class="reference external" href="https://salmon.readthedocs.io/en/stable/library_type.html">https://salmon.readthedocs.io/en/stable/library_type.html</a>.</p></li>
<li><p><strong>overhang_error</strong> (: <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Determines how build_jralign handles reads with &gt;255 length
overhangs. “error” stops and reports the error. “clamp” limits
the read to the maximum 255 positions on each side.
This only affects overhangs and will not clip or filter variants.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JReadAlignments.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JReadAlignments.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JReadAlignments.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="junctionreadalignmentstable">
<h3>JunctionReadAlignmentsTable<a class="headerlink" href="#junctionreadalignmentstable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">JunctionReadAlignmentsTable</span></span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Access to all junctions.</p>
<p>Allows iteration over all junctions in the read distribution table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The results will be ordered as in the source file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested junction.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The junction read alignment object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignmentsTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignmentsTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignmentsTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="junctionreadalignments">
<h3>JunctionReadAlignments<a class="headerlink" href="#junctionreadalignments" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignments">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">JunctionReadAlignments</span></span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignments" title="Link to this definition"></a></dt>
<dd><p>A set of read alignments across a junction.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignments.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignments.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignments.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignments.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.JunctionReadAlignments.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this junction.</p>
<p>Note: All junction intervals are on the reference strand, even if they comprise reads that were mapped to the
reverse strand.
When querying junctions, your query intervals must therefore (currently) be on the positive strand.</p>
<p>Note that <a class="reference internal" href="#genome_kit.JunctionReadAlignments" title="genome_kit.JunctionReadAlignments"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignments</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this junction.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignments.num_reads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_reads</span></span><a class="headerlink" href="#genome_kit.JunctionReadAlignments.num_reads" title="Link to this definition"></a></dt>
<dd><p>The number of reads across this junction.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of reads across this junction, equivalent to the <cite>len</cite> of this object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignments.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignments.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignments.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Access to a junction read alignment.</p>
<p>Allows iteration over all read alignments across this junction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">junction</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested alignment entry.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The junction read alignment identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.JunctionReadAlignment" title="genome_kit.JunctionReadAlignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadAlignment</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignments.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignments.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignments.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="junctionreadalignment">
<h3>JunctionReadAlignment<a class="headerlink" href="#junctionreadalignment" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">JunctionReadAlignment</span></span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignment" title="Link to this definition"></a></dt>
<dd><p>A junction read alignment.</p>
<p>The <cite>left</cite> and <cite>right</cite> overhangs are positive integers representing the number
of matched positions on either side of the junction.</p>
<p>Overhangs are counted on the private genome from which the data was sequenced,
and may not map to reference genome coordinates.
As such, one should not expect meaningful reference genome coordinates by
adding overhangs to the junction start/end.</p>
<p>If <cite>strand</cite> is positive, then the read started <cite>left</cite> positions before the
first intronic position on the private genome.
If <cite>strand</cite> is negative, then the read started
<cite>right</cite> positions after the last intronic position on the private genome.</p>
<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignment.left">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">left</span></span><a class="headerlink" href="#genome_kit.JunctionReadAlignment.left" title="Link to this definition"></a></dt>
<dd><p>The left overhang relative to the junction.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of matched positions left of the junction.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignment.right">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">right</span></span><a class="headerlink" href="#genome_kit.JunctionReadAlignment.right" title="Link to this definition"></a></dt>
<dd><p>The right overhang relative to the junction.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of matched positions right of the junction.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignment.strand">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#genome_kit.JunctionReadAlignment.strand" title="Link to this definition"></a></dt>
<dd><p>The strand (“+” or “-”) that this read was aligned to.</p>
<p>Not to be confused with the strand of the junction itself, which may or
may not have been inferred.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The strand the read was aligned to by the original aligner.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">chr</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignment.num_variants">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_variants</span></span><a class="headerlink" href="#genome_kit.JunctionReadAlignment.num_variants" title="Link to this definition"></a></dt>
<dd><p>The number of variants observed in this read alignment.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of variants in the alignment.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignment.variants">
<span class="sig-name descname"><span class="pre">variants</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignment.variants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignment.variants" title="Link to this definition"></a></dt>
<dd><p>A list of variants observed in this read alignment.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A tuple of variants that this read contains</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#tuple" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadAlignment.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jralign.html#JunctionReadAlignment.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadAlignment.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="readdistributions">
<h3>ReadDistributions<a class="headerlink" href="#readdistributions" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ReadDistributions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#ReadDistributions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadDistributions" title="Link to this definition"></a></dt>
<dd><p>Access to read distributions.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#ReadDistributions.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadDistributions.__init__" title="Link to this definition"></a></dt>
<dd><p>Open a .rdist file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>infile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The .rdist file to open.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions.junctions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">junctions</span></span><a class="headerlink" href="#genome_kit.ReadDistributions.junctions" title="Link to this definition"></a></dt>
<dd><p>Access to read distributions that are split across junctions.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object that supports looping or queries over junctions, where
each junction provides a read distribution.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.JunctionReadDistributionTable" title="genome_kit.JunctionReadDistributionTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistributionTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions.filename">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filename</span></span><a class="headerlink" href="#genome_kit.ReadDistributions.filename" title="Link to this definition"></a></dt>
<dd><p>The path to the file from which read distributions are retrieved.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the file, e.g. “/mnt/data/sample000132.rdist”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#ReadDistributions.close"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadDistributions.close" title="Link to this definition"></a></dt>
<dd><p>Close the file handle.</p>
<p>Note that if you use a <cite>with</cite> statement the file handle is automatically closed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Open the file</span>
<span class="k">with</span> <span class="n">ReadDistributions</span><span class="p">(</span><span class="s1">&#39;my_file.ralign&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">raligns</span><span class="p">:</span>
    <span class="o">...</span>

<span class="c1"># &lt;-- File is now closed</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions.rdist_version">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rdist_version</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#ReadDistributions.rdist_version"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadDistributions.rdist_version" title="Link to this definition"></a></dt>
<dd><p>The <cite>rdist</cite> file format version that this build supports.</p>
<p>Attempting to open an <cite>rdist</cite> file of a mismatched version will raise an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IOError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IOError</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The version number.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions.build_rdist">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_rdist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_reads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overhang</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pass_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#ReadDistributions.build_rdist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadDistributions.build_rdist" title="Link to this definition"></a></dt>
<dd><p>Build an <cite>rdist</cite> file from one or more <cite>jralign</cite> files.</p>
<p>If multiple input files are specified, their read counts will be pooled.</p>
<p>Once an <cite>rdist</cite> file is created, it can be opened by creating
a <a class="reference internal" href="#genome_kit.ReadDistributions" title="genome_kit.ReadDistributions"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadDistributions</span></code></a> object .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the destination <cite>rdist</cite> file.</p></li>
<li><p><strong>infiles</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The paths to the source <cite>jralign</cite> files.</p></li>
<li><p><strong>min_reads</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The minimum number of reads across a junction to
be considered for inclusion in the output file.</p></li>
<li><p><strong>min_overhang</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The minimum overhang for a read to be included in a count.</p></li>
<li><p><strong>exclude</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Optional. Junctions within these intervals will be excluded.</p></li>
<li><p><strong>allow</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Optional. Only junctions within these intervals will be included,
so long as they are not excluded.</p></li>
<li><p><strong>pass_filter</strong> (<em>Callable</em><em>[</em><em>[</em><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><em>bool</em></a><em>]</em><em>, </em><em>optional</em>) – <p>A pass filter to be applied while iterating over the
(infile, junction_index, read_index). For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># downsample to 10%</span>
<span class="n">pass_filter</span><span class="o">=</span><span class="k">lambda</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span>

<span class="c1"># only take reads that have variants</span>
<span class="n">jraligns</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="n">JReadAlignments</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">infiles</span><span class="p">}</span>
<span class="n">pass_filter</span><span class="o">=</span><span class="k">lambda</span> <span class="n">infile</span><span class="p">,</span> <span class="n">junction</span><span class="p">,</span> <span class="n">read</span><span class="p">:</span> <span class="n">jraligns</span><span class="p">[</span><span class="n">infile</span><span class="p">]</span><span class="o">.</span><span class="n">junctions</span><span class="p">[</span><span class="n">junction</span><span class="p">][</span><span class="n">read</span><span class="p">]</span><span class="o">.</span><span class="n">num_variants</span> <span class="o">!=</span> <span class="mi">0</span>
<span class="o">...</span>
<span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">jraligns</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
    <span class="n">x</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions.from_jraligns">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_jraligns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_reads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overhang</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pass_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#ReadDistributions.from_jraligns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadDistributions.from_jraligns" title="Link to this definition"></a></dt>
<dd><p>Returns a <a class="reference internal" href="#genome_kit.ReadDistributions" title="genome_kit.ReadDistributions"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadDistributions</span></code></a> object opened from a
set of jralign files.</p>
<p>This convenience function builds and then returns ReadDistributions in
one step, by default the binary file (<cite>.rdist</cite>) is adjacent to the
first input file.</p>
<p>It is intended to be used as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rdist</span> <span class="o">=</span> <span class="n">ReadDistributions</span><span class="o">.</span><span class="n">from_jraligns</span><span class="p">([</span><span class="s2">&quot;input.jralign&quot;</span><span class="p">])</span>
<span class="n">junctions</span> <span class="o">=</span> <span class="n">rdist</span><span class="o">.</span><span class="n">find_within</span><span class="p">(</span><span class="n">brca1_interval</span><span class="p">)</span>
<span class="o">...</span>
</pre></div>
</div>
<p>If any of the filtering parameters or <cite>infiles</cite> contents change,
<cite>cache</cite> should be set to <cite>False</cite> to regenerate the <cite>ReadDistribution</cite>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#genome_kit.ReadDistributions.build_rdist" title="genome_kit.ReadDistributions.build_rdist"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_rdist()</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infiles</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The paths to the source <cite>jralign</cite> files.</p></li>
<li><p><strong>min_reads</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The minimum number of reads across a junction to
be considered for inclusion in the output file.</p></li>
<li><p><strong>min_overhang</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Optional. The minimum overhang for a read to be included in a count.</p></li>
<li><p><strong>exclude</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Optional. Junctions within these intervals will be excluded.</p></li>
<li><p><strong>allow</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Optional. Only junctions within these intervals will be included,
so long as they are not excluded.</p></li>
<li><p><strong>pass_filter</strong> (<em>Callable</em><em>[</em><em>[</em><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><em>bool</em></a><em>]</em><em>, </em><em>optional</em>) – <p>A pass filter to be applied while iterating over the
(infile, junction_index, read_index). For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># downsample to 10%</span>
<span class="n">pass_filter</span><span class="o">=</span><span class="k">lambda</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span>

<span class="c1"># only take reads that have variants</span>
<span class="n">jraligns</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="n">JReadAlignments</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">infiles</span><span class="p">}</span>
<span class="n">pass_filter</span><span class="o">=</span><span class="k">lambda</span> <span class="n">infile</span><span class="p">,</span> <span class="n">junction</span><span class="p">,</span> <span class="n">read</span><span class="p">:</span> <span class="n">jraligns</span><span class="p">[</span><span class="n">infile</span><span class="p">]</span><span class="o">.</span><span class="n">junctions</span><span class="p">[</span><span class="n">junction</span><span class="p">][</span><span class="n">read</span><span class="p">]</span><span class="o">.</span><span class="n">num_variants</span> <span class="o">!=</span> <span class="mi">0</span>
<span class="o">...</span>
<span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">jraligns</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
    <span class="n">x</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</p></li>
<li><p><strong>cache</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, optional) – Whether to re-use the binary or to re-generate it from <cite>infiles</cite>.</p></li>
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, optional) – The path to the destination <cite>rdist</cite> file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The opened <cite>rdist</cite> file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.ReadDistributions" title="genome_kit.ReadDistributions"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadDistributions</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.ReadDistributions.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#ReadDistributions.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.ReadDistributions.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="junctionreaddistributiontable">
<h3>JunctionReadDistributionTable<a class="headerlink" href="#junctionreaddistributiontable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">JunctionReadDistributionTable</span></span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p></li>
<li><p><strong>5'.</strong> (<em>The results will also be sorted by</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all junctions that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">junc</span> <span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span> <span class="k">if</span> <span class="n">junc</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All junctions that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Access to all junctions.</p>
<p>Allows iteration over all junctions in the read distribution table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">junc</span> <span class="ow">in</span> <span class="n">junctions</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The results will be ordered as in the source file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested junction.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The junction read distribution object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistributionTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistributionTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistributionTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="junctionreaddistribution">
<h3>JunctionReadDistribution<a class="headerlink" href="#junctionreaddistribution" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistribution">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">JunctionReadDistribution</span></span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistribution"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistribution" title="Link to this definition"></a></dt>
<dd><p>A read distribution across a junction.
A junction read distribution comprises a list of non-zero counts across the junction,
where the shift and strand of each count is preserved.</p>
<p>Each read mapping across the junction corresponds to two counts:
one for the part mapping upstream of the junction and another for the part mapping downstream of it.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistribution.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistribution.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistribution.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistribution.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#genome_kit.JunctionReadDistribution.interval" title="Link to this definition"></a></dt>
<dd><p>The interval spanned by this junction.</p>
<p>Note: All junction intervals are on the reference strand, even if they comprise reads that were mapped to the
reverse strand.
When querying junctions, your query intervals must therefore (currently) be on the positive strand.</p>
<p>Note that <a class="reference internal" href="#genome_kit.JunctionReadDistribution" title="genome_kit.JunctionReadDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadDistribution</span></code></a> also inherits from <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval spanned by this junction.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistribution.num_counts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_counts</span></span><a class="headerlink" href="#genome_kit.JunctionReadDistribution.num_counts" title="Link to this definition"></a></dt>
<dd><p>The number of counts on this junction.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of unique non-zero counts stored on this junction, equivalent to the <cite>len</cite> of this object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistribution.num_reads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_reads</span></span><a class="headerlink" href="#genome_kit.JunctionReadDistribution.num_reads" title="Link to this definition"></a></dt>
<dd><p>The number of reads across this junction.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of reads across this junction, equal to the summing all counts and dividing by two.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistribution.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistribution.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistribution.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Access to a junction read count.</p>
<p>Allows iteration over all read counts across this junction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">junction</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested count entry.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The junction read count identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.JunctionReadCount" title="genome_kit.JunctionReadCount"><code class="xref py py-class docutils literal notranslate"><span class="pre">JunctionReadCount</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadDistribution.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadDistribution.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadDistribution.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="junctionreadcount">
<h3>JunctionReadCount<a class="headerlink" href="#junctionreadcount" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.JunctionReadCount">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">JunctionReadCount</span></span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadCount"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadCount" title="Link to this definition"></a></dt>
<dd><p>A junction read count at a specific shift from the junction.</p>
<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadCount.strand">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#genome_kit.JunctionReadCount.strand" title="Link to this definition"></a></dt>
<dd><p>The strand that this count refers to.
:returns: The strand that this count refers to.
:rtype: <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadCount.shift">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shift</span></span><a class="headerlink" href="#genome_kit.JunctionReadCount.shift" title="Link to this definition"></a></dt>
<dd><p>The number of aligned positions between the far end (start/ end) of the alignment and the junction.</p>
<p>If negative, the count refers to an alignment upstream of the junction interval
(always defined on the positive strand). Does not count insertions, deletions, clippings, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The shift relative to the junction.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.JunctionReadCount.count">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">count</span></span><a class="headerlink" href="#genome_kit.JunctionReadCount.count" title="Link to this definition"></a></dt>
<dd><p>The number of reads at the given strand and shift from junction.
:returns: The number of reads merged into the count.
:rtype: <a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.JunctionReadCount.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/jrdist.html#JunctionReadCount.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.JunctionReadCount.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="vcf-tables">
<h2>VCF Tables<a class="headerlink" href="#vcf-tables" title="Link to this heading"></a></h2>
<section id="vcftable">
<h3>VCFTable<a class="headerlink" href="#vcftable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.VCFTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">VCFTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable" title="Link to this definition"></a></dt>
<dd><p>Open a <code class="docutils literal notranslate"><span class="pre">.vcfbin</span></code> file for querying.</p>
<p>See the VCF section of <a class="reference internal" href="quickstart.html#quickstart"><span class="std std-ref">Quickstart</span></a> for examples of how to use this class,
or see <code class="docutils literal notranslate"><span class="pre">demos/query_vcf.py</span></code>.</p>
<p>To convert from <code class="docutils literal notranslate"><span class="pre">.vcf[.gz]</span></code> see <a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>infile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the <code class="docutils literal notranslate"><span class="pre">.vcfbin</span></code> file.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.from_vcf">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vcfpath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genome_kit.Genome" title="genome_kit.genome.Genome"><span class="pre">Genome</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><span class="pre">bool</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'EBI'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'UCSC'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcfbinpath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#genome_kit.VCFTable" title="genome_kit.vcf_table.VCFTable"><span class="pre">VCFTable</span></a></span></span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.from_vcf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.from_vcf" title="Link to this definition"></a></dt>
<dd><p>Returns a <a class="reference internal" href="#genome_kit.VCFTable" title="genome_kit.VCFTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFTable</span></code></a> object opened to a VCF file.</p>
<p>The function internally converts the VCF to a binary format and then returns a <cite>VCFTable</cite> object
for fast access to the variants and fields within. By default the binary file (<cite>.vcf.bin</cite>) will be
written adjacent to the input file.</p>
<p>It is intended to be used as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">open_clinvar</span><span class="p">(</span><span class="n">vcfpath</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">VCFTable</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="n">vcfpath</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">info_ids</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AF_EXAC&quot;</span><span class="p">,</span> <span class="s2">&quot;CLNSIG&quot;</span><span class="p">])</span>

<span class="n">vcf</span> <span class="o">=</span> <span class="n">open_clinvar</span><span class="p">(</span><span class="s2">&quot;clinvar_20180128.vcf.gz&quot;</span><span class="p">)</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">find_within</span><span class="p">(</span><span class="n">brca1_interval</span><span class="p">)</span>
<span class="o">...</span>
</pre></div>
</div>
<p>If the input VCF contains multi-allelic sites, then use <cite>normalize</cite> to automatically
split them on conversion. This step requires <cite>samtools</cite> and <cite>bcftools</cite> to be installed.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vcfpath</strong> – Path to input <cite>.vcf or `.vcf.gz</cite> file.</p></li>
<li><p><strong>reference_genome</strong> – The reference genome (e.g. <cite>“hg19”</cite>). Forwarded to <a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a>.</p></li>
<li><p><strong>args</strong> – Parameters forwarded to <a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a>.</p></li>
<li><p><strong>normalize</strong> – Whether to normalize the VCF during conversion (if applicable).
If <cite>True</cite>, <cite>bcftools</cite> is used for normalization of all the EBI
(Ensembl) chromosomes; alternatively, <cite>EBI</cite> or <cite>UCSC</cite> can be used
to specify the naming scheme used by the VCF (eg. GTEx uses the
UCSC format).
Unsupported on Windows due to bcftools and samtools lack of support.</p></li>
<li><p><strong>cache</strong> – Whether to re-use the binary or to re-generate it from <cite>vcfpath</cite>.</p></li>
<li><p><strong>vcfbinpath</strong> – Path to the intermediate binary. Default is adjacent to <cite>vcfpath</cite>.</p></li>
<li><p><strong>kwargs</strong> – Parameters forwarded to <a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The opened VCF file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.VCFTable" title="genome_kit.VCFTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">variant</span> <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span> <span class="k">if</span> <span class="n">variant</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">variant</span> <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span> <span class="k">if</span> <span class="n">variant</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that have 5’ end within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">variant</span> <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span> <span class="k">if</span> <span class="n">variant</span><span class="o">.</span><span class="n">end5</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that have 5’ end falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that have 3’ end within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">variant</span> <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span> <span class="k">if</span> <span class="n">variant</span><span class="o">.</span><span class="n">end3</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that have 3’ end falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">variant</span> <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span> <span class="k">if</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">variant</span> <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span> <span class="k">if</span> <span class="n">variant</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">variant</span> <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span> <span class="k">if</span> <span class="n">variant</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.VCFTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.VCFTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Access to all variants.</p>
<p>Allows iteration over all variants:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The results will be ordered as in the source file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested variant.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The variant object identified by the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The VCFVariant will be invalidated after this table has been closed via <a class="reference internal" href="#genome_kit.VCFTable.close" title="genome_kit.VCFTable.close"><code class="xref py py-meth docutils literal notranslate"><span class="pre">close()</span></code></a> or
a <code class="docutils literal notranslate"><span class="pre">with</span></code> statement.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.__len__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.__len__" title="Link to this definition"></a></dt>
<dd><p>Return len(self).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.where">
<span class="sig-name descname"><span class="pre">where</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.where"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.where" title="Link to this definition"></a></dt>
<dd><p>Filter variants by numpy mask.</p>
<p>Fast extraction of table elements based on a mask:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Intended to be used like this.</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>

<span class="c1"># It is faster but equivalent to this.</span>
<span class="n">variants</span> <span class="o">=</span> <span class="p">[</span><span class="n">vcf</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mask</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>) – A boolean mask the same size as the table.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of variants selected by the mask.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.masked">
<span class="sig-name descname"><span class="pre">masked</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.masked"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.masked" title="Link to this definition"></a></dt>
<dd><p>Return a new VCFTable which is filtered by a numpy mask</p>
<p>For example, to do a random downsample:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">downsampled</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">masked</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcf</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">))</span>
</pre></div>
</div>
<p>Or to subset by only ABCA4 genes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">abca4_vcf</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">masked</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">index_of</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">vcf</span><span class="o">.</span><span class="n">find_overlapping</span><span class="p">(</span><span class="n">gene</span><span class="p">))</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>The masked VCFTable does not share resources with the original, so <a class="reference internal" href="#genome_kit.VCFTable.close" title="genome_kit.VCFTable.close"><code class="xref py py-meth docutils literal notranslate"><span class="pre">close()</span></code></a>
should be called on both instances when required.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mask</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code> of <a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a> or <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code> of <a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – A boolean mask the same size as the table or a list of indices.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of variants selected by the mask.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.VCFTable.GT_HOMOZYGOUS_REF">
<span class="sig-name descname"><span class="pre">GT_HOMOZYGOUS_REF</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#genome_kit.VCFTable.GT_HOMOZYGOUS_REF" title="Link to this definition"></a></dt>
<dd><p>Constant indicating a <code class="docutils literal notranslate"><span class="pre">0/0</span></code> or <code class="docutils literal notranslate"><span class="pre">0|0</span></code> in GT format value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED">
<span class="sig-name descname"><span class="pre">GT_HETEROZYGOUS_UNPHASED</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span></em><a class="headerlink" href="#genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED" title="Link to this definition"></a></dt>
<dd><p>Constant indicating a <code class="docutils literal notranslate"><span class="pre">0/1</span></code> in GT format value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.VCFTable.GT_HOMOZYGOUS_ALT">
<span class="sig-name descname"><span class="pre">GT_HOMOZYGOUS_ALT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">2</span></em><a class="headerlink" href="#genome_kit.VCFTable.GT_HOMOZYGOUS_ALT" title="Link to this definition"></a></dt>
<dd><p>Constant indicating a <code class="docutils literal notranslate"><span class="pre">1/1</span></code> or <code class="docutils literal notranslate"><span class="pre">1|1</span></code> in GT format value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.VCFTable.GT_UNKNOWN">
<span class="sig-name descname"><span class="pre">GT_UNKNOWN</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">3</span></em><a class="headerlink" href="#genome_kit.VCFTable.GT_UNKNOWN" title="Link to this definition"></a></dt>
<dd><p>Constant indicating a <code class="docutils literal notranslate"><span class="pre">?/?</span></code> or <code class="docutils literal notranslate"><span class="pre">?|?</span></code> in GT format value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_0_1">
<span class="sig-name descname"><span class="pre">GT_HETEROZYGOUS_PHASED_0_1</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">4</span></em><a class="headerlink" href="#genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_0_1" title="Link to this definition"></a></dt>
<dd><p>Constant indicating a <code class="docutils literal notranslate"><span class="pre">0|1</span></code> in GT format value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_1_0">
<span class="sig-name descname"><span class="pre">GT_HETEROZYGOUS_PHASED_1_0</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">5</span></em><a class="headerlink" href="#genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_1_0" title="Link to this definition"></a></dt>
<dd><p>Constant indicating a <code class="docutils literal notranslate"><span class="pre">1|0</span></code> in GT format value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.VCFTable.SVTYPE_NA">
<span class="sig-name descname"><span class="pre">SVTYPE_NA</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#genome_kit.VCFTable.SVTYPE_NA" title="Link to this definition"></a></dt>
<dd><p>Constant indicating the SVTYPE was missing.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.VCFTable.num_samples">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_samples</span></span><a class="headerlink" href="#genome_kit.VCFTable.num_samples" title="Link to this definition"></a></dt>
<dd><p>The number of samples represented in the VCF.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Number of samples represented in the VCF file, for which FORMAT data was provided.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.VCFTable.sample_names">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sample_names</span></span><a class="headerlink" href="#genome_kit.VCFTable.sample_names" title="Link to this definition"></a></dt>
<dd><p>The names of the samples represented in the VCF if it was encoded with format IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A list of the sample IDs in the same order as the original VCF. The indices correspond to the format values.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.VCFTable.info_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">info_ids</span></span><a class="headerlink" href="#genome_kit.VCFTable.info_ids" title="Link to this definition"></a></dt>
<dd><p>The INFO fields available for <a class="reference internal" href="#genome_kit.VCFTable.info" title="genome_kit.VCFTable.info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">info()</span></code></a> that were encoded in <a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a> or
<a class="reference internal" href="#genome_kit.VCFTable.from_vcf" title="genome_kit.VCFTable.from_vcf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_vcf()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A list of INFO ids that can be passed into <a class="reference internal" href="#genome_kit.VCFTable.info" title="genome_kit.VCFTable.info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">info()</span></code></a>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.info">
<span class="sig-name descname"><span class="pre">info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">info_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.info" title="Link to this definition"></a></dt>
<dd><p>The full per-variant data array associated with an INFO field. String INFO fields
are accessed by <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a> attribute, eg. <code class="docutils literal notranslate"><span class="pre">variant.CIGAR</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>info_id</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The INFO ID from the VCF, such as <cite>AD</cite> or <cite>DP</cite>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A complete array of INFO data values associated with <cite>info_id</cite>.</p>
<p>If there is one value per variant, the results is 1-dimensional:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span>
<span class="n">DP</span><span class="o">=</span><span class="mi">1</span>
<span class="n">DP</span><span class="o">=</span><span class="mi">4</span>

<span class="p">[</span><span class="mi">1</span> <span class="mi">4</span><span class="p">]</span>
</pre></div>
</div>
<p>If there are multiple values per variant, the results is 2-dimensional:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span>
<span class="n">AD</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span>
<span class="n">AD</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span>

<span class="p">[[</span><span class="mi">0</span> <span class="mi">1</span><span class="p">]</span>
 <span class="p">[</span><span class="mi">2</span> <span class="mi">2</span><span class="p">]]</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#KeyError" title="(in Python v3.7)"><strong>KeyError</strong></a> – Raised if <cite>info_id</cite> is unrecognized, because it was not in the original VCF
    or included by <a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a></p></li>
<li><p><a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#TypeError" title="(in Python v3.7)"><strong>TypeError</strong></a> – Raised if <cite>info_id</cite> refers to a String valued column. String values are
    accessed by <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a> attribute.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.VCFTable.format_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">format_ids</span></span><a class="headerlink" href="#genome_kit.VCFTable.format_ids" title="Link to this definition"></a></dt>
<dd><p>The INFO fields available for <a class="reference internal" href="#genome_kit.VCFTable.format" title="genome_kit.VCFTable.format"><code class="xref py py-meth docutils literal notranslate"><span class="pre">format()</span></code></a> that were encoded in <a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a> or
<a class="reference internal" href="#genome_kit.VCFTable.from_vcf" title="genome_kit.VCFTable.from_vcf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_vcf()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A list of FORMAT ids that can be passed into <a class="reference internal" href="#genome_kit.VCFTable.format" title="genome_kit.VCFTable.format"><code class="xref py py-meth docutils literal notranslate"><span class="pre">format()</span></code></a>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.format">
<span class="sig-name descname"><span class="pre">format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">format_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.format"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.format" title="Link to this definition"></a></dt>
<dd><p>The full per-variant per-sample data matrix associated with a FORMAT field.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>format_id</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The FORMAT ID from the VCF, such as <cite>AD</cite> or <cite>GT</cite>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A complete array of FORMAT data values associated with <cite>format_id</cite>.</p>
<p>If there is one value per variant and per sample, the results is 2-dimensional:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span>
<span class="mi">1</span>       <span class="mi">2</span>       <span class="mi">3</span>
<span class="mi">4</span>       <span class="mi">5</span>       <span class="mi">6</span>

<span class="p">[[</span><span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span><span class="p">]</span>
 <span class="p">[</span><span class="mi">4</span> <span class="mi">5</span> <span class="mi">6</span><span class="p">]]</span>
</pre></div>
</div>
<p>If there are multiple values per variant and per sample, the results is 3-dimensional:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span>
<span class="mi">1</span><span class="p">,</span><span class="mi">2</span>     <span class="mi">3</span><span class="p">,</span><span class="mi">4</span>     <span class="mi">5</span><span class="p">,</span><span class="mi">6</span>
<span class="mi">7</span><span class="p">,</span><span class="mi">8</span>     <span class="mi">9</span><span class="p">,</span><span class="mi">10</span>    <span class="mi">11</span><span class="p">,</span><span class="mi">12</span>

<span class="p">[[[</span><span class="mi">1</span> <span class="mi">2</span><span class="p">]</span> <span class="p">[</span><span class="mi">3</span> <span class="mi">4</span><span class="p">]</span>  <span class="p">[</span><span class="mi">5</span>  <span class="mi">6</span> <span class="p">]]</span>
 <span class="p">[[</span><span class="mi">7</span> <span class="mi">8</span><span class="p">]</span> <span class="p">[</span><span class="mi">9</span> <span class="mi">10</span><span class="p">]</span> <span class="p">[</span><span class="mi">11</span> <span class="mi">12</span><span class="p">]]]</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#KeyError" title="(in Python v3.7)"><strong>KeyError</strong></a> – Raised if <cite>format_id</cite> is unrecognized, either because it was not in the original VCF
    or included by <a class="reference internal" href="#genome_kit.VCFTable.build_vcfbin" title="genome_kit.VCFTable.build_vcfbin"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_vcfbin()</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.build_vcfbin">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_vcfbin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">info_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fmt_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'error'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.build_vcfbin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.build_vcfbin" title="Link to this definition"></a></dt>
<dd><p>Builds a binary <code class="docutils literal notranslate"><span class="pre">.vcfbin</span></code> file from a <code class="docutils literal notranslate"><span class="pre">.vcf[.gz]</span></code>.</p>
<p>The resulting file can be opened by <a class="reference internal" href="#genome_kit.VCFTable" title="genome_kit.VCFTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFTable</span></code></a>.</p>
<p><strong>VCF normalization.</strong> The input VCF should be normalized, and multi-allelic sites must be split.
Consider running <cite>infile</cite> through <code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">norm</span> <span class="pre">-m</span> <span class="pre">-</span> <span class="pre">-f</span></code>. Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">REF</span>    <span class="n">ALT</span>
<span class="n">CT</span>     <span class="n">CAT</span>     <span class="c1"># FAIL (not normalized)</span>
       <span class="n">A</span>       <span class="c1"># OK</span>
<span class="o">.</span>      <span class="n">A</span>       <span class="c1"># OK</span>
<span class="o">-</span>      <span class="n">A</span>       <span class="c1"># OK</span>
<span class="n">C</span>      <span class="n">CA</span>      <span class="c1"># OK</span>
<span class="n">C</span>      <span class="n">A</span><span class="p">,</span><span class="n">G</span>     <span class="c1"># FAIL (multi-allelic)</span>
</pre></div>
</div>
<p><strong>Extracting INFO/FORMAT columns.</strong> By default, INFO and FORMAT columns such as DP and AD are omitted.
Use <cite>info_ids</cite> and <cite>fmt_ids</cite> arguments to specify which columns you want included:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="s2">&quot;GT&quot;</span><span class="p">,</span> <span class="s2">&quot;AD&quot;</span><span class="p">]</span>                        <span class="c1"># Keep GT and AD (as int8 and int32 respectively)</span>
<span class="p">{</span><span class="s2">&quot;GT&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;AD&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">}</span>          <span class="c1"># Same</span>
<span class="p">{</span><span class="s2">&quot;GT&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;AD&quot;</span> <span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">}</span>            <span class="c1"># Same, but store -1 for any missing AD value</span>
<span class="p">{</span><span class="s2">&quot;GT&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;AD&quot;</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="p">}</span>  <span class="c1"># Keep GT and AD (as int8 and int16 respectively)</span>
</pre></div>
</div>
<p>The GT (genotype) FORMAT and SVTYPE (structural variant type) INFO columns are special. They default to
<code class="xref py py-class docutils literal notranslate"><span class="pre">int8</span></code>, and the string values are stored as integer constants:</p>
<p><strong>*GT*</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">0/0</span></code> and <code class="docutils literal notranslate"><span class="pre">0|0</span></code> are stored as <a class="reference internal" href="#genome_kit.VCFTable.GT_HOMOZYGOUS_REF" title="genome_kit.VCFTable.GT_HOMOZYGOUS_REF"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_HOMOZYGOUS_REF</span></code></a> (0)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">0/1</span></code> is stored as <a class="reference internal" href="#genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED" title="genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_HETEROZYGOUS_UNPHASED</span></code></a> (1)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">1/1</span></code> and <code class="docutils literal notranslate"><span class="pre">1|1</span></code> are stored as <a class="reference internal" href="#genome_kit.VCFTable.GT_HOMOZYGOUS_ALT" title="genome_kit.VCFTable.GT_HOMOZYGOUS_ALT"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_HOMOZYGOUS_ALT</span></code></a> (2)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">?/?</span></code> and <code class="docutils literal notranslate"><span class="pre">?|?</span></code> are stored as <a class="reference internal" href="#genome_kit.VCFTable.GT_UNKNOWN" title="genome_kit.VCFTable.GT_UNKNOWN"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_UNKNOWN</span></code></a> (3)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">0|1</span></code> is stored as <a class="reference internal" href="#genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_0_1" title="genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_0_1"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_HETEROZYGOUS_PHASED_0_1</span></code></a> (4)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">1|0</span></code> is stored as <a class="reference internal" href="#genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_1_0" title="genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_1_0"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_HETEROZYGOUS_PHASED_1_0</span></code></a> (5)</p></li>
</ul>
<p><strong>*SVTYPE*</strong></p>
<ul class="simple">
<li><p>A missing SVTYPE is represented as <a class="reference internal" href="#genome_kit.VCFTable.SVTYPE_NA" title="genome_kit.VCFTable.SVTYPE_NA"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.SVTYPE_NA</span></code></a> (0)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">DEL</span></code> is stored as <code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.SVTYPE_DEL</span></code> (1)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">INS</span></code> is stored as <code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.SVTYPE_INS</span></code> (2)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">DUP</span></code> is stored as <code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.SVTYPE_DUP</span></code> (3)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">INV</span></code> is stored as <code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.SVTYPE_INV</span></code> (4)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">CNV</span></code> is stored as <code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.SVTYPE_CNV</span></code> (5)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">BND</span></code> is stored as <code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.SVTYPE_BND</span></code> (6)</p></li>
</ul>
<p>The ID INFO column is stored as an INFO ID. If it is required (e.g., you need to find a MATEID pair), pass
<cite>“ID”</cite> into <cite>info_ids</cite>.</p>
<p>Missing values are also treated differently: GT values must be specified as a diploid.
If a single missing allele (e.g. <code class="docutils literal notranslate"><span class="pre">0/.</span></code>, <code class="docutils literal notranslate"><span class="pre">./0</span></code>, <code class="docutils literal notranslate"><span class="pre">1/.</span></code>, or <code class="docutils literal notranslate"><span class="pre">./1</span></code>) should be encoded as a <em>possible</em>
<a class="reference internal" href="#genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED" title="genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_HETEROZYGOUS_UNPHASED</span></code></a> instead of <a class="reference internal" href="#genome_kit.VCFTable.GT_UNKNOWN" title="genome_kit.VCFTable.GT_UNKNOWN"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_UNKNOWN</span></code></a>, the missing value should
be set to <a class="reference internal" href="#genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED" title="genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_HETEROZYGOUS_UNPHASED</span></code></a>. <code class="docutils literal notranslate"><span class="pre">./.</span></code> will still be interpreted as
<a class="reference internal" href="#genome_kit.VCFTable.GT_UNKNOWN" title="genome_kit.VCFTable.GT_UNKNOWN"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_UNKNOWN</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s2">&quot;GT&quot;</span> <span class="p">:</span> <span class="n">VCFTable</span><span class="o">.</span><span class="n">GT_HETEROZYGOUS_UNPHASED</span><span class="p">}</span>
</pre></div>
</div>
<p>For phased values, the default value is ignored. For example, <code class="docutils literal notranslate"><span class="pre">.|1</span></code> will be encoded as
<a class="reference internal" href="#genome_kit.VCFTable.GT_UNKNOWN" title="genome_kit.VCFTable.GT_UNKNOWN"><code class="xref py py-const docutils literal notranslate"><span class="pre">VCFTable.GT_UNKNOWN</span></code></a>.</p>
<p>All other columns default to <code class="xref py py-class docutils literal notranslate"><span class="pre">int32</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float32</span></code>, or <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> as defined
by the VCF Type header (can override bitwidth for performance or to save disk space.)
For integer/float columns, only constant value dimensions are currently supported.
By default, a missing value is substituted with 0, 0.0, or “”. If missing integer values must be
distinguishable, consider using <code class="docutils literal notranslate"><span class="pre">-(2**31)</span></code> to <code class="docutils literal notranslate"><span class="pre">(2**31+7)</span></code> as these are reserved by VCF.</p>
<p><strong>String-valued columns</strong> String values can be accessed by attribute (<code class="docutils literal notranslate"><span class="pre">variant.CIGAR</span></code>).
Unlike numeric columns, a string column cannot be accessed as a numpy array.</p>
<p>Caveats: Multi-dimensional values are not split, and ASCII-encoded VCF values,
eg. <code class="docutils literal notranslate"><span class="pre">%3A</span></code> for <code class="docutils literal notranslate"><span class="pre">:</span></code>, will not be decoded. <code class="docutils literal notranslate"><span class="pre">.</span></code> are not decoded as empty strings
unless it is the sole value.</p>
<p><strong>Value multiplicity.</strong> Each INFO/FORMAT field has an expected number of values per sample,
following the VCF 4.2 Specification sections 1.4.2 and 1.6.2. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;</span>
<span class="c1">##FORMAT=&lt;ID=AD,Number=R,Type=Integer,Description=&quot;Allelic depths&quot;&gt;</span>
</pre></div>
</div>
<p>Here, a FORMAT value of <code class="docutils literal notranslate"><span class="pre">0/1:5,8</span></code> for a single sample would be valid because we expect one
GT value per (<code class="docutils literal notranslate"><span class="pre">Number=1</span></code>) and two AD values (<code class="docutils literal notranslate"><span class="pre">Number=R</span></code>).</p>
<p>FORMAT values with multiple samples or INFO/FORMAT values that are multidimensional can only
be accessed via <a class="reference internal" href="#genome_kit.VCFTable.info" title="genome_kit.VCFTable.info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">info()</span></code></a> or <a class="reference internal" href="#genome_kit.VCFTable.format" title="genome_kit.VCFTable.format"><code class="xref py py-meth docutils literal notranslate"><span class="pre">format()</span></code></a> and cannot be accessed by <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a>
attributes.</p>
<p>Beware VCF headers that incorrectly specify <code class="docutils literal notranslate"><span class="pre">Number=.</span></code> instead of <code class="docutils literal notranslate"><span class="pre">R</span></code>, as their
multi-allelic variants will be split incorrectly by <code class="docutils literal notranslate"><span class="pre">bcftools</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The binary output file. This file can then be opened by <a class="reference internal" href="#genome_kit.VCFTable" title="genome_kit.VCFTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFTable</span></code></a>.</p></li>
<li><p><strong>infile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the source <code class="docutils literal notranslate"><span class="pre">.vcf[.gz]</span></code> file.</p></li>
<li><p><strong>reference_genome</strong> (<a class="reference internal" href="#genome_kit.Genome" title="genome_kit.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>) – The reference genome that <cite>infile</cite> uses as <code class="docutils literal notranslate"><span class="pre">#reference</span></code> field.
The parameter is not validated against the VCF, so it is up to the caller to ensure it is correct.
For example, a <code class="docutils literal notranslate"><span class="pre">#reference=b37.fa</span></code> or <code class="docutils literal notranslate"><span class="pre">=NCBI37</span></code> is a match to <code class="docutils literal notranslate"><span class="pre">&quot;hg19&quot;</span></code>.</p></li>
<li><p><strong>info_ids</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#dict" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> to <a class="reference external" href="https://docs.python.org/3.7/library/functions.html#object" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a>,                 optional) – Controls how INFO columns are preserved in <cite>outfile</cite>. See above.</p></li>
<li><p><strong>fmt_ids</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#dict" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a> of <a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> to <a class="reference external" href="https://docs.python.org/3.7/library/functions.html#object" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a>,                 optional) – Controls how FORMAT columns are preserved in <cite>outfile</cite>. See above.</p></li>
<li><p><strong>validate</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, optional) – Whether to validate <cite>REF</cite> sequences against the reference genome. Default is True.</p></li>
<li><p><strong>exclude</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Optional. Variants within these intervals will be excluded.</p></li>
<li><p><strong>allow</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – Optional. Only variants within these intervals will be included,
so long as they are not excluded.</p></li>
<li><p><strong>ancestral</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Behaviour when encountering ancestral alleles. Valid values are “error”, “warn”, and “exclude”. Defaults
to “error”.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#ValueError" title="(in Python v3.7)"><strong>ValueError</strong></a> – <p>Under the following conditions:</p>
<ul class="simple">
<li><p>The VCF is malformed.</p></li>
<li><p>A row in <cite>infile</cite> does not contain every FORMAT ID specified by <cite>fmt_ids</cite>.</p></li>
<li><p>An INFO/FORMAT ID is unsupported.</p></li>
<li><p>A <cite>REF</cite> or <cite>ALT</cite> sequence contained a VCF non-standard nucleotide code.</p></li>
<li><p>A <cite>REF</cite> sequence failed to validate.</p></li>
<li><p>An INFO/FORMAT field type is incompatible with the VCF Type header definition.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.vcfbin_version">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">vcfbin_version</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.vcfbin_version"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.vcfbin_version" title="Link to this definition"></a></dt>
<dd><p>The <cite>vcfbin</cite> file format version that this build supports.</p>
<p>Attempting to open an <cite>vcfbin</cite> file of a mismatched version will raise an <a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IOError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">IOError</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The version number.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.sequence_variations">
<span class="sig-name descname"><span class="pre">sequence_variations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.sequence_variations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.sequence_variations" title="Link to this definition"></a></dt>
<dd><p>Find variants that could affect the given <cite>interval</cite>’s DNA sequence.
When specified with an <cite>anchor</cite>, deletion variants will increase the
interval’s span; this can be overestimated (due to overlapping
deletions), but the spurious variants will be filtered during DNA
sequence retrieval.</p>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The dna sequence interval. Anchoring will expand the
interval when deletion variants are encountered.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of variants that possibly affect the given <cite>interval</cite>’s DNA
sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.VCFVariant" title="genome_kit.VCFVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.close"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.close" title="Link to this definition"></a></dt>
<dd><p>Close the file handle.</p>
<p>Note that if you use a <cite>with</cite> statement the file handle is automatically closed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Open the file</span>
<span class="k">with</span> <span class="n">VCFTable</span><span class="p">(</span><span class="s1">&#39;foo.vcfbin&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">vcfbin</span><span class="p">:</span>
    <span class="o">...</span>

<span class="c1"># &lt;-- File is now closed</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>Closing this table will invalidate all VCFVariants previously accessed via __getitem__.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.VCFTable.filename">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filename</span></span><a class="headerlink" href="#genome_kit.VCFTable.filename" title="Link to this definition"></a></dt>
<dd><p>The path to the file from which FORMAT field matrices are retrieved.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the VCF file.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="vcfvariant">
<h3>VCFVariant<a class="headerlink" href="#vcfvariant" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.VCFVariant">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">VCFVariant</span></span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFVariant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFVariant" title="Link to this definition"></a></dt>
<dd><p>A VCF variant. Scalar INFO/FORMAT fields can be accessed as a dynamic
attribute, eg. <code class="docutils literal notranslate"><span class="pre">variant.AD</span></code>.</p>
<p>Bases: <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFVariant.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFVariant.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFVariant.__init__" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFVariant.as_variant_string">
<span class="sig-name descname"><span class="pre">as_variant_string</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFVariant.as_variant_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFVariant.as_variant_string" title="Link to this definition"></a></dt>
<dd><p>Returns a string representation of the variant.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#genome_kit.Variant.from_string" title="genome_kit.Variant.from_string"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_string()</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The variant in the format <code class="docutils literal notranslate"><span class="pre">&quot;chromosome:start:ref:alt&quot;</span></code>, where
start is in DNA1 coordinates.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFVariant.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genome_kit.VCFVariant.__str__" title="Link to this definition"></a></dt>
<dd><p>Returns a string representation of the variant.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#genome_kit.Variant.from_string" title="genome_kit.Variant.from_string"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_string()</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The variant in the format <code class="docutils literal notranslate"><span class="pre">&quot;chromosome:start:ref:alt&quot;</span></code>, where
start is in DNA1 coordinates.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VCFVariant.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/vcf_table.html#VCFVariant.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VCFVariant.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
<section id="varianttable">
<h3>VariantTable<a class="headerlink" href="#varianttable" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.VariantTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">VariantTable</span></span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.find_5p_aligned">
<span class="sig-name descname"><span class="pre">find_5p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.find_5p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.find_5p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that have 5’ end aligned with 5’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variants</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">end5</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end5</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that have 5’ end aligned to the 5’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.find_3p_aligned">
<span class="sig-name descname"><span class="pre">find_3p_aligned</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.find_3p_aligned"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.find_3p_aligned" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that have 3’ end aligned with 3’ end of <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variants</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">end3</span> <span class="o">==</span> <span class="n">interval</span><span class="o">.</span><span class="n">end3</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that have 3’ end aligned to the 3’ end of <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.find_5p_within">
<span class="sig-name descname"><span class="pre">find_5p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.find_5p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.find_5p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that have 5’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variants</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">end5</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that have 5’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.find_3p_within">
<span class="sig-name descname"><span class="pre">find_3p_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.find_3p_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.find_3p_within" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that have 3’-most position within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variants</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">end3</span><span class="o">.</span><span class="n">expand</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that have 3’-most position falling entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.find_within">
<span class="sig-name descname"><span class="pre">find_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.find_within"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.find_within" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that fall entirely within <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variants</span> <span class="k">if</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that fall entirely within <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.find_overlapping">
<span class="sig-name descname"><span class="pre">find_overlapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.find_overlapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.find_overlapping" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that overlap <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variants</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">interval</span><span class="p">)]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that overlap <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.find_exact">
<span class="sig-name descname"><span class="pre">find_exact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.find_exact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.find_exact" title="Link to this definition"></a></dt>
<dd><p>Returns all variants that span exactly <cite>interval</cite>.</p>
<p>This function is a faster version of the brute-force filter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variants</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="n">interval</span><span class="p">]</span>
</pre></div>
</div>
<p>The results will also be sorted by 5’ and then 3’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>) – The query interval.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All variants that span exactly <cite>interval</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genome_kit.VariantTable.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#genome_kit.VariantTable.stranded" title="Link to this definition"></a></dt>
<dd><p>If <cite>True</cite> then the strand is significant when calling the <cite>find_x</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Whether this table can contain negative stranded intervals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Access to all variants.</p>
<p>Allows iteration over all variants in the table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">table</span><span class="p">:</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The results will be ordered as in the source file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The integer index of the requested junction.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The variant at the given index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.where">
<span class="sig-name descname"><span class="pre">where</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.where"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.where" title="Link to this definition"></a></dt>
<dd><p>Filter variants by numpy mask.</p>
<p>Fast extraction of table elements based on a mask:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Intended to be used like this.</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>

<span class="c1"># It is faster but equivalent to this.</span>
<span class="n">variants</span> <span class="o">=</span> <span class="p">[</span><span class="n">table</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mask</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>) – A boolean mask the same size as the table.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of variants selected by the mask.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.VariantTable.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/variant.html#VariantTable.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.VariantTable.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="data-access">
<h2>Data Access<a class="headerlink" href="#data-access" title="Link to this heading"></a></h2>
<section id="genome-kit-datamanager">
<h3>genome_kit.DataManager<a class="headerlink" href="#genome-kit-datamanager" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.DataManager">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DataManager</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/data_manager.html#DataManager"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.DataManager" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="genome_kit.DataManager.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/data_manager.html#DataManager.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.DataManager.__init__" title="Link to this definition"></a></dt>
<dd><p>Implementations should accept a data directory.
:param data_dir: location where local files are cached</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.DataManager.get_file">
<span class="sig-name descname"><span class="pre">get_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></span><a class="reference internal" href="_modules/genome_kit/data_manager.html#DataManager.get_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.DataManager.get_file" title="Link to this definition"></a></dt>
<dd><p>Get path to a managed data file and download if necessary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filename</strong> – A single filename, without a path.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the local version of the file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.DataManager.upload_file">
<span class="sig-name descname"><span class="pre">upload_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/typing.html#typing.Dict" title="(in Python v3.7)"><span class="pre">Dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/data_manager.html#DataManager.upload_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.DataManager.upload_file" title="Link to this definition"></a></dt>
<dd><p>Upload a file to make it available in GenomeKit.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> – Full path to the file to be uploaded.</p></li>
<li><p><strong>filename</strong> – The remote filename.</p></li>
<li><p><strong>metadata</strong> – Extra metadata to attach to the file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.DataManager.list_available_genomes">
<span class="sig-name descname"><span class="pre">list_available_genomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/data_manager.html#DataManager.list_available_genomes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.DataManager.list_available_genomes" title="Link to this definition"></a></dt>
<dd><p>List all available genomes in the data manager</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.DataManager.__weakref__">
<span class="sig-name descname"><span class="pre">__weakref__</span></span><a class="headerlink" href="#genome_kit.DataManager.__weakref__" title="Link to this definition"></a></dt>
<dd><p>list of weak references to the object (if defined)</p>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="internals">
<h2>Internals<a class="headerlink" href="#internals" title="Link to this heading"></a></h2>
<section id="module-genome_kit.gk_data">
<span id="genome-kit-gk-data"></span><h3>genome_kit.gk_data<a class="headerlink" href="#module-genome_kit.gk_data" title="Link to this heading"></a></h3>
<p>This module automatically fetches GenomeKit data.</p>
<p>A user should first use <a class="reference internal" href="#genome_kit.gk_data.upload_file" title="genome_kit.gk_data.upload_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">upload_file()</span></code></a> to upload a
file to make it available other GenomeKit users.</p>
<p>After the upload, the file can be used by any GenomeKit function via
<a class="reference internal" href="#genome_kit.gk_data.get_file" title="genome_kit.gk_data.get_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_file()</span></code></a>, which will download the file on demand
and return its local path.</p>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">upload_file</span><span class="p">(</span><span class="s1">&#39;/local/path/hg38.2bit&#39;</span><span class="p">,</span> <span class="s1">&#39;hg38.2bit&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">get_file</span><span class="p">(</span><span class="s1">&#39;hg38.2bit&#39;</span><span class="p">)</span>
<span class="go">&quot;/Users/example/Application Support/genome_kit/hg38.2bit&quot;</span>
</pre></div>
</div>
<dl class="py function">
<dt class="sig sig-object py" id="genome_kit.gk_data.get_file">
<span class="sig-name descname"><span class="pre">get_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/gk_data.html#get_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.gk_data.get_file" title="Link to this definition"></a></dt>
<dd><p>compatibility wrapper for <a class="reference internal" href="#genome_kit.DataManager.get_file" title="genome_kit.DataManager.get_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_file()</span></code></a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genome_kit.gk_data.upload_file">
<span class="sig-name descname"><span class="pre">upload_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.7/library/typing.html#typing.Dict" title="(in Python v3.7)"><span class="pre">Dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.7/library/constants.html#None" title="(in Python v3.7)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/gk_data.html#upload_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.gk_data.upload_file" title="Link to this definition"></a></dt>
<dd><p>compatibility wrapper for <a class="reference internal" href="#genome_kit.DataManager.upload_file" title="genome_kit.DataManager.upload_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">upload_file()</span></code></a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genome_kit.gk_data.wget">
<span class="sig-name descname"><span class="pre">wget</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dst</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timestamping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/gk_data.html#wget"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.gk_data.wget" title="Link to this definition"></a></dt>
<dd><p>Download a file from the URL.</p>
<p>If <cite>dst</cite> is unspecified, the file will be downloaded to
the system temp directory and given the same name.</p>
<p>If the last-modified time of the remote file is available, it
will be assigned to the downloaded file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>url</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The URL of the source file.</p></li>
<li><p><strong>dst</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – The path to the local destination.</p></li>
<li><p><strong>timestamping</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – If true, only download when last-modified date if the remote file
is newer than the local file. If the server does not provide the
last-modified date, the local file will be kept by default.</p></li>
<li><p><strong>progress</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – If true, show a progress bar.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the local version of the file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3.7/library/exceptions.html#IOError" title="(in Python v3.7)"><strong>IOError</strong></a> – There was an error opening <cite>dst</cite>.</p></li>
<li><p><strong>HTTPError</strong> – There was an error opening the URL.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3.7/library/socket.html#socket.timeout" title="(in Python v3.7)"><strong>socket.timeout</strong></a> – The connection timed out or was interrupted.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="genome-kit-data-manager">
<h3>genome_kit.data_manager<a class="headerlink" href="#genome-kit-data-manager" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="genome_kit.data_manager.ProgressPercentage">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ProgressPercentage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filesize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">progress_msg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Progress'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/data_manager.html#ProgressPercentage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.data_manager.ProgressPercentage" title="Link to this definition"></a></dt>
<dd><p>Display a progress bar on a text terminal showing file being transferred, completion
percentage, estimated transfer time and throughput.</p>
<p><code class="docutils literal notranslate"><span class="pre">`Progress</span> <span class="pre">&quot;myfile.bin&quot;</span> <span class="pre">[</span> <span class="pre">20.0</span> <span class="pre">B/</span> <span class="pre">20.0</span> <span class="pre">B]</span> <span class="pre">100%</span> <span class="pre">|###############|</span> <span class="pre">ETA:</span>&#160; <span class="pre">0:00:00</span> <span class="pre">147.8</span> <span class="pre">B/s`</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.data_manager.ProgressPercentage.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filesize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">progress_msg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Progress'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/data_manager.html#ProgressPercentage.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.data_manager.ProgressPercentage.__init__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – file name to display, will be truncated for display if necessary</p></li>
<li><p><strong>filesize</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#int" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – file size in bytes</p></li>
<li><p><strong>progress_msg</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – message to display before filename, e.g. “[MyPackage] downloading”, defaults to “”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genome_kit.data_manager.ProgressPercentage.__call__">
<span class="sig-name descname"><span class="pre">__call__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bytes_amount</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/data_manager.html#ProgressPercentage.__call__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit.data_manager.ProgressPercentage.__call__" title="Link to this definition"></a></dt>
<dd><p>Call self as a function.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genome_kit.data_manager.ProgressPercentage.__weakref__">
<span class="sig-name descname"><span class="pre">__weakref__</span></span><a class="headerlink" href="#genome_kit.data_manager.ProgressPercentage.__weakref__" title="Link to this definition"></a></dt>
<dd><p>list of weak references to the object (if defined)</p>
</dd></dl>

</dd></dl>

</section>
<section id="genome-kit-twobitutils">
<h3>genome_kit._twobitutils<a class="headerlink" href="#genome-kit-twobitutils" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="genome_kit._twobitutils.write2bit">
<span class="sig-name descname"><span class="pre">write2bit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/_twobitutils.html#write2bit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit._twobitutils.write2bit" title="Link to this definition"></a></dt>
<dd><p>Write a 2bit file containing the given DNA sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a>) – The name of the output file.</p></li>
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#dict" title="(in Python v3.7)"><em>dict</em></a>) – A dictionary of named DNA sequence strings each of which will be
represented by name. Sequence names can be arbitrary. Sequences may
contain a mix of upper/lower case ACGTN characters.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>  <span class="s2">&quot;chr1&quot;</span> <span class="p">:</span> <span class="s2">&quot;AAAANNnnCCAAaattTT&quot;</span><span class="p">,</span>
<span class="gp">... </span>  <span class="s2">&quot;chr2&quot;</span> <span class="p">:</span> <span class="s2">&quot;NNNnnnnggggGGGG&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">write2bit</span><span class="p">(</span><span class="s1">&#39;mydna.2bit&#39;</span><span class="p">,</span> <span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="genome-kit-apply-variants">
<h3>genome_kit._apply_variants<a class="headerlink" href="#genome-kit-apply-variants" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="genome_kit._apply_variants.apply_variants">
<span class="sig-name descname"><span class="pre">apply_variants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variants</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_alignment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/_apply_variants.html#apply_variants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit._apply_variants.apply_variants" title="Link to this definition"></a></dt>
<dd><p>Apply variants to a sequence interval.</p>
<p>This function ignores the interval’s strand attribute and always returns
the sequence on the reference strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<a class="reference internal" href="#genome_kit.GenomeDNA" title="genome_kit.GenomeDNA"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomeDNA</span></code></a>) – The source from which to extract DNA.</p></li>
<li><p><strong>variants</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of <a class="reference internal" href="#genome_kit.Variant" title="genome_kit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a>) – A list of variants.</p></li>
<li><p><strong>interval</strong> (<a class="reference internal" href="#genome_kit.Interval" title="genome_kit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a> | <code class="xref py py-class docutils literal notranslate"><span class="pre">Coordinate</span></code>) – An anchored interval/coordinate specifying what to extract.</p></li>
<li><p><strong>reference_alignment</strong> (<a class="reference external" href="https://docs.python.org/3.7/library/functions.html#bool" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to compute the alignment between the variant and reference sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#str" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> – The variant DNA sequence.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3.7/library/stdtypes.html#list" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> – The alignment of the variant sequence with the reference sequence
(if <code class="docutils literal notranslate"><span class="pre">reference_alignment=True</span></code>).</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="genome-kit-cxx-util">
<h3>genome_kit._cxx_util<a class="headerlink" href="#genome-kit-cxx-util" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="genome_kit._cxx_util.mock">
<span class="sig-name descname"><span class="pre">mock</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/_cxx_util.html#mock"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit._cxx_util.mock" title="Link to this definition"></a></dt>
<dd><p>Marks a class, attribute, or method as being a ‘mock’.</p>
<p>When Python-defined class <cite>A</cite> inherits from C++-defined class <cite>B</cite>,
the attributes and methods <cite>B</cite> are only partially visible to the IDE.</p>
<p>The <cite>mock</cite> function allows the definition of <cite>A</cite> to mock the attributes
and methods of <cite>B</cite>, making them accessible the IDE (for autocomplete)
and to automatic doc extraction (e.g. sphinx).</p>
<p>For example, the C++-defined <code class="xref py py-class docutils literal notranslate"><span class="pre">genome_kit._cxx.GeneTable</span></code> implements
<cite>__getitem__</cite>, which returns an instance of the Python-defined type
<a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">genome_kit.Gene</span></code></a>. The Python-defined class
<a class="reference internal" href="#genome_kit.GeneTable" title="genome_kit.GeneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">genome_kit.GeneTable</span></code></a> is defined as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@_cxx</span><span class="o">.</span><span class="n">register</span>
<span class="k">class</span> <span class="nc">GeneTable</span><span class="p">(</span><span class="n">_cxx</span><span class="o">.</span><span class="n">GeneTable</span><span class="p">):</span>

    <span class="nd">@mock</span>
    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
        <span class="s2">&quot;Returns a Gene object&quot;</span>
        <span class="k">return</span> <span class="n">mock_result</span><span class="p">(</span><span class="n">Gene</span><span class="p">)</span>
</pre></div>
</div>
<p>The IDE and autodocs ignore the <code class="docutils literal notranslate"><span class="pre">&#64;mock</span></code> decorator, so they treat
this definition at face value and do static analysis, propagating
the return type (<a class="reference internal" href="#genome_kit.Gene" title="genome_kit.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">genome_kit.Gene</span></code></a>) when doing autocomplete:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>      <span class="c1"># genes is of type GeneTable</span>
    <span class="n">gene</span><span class="o">.&lt;</span><span class="n">autocomplete</span><span class="o">&gt;</span>        <span class="c1"># gene is of type Gene</span>
</pre></div>
</div>
<p>When a class is registered with the C++ backend, all mock
properties and methods are deleted from the type, so that all
requests for those attributes will fall through directly to
the C++ implementation in the base class.</p>
<p>See the <code class="docutils literal notranslate"><span class="pre">PyDeleteMockAttrs</span></code> C++ function for the code that strips
some mock symbols from each registered type.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genome_kit._cxx_util.mock_result">
<span class="sig-name descname"><span class="pre">mock_result</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">result_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/_cxx_util.html#mock_result"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit._cxx_util.mock_result" title="Link to this definition"></a></dt>
<dd><p>Return a mock result of one or more types.</p>
<p>This function is used to give IDEs a hint during static analysis.</p>
<p>This function merely calls <code class="xref py py-func docutils literal notranslate"><span class="pre">mock_unreachable()</span></code>
before returning an instance of <cite>result_type</cite>. It’s important to print
the error message before trying to instantiate the return type.</p>
<p>See also: <a class="reference internal" href="#genome_kit._cxx_util.mock" title="genome_kit._cxx_util.mock"><code class="xref py py-func docutils literal notranslate"><span class="pre">mock()</span></code></a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genome_kit._cxx_util.mock_unreachable">
<span class="sig-name descname"><span class="pre">mock_unreachable</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/_cxx_util.html#mock_unreachable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit._cxx_util.mock_unreachable" title="Link to this definition"></a></dt>
<dd><p>Prints an unreachable code warning message.</p>
<p>If a mock attribute or method is ever called, it means there’s
some mismatch between a Python-defined type and a C++-defined type.</p>
<p>An alternative behaviour would be to raise an exception, but that
causes headaches when pylint and PyCharm try to warn about
unreachable code.</p>
<p>The reason this function is provided separately from <cite>mock_result</cite>
is that it allows the user to either return nothing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">foo</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="n">mock_unreachable</span><span class="p">()</span>
</pre></div>
</div>
<p>or to return an object that provides more complex type hints:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">foo</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="n">mock_unreachable</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">[</span><span class="nb">str</span><span class="p">()]</span>  <span class="c1"># type: List[str]</span>
</pre></div>
</div>
<p>See also: <a class="reference internal" href="#genome_kit._cxx_util.mock_result" title="genome_kit._cxx_util.mock_result"><code class="xref py py-func docutils literal notranslate"><span class="pre">mock_result()</span></code></a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genome_kit._cxx_util.strip_mock_bases">
<span class="sig-name descname"><span class="pre">strip_mock_bases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subtype</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/genome_kit/_cxx_util.html#strip_mock_bases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genome_kit._cxx_util.strip_mock_bases" title="Link to this definition"></a></dt>
<dd><p>Strips the last base classes at runtime from the given type.</p>
<p>Given a class hierarchy <cite>X-&gt;Y-&gt;Z</cite> where <cite>X</cite> and <cite>Z</cite> are defined in Python
but <cite>Y</cite> is defined in C++, then the IDE does not know that <cite>Z</cite> inherits
from <cite>X</cite>, and so it cannot provide autocomplete for members of <cite>X</cite>.</p>
<p>This decorator is a hacky way to trick IDEs into statically parsing out
the inheritance structure, while removing the ‘trick’ at runtime.</p>
<p>For example, the class hierarchy we want for <cite>Gene</cite> is simply:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@_cxx</span><span class="o">.</span><span class="n">register</span>
<span class="k">class</span> <span class="nc">Gene</span><span class="p">(</span><span class="n">_cxx</span><span class="o">.</span><span class="n">Gene</span><span class="p">):</span>  <span class="c1"># Interval -&gt; _cxx.Gene -&gt; Gene</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>However, IDEs will not see that <cite>Gene</cite> inherits from <cite>Interval</cite>.
So, instead we declare <cite>Gene</cite> as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@strip_mock_bases</span>
<span class="nd">@_cxx</span><span class="o">.</span><span class="n">register</span>
<span class="k">class</span> <span class="nc">Gene</span><span class="p">(</span><span class="n">_cxx</span><span class="o">.</span><span class="n">Gene</span><span class="p">,</span> <span class="n">Interval</span><span class="p">):</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>The IDE will now think that <cite>Gene</cite> inherits <cite>Interval</cite> via multiple
inheritance, and will IDE autocomplete accordingly, even though at
runtime we’ll get the linear inheritance scheme that we want.</p>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="anchors.html" class="btn btn-neutral float-left" title="Anchors" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="genomes.html" class="btn btn-neutral float-right" title="Genomes" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Deep Genomics.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>