Search.setIndex({"alltitles": {"API Documentation": [[1, null]], "Adding tests": [[2, "adding-tests"]], "Alignment": [[1, "alignment"]], "AlignmentMatch": [[1, "alignmentmatch"]], "AlignmentMatchTable": [[1, "alignmentmatchtable"]], "AlignmentTable": [[1, "alignmenttable"]], "Anchors": [[0, null]], "Anchors outside an interval": [[0, "anchors-outside-an-interval"]], "Annotations": [[1, "annotations"], [3, "annotations"], [5, "annotations"]], "Assemblies": [[3, "assemblies"]], "Basics": [[5, "basics"]], "Building data files": [[2, "building-data-files"]], "Building full data files": [[2, "building-full-data-files"]], "Building test data files": [[2, "building-test-data-files"]], "C++ tests": [[2, "c-tests"]], "Cds": [[1, "cds"]], "CdsTable": [[1, "cdstable"]], "Data Access": [[1, "data-access"]], "Debugging tests": [[2, "debugging-tests"]], "Deletions at the anchor": [[0, "deletions-at-the-anchor"]], "Deletions with an anchored interval": [[0, "deletions-with-an-anchored-interval"]], "Developers": [[2, null]], "Examples": [[3, "examples"]], "Exon": [[1, "exon"]], "ExonTable": [[1, "exontable"]], "Feature Extraction with Variants": [[5, "feature-extraction-with-variants"]], "Feature extraction basics": [[5, "feature-extraction-basics"]], "Gene": [[1, "gene"]], "GeneTable": [[1, "genetable"]], "Genome": [[1, "genome"]], "GenomeAnnotation": [[1, "genomeannotation"]], "GenomeDNA": [[1, "genomedna"]], "GenomeKit": [[4, null]], "GenomeTrack": [[1, "genometrack"]], "GenomeTrackBuilder": [[1, "genometrackbuilder"]], "Genomes": [[3, null]], "Indices and tables": [[4, "indices-and-tables"]], "Initial Data Access": [[5, "initial-data-access"]], "Insertions at the anchor": [[0, "insertions-at-the-anchor"]], "Insertions with an anchored interval": [[0, "insertions-with-an-anchored-interval"]], "Install via conda": [[5, "install-via-conda"]], "Installation": [[5, "installation"]], "Internals": [[1, "internals"]], "Interval": [[1, "interval"]], "Intervals without anchors": [[0, "intervals-without-anchors"]], "Intron": [[1, "intron"]], "IntronTable": [[1, "introntable"]], "JReadAlignments": [[1, "jreadalignments"]], "Jetbrains CLion setup": [[2, "jetbrains-clion-setup"]], "Junction": [[1, "junction"]], "Junction Read Alignments": [[1, "junction-read-alignments"]], "JunctionReadAlignment": [[1, "junctionreadalignment"]], "JunctionReadAlignments": [[1, "junctionreadalignments"]], "JunctionReadAlignmentsTable": [[1, "junctionreadalignmentstable"]], "JunctionReadCount": [[1, "junctionreadcount"]], "JunctionReadDistribution": [[1, "junctionreaddistribution"]], "JunctionReadDistributionTable": [[1, "junctionreaddistributiontable"]], "JunctionTable": [[1, "junctiontable"]], "Length-changing variants": [[5, "length-changing-variants"]], "Logging and troubleshooting": [[5, "logging-and-troubleshooting"]], "Making changes": [[2, "making-changes"]], "Motif finding": [[5, "motif-finding"]], "Next steps": [[4, "next-steps"]], "Quickstart": [[5, null]], "Read Alignments": [[1, "read-alignments"]], "ReadAlignments": [[1, "readalignments"]], "ReadDistributions": [[1, "readdistributions"]], "Releasing GenomeKit": [[2, "releasing-genomekit"]], "Resources Available": [[4, "resources-available"]], "Setting up": [[2, "setting-up"]], "Sharing new data files": [[5, "sharing-new-data-files"]], "Specifying anchors": [[0, "specifying-anchors"]], "Todo": [[0, "id3"]], "Tracks": [[1, "tracks"], [5, "tracks"]], "Transcript": [[1, "transcript"]], "TranscriptTable": [[1, "transcripttable"]], "Usage": [[5, "usage"]], "Useful Features": [[4, "useful-features"]], "VCF Tables": [[1, "vcf-tables"]], "VCFTable": [[1, "vcftable"]], "VCFVariant": [[1, "vcfvariant"]], "Variant": [[1, "variant"]], "Variant genomes": [[5, "variant-genomes"]], "VariantGenome": [[1, "variantgenome"]], "VariantTable": [[1, "varianttable"]], "Variants": [[5, "variants"]], "Variants from VCF Files": [[5, "variants-from-vcf-files"]], "What is GenomeKit?": [[4, "what-is-genomekit"]], "Working with Intervals": [[5, "working-with-intervals"]], "genome_kit.DataManager": [[1, "genome-kit-datamanager"]], "genome_kit._apply_variants": [[1, "genome-kit-apply-variants"]], "genome_kit._cxx_util": [[1, "genome-kit-cxx-util"]], "genome_kit._twobitutils": [[1, "genome-kit-twobitutils"]], "genome_kit.data_manager": [[1, "genome-kit-data-manager"]], "genome_kit.gk_data": [[1, "module-genome_kit.gk_data"]]}, "docnames": ["anchors", "api", "develop", "genomes", "index", "quickstart"], "envversion": {"sphinx": 62, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.todo": 2, "sphinx.ext.viewcode": 1}, "filenames": ["anchors.rst", "api.rst", "develop.rst", "genomes.rst", "index.rst", "quickstart.rst"], "indexentries": {"__call__() (genomedna method)": [[1, "genome_kit.GenomeDNA.__call__", false]], "__call__() (genometrack method)": [[1, "genome_kit.GenomeTrack.__call__", false]], "__call__() (progresspercentage method)": [[1, "genome_kit.data_manager.ProgressPercentage.__call__", false]], "__getitem__() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.__getitem__", false]], "__getitem__() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.__getitem__", false]], "__getitem__() (cdstable method)": [[1, "genome_kit.CdsTable.__getitem__", false]], "__getitem__() (exontable method)": [[1, "genome_kit.ExonTable.__getitem__", false]], "__getitem__() (genetable method)": [[1, "genome_kit.GeneTable.__getitem__", false]], "__getitem__() (introntable method)": [[1, "genome_kit.IntronTable.__getitem__", false]], "__getitem__() (junctionreadalignments method)": [[1, "genome_kit.JunctionReadAlignments.__getitem__", false]], "__getitem__() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.__getitem__", false]], "__getitem__() (junctionreaddistribution method)": [[1, "genome_kit.JunctionReadDistribution.__getitem__", false]], "__getitem__() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.__getitem__", false]], "__getitem__() (junctiontable method)": [[1, "genome_kit.JunctionTable.__getitem__", false]], "__getitem__() (transcripttable method)": [[1, "genome_kit.TranscriptTable.__getitem__", false]], "__getitem__() (varianttable method)": [[1, "genome_kit.VariantTable.__getitem__", false]], "__getitem__() (vcftable method)": [[1, "genome_kit.VCFTable.__getitem__", false]], "__init__() (alignment method)": [[1, "genome_kit.Alignment.__init__", false]], "__init__() (alignmentmatch method)": [[1, "genome_kit.AlignmentMatch.__init__", false]], "__init__() (cds method)": [[1, "genome_kit.Cds.__init__", false]], "__init__() (datamanager method)": [[1, "genome_kit.DataManager.__init__", false]], "__init__() (exon method)": [[1, "genome_kit.Exon.__init__", false]], "__init__() (gene method)": [[1, "genome_kit.Gene.__init__", false]], "__init__() (genome method)": [[1, "genome_kit.Genome.__init__", false]], "__init__() (genometrack method)": [[1, "genome_kit.GenomeTrack.__init__", false]], "__init__() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.__init__", false]], "__init__() (interval method)": [[1, "genome_kit.Interval.__init__", false]], "__init__() (intron method)": [[1, "genome_kit.Intron.__init__", false]], "__init__() (jreadalignments method)": [[1, "genome_kit.JReadAlignments.__init__", false]], "__init__() (junction method)": [[1, "genome_kit.Junction.__init__", false]], "__init__() (junctionreadalignments method)": [[1, "genome_kit.JunctionReadAlignments.__init__", false]], "__init__() (junctionreaddistribution method)": [[1, "genome_kit.JunctionReadDistribution.__init__", false]], "__init__() (progresspercentage method)": [[1, "genome_kit.data_manager.ProgressPercentage.__init__", false]], "__init__() (readalignments method)": [[1, "genome_kit.ReadAlignments.__init__", false]], "__init__() (readdistributions method)": [[1, "genome_kit.ReadDistributions.__init__", false]], "__init__() (transcript method)": [[1, "genome_kit.Transcript.__init__", false]], "__init__() (variant method)": [[1, "genome_kit.Variant.__init__", false]], "__init__() (variantgenome method)": [[1, "genome_kit.VariantGenome.__init__", false]], "__init__() (vcftable method)": [[1, "genome_kit.VCFTable.__init__", false]], "__init__() (vcfvariant method)": [[1, "genome_kit.VCFVariant.__init__", false]], "__len__() (vcftable method)": [[1, "genome_kit.VCFTable.__len__", false]], "__new__() (genome static method)": [[1, "genome_kit.Genome.__new__", false]], "__reduce__() (genome method)": [[1, "genome_kit.Genome.__reduce__", false]], "__repr__() (alignment method)": [[1, "genome_kit.Alignment.__repr__", false]], "__repr__() (alignmentmatch method)": [[1, "genome_kit.AlignmentMatch.__repr__", false]], "__repr__() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.__repr__", false]], "__repr__() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.__repr__", false]], "__repr__() (cds method)": [[1, "genome_kit.Cds.__repr__", false]], "__repr__() (cdstable method)": [[1, "genome_kit.CdsTable.__repr__", false]], "__repr__() (exon method)": [[1, "genome_kit.Exon.__repr__", false]], "__repr__() (exontable method)": [[1, "genome_kit.ExonTable.__repr__", false]], "__repr__() (gene method)": [[1, "genome_kit.Gene.__repr__", false]], "__repr__() (genetable method)": [[1, "genome_kit.GeneTable.__repr__", false]], "__repr__() (genome method)": [[1, "genome_kit.Genome.__repr__", false]], "__repr__() (genomeannotation method)": [[1, "genome_kit.GenomeAnnotation.__repr__", false]], "__repr__() (genomedna method)": [[1, "genome_kit.GenomeDNA.__repr__", false]], "__repr__() (genometrack method)": [[1, "genome_kit.GenomeTrack.__repr__", false]], "__repr__() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.__repr__", false]], "__repr__() (interval method)": [[1, "genome_kit.Interval.__repr__", false]], "__repr__() (intron method)": [[1, "genome_kit.Intron.__repr__", false]], "__repr__() (introntable method)": [[1, "genome_kit.IntronTable.__repr__", false]], "__repr__() (jreadalignments method)": [[1, "genome_kit.JReadAlignments.__repr__", false]], "__repr__() (junction method)": [[1, "genome_kit.Junction.__repr__", false]], "__repr__() (junctionreadalignment method)": [[1, "genome_kit.JunctionReadAlignment.__repr__", false]], "__repr__() (junctionreadalignments method)": [[1, "genome_kit.JunctionReadAlignments.__repr__", false]], "__repr__() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.__repr__", false]], "__repr__() (junctionreadcount method)": [[1, "genome_kit.JunctionReadCount.__repr__", false]], "__repr__() (junctionreaddistribution method)": [[1, "genome_kit.JunctionReadDistribution.__repr__", false]], "__repr__() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.__repr__", false]], "__repr__() (junctiontable method)": [[1, "genome_kit.JunctionTable.__repr__", false]], "__repr__() (readalignments method)": [[1, "genome_kit.ReadAlignments.__repr__", false]], "__repr__() (readdistributions method)": [[1, "genome_kit.ReadDistributions.__repr__", false]], "__repr__() (transcript method)": [[1, "genome_kit.Transcript.__repr__", false]], "__repr__() (transcripttable method)": [[1, "genome_kit.TranscriptTable.__repr__", false]], "__repr__() (variant method)": [[1, "genome_kit.Variant.__repr__", false]], "__repr__() (variantgenome method)": [[1, "genome_kit.VariantGenome.__repr__", false]], "__repr__() (varianttable method)": [[1, "genome_kit.VariantTable.__repr__", false]], "__repr__() (vcftable method)": [[1, "genome_kit.VCFTable.__repr__", false]], "__repr__() (vcfvariant method)": [[1, "genome_kit.VCFVariant.__repr__", false]], "__str__() (interval method)": [[1, "genome_kit.Interval.__str__", false]], "__str__() (variant method)": [[1, "genome_kit.Variant.__str__", false]], "__str__() (vcfvariant method)": [[1, "genome_kit.VCFVariant.__str__", false]], "__weakref__ (datamanager attribute)": [[1, "genome_kit.DataManager.__weakref__", false]], "__weakref__ (genome attribute)": [[1, "genome_kit.Genome.__weakref__", false]], "__weakref__ (progresspercentage attribute)": [[1, "genome_kit.data_manager.ProgressPercentage.__weakref__", false]], "alignment (class in genome_kit)": [[1, "genome_kit.Alignment", false]], "alignmentmatch (class in genome_kit)": [[1, "genome_kit.AlignmentMatch", false]], "alignmentmatchtable (class in genome_kit)": [[1, "genome_kit.AlignmentMatchTable", false]], "alignments (junction property)": [[1, "genome_kit.Junction.alignments", false]], "alignments (readalignments property)": [[1, "genome_kit.ReadAlignments.alignments", false]], "alignmenttable (class in genome_kit)": [[1, "genome_kit.AlignmentTable", false]], "alt (variant property)": [[1, "genome_kit.Variant.alt", false]], "anchor (interval property)": [[1, "genome_kit.Interval.anchor", false]], "anchor_offset (interval property)": [[1, "genome_kit.Interval.anchor_offset", false]], "annotation (genome property)": [[1, "genome_kit.Genome.annotation", false]], "apply_variants() (in module genome_kit._apply_variants)": [[1, "genome_kit._apply_variants.apply_variants", false]], "appris_principality() (genome method)": [[1, "genome_kit.Genome.appris_principality", false]], "appris_transcripts() (genome method)": [[1, "genome_kit.Genome.appris_transcripts", false]], "as_dna0() (interval method)": [[1, "genome_kit.Interval.as_dna0", false]], "as_dna1() (interval method)": [[1, "genome_kit.Interval.as_dna1", false]], "as_negative_strand() (interval method)": [[1, "genome_kit.Interval.as_negative_strand", false]], "as_opposite_strand() (interval method)": [[1, "genome_kit.Interval.as_opposite_strand", false]], "as_positive_strand() (interval method)": [[1, "genome_kit.Interval.as_positive_strand", false]], "as_rna1() (interval method)": [[1, "genome_kit.Interval.as_rna1", false]], "as_ucsc() (interval method)": [[1, "genome_kit.Interval.as_ucsc", false]], "as_variant_string() (variant method)": [[1, "genome_kit.Variant.as_variant_string", false]], "as_variant_string() (vcfvariant method)": [[1, "genome_kit.VCFVariant.as_variant_string", false]], "binary_version() (genomeannotation static method)": [[1, "genome_kit.GenomeAnnotation.binary_version", false]], "build_gencode() (genomeannotation static method)": [[1, "genome_kit.GenomeAnnotation.build_gencode", false]], "build_jralign() (jreadalignments static method)": [[1, "genome_kit.JReadAlignments.build_jralign", false]], "build_ncbi_refseq() (genomeannotation static method)": [[1, "genome_kit.GenomeAnnotation.build_ncbi_refseq", false]], "build_ralign() (readalignments static method)": [[1, "genome_kit.ReadAlignments.build_ralign", false]], "build_rdist() (readdistributions static method)": [[1, "genome_kit.ReadDistributions.build_rdist", false]], "build_ucsc_refseq() (genomeannotation static method)": [[1, "genome_kit.GenomeAnnotation.build_ucsc_refseq", false]], "build_vcfbin() (vcftable static method)": [[1, "genome_kit.VCFTable.build_vcfbin", false]], "ccds_id (transcript property)": [[1, "genome_kit.Transcript.ccds_id", false]], "cds (class in genome_kit)": [[1, "genome_kit.Cds", false]], "cds (exon property)": [[1, "genome_kit.Exon.cds", false]], "cdss (genome property)": [[1, "genome_kit.Genome.cdss", false]], "cdss (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.cdss", false]], "cdss (transcript property)": [[1, "genome_kit.Transcript.cdss", false]], "cdstable (class in genome_kit)": [[1, "genome_kit.CdsTable", false]], "chromosome (interval property)": [[1, "genome_kit.Interval.chromosome", false]], "chromosome_size() (genome method)": [[1, "genome_kit.Genome.chromosome_size", false]], "chromosomes (genome property)": [[1, "genome_kit.Genome.chromosomes", false]], "close() (genometrack method)": [[1, "genome_kit.GenomeTrack.close", false]], "close() (jreadalignments method)": [[1, "genome_kit.JReadAlignments.close", false]], "close() (readalignments method)": [[1, "genome_kit.ReadAlignments.close", false]], "close() (readdistributions method)": [[1, "genome_kit.ReadDistributions.close", false]], "close() (vcftable method)": [[1, "genome_kit.VCFTable.close", false]], "config (genome property)": [[1, "genome_kit.Genome.config", false]], "contains() (interval method)": [[1, "genome_kit.Interval.contains", false]], "count (junctionreadcount property)": [[1, "genome_kit.JunctionReadCount.count", false]], "data_dir (genome property)": [[1, "genome_kit.Genome.data_dir", false]], "data_size (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.data_size", false]], "datamanager (class in genome_kit)": [[1, "genome_kit.DataManager", false]], "dim (genometrack property)": [[1, "genome_kit.GenomeTrack.dim", false]], "dim (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.dim", false]], "distance() (interval method)": [[1, "genome_kit.Interval.distance", false]], "dna() (genome method)": [[1, "genome_kit.Genome.dna", false]], "dna() (variantgenome method)": [[1, "genome_kit.VariantGenome.dna", false]], "dnstream_of() (interval method)": [[1, "genome_kit.Interval.dnstream_of", false]], "dtype (genometrack property)": [[1, "genome_kit.GenomeTrack.dtype", false]], "dtype (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.dtype", false]], "end (interval property)": [[1, "genome_kit.Interval.end", false]], "end3 (interval property)": [[1, "genome_kit.Interval.end3", false]], "end5 (interval property)": [[1, "genome_kit.Interval.end5", false]], "etype (genometrack property)": [[1, "genome_kit.GenomeTrack.etype", false]], "etype (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.etype", false]], "exon (cds property)": [[1, "genome_kit.Cds.exon", false]], "exon (class in genome_kit)": [[1, "genome_kit.Exon", false]], "exons (genome property)": [[1, "genome_kit.Genome.exons", false]], "exons (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.exons", false]], "exons (transcript property)": [[1, "genome_kit.Transcript.exons", false]], "exontable (class in genome_kit)": [[1, "genome_kit.ExonTable", false]], "expand() (interval method)": [[1, "genome_kit.Interval.expand", false]], "filename (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.filename", false]], "filename (genomedna property)": [[1, "genome_kit.GenomeDNA.filename", false]], "filename (genometrack property)": [[1, "genome_kit.GenomeTrack.filename", false]], "filename (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.filename", false]], "filename (jreadalignments property)": [[1, "genome_kit.JReadAlignments.filename", false]], "filename (readalignments property)": [[1, "genome_kit.ReadAlignments.filename", false]], "filename (readdistributions property)": [[1, "genome_kit.ReadDistributions.filename", false]], "filename (vcftable property)": [[1, "genome_kit.VCFTable.filename", false]], "finalize() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.finalize", false]], "find_3p_aligned() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_3p_aligned", false]], "find_3p_aligned() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_3p_aligned", false]], "find_3p_aligned() (cdstable method)": [[1, "genome_kit.CdsTable.find_3p_aligned", false]], "find_3p_aligned() (exontable method)": [[1, "genome_kit.ExonTable.find_3p_aligned", false]], "find_3p_aligned() (genetable method)": [[1, "genome_kit.GeneTable.find_3p_aligned", false]], "find_3p_aligned() (introntable method)": [[1, "genome_kit.IntronTable.find_3p_aligned", false]], "find_3p_aligned() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_3p_aligned", false]], "find_3p_aligned() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_3p_aligned", false]], "find_3p_aligned() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_3p_aligned", false]], "find_3p_aligned() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_3p_aligned", false]], "find_3p_aligned() (varianttable method)": [[1, "genome_kit.VariantTable.find_3p_aligned", false]], "find_3p_aligned() (vcftable method)": [[1, "genome_kit.VCFTable.find_3p_aligned", false]], "find_3p_within() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_3p_within", false]], "find_3p_within() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_3p_within", false]], "find_3p_within() (cdstable method)": [[1, "genome_kit.CdsTable.find_3p_within", false]], "find_3p_within() (exontable method)": [[1, "genome_kit.ExonTable.find_3p_within", false]], "find_3p_within() (genetable method)": [[1, "genome_kit.GeneTable.find_3p_within", false]], "find_3p_within() (introntable method)": [[1, "genome_kit.IntronTable.find_3p_within", false]], "find_3p_within() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_3p_within", false]], "find_3p_within() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_3p_within", false]], "find_3p_within() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_3p_within", false]], "find_3p_within() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_3p_within", false]], "find_3p_within() (varianttable method)": [[1, "genome_kit.VariantTable.find_3p_within", false]], "find_3p_within() (vcftable method)": [[1, "genome_kit.VCFTable.find_3p_within", false]], "find_5p_aligned() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_5p_aligned", false]], "find_5p_aligned() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_5p_aligned", false]], "find_5p_aligned() (cdstable method)": [[1, "genome_kit.CdsTable.find_5p_aligned", false]], "find_5p_aligned() (exontable method)": [[1, "genome_kit.ExonTable.find_5p_aligned", false]], "find_5p_aligned() (genetable method)": [[1, "genome_kit.GeneTable.find_5p_aligned", false]], "find_5p_aligned() (introntable method)": [[1, "genome_kit.IntronTable.find_5p_aligned", false]], "find_5p_aligned() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_5p_aligned", false]], "find_5p_aligned() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_5p_aligned", false]], "find_5p_aligned() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_5p_aligned", false]], "find_5p_aligned() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_5p_aligned", false]], "find_5p_aligned() (varianttable method)": [[1, "genome_kit.VariantTable.find_5p_aligned", false]], "find_5p_aligned() (vcftable method)": [[1, "genome_kit.VCFTable.find_5p_aligned", false]], "find_5p_within() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_5p_within", false]], "find_5p_within() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_5p_within", false]], "find_5p_within() (cdstable method)": [[1, "genome_kit.CdsTable.find_5p_within", false]], "find_5p_within() (exontable method)": [[1, "genome_kit.ExonTable.find_5p_within", false]], "find_5p_within() (genetable method)": [[1, "genome_kit.GeneTable.find_5p_within", false]], "find_5p_within() (introntable method)": [[1, "genome_kit.IntronTable.find_5p_within", false]], "find_5p_within() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_5p_within", false]], "find_5p_within() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_5p_within", false]], "find_5p_within() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_5p_within", false]], "find_5p_within() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_5p_within", false]], "find_5p_within() (varianttable method)": [[1, "genome_kit.VariantTable.find_5p_within", false]], "find_5p_within() (vcftable method)": [[1, "genome_kit.VCFTable.find_5p_within", false]], "find_exact() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_exact", false]], "find_exact() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_exact", false]], "find_exact() (cdstable method)": [[1, "genome_kit.CdsTable.find_exact", false]], "find_exact() (exontable method)": [[1, "genome_kit.ExonTable.find_exact", false]], "find_exact() (genetable method)": [[1, "genome_kit.GeneTable.find_exact", false]], "find_exact() (introntable method)": [[1, "genome_kit.IntronTable.find_exact", false]], "find_exact() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_exact", false]], "find_exact() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_exact", false]], "find_exact() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_exact", false]], "find_exact() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_exact", false]], "find_exact() (varianttable method)": [[1, "genome_kit.VariantTable.find_exact", false]], "find_exact() (vcftable method)": [[1, "genome_kit.VCFTable.find_exact", false]], "find_motif() (genome method)": [[1, "genome_kit.Genome.find_motif", false]], "find_motif() (variantgenome method)": [[1, "genome_kit.VariantGenome.find_motif", false]], "find_overlapping() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_overlapping", false]], "find_overlapping() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_overlapping", false]], "find_overlapping() (cdstable method)": [[1, "genome_kit.CdsTable.find_overlapping", false]], "find_overlapping() (exontable method)": [[1, "genome_kit.ExonTable.find_overlapping", false]], "find_overlapping() (genetable method)": [[1, "genome_kit.GeneTable.find_overlapping", false]], "find_overlapping() (introntable method)": [[1, "genome_kit.IntronTable.find_overlapping", false]], "find_overlapping() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_overlapping", false]], "find_overlapping() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_overlapping", false]], "find_overlapping() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_overlapping", false]], "find_overlapping() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_overlapping", false]], "find_overlapping() (varianttable method)": [[1, "genome_kit.VariantTable.find_overlapping", false]], "find_overlapping() (vcftable method)": [[1, "genome_kit.VCFTable.find_overlapping", false]], "find_within() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_within", false]], "find_within() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_within", false]], "find_within() (cdstable method)": [[1, "genome_kit.CdsTable.find_within", false]], "find_within() (exontable method)": [[1, "genome_kit.ExonTable.find_within", false]], "find_within() (genetable method)": [[1, "genome_kit.GeneTable.find_within", false]], "find_within() (introntable method)": [[1, "genome_kit.IntronTable.find_within", false]], "find_within() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_within", false]], "find_within() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_within", false]], "find_within() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_within", false]], "find_within() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_within", false]], "find_within() (varianttable method)": [[1, "genome_kit.VariantTable.find_within", false]], "find_within() (vcftable method)": [[1, "genome_kit.VCFTable.find_within", false]], "first_by_name() (genetable method)": [[1, "genome_kit.GeneTable.first_by_name", false]], "flush() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.flush", false]], "format() (vcftable method)": [[1, "genome_kit.VCFTable.format", false]], "format_ids (vcftable property)": [[1, "genome_kit.VCFTable.format_ids", false]], "from_dna0() (interval static method)": [[1, "genome_kit.Interval.from_dna0", false]], "from_dna0_coord() (interval static method)": [[1, "genome_kit.Interval.from_dna0_coord", false]], "from_dna1() (interval static method)": [[1, "genome_kit.Interval.from_dna1", false]], "from_jraligns() (readdistributions static method)": [[1, "genome_kit.ReadDistributions.from_jraligns", false]], "from_rna1() (interval static method)": [[1, "genome_kit.Interval.from_rna1", false]], "from_string() (variant static method)": [[1, "genome_kit.Variant.from_string", false]], "from_vcf() (vcftable static method)": [[1, "genome_kit.VCFTable.from_vcf", false]], "gene (class in genome_kit)": [[1, "genome_kit.Gene", false]], "gene (transcript property)": [[1, "genome_kit.Transcript.gene", false]], "genes (genome property)": [[1, "genome_kit.Genome.genes", false]], "genes (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.genes", false]], "genetable (class in genome_kit)": [[1, "genome_kit.GeneTable", false]], "genome (class in genome_kit)": [[1, "genome_kit.Genome", false]], "genome_kit.gk_data": [[1, "module-genome_kit.gk_data", false]], "genomeannotation (class in genome_kit)": [[1, "genome_kit.GenomeAnnotation", false]], "genomedna (class in genome_kit)": [[1, "genome_kit.GenomeDNA", false]], "genometrack (class in genome_kit)": [[1, "genome_kit.GenomeTrack", false]], "genometrackbuilder (class in genome_kit)": [[1, "genome_kit.GenomeTrackBuilder", false]], "get_file() (datamanager method)": [[1, "genome_kit.DataManager.get_file", false]], "get_file() (in module genome_kit.gk_data)": [[1, "genome_kit.gk_data.get_file", false]], "gt_heterozygous_phased_0_1 (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_0_1", false]], "gt_heterozygous_phased_1_0 (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_1_0", false]], "gt_heterozygous_unphased (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED", false]], "gt_homozygous_alt (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HOMOZYGOUS_ALT", false]], "gt_homozygous_ref (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HOMOZYGOUS_REF", false]], "gt_unknown (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_UNKNOWN", false]], "gtrack_version() (genometrack static method)": [[1, "genome_kit.GenomeTrack.gtrack_version", false]], "histogram() (genometrack method)": [[1, "genome_kit.GenomeTrack.histogram", false]], "id (exon property)": [[1, "genome_kit.Exon.id", false]], "id (gene property)": [[1, "genome_kit.Gene.id", false]], "id (transcript property)": [[1, "genome_kit.Transcript.id", false]], "index (exon property)": [[1, "genome_kit.Exon.index", false]], "index (intron property)": [[1, "genome_kit.Intron.index", false]], "index_size (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.index_size", false]], "info() (vcftable method)": [[1, "genome_kit.VCFTable.info", false]], "info_ids (vcftable property)": [[1, "genome_kit.VCFTable.info_ids", false]], "intersect() (interval method)": [[1, "genome_kit.Interval.intersect", false]], "interval (alignment property)": [[1, "genome_kit.Alignment.interval", false]], "interval (alignmentmatch property)": [[1, "genome_kit.AlignmentMatch.interval", false]], "interval (cds property)": [[1, "genome_kit.Cds.interval", false]], "interval (class in genome_kit)": [[1, "genome_kit.Interval", false]], "interval (exon property)": [[1, "genome_kit.Exon.interval", false]], "interval (gene property)": [[1, "genome_kit.Gene.interval", false]], "interval (intron property)": [[1, "genome_kit.Intron.interval", false]], "interval (junction property)": [[1, "genome_kit.Junction.interval", false]], "interval (junctionreadalignments property)": [[1, "genome_kit.JunctionReadAlignments.interval", false]], "interval (junctionreaddistribution property)": [[1, "genome_kit.JunctionReadDistribution.interval", false]], "interval (transcript property)": [[1, "genome_kit.Transcript.interval", false]], "interval (variant property)": [[1, "genome_kit.Variant.interval", false]], "interval() (genome method)": [[1, "genome_kit.Genome.interval", false]], "intervals() (genometrack method)": [[1, "genome_kit.GenomeTrack.intervals", false]], "intron (class in genome_kit)": [[1, "genome_kit.Intron", false]], "introns (genome property)": [[1, "genome_kit.Genome.introns", false]], "introns (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.introns", false]], "introns (transcript property)": [[1, "genome_kit.Transcript.introns", false]], "introntable (class in genome_kit)": [[1, "genome_kit.IntronTable", false]], "is_positive_strand() (interval method)": [[1, "genome_kit.Interval.is_positive_strand", false]], "jralign_version() (jreadalignments static method)": [[1, "genome_kit.JReadAlignments.jralign_version", false]], "jreadalignments (class in genome_kit)": [[1, "genome_kit.JReadAlignments", false]], "junction (class in genome_kit)": [[1, "genome_kit.Junction", false]], "junctionreadalignment (class in genome_kit)": [[1, "genome_kit.JunctionReadAlignment", false]], "junctionreadalignments (class in genome_kit)": [[1, "genome_kit.JunctionReadAlignments", false]], "junctionreadalignmentstable (class in genome_kit)": [[1, "genome_kit.JunctionReadAlignmentsTable", false]], "junctionreadcount (class in genome_kit)": [[1, "genome_kit.JunctionReadCount", false]], "junctionreaddistribution (class in genome_kit)": [[1, "genome_kit.JunctionReadDistribution", false]], "junctionreaddistributiontable (class in genome_kit)": [[1, "genome_kit.JunctionReadDistributionTable", false]], "junctions (jreadalignments property)": [[1, "genome_kit.JReadAlignments.junctions", false]], "junctions (readalignments property)": [[1, "genome_kit.ReadAlignments.junctions", false]], "junctions (readdistributions property)": [[1, "genome_kit.ReadDistributions.junctions", false]], "junctiontable (class in genome_kit)": [[1, "genome_kit.JunctionTable", false]], "left (junctionreadalignment property)": [[1, "genome_kit.JunctionReadAlignment.left", false]], "level (gene property)": [[1, "genome_kit.Gene.level", false]], "level (transcript property)": [[1, "genome_kit.Transcript.level", false]], "list_available_genomes() (datamanager method)": [[1, "genome_kit.DataManager.list_available_genomes", false]], "masked() (vcftable method)": [[1, "genome_kit.VCFTable.masked", false]], "matches (alignment property)": [[1, "genome_kit.Alignment.matches", false]], "matches (readalignments property)": [[1, "genome_kit.ReadAlignments.matches", false]], "midpoint (interval property)": [[1, "genome_kit.Interval.midpoint", false]], "mock() (in module genome_kit._cxx_util)": [[1, "genome_kit._cxx_util.mock", false]], "mock_result() (in module genome_kit._cxx_util)": [[1, "genome_kit._cxx_util.mock_result", false]], "mock_unreachable() (in module genome_kit._cxx_util)": [[1, "genome_kit._cxx_util.mock_unreachable", false]], "module": [[1, "module-genome_kit.gk_data", false]], "name (gene property)": [[1, "genome_kit.Gene.name", false]], "next_cds (cds property)": [[1, "genome_kit.Cds.next_cds", false]], "next_exon (exon property)": [[1, "genome_kit.Exon.next_exon", false]], "next_exon (intron property)": [[1, "genome_kit.Intron.next_exon", false]], "next_intron (exon property)": [[1, "genome_kit.Exon.next_intron", false]], "next_intron (intron property)": [[1, "genome_kit.Intron.next_intron", false]], "num_counts (junctionreaddistribution property)": [[1, "genome_kit.JunctionReadDistribution.num_counts", false]], "num_reads (junctionreadalignments property)": [[1, "genome_kit.JunctionReadAlignments.num_reads", false]], "num_reads (junctionreaddistribution property)": [[1, "genome_kit.JunctionReadDistribution.num_reads", false]], "num_samples (vcftable property)": [[1, "genome_kit.VCFTable.num_samples", false]], "num_variants (junctionreadalignment property)": [[1, "genome_kit.JunctionReadAlignment.num_variants", false]], "overlaps() (interval method)": [[1, "genome_kit.Interval.overlaps", false]], "phase (cds property)": [[1, "genome_kit.Cds.phase", false]], "pileup() (readalignments method)": [[1, "genome_kit.ReadAlignments.pileup", false]], "position (variant property)": [[1, "genome_kit.Variant.position", false]], "prev_cds (cds property)": [[1, "genome_kit.Cds.prev_cds", false]], "prev_exon (exon property)": [[1, "genome_kit.Exon.prev_exon", false]], "prev_exon (intron property)": [[1, "genome_kit.Intron.prev_exon", false]], "prev_intron (exon property)": [[1, "genome_kit.Exon.prev_intron", false]], "prev_intron (intron property)": [[1, "genome_kit.Intron.prev_intron", false]], "product (transcript property)": [[1, "genome_kit.Transcript.product", false]], "progresspercentage (class in genome_kit.data_manager)": [[1, "genome_kit.data_manager.ProgressPercentage", false]], "protein_id (transcript property)": [[1, "genome_kit.Transcript.protein_id", false]], "ralign_version() (readalignments static method)": [[1, "genome_kit.ReadAlignments.ralign_version", false]], "rdist_version() (readdistributions static method)": [[1, "genome_kit.ReadDistributions.rdist_version", false]], "readalignments (class in genome_kit)": [[1, "genome_kit.ReadAlignments", false]], "readdistributions (class in genome_kit)": [[1, "genome_kit.ReadDistributions", false]], "ref (variant property)": [[1, "genome_kit.Variant.ref", false]], "reference_genome (genome property)": [[1, "genome_kit.Genome.reference_genome", false]], "reference_genome (genomedna property)": [[1, "genome_kit.GenomeDNA.reference_genome", false]], "reference_genome (genometrack property)": [[1, "genome_kit.GenomeTrack.reference_genome", false]], "reference_genome (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.reference_genome", false]], "reference_genome (interval property)": [[1, "genome_kit.Interval.reference_genome", false]], "resolution (genometrack property)": [[1, "genome_kit.GenomeTrack.resolution", false]], "resolution (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.resolution", false]], "right (junctionreadalignment property)": [[1, "genome_kit.JunctionReadAlignment.right", false]], "sample_names (vcftable property)": [[1, "genome_kit.VCFTable.sample_names", false]], "sequence_variations() (vcftable method)": [[1, "genome_kit.VCFTable.sequence_variations", false]], "set_clamping() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_clamping", false]], "set_data() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_data", false]], "set_data_from_bed() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_data_from_bed", false]], "set_data_from_bedgraph() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_data_from_bedgraph", false]], "set_data_from_wig() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_data_from_wig", false]], "set_default_value() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_default_value", false]], "set_dict() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_dict", false]], "set_restriction() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_restriction", false]], "set_sparsity() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_sparsity", false]], "set_transform() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_transform", false]], "shift (junctionreadcount property)": [[1, "genome_kit.JunctionReadCount.shift", false]], "shift() (interval method)": [[1, "genome_kit.Interval.shift", false]], "spanning() (interval static method)": [[1, "genome_kit.Interval.spanning", false]], "spanning() (variant static method)": [[1, "genome_kit.Variant.spanning", false]], "start (interval property)": [[1, "genome_kit.Interval.start", false]], "strand (interval property)": [[1, "genome_kit.Interval.strand", false]], "strand (junctionreadalignment property)": [[1, "genome_kit.JunctionReadAlignment.strand", false]], "strand (junctionreadcount property)": [[1, "genome_kit.JunctionReadCount.strand", false]], "stranded (alignmentmatchtable property)": [[1, "genome_kit.AlignmentMatchTable.stranded", false]], "stranded (alignmenttable property)": [[1, "genome_kit.AlignmentTable.stranded", false]], "stranded (cdstable property)": [[1, "genome_kit.CdsTable.stranded", false]], "stranded (exontable property)": [[1, "genome_kit.ExonTable.stranded", false]], "stranded (genetable property)": [[1, "genome_kit.GeneTable.stranded", false]], "stranded (genometrack property)": [[1, "genome_kit.GenomeTrack.stranded", false]], "stranded (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.stranded", false]], "stranded (introntable property)": [[1, "genome_kit.IntronTable.stranded", false]], "stranded (junctionreadalignmentstable property)": [[1, "genome_kit.JunctionReadAlignmentsTable.stranded", false]], "stranded (junctionreaddistributiontable property)": [[1, "genome_kit.JunctionReadDistributionTable.stranded", false]], "stranded (junctiontable property)": [[1, "genome_kit.JunctionTable.stranded", false]], "stranded (transcripttable property)": [[1, "genome_kit.TranscriptTable.stranded", false]], "stranded (varianttable property)": [[1, "genome_kit.VariantTable.stranded", false]], "stranded (vcftable property)": [[1, "genome_kit.VCFTable.stranded", false]], "strip_mock_bases() (in module genome_kit._cxx_util)": [[1, "genome_kit._cxx_util.strip_mock_bases", false]], "subtract() (interval method)": [[1, "genome_kit.Interval.subtract", false]], "svtype_na (vcftable attribute)": [[1, "genome_kit.VCFTable.SVTYPE_NA", false]], "transcript (cds property)": [[1, "genome_kit.Cds.transcript", false]], "transcript (class in genome_kit)": [[1, "genome_kit.Transcript", false]], "transcript (exon property)": [[1, "genome_kit.Exon.transcript", false]], "transcript (intron property)": [[1, "genome_kit.Intron.transcript", false]], "transcripts (gene property)": [[1, "genome_kit.Gene.transcripts", false]], "transcripts (genome property)": [[1, "genome_kit.Genome.transcripts", false]], "transcripts (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.transcripts", false]], "transcripttable (class in genome_kit)": [[1, "genome_kit.TranscriptTable", false]], "tsl (transcript property)": [[1, "genome_kit.Transcript.tsl", false]], "type (gene property)": [[1, "genome_kit.Gene.type", false]], "type (transcript property)": [[1, "genome_kit.Transcript.type", false]], "upload_file() (datamanager method)": [[1, "genome_kit.DataManager.upload_file", false]], "upload_file() (in module genome_kit.gk_data)": [[1, "genome_kit.gk_data.upload_file", false]], "upstream_of() (interval method)": [[1, "genome_kit.Interval.upstream_of", false]], "utr3 (exon property)": [[1, "genome_kit.Exon.utr3", false]], "utr3s (genome property)": [[1, "genome_kit.Genome.utr3s", false]], "utr3s (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.utr3s", false]], "utr3s (transcript property)": [[1, "genome_kit.Transcript.utr3s", false]], "utr5 (exon property)": [[1, "genome_kit.Exon.utr5", false]], "utr5s (genome property)": [[1, "genome_kit.Genome.utr5s", false]], "utr5s (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.utr5s", false]], "utr5s (transcript property)": [[1, "genome_kit.Transcript.utr5s", false]], "variant (class in genome_kit)": [[1, "genome_kit.Variant", false]], "variant() (genome method)": [[1, "genome_kit.Genome.variant", false]], "variant_dna() (genome method)": [[1, "genome_kit.Genome.variant_dna", false]], "variantgenome (class in genome_kit)": [[1, "genome_kit.VariantGenome", false]], "variants (alignmentmatch property)": [[1, "genome_kit.AlignmentMatch.variants", false]], "variants (jreadalignments property)": [[1, "genome_kit.JReadAlignments.variants", false]], "variants (readalignments property)": [[1, "genome_kit.ReadAlignments.variants", false]], "variants() (junctionreadalignment method)": [[1, "genome_kit.JunctionReadAlignment.variants", false]], "varianttable (class in genome_kit)": [[1, "genome_kit.VariantTable", false]], "vcfbin_version() (vcftable static method)": [[1, "genome_kit.VCFTable.vcfbin_version", false]], "vcftable (class in genome_kit)": [[1, "genome_kit.VCFTable", false]], "vcfvariant (class in genome_kit)": [[1, "genome_kit.VCFVariant", false]], "wget() (in module genome_kit.gk_data)": [[1, "genome_kit.gk_data.wget", false]], "where() (varianttable method)": [[1, "genome_kit.VariantTable.where", false]], "where() (vcftable method)": [[1, "genome_kit.VCFTable.where", false]], "with_anchor() (interval method)": [[1, "genome_kit.Interval.with_anchor", false]], "within() (interval method)": [[1, "genome_kit.Interval.within", false]], "write2bit() (in module genome_kit._twobitutils)": [[1, "genome_kit._twobitutils.write2bit", false]]}, "objects": {"genome_kit": [[1, 0, 1, "", "Alignment"], [1, 0, 1, "", "AlignmentMatch"], [1, 0, 1, "", "AlignmentMatchTable"], [1, 0, 1, "", "AlignmentTable"], [1, 0, 1, "", "Cds"], [1, 0, 1, "", "CdsTable"], [1, 0, 1, "", "DataManager"], [1, 0, 1, "", "Exon"], [1, 0, 1, "", "ExonTable"], [1, 0, 1, "", "Gene"], [1, 0, 1, "", "GeneTable"], [1, 0, 1, "", "Genome"], [1, 0, 1, "", "GenomeAnnotation"], [1, 0, 1, "", "GenomeDNA"], [1, 0, 1, "", "GenomeTrack"], [1, 0, 1, "", "GenomeTrackBuilder"], [1, 0, 1, "", "Interval"], [1, 0, 1, "", "Intron"], [1, 0, 1, "", "IntronTable"], [1, 0, 1, "", "JReadAlignments"], [1, 0, 1, "", "Junction"], [1, 0, 1, "", "JunctionReadAlignment"], [1, 0, 1, "", "JunctionReadAlignments"], [1, 0, 1, "", "JunctionReadAlignmentsTable"], [1, 0, 1, "", "JunctionReadCount"], [1, 0, 1, "", "JunctionReadDistribution"], [1, 0, 1, "", "JunctionReadDistributionTable"], [1, 0, 1, "", "JunctionTable"], [1, 0, 1, "", "ReadAlignments"], [1, 0, 1, "", "ReadDistributions"], [1, 0, 1, "", "Transcript"], [1, 0, 1, "", "TranscriptTable"], [1, 0, 1, "", "VCFTable"], [1, 0, 1, "", "VCFVariant"], [1, 0, 1, "", "Variant"], [1, 0, 1, "", "VariantGenome"], [1, 0, 1, "", "VariantTable"], [1, 5, 0, "-", "gk_data"]], "genome_kit.Alignment": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "matches"]], "genome_kit.AlignmentMatch": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "variants"]], "genome_kit.AlignmentMatchTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.AlignmentTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.Cds": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "exon"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_cds"], [1, 2, 1, "", "phase"], [1, 2, 1, "", "prev_cds"], [1, 2, 1, "", "transcript"]], "genome_kit.CdsTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.DataManager": [[1, 1, 1, "", "__init__"], [1, 3, 1, "", "__weakref__"], [1, 1, 1, "", "get_file"], [1, 1, 1, "", "list_available_genomes"], [1, 1, 1, "", "upload_file"]], "genome_kit.Exon": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "cds"], [1, 2, 1, "", "id"], [1, 2, 1, "", "index"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_exon"], [1, 2, 1, "", "next_intron"], [1, 2, 1, "", "prev_exon"], [1, 2, 1, "", "prev_intron"], [1, 2, 1, "", "transcript"], [1, 2, 1, "", "utr3"], [1, 2, 1, "", "utr5"]], "genome_kit.ExonTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.Gene": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "id"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "level"], [1, 2, 1, "", "name"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "type"]], "genome_kit.GeneTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 1, 1, "", "first_by_name"], [1, 2, 1, "", "stranded"]], "genome_kit.Genome": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__new__"], [1, 1, 1, "", "__reduce__"], [1, 1, 1, "", "__repr__"], [1, 3, 1, "", "__weakref__"], [1, 2, 1, "", "annotation"], [1, 1, 1, "", "appris_principality"], [1, 1, 1, "", "appris_transcripts"], [1, 2, 1, "", "cdss"], [1, 1, 1, "", "chromosome_size"], [1, 2, 1, "", "chromosomes"], [1, 2, 1, "", "config"], [1, 2, 1, "", "data_dir"], [1, 1, 1, "", "dna"], [1, 2, 1, "", "exons"], [1, 1, 1, "", "find_motif"], [1, 2, 1, "", "genes"], [1, 1, 1, "", "interval"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"], [1, 1, 1, "", "variant"], [1, 1, 1, "", "variant_dna"]], "genome_kit.GenomeAnnotation": [[1, 1, 1, "", "__repr__"], [1, 1, 1, "", "binary_version"], [1, 1, 1, "", "build_gencode"], [1, 1, 1, "", "build_ncbi_refseq"], [1, 1, 1, "", "build_ucsc_refseq"], [1, 2, 1, "", "cdss"], [1, 2, 1, "", "exons"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "genes"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"]], "genome_kit.GenomeDNA": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "reference_genome"]], "genome_kit.GenomeTrack": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "close"], [1, 2, 1, "", "dim"], [1, 2, 1, "", "dtype"], [1, 2, 1, "", "etype"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "gtrack_version"], [1, 1, 1, "", "histogram"], [1, 1, 1, "", "intervals"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "resolution"], [1, 2, 1, "", "stranded"]], "genome_kit.GenomeTrackBuilder": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "data_size"], [1, 2, 1, "", "dim"], [1, 2, 1, "", "dtype"], [1, 2, 1, "", "etype"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "finalize"], [1, 1, 1, "", "flush"], [1, 2, 1, "", "index_size"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "resolution"], [1, 1, 1, "", "set_clamping"], [1, 1, 1, "", "set_data"], [1, 1, 1, "", "set_data_from_bed"], [1, 1, 1, "", "set_data_from_bedgraph"], [1, 1, 1, "", "set_data_from_wig"], [1, 1, 1, "", "set_default_value"], [1, 1, 1, "", "set_dict"], [1, 1, 1, "", "set_restriction"], [1, 1, 1, "", "set_sparsity"], [1, 1, 1, "", "set_transform"], [1, 2, 1, "", "stranded"]], "genome_kit.Interval": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 2, 1, "", "anchor"], [1, 2, 1, "", "anchor_offset"], [1, 1, 1, "", "as_dna0"], [1, 1, 1, "", "as_dna1"], [1, 1, 1, "", "as_negative_strand"], [1, 1, 1, "", "as_opposite_strand"], [1, 1, 1, "", "as_positive_strand"], [1, 1, 1, "", "as_rna1"], [1, 1, 1, "", "as_ucsc"], [1, 2, 1, "", "chromosome"], [1, 1, 1, "", "contains"], [1, 1, 1, "", "distance"], [1, 1, 1, "", "dnstream_of"], [1, 2, 1, "", "end"], [1, 2, 1, "", "end3"], [1, 2, 1, "", "end5"], [1, 1, 1, "", "expand"], [1, 1, 1, "", "from_dna0"], [1, 1, 1, "", "from_dna0_coord"], [1, 1, 1, "", "from_dna1"], [1, 1, 1, "", "from_rna1"], [1, 1, 1, "", "intersect"], [1, 1, 1, "", "is_positive_strand"], [1, 2, 1, "", "midpoint"], [1, 1, 1, "", "overlaps"], [1, 2, 1, "", "reference_genome"], [1, 1, 1, "", "shift"], [1, 1, 1, "", "spanning"], [1, 2, 1, "", "start"], [1, 2, 1, "", "strand"], [1, 1, 1, "", "subtract"], [1, 1, 1, "", "upstream_of"], [1, 1, 1, "", "with_anchor"], [1, 1, 1, "", "within"]], "genome_kit.Intron": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "index"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_exon"], [1, 2, 1, "", "next_intron"], [1, 2, 1, "", "prev_exon"], [1, 2, 1, "", "prev_intron"], [1, 2, 1, "", "transcript"]], "genome_kit.IntronTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JReadAlignments": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_jralign"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "jralign_version"], [1, 2, 1, "", "junctions"], [1, 2, 1, "", "variants"]], "genome_kit.Junction": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "alignments"], [1, 2, 1, "", "interval"]], "genome_kit.JunctionReadAlignment": [[1, 1, 1, "", "__repr__"], [1, 2, 1, "", "left"], [1, 2, 1, "", "num_variants"], [1, 2, 1, "", "right"], [1, 2, 1, "", "strand"], [1, 1, 1, "", "variants"]], "genome_kit.JunctionReadAlignments": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "num_reads"]], "genome_kit.JunctionReadAlignmentsTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JunctionReadCount": [[1, 1, 1, "", "__repr__"], [1, 2, 1, "", "count"], [1, 2, 1, "", "shift"], [1, 2, 1, "", "strand"]], "genome_kit.JunctionReadDistribution": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "num_counts"], [1, 2, 1, "", "num_reads"]], "genome_kit.JunctionReadDistributionTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JunctionTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.ReadAlignments": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "alignments"], [1, 1, 1, "", "build_ralign"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "junctions"], [1, 2, 1, "", "matches"], [1, 1, 1, "", "pileup"], [1, 1, 1, "", "ralign_version"], [1, 2, 1, "", "variants"]], "genome_kit.ReadDistributions": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_rdist"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "from_jraligns"], [1, 2, 1, "", "junctions"], [1, 1, 1, "", "rdist_version"]], "genome_kit.Transcript": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "ccds_id"], [1, 2, 1, "", "cdss"], [1, 2, 1, "", "exons"], [1, 2, 1, "", "gene"], [1, 2, 1, "", "id"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "level"], [1, 2, 1, "", "product"], [1, 2, 1, "", "protein_id"], [1, 2, 1, "", "tsl"], [1, 2, 1, "", "type"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"]], "genome_kit.TranscriptTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.VCFTable": [[1, 3, 1, "", "GT_HETEROZYGOUS_PHASED_0_1"], [1, 3, 1, "", "GT_HETEROZYGOUS_PHASED_1_0"], [1, 3, 1, "", "GT_HETEROZYGOUS_UNPHASED"], [1, 3, 1, "", "GT_HOMOZYGOUS_ALT"], [1, 3, 1, "", "GT_HOMOZYGOUS_REF"], [1, 3, 1, "", "GT_UNKNOWN"], [1, 3, 1, "", "SVTYPE_NA"], [1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__len__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_vcfbin"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 1, 1, "", "format"], [1, 2, 1, "", "format_ids"], [1, 1, 1, "", "from_vcf"], [1, 1, 1, "", "info"], [1, 2, 1, "", "info_ids"], [1, 1, 1, "", "masked"], [1, 2, 1, "", "num_samples"], [1, 2, 1, "", "sample_names"], [1, 1, 1, "", "sequence_variations"], [1, 2, 1, "", "stranded"], [1, 1, 1, "", "vcfbin_version"], [1, 1, 1, "", "where"]], "genome_kit.VCFVariant": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 1, 1, "", "as_variant_string"]], "genome_kit.Variant": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 2, 1, "", "alt"], [1, 1, 1, "", "as_variant_string"], [1, 1, 1, "", "from_string"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "position"], [1, 2, 1, "", "ref"], [1, 1, 1, "", "spanning"]], "genome_kit.VariantGenome": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "dna"], [1, 1, 1, "", "find_motif"]], "genome_kit.VariantTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"], [1, 1, 1, "", "where"]], "genome_kit._apply_variants": [[1, 4, 1, "", "apply_variants"]], "genome_kit._cxx_util": [[1, 4, 1, "", "mock"], [1, 4, 1, "", "mock_result"], [1, 4, 1, "", "mock_unreachable"], [1, 4, 1, "", "strip_mock_bases"]], "genome_kit._twobitutils": [[1, 4, 1, "", "write2bit"]], "genome_kit.data_manager": [[1, 0, 1, "", "ProgressPercentage"]], "genome_kit.data_manager.ProgressPercentage": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__init__"], [1, 3, 1, "", "__weakref__"]], "genome_kit.gk_data": [[1, 4, 1, "", "get_file"], [1, 4, 1, "", "upload_file"], [1, 4, 1, "", "wget"]]}, "objnames": {"0": ["py", "class", "Python class"], "1": ["py", "method", "Python method"], "2": ["py", "property", "Python property"], "3": ["py", "attribute", "Python attribute"], "4": ["py", "function", "Python function"], "5": ["py", "module", "Python module"]}, "objtypes": {"0": "py:class", "1": "py:method", "2": "py:property", "3": "py:attribute", "4": "py:function", "5": "py:module"}, "terms": {"": [0, 1, 2, 5], "0": [0, 1, 4, 5], "00": [1, 5], "01": 5, "0123456789": 5, "02": 5, "03": 5, "04": 5, "05": 5, "0500000007451": 5, "06": [1, 5], "07": 5, "08": 5, "09": 5, "1": [0, 1, 2, 4, 5], "10": [0, 1, 5], "100": [1, 5], "100nt": 5, "10l": 5, "10nt": 5, "11": [0, 1, 5], "117": [0, 5], "117120016": 5, "117120017": 5, "117120148": 5, "117120149": 5, "117120150": 5, "117120151": 5, "117120185": 5, "117120188": 5, "117120190": 5, "117120192": 5, "117120201": 5, "117231957": 5, "117231963": 5, "117231964": 5, "117231967": 0, "117231970": 0, "117231971": 5, "117231972": 5, "117231973": 5, "117231974": 5, "117231977": 0, "117231979": 5, "117231980": 5, "117231981": 5, "117231987": [0, 5], "117231997": 0, "117232004": 5, "117232009": 5, "117232014": 5, "117232020": 5, "117232022": 5, "117232023": 5, "117232024": 5, "117232030": 5, "1196293": 5, "12": [0, 1, 5], "120": 5, "125": 1, "126": 1, "127": 1, "128": 1, "13": [0, 4, 5], "14": [0, 5], "147": 1, "149": 5, "15": [0, 1, 5], "150": [1, 5], "151": 1, "16": [0, 1, 5], "169295": 5, "169296": 5, "17": [0, 5], "18": [0, 5], "185": 5, "19": [0, 5], "198992": 5, "1bp": 1, "1m1d1m2n1m": 1, "1m1i1m2n1m": 1, "1s2m2n1m1n1m": 1, "1st": [1, 5], "2": [0, 1, 5], "20": [0, 1, 2, 5], "200": 1, "2017": 1, "2017_05": 1, "2019": 2, "20190906": [1, 4], "20m100n10m100n3m": 1, "21": 5, "22": 5, "23": [1, 5], "231": [0, 5], "24": 5, "25": [1, 5], "250": 1, "255": 1, "26": 5, "260": 1, "27": 5, "28": 5, "286": 1, "29": 5, "2bit": [1, 2, 3], "2e": 1, "2nd": [1, 5], "3": [0, 1, 5], "30": 5, "300": [1, 5], "31": 1, "32": 1, "320861": 5, "320862": 5, "325": 1, "357": 1, "38": 1, "3a": 1, "3l": 5, "3p": [0, 1, 5], "3rd": 5, "4": [0, 1, 5], "400": 1, "40000": 1, "40030": 1, "40031": 1, "40033": 1, "40061": 1, "40080": 1, "41246881": 1, "41246886": 1, "48": 1, "4e38": 1, "5": [0, 1, 5], "500": 1, "50000": 1, "50010": 1, "500201": 1, "5002010": 1, "558099": 5, "57": 1, "571": 1, "5bp": 1, "5m100n10m50n5m": 1, "5m8d5m": 1, "5p": [0, 1, 5], "6": [0, 1, 4, 5], "605573": 5, "63": 1, "64": 2, "65504": 1, "69345349": 1, "69373418": 1, "69373422": 1, "69374349": 1, "7": [0, 1, 5], "70220767": 1, "70248838": 1, "70248842": 1, "70249769": 1, "724078": 5, "754": 1, "7841": 1, "7nt": 5, "8": [0, 1, 5], "857": 1, "88": [1, 4], "9": [0, 1, 5], "91661": 5, "91662": 5, "949522": 5, "949523": 5, "949608": 5, "949696": 5, "949739": 5, "95": [1, 4], "967": 0, "977": 0, "977028": 5, "977030": 5, "977329": 5, "977330": 5, "977515": 5, "977516": 5, "977569": 5, "977570": 5, "977630": 5, "978603": 5, "978604": 5, "978627": 5, "978628": 5, "98": [1, 4], "982": 0, "984": 0, "987": 0, "993": [0, 5], "994": 5, "996": 0, "A": [0, 1, 5], "AT": 5, "As": [0, 1], "Be": 1, "But": 5, "By": [0, 1, 5], "For": [0, 1, 2, 3, 5], "INS": 1, "If": [0, 1, 2, 3, 5], "In": [0, 1, 2, 4, 5], "It": [0, 1, 2, 5], "Not": 1, "Of": 0, "On": 2, "One": 1, "Or": 1, "The": [0, 1, 2, 3, 4, 5], "Then": 0, "There": [0, 1], "These": [1, 5], "To": [0, 1, 2, 5], "_": 1, "__________": 5, "_____________": 5, "__call__": 1, "__getitem__": 1, "__init__": [1, 5], "__len__": 1, "__new__": 1, "__reduce__": 1, "__repr__": 1, "__str__": 1, "__weakref__": 1, "_build_appri": 1, "_cxx": [1, 2], "aa": 1, "aaaaabbbbb": 5, "aaaannnnccaaaatttt": 1, "aaagct": 0, "aacaa": 5, "aactttttttcag": 5, "aattggaagcaaa": 5, "ab": 1, "abca4": 1, "abca4_vcf": 1, "abl": [1, 2], "about": [1, 4, 5], "abov": [0, 1, 5], "accept": [1, 5], "acceptor": [0, 1, 5], "access": 4, "accommod": 0, "accord": [1, 5], "accordingli": [0, 1], "accumul": 1, "acgtacgt": 1, "acgtn": 1, "achiev": 0, "across": 1, "act": 1, "activ": [2, 3, 5], "actual": [0, 1], "ad": [1, 4, 5], "add": 1, "addd": 1, "addit": [0, 1], "adjac": 1, "advertis": 5, "af": 5, "af_exac": 1, "affect": [0, 1], "affin": 1, "after": 1, "ag": [1, 5], "again": 0, "against": [1, 5], "alias": 3, "align": [0, 2, 4, 5], "all": [1, 2, 4, 5], "allel": [1, 5], "alloc": 1, "allow": [0, 1, 5], "allow_outside_chromosom": 1, "alreadi": 1, "also": [0, 1, 2, 4, 5], "alt": [1, 5], "alter": 1, "altern": [1, 5], "alwai": [0, 1, 5], "ambigu": [1, 5], "amount": [0, 1], "an": [1, 3, 4, 5], "anaconda": 5, "analogu": 1, "analysi": 1, "analyz": 1, "ancestr": 1, "anchor": [1, 4, 5], "anchor_offset": [0, 1, 5], "anchored_3p": 5, "anchored_5p": 5, "ani": [0, 1, 2, 3, 5], "anno": [1, 2], "annot": [2, 4], "anoth": 1, "anywher": 0, "api": [2, 4, 5], "appdir": [3, 5], "appear": 1, "appl": 2, "appli": [0, 1, 5], "applic": [1, 5], "apply_vari": 1, "appri": [1, 2], "appris_index_nam": 1, "appris_princip": 1, "appris_transcript": 1, "apprisnotavailableerror": 1, "appropri": 0, "approxim": 1, "ar": [0, 1, 2, 4, 5], "arang": 5, "arbitrari": [0, 1], "arg": 1, "argument": [0, 1, 5], "around": [0, 5], "arrai": [1, 5], "as_dna0": 1, "as_dna1": 1, "as_negative_strand": 1, "as_opposite_strand": [1, 5], "as_positive_strand": 1, "as_rna1": 1, "as_str": 1, "as_ucsc": [1, 5], "as_variant_str": 1, "ascii": 1, "assembli": [1, 2, 5], "assign": 1, "associ": [1, 5], "assum": [1, 5], "astyp": 1, "atcttaaagtttctgtgtct": 0, "attach": 1, "attempt": [1, 5], "attribut": [1, 5], "auth": 5, "autocomplet": 1, "autodoc": 1, "automat": [1, 2, 5], "autotick": 2, "avail": [1, 5], "avoid": 1, "awai": 1, "awar": 1, "axi": 5, "b": [1, 2, 5], "b37": 1, "back": [0, 5], "backend": 1, "bam": [1, 4], "bar": [0, 1], "base": [0, 1, 2, 5], "basepair": 1, "bash": [3, 5], "basic": [1, 4], "bat": 2, "bcftool": 1, "becaus": [1, 5], "becom": 1, "bed": 1, "bedfil": 1, "bedgraph": 1, "been": 1, "befor": [0, 1, 2, 5], "begin": 1, "behav": 1, "behaviour": [0, 1, 2, 5], "being": 1, "below": [1, 4, 5], "besid": 5, "best": [0, 1, 5], "between": [0, 1, 5], "bewar": 1, "bigzip": 3, "bin": 1, "binari": [1, 2, 5], "binary_vers": 1, "binomi": 1, "bioconda": 2, "biotyp": 1, "bit": 1, "bits_per_datum": 1, "bitwidth": 1, "blob": 1, "block": 1, "bnd": 1, "bodi": 1, "bool": 1, "bool_": 1, "boolean": 1, "bootstrap": 1, "bot": 2, "both": [1, 2, 5], "bp": 1, "branch": [2, 5], "brca1": 1, "brca1_interv": 1, "break": 2, "breast": 1, "breviti": 5, "broadcast": 1, "browser": [1, 5], "brute": 1, "bucket": 5, "build": [1, 3, 4, 5], "build_gencod": [1, 3], "build_jralign": 1, "build_ncbi_refseq": 1, "build_ralign": 1, "build_rdist": 1, "build_ucsc_refseq": 1, "build_vcfbin": 1, "builder": [1, 5], "built": [1, 2, 5], "byte": 1, "bytes_amount": 1, "c": [0, 1, 3, 5], "ca": [1, 5], "cach": 1, "call": [0, 1, 4, 5], "callabl": 1, "caller": 1, "can": [0, 1, 2, 4, 5], "cancer": 1, "candid": 5, "cannot": [0, 1], "canon": 1, "carri": 5, "case": [0, 1, 5], "cat": 1, "categori": 1, "caus": 1, "caveat": 1, "ccaaact": 5, "ccaact": 5, "ccaactt": 5, "ccach": 2, "ccatgca": 5, "cccagattgg": 5, "cccc": 5, "ccds53759": 1, "ccds_id": 1, "ccdsid": 1, "ccgttca": 5, "cctaact": 5, "cctact": 5, "cd": 5, "cdss": [1, 5], "center": [0, 1, 5], "certain": 5, "cf": 2, "cfg": [1, 3], "cftr": [0, 5], "chang": [0, 1, 4], "chapter": 0, "charact": 1, "check": [0, 2, 5], "child": 1, "chmod": 5, "chr": [1, 5], "chr1": [1, 5], "chr17": 1, "chr2": 1, "chr5": 1, "chr7": [0, 5], "chrom": [1, 3, 5], "chromalia": 3, "chromosom": [0, 1, 5], "chromosome_s": 1, "ci": 2, "cigar": 1, "cl": 1, "clamp": 1, "clangxx_osx": 2, "class": [1, 5], "clear": 0, "clinvar": [0, 5], "clinvar_20180128": 1, "clion": 4, "clip": 1, "clnsig": 1, "clone": 2, "close": 1, "cloud": 5, "cmake": 2, "cnv": 1, "coars": 1, "coarsen": 1, "coarser": 1, "code": [0, 1, 2, 5], "codon": 1, "column": [1, 5], "com": [1, 2, 5], "come": [0, 5], "comma": 5, "command": 2, "comment": 2, "common": 1, "commonli": 4, "compact": 5, "compar": 1, "compat": [1, 2], "compil": [2, 3, 5], "complement": [1, 5], "complet": [0, 1], "complex": [1, 5], "complic": 0, "comprehens": 4, "compris": 1, "comput": 1, "concept": [0, 5], "conceptu": [0, 1], "concret": 0, "conda": [2, 3], "conda_prefix": 2, "condit": 1, "config": 1, "configur": [1, 5], "confirm": 0, "confus": 1, "connect": 1, "consecut": 5, "conserv": 4, "consid": [0, 1, 5], "consist": 1, "constant": [0, 1], "constructor": [1, 5], "contain": [0, 1, 2, 3, 5], "content": [1, 4], "context": 1, "contig": 3, "contigu": 1, "continu": 2, "contribut": 1, "control": [1, 2, 5], "conveni": [0, 1, 4, 5], "convent": [0, 1, 4, 5], "convers": 1, "convert": [1, 5], "coord": 5, "coordin": [0, 1, 5], "copi": [1, 2, 3], "core": [0, 1, 5], "correct": [0, 1], "correctli": 1, "correspond": [1, 5], "could": [0, 1, 5], "count": 1, "cours": 0, "cover": [1, 5], "cpp": 2, "creat": [1, 2, 3, 5], "creation": 1, "creativ": 4, "credenti": 5, "criteria": 5, "crop": 1, "ct": [0, 1, 5], "ctgaaaaaaagtt": 5, "ctgggtaaagctgtgtctgt": 0, "cttaaagtttctgtgtctgt": 0, "curl": 5, "current": [1, 2, 5], "custom": [1, 5], "cxx": 2, "d": [1, 5], "data": 4, "data_dir": [1, 5], "data_manag": [4, 5], "data_s": 1, "databas": [1, 5], "datamanag": 5, "date": 1, "datum": 1, "dcmake_build_typ": 2, "ddx11l1": 5, "de": 0, "debugg": 2, "declar": 1, "decod": 1, "decor": 1, "decreas": 1, "dedic": 1, "deep": 4, "deepgenom": [2, 5], "def": [1, 5], "default": [0, 1, 2, 5], "default_valu": 1, "defin": [0, 1, 2, 5], "definit": 1, "del": 1, "dela": 5, "delet": [1, 4, 5], "delta": 1, "demand": [1, 2], "demo": [1, 2, 5], "denot": [0, 5], "dep": 2, "depend": [1, 2, 4, 5], "depth": [1, 5], "describ": 5, "descript": [1, 5], "design": 4, "desir": 1, "destin": 1, "detail": [1, 4, 5], "determin": 1, "develop": [1, 4], "dganno": [1, 2, 3], "dict": [1, 5], "dictionari": 1, "differ": [0, 1, 2, 4, 5], "dim": 1, "dimens": 1, "dimension": 1, "dinucleotid": 0, "diploid": 1, "direct": 1, "directli": [0, 1, 2, 5], "directori": [1, 2, 3, 5], "disabl": 1, "disambigu": [1, 4], "discov": 2, "disjoint": 1, "disk": 1, "displai": 1, "distanc": 1, "distinct": 1, "distinguish": [0, 1], "distribut": [1, 2], "diverg": 1, "divid": 1, "dna": [0, 1, 4, 5], "dna0": [1, 5], "dna1": [1, 5], "dnstream": 1, "dnstream_of": [1, 5], "do": [0, 1, 5], "doc": 1, "document": [2, 4], "doe": [1, 5], "doesn": 1, "domin": 1, "done": 0, "download": [1, 2, 5], "downsampl": 1, "downstream": [1, 5], "dp": 1, "dst": 1, "dtype": [1, 5], "due": 1, "dup": 1, "duplic": 1, "dure": 1, "dynam": 1, "e": [0, 1, 2, 5], "each": [1, 5], "earlier": 5, "easi": [4, 5], "easili": [0, 1, 4], "ebi": 1, "edit": 2, "edu": 3, "effect": [0, 5], "eg": 1, "either": [1, 2, 5], "element": [0, 1, 5], "empti": [1, 5], "en": [1, 5], "enabl": [1, 5], "encod": 1, "encount": 1, "end": [0, 1, 5], "end3": [1, 5], "end5": [1, 5], "endpoint": 5, "enough": 2, "ense00000846737": 1, "ensembl": [1, 3, 4, 5], "ensg00000001626": 5, "ensg00000115275": 1, "ensg00000223972": 5, "ensp00000233616": 1, "enst00000003084": 5, "enst00000233616": 1, "enst00000426809": 5, "enst00000432964": 5, "enst00000454343": 5, "enst00000456328": 5, "enst00000466430": 5, "enst00000466557": 5, "enst00000599771": 5, "enst00000601486": 5, "ensur": [1, 2], "entir": [0, 1, 5], "entrez": 1, "entri": [1, 5], "entry_point": 5, "env": [2, 5], "environ": [2, 5], "envvar": 2, "equal": [0, 1, 5], "equival": [0, 1, 5], "error": 1, "especi": 1, "est": 1, "estim": 1, "eta": 1, "etc": [1, 5], "etyp": 1, "even": [0, 1], "ever": 1, "everi": [1, 5], "evid": 1, "exact": 1, "exactli": [1, 5], "examin": 0, "exampl": [0, 1, 2, 4, 5], "except": [1, 5], "excerpt": 2, "exclud": [1, 5], "exclus": [1, 5], "exist": 1, "exon": [0, 5], "exonic_interv": 1, "exont": 5, "expand": [0, 1, 5], "expect": [0, 1, 5], "explain": [0, 4, 5], "explan": [1, 5], "explicitli": 1, "export": 5, "expos": 1, "express": 1, "extend": [0, 1], "extens": 2, "extent": 1, "extra": [0, 1], "extract": [0, 1, 4], "extract_featur": 5, "f": [1, 2], "f16": 1, "f2": 1, "f3": 1, "f32": 1, "f4": 1, "f5": 1, "f6": 1, "f8": 1, "fa": 1, "face": 1, "facilit": 1, "fail": 1, "failur": 2, "fall": [0, 1, 5], "fals": [1, 5], "far": 1, "fast": [1, 2, 4, 5], "fasta": 3, "faster": 1, "fastest": 5, "fatotwobit": 3, "featur": [0, 2], "feedstock": 2, "fetch": 1, "few": [3, 5], "field": [1, 5], "file": [1, 3, 4], "fileformat": 5, "filenam": [1, 5], "filepath": [1, 5], "files": 1, "fill": [1, 5], "filter": [1, 4, 5], "final": [1, 2, 5], "find": [1, 2, 4], "find_3p_align": [1, 5], "find_3p_within": [1, 5], "find_5p_align": [1, 5], "find_5p_within": [1, 5], "find_exact": [1, 5], "find_motif": [1, 5], "find_overlap": [1, 5], "find_overlapping_match": 5, "find_overlapping_motif": [1, 5], "find_within": [1, 5], "find_x": 1, "fine": 1, "finish": 1, "first": [0, 1, 5], "first_by_nam": 1, "five": [0, 5], "fix": [0, 1], "fixedstep": 1, "flag": [1, 2], "flank": 5, "float": [1, 5], "float16": 1, "float32": [1, 5], "flush": 1, "fmt": 2, "fmt_id": [1, 5], "follow": [0, 1, 2, 3, 5], "foo": 1, "forc": 1, "forg": [2, 5], "forget": 2, "format": [0, 1, 2, 3, 5], "format_id": 1, "forward": [1, 5], "found": [1, 5], "four": [1, 5], "fourth": 5, "frame": 1, "free": 1, "frequenc": 5, "freshli": 1, "from": [0, 1, 2, 3, 4], "from_dna0": 1, "from_dna0_coord": [1, 5], "from_dna1": 1, "from_jralign": 1, "from_rna1": 1, "from_str": [1, 5], "from_vcf": [1, 5], "full": 1, "function": [0, 1, 5], "further": 1, "furthermor": 0, "futur": 4, "g": [0, 1, 5], "gacagaacca": 5, "gap": [0, 1], "gc": 5, "gccagaggac": 5, "gcloud": 5, "gencod": [1, 2, 3, 4, 5], "gene": 5, "gene_typ": 1, "geneid": 1, "gener": [0, 1, 3, 5], "genom": [0, 2, 4], "genome37": 1, "genome_kit": [0, 2, 3, 5], "genomeannot": [3, 5], "genomekit": [0, 1, 3, 5], "genomekit_data_dir": 5, "genomekit_dev": 2, "genomekit_gcs_billing_project": 5, "genomekit_gcs_bucket": 5, "genomekit_quiet": 5, "genomekit_trac": 5, "genometrack": 5, "genometrackbuild": 5, "genotyp": [1, 4, 5], "get": [0, 1, 2, 3, 4, 5], "get_fil": [1, 5], "gff3": [1, 2, 3], "ggg": 0, "git": 2, "github": [1, 2], "githubusercont": 5, "give": [1, 4], "given": [0, 1, 5], "gk": [3, 5], "gk_check": 2, "gk_data": [4, 5], "gk_debugbreak": 2, "gmod": 1, "go": 0, "goal": 4, "goldenpath": 3, "good": 5, "googl": 5, "google_cloud_project": 5, "grch37": [4, 5], "grch38": 4, "grcm38": 4, "grcm39": 4, "greater": [0, 1], "group": 5, "grow": 0, "gt": [1, 5], "gt_heterozygous_phased_0_1": 1, "gt_heterozygous_phased_1_0": 1, "gt_heterozygous_unphas": 1, "gt_homozygous_alt": 1, "gt_homozygous_ref": 1, "gt_unknown": 1, "gtex": 1, "gtrack": [1, 5], "gtrack_vers": 1, "gttt": 0, "gxx_linux": 2, "gz": [1, 2, 5], "h": 1, "ha": [1, 2, 5], "hacki": 1, "had": [0, 1], "half": 1, "handl": [0, 1, 5], "happen": 0, "has_coding_seq": 5, "has_good_level": 5, "have": [0, 1, 2, 5], "headach": 1, "header": 1, "help": [2, 3], "helper": 1, "her": 1, "here": [0, 1, 5], "hg19": [0, 1, 4, 5], "hg38": [1, 3, 4, 5], "hgdownload": 3, "hi": 1, "hierarchi": 1, "higher": 1, "highli": 1, "hint": 1, "histogram": 1, "histor": 1, "hold": 1, "home": 2, "how": [0, 1, 2, 4, 5], "howev": [0, 1, 5], "ht": 1, "html": [1, 3, 5], "http": [1, 2, 3, 5], "httperror": 1, "i": [0, 1, 2, 3, 5], "i8": 1, "id": [1, 5], "idea": 5, "ideal": 0, "ident": [1, 5], "identifi": 1, "ieee": 1, "ignor": [1, 5], "immedi": [1, 5], "implement": [1, 5], "impli": [0, 1, 5], "import": [0, 1, 2, 3, 5], "importantli": [0, 5], "in_clion": 2, "includ": [0, 1, 2, 4, 5], "include_dupl": 1, "include_vari": 1, "inclus": [1, 5], "incompat": 1, "incorpor": 1, "incorrectli": 1, "increas": 1, "inde": 1, "indel": [0, 1, 5], "independ": [0, 1], "index": [1, 4, 5], "index_of": [1, 5], "index_s": 1, "indexerror": 1, "indic": [0, 1, 5], "individu": 5, "infer": 1, "infil": 1, "info": [1, 5], "info_id": [1, 5], "inform": 1, "inherit": 1, "initi": 1, "input": [0, 1], "insert": [1, 4, 5], "insid": [0, 1], "instal": [1, 2, 4], "instanc": [1, 5], "instanti": 1, "instead": [0, 1], "instruct": [2, 3], "int": 1, "int16": 1, "int32": [1, 5], "int8": [1, 5], "integ": [0, 1, 5], "integr": 2, "intend": [1, 5], "intent": 1, "interest": [0, 5], "intermedi": 1, "intern": [4, 5], "interpret": [1, 5], "interrupt": 1, "intersect": 1, "interv": 4, "interval1": 1, "interval2": 1, "interval_a": 1, "interval_b": 1, "interval_c": 1, "intr": 1, "intron": 5, "intronic_interv": 1, "inv": 1, "invalid": 1, "involv": 1, "io": [1, 5], "ioerror": 1, "irrespect": 5, "is_positive_strand": 1, "isoform": 1, "iter": 1, "its": [0, 1, 5], "itself": 1, "jetbrain": 4, "jralign": 1, "jralign_vers": 1, "junc": 1, "junction": [2, 4], "junction_index": 1, "just": [2, 5], "keep": [0, 1], "kei": 1, "kept": [0, 1], "keyerror": 1, "kick": 2, "kind": 1, "know": [1, 5], "kwarg": 1, "lack": 1, "lambda": 1, "larg": 1, "larger": 5, "last": [0, 1], "latest": 5, "lead": [2, 5], "learn": [0, 4, 5], "least": [1, 5], "left": [0, 1], "len": [0, 1, 5], "length": [0, 1, 4], "let": [0, 4, 5], "level": [1, 5], "li": 1, "librari": [1, 4], "library_format": 1, "library_typ": 1, "lift": 5, "lightweight": 4, "like": [0, 1, 4, 5], "limit": 1, "line": [0, 1, 5], "linear": 1, "link": [1, 2], "linspac": 1, "linux": 2, "list": [1, 4, 5], "list_available_genom": [1, 4], "liter": 1, "live": 5, "ll": [0, 1, 2], "load": 1, "local": [1, 2, 5], "locat": [1, 2, 5], "log": 4, "login": 5, "long": 1, "longer": 1, "look": 0, "lookup": 1, "loop": 1, "lower": 1, "m": [1, 2], "m0": 1, "m1": 2, "m39": [1, 4], "mac": 2, "macaca_fascicularis_5": 4, "macfas5": [1, 4], "macfas_5": [1, 4], "machin": [0, 4], "made": [1, 4, 5], "mai": [1, 2, 5], "main": [2, 5], "make": [1, 4, 5], "malform": 1, "manag": [1, 5], "mani": [0, 2, 5], "manner": 5, "manual": 0, "map": [1, 5], "mark": 1, "mask": [1, 5], "master": 1, "match": [1, 5], "match_posit": [1, 5], "mateid": 1, "matric": 1, "matrix": 1, "max": 1, "max_int": 0, "maximum": 1, "mean": [1, 5], "meaning": 1, "mechan": [0, 5], "meet": 5, "member": 1, "memori": 1, "menu": 2, "mere": 1, "merg": [1, 2], "messag": 1, "metadata": [1, 3, 5], "method": [0, 1, 2, 5], "middl": [0, 1], "midpoint": 1, "might": 2, "min": 1, "min_delta": 1, "min_overhang": 1, "min_read": 1, "min_run": 1, "mind": 0, "minimum": 1, "mismatch": 1, "miss": 1, "mix": 1, "mixtur": 5, "mm10": [1, 4], "mm39": [1, 4], "mnt": 1, "mock": 1, "mock_result": 1, "mock_unreach": 1, "mode": [2, 5], "model": 1, "modif": 1, "modifi": [0, 1], "modul": [1, 2, 4], "mog": 1, "more": [0, 1, 4, 5], "most": [0, 1, 5], "motif": [0, 1, 4], "mrna": 1, "mt": [0, 4], "multi": 1, "multidimension": 1, "multipl": 1, "must": [1, 2, 3], "mutant": [0, 5], "my_anno": 3, "my_assembli": 3, "my_fil": 1, "mydatamanag": 5, "mydna": 1, "myfil": 1, "mypackag": 1, "n": [1, 2, 3], "name": [1, 5], "namespac": 5, "nan": 1, "nativ": 2, "na\u00efv": 0, "ncbi": [1, 3, 4], "ncbi37": 1, "ncbi_refseq": 1, "ndarrai": 1, "nearest": 5, "necessari": [0, 1], "need": [0, 1, 2, 3, 5], "neg": [1, 5], "neg_strand_fil": 1, "never": 1, "new": [1, 2, 4], "newer": [1, 2], "next": [1, 5], "next_cd": 1, "next_exon": [1, 5], "next_intron": 1, "nm_000344": 1, "nm_001297715": 1, "nm_006302": 1, "nm_022874": 1, "nnnnnnngggggggg": 1, "non": [1, 5], "non_cod": 1, "none": [1, 5], "norm": 1, "normal": [1, 2], "not_first_exon": 5, "note": [1, 5], "noth": 1, "notic": 5, "notifi": 1, "notion": 0, "novo": 0, "now": [0, 1, 2, 3, 5], "np": [1, 5], "np_006293": 1, "nuclear": 5, "nucleotid": [1, 5], "num_bit": 1, "num_count": 1, "num_read": 1, "num_sampl": 1, "num_vari": 1, "number": [1, 5], "numer": 1, "numpi": [1, 5], "nxm": 1, "o": 5, "object": [0, 1, 4, 5], "observ": 1, "occupi": 1, "occur": [0, 1], "occurr": 1, "odd": 1, "off": [0, 2, 5], "offset": [0, 1], "often": [0, 1], "ok": 1, "omit": 1, "onc": [1, 2], "one": [0, 1, 5], "ones": [1, 4], "onli": [0, 1, 2, 5], "open": [1, 2, 5], "open_clinvar": 1, "oper": 5, "opposit": [0, 1, 5], "optim": 1, "option": [1, 2, 3, 5], "order": [1, 5], "org": [1, 5], "origin": 1, "osx": 2, "other": [0, 1, 2, 4, 5], "otherwis": 1, "our": [0, 1], "out": [0, 1, 2], "outfil": 1, "output": [1, 5], "outsid": [1, 4], "over": [1, 5], "overestim": 1, "overhang": 1, "overhang_error": 1, "overlap": [0, 1, 5], "overrid": 1, "own": [1, 5], "p12": [1, 4], "p13": [1, 4], "p14": [1, 4], "p6": [1, 4], "packag": [2, 5], "pad": 5, "page": 4, "pai": 5, "pair": 1, "parallel": 2, "param": 1, "paramet": [1, 5], "parent": 1, "parenthes": 0, "pars": 1, "part": [1, 5], "partial": 1, "particular": [0, 1], "pass": [0, 1, 5], "pass_filt": 1, "patch": 4, "path": 1, "pdf": 1, "per": [1, 5], "percentag": 1, "perform": 1, "phase": 1, "phastcon": 1, "phastcons_mamm": 1, "pickl": 1, "pileup": 1, "pileup_a": 1, "pileup_c": 1, "pileup_del": 1, "pileup_g": 1, "pileup_t": 1, "pip": 2, "pipelin": [1, 2], "place": 5, "plan": 5, "pleas": [1, 2], "plu": 5, "plugin": 5, "plusmt": [1, 4], "po": 5, "point": [1, 5], "polici": 1, "pool": 1, "portion": 1, "pos_strand_fil": 1, "posit": [0, 1, 5], "possibl": [0, 1], "possibli": 1, "potenti": 5, "power": [1, 4], "pr": 2, "practic": 0, "pre": [2, 5], "preced": [1, 5], "precis": 1, "preconfigur": 2, "predict": 0, "prefer": 1, "prerequisit": 2, "preserv": 1, "prev_cd": 1, "prev_exon": 1, "prev_intron": 1, "previou": [0, 1, 5], "previous": 1, "primarili": 0, "princip": 1, "principl": [0, 1], "prinicp": 1, "print": [1, 3, 5], "prioriti": 1, "privat": 1, "problem": 0, "process": [1, 2], "product": 1, "progress": 1, "progress_msg": 1, "progresspercentag": 1, "project": [2, 5], "prompt": 2, "propag": 1, "properli": 1, "properti": 1, "protein": 1, "protein_cod": 1, "protein_id": 1, "provid": [1, 2, 5], "public": 5, "publish": 2, "pull": [0, 2], "purpos": 1, "push": 0, "put": 1, "pwm": 5, "py": [1, 2, 5], "pycharm": 1, "pydeletemockattr": 1, "pylint": 1, "pypa": 5, "python": [1, 2, 3, 4, 5], "qual": 5, "qualiti": 1, "quantiti": 1, "queri": [1, 4, 5], "query_vcf": [1, 5], "quick": 5, "quickstart": [1, 4], "r": [1, 5], "rais": [1, 5], "ralign": 1, "ralign_vers": 1, "random": 1, "rang": [1, 5], "rare": 0, "rather": [0, 5], "raw": 5, "rcr": 4, "rdist": 1, "rdist_vers": 1, "re": [1, 2], "reach": 1, "read": [2, 4, 5], "read_index": 1, "readthedoc": 1, "reason": [1, 2, 5], "recogn": 1, "recommend": 1, "reduc": 1, "ref": [1, 5], "refer": [0, 1, 4, 5], "reference_align": 1, "reference_genom": [1, 5], "refg": [1, 3], "refgen": 1, "reflink": 1, "refseq": [1, 4, 5], "regener": 1, "region": 1, "regist": [1, 2, 5], "regro": 2, "rel": [1, 5], "releas": 4, "relev": 1, "reli": 2, "remain": 5, "remot": [1, 2, 5], "remote_fil": 1, "remov": [0, 1], "repeat": [1, 5], "repeatedli": 2, "repo": 2, "report": 1, "repr": 1, "repres": [1, 5], "represent": 1, "request": [1, 5], "requir": [1, 2, 3, 5], "research": 4, "reserv": [1, 2], "resid": 2, "resolut": 1, "resourc": [1, 5], "respect": [1, 5], "respons": 1, "rest": 1, "restrict": 1, "result": [0, 1, 3, 5], "result_typ": 1, "retain": 1, "retriev": 1, "return": [0, 1, 5], "revers": [1, 5], "rgsc": 4, "ride": 0, "right": [0, 1], "rn6": [1, 4], "rna1": 1, "rnor_6": [1, 4], "rough": 1, "round": 1, "row": [1, 5], "rs397508328": 5, "rs397508657": 5, "rs397508673": 5, "rs397508710": 5, "rtype": 1, "rule": 1, "run": [1, 2, 5], "runtim": 1, "sake": 1, "salmon": 1, "sam": [1, 4], "same": [0, 1, 5], "sampl": [1, 5], "sample000132": 1, "sample1": [1, 5], "sample2": [1, 5], "sample3": [1, 5], "sample_nam": 1, "samtool": 1, "samv1": 1, "save": [1, 5], "scalar": 1, "scan": 4, "scenario": 1, "schema": 3, "scheme": 1, "scope": 1, "score": 1, "script": [3, 5], "sdk": 2, "search": [1, 4, 5], "second": [1, 5], "secondari": 4, "section": [1, 5], "see": [0, 1, 2, 5], "seen": 0, "segment": 1, "select": [1, 2, 5], "self": [1, 5], "sens": [1, 5], "sensit": 1, "separ": [1, 5], "separate_dim": 1, "seq": 1, "sequenc": [0, 1, 4, 5], "sequence_vari": 1, "serv": [1, 5], "server": 1, "set": [0, 1, 4, 5], "set_clamp": 1, "set_data": [1, 5], "set_data_from_b": 1, "set_data_from_bedgraph": 1, "set_data_from_wig": 1, "set_default_valu": 1, "set_dict": 1, "set_restrict": 1, "set_spars": 1, "set_transform": 1, "setup": 4, "setuptool": 5, "sever": 5, "sf": 1, "sgsc": 4, "sh": [3, 5], "shape": [1, 5], "share": [1, 4], "shell": 2, "shift": 1, "short": [0, 4, 5], "shortcut": 1, "shorter": 0, "shorthand": 1, "should": [0, 1, 5], "show": [0, 1, 5], "shrink": [0, 5], "side": [0, 1], "sign": 1, "signific": 1, "similarli": 1, "simpli": 1, "sinc": [0, 1, 5], "singl": [0, 1, 5], "single_strand": [1, 5], "site": [0, 1, 2, 5], "size": [0, 1, 2, 3], "slice": 5, "slightli": 0, "sln": 2, "small": [1, 2], "smallest": 1, "smn1": 1, "smn2": 1, "snapshot": 4, "snv": 5, "so": [0, 1, 2, 5], "socket": 1, "soe": 3, "sole": 1, "solv": 0, "some": [0, 1, 4, 5], "sort": 1, "sourc": [1, 2, 5], "space": [1, 5], "span": [1, 5], "spars": 1, "sparsif": 1, "sparsiti": 1, "spec": 1, "special": [0, 1, 5], "specif": [0, 1, 5], "specifi": [1, 3, 4, 5], "speed": [1, 2], "sphinx": 1, "splice": [0, 1, 5], "split": [0, 1], "spuriou": 1, "sr": 1, "src": 2, "sscrofa11": [1, 4], "stabl": 1, "stai": [0, 5], "standard": [1, 4, 5], "start": [0, 1, 2, 3, 4, 5], "starter": [3, 5], "starter_build": 5, "statement": 1, "static": 1, "stdin": 1, "step": [1, 2], "stick": 0, "still": [0, 1], "stop": 1, "storag": [1, 5], "store": [1, 2, 5], "str": [1, 5], "straight": 5, "strand": [1, 5], "strand_awar": [1, 5], "strand_unawar": [1, 5], "stranded": [1, 5], "stream": 1, "strictli": 1, "string": [0, 1, 5], "strip": 1, "strip_mock_bas": 1, "structur": [1, 4, 5], "studi": [0, 4], "style": 5, "subclass": [1, 5], "subcommand": 2, "subsect": 1, "subset": 1, "substitut": 1, "subtract": 1, "subtyp": 1, "succeed": 1, "suffix": 1, "suit": 5, "sum": [1, 5], "suppli": [1, 5], "support": [1, 2, 3, 5], "suppos": 5, "suppress": 5, "sure": [1, 5], "surround": 5, "suscept": 1, "suspect": 1, "susscr11": [1, 4], "svtype": 1, "svtype_bnd": 1, "svtype_cnv": 1, "svtype_del": 1, "svtype_dup": 1, "svtype_in": 1, "svtype_inv": 1, "svtype_na": 1, "switch": 2, "sy": 1, "symbol": 1, "symmetr": 1, "synonym": 1, "system": [1, 2, 5], "t": [0, 1, 5], "taaaccacat": 1, "tabl": [2, 5], "tag": 1, "take": [1, 5], "target": [1, 2], "tatcttaaagctgtgtgt": 0, "tatcttaaagctgtgtgtaa": 0, "tatcttaaagtttctgtgtc": 0, "tc": 5, "tccaact": 5, "tccaggctat": 5, "tct": [0, 5], "tcttaaagctgtgtctgtaa": 0, "technic": 5, "temp": 1, "ten": 0, "tend": 1, "termin": 1, "test": [1, 4, 5], "test_interv": 2, "test_seri": 2, "testinterv": 2, "text": 1, "tgcagggaac": 5, "tggg": 0, "tgt": 0, "tgtaggaaac": 5, "than": [0, 1, 5], "thei": [1, 2, 3, 5], "them": [1, 2, 5], "therefor": [0, 1, 5], "thi": [0, 1, 2, 5], "think": 1, "third": 5, "those": [1, 4, 5], "though": 1, "thought": 5, "three": 0, "through": [0, 1], "throughput": 1, "time": 1, "timeout": 1, "timestamp": 1, "tini": 2, "todo": 1, "togeth": 5, "toggl": 1, "toi": 5, "tool": [2, 3], "top": 5, "toward": 1, "trace": 5, "track": [2, 4], "trade": 1, "tran": [1, 5], "transcript": 5, "transcript_typ": 1, "transfer": 1, "transform": 1, "transpar": 5, "treat": 1, "tree": 2, "trick": 1, "trigger": 1, "trim": [0, 1, 5], "trivial": 1, "troubleshoot": 4, "true": [1, 5], "truncat": 1, "try": 1, "tsl": 1, "tsl1": 1, "tsl2": 1, "tsl3": 1, "tsl4": 1, "tsl5": 1, "tslna": 1, "tt": 5, "ttagtt": 5, "ttatatcttagtgtctgtaa": 0, "ttcagctgct": 5, "ttgatatttatatgttttta": 5, "ttgatatttatatgtttttatatcttaaagctgtgtctgtaaactgatggctaacaaaactaggattttggtc": 5, "ttt": 0, "tttgcttccaatt": 5, "tupl": 1, "tutori": [4, 5], "two": [0, 1, 2, 5], "twobit": 3, "twobitinfo": 3, "txt": [1, 2, 3], "type": [0, 1, 2, 5], "typeerror": 1, "typic": 0, "u": 1, "u1": 1, "u2": 1, "u3": [1, 5], "u4": 1, "u5": 1, "u6": 1, "u8": 1, "ubuntu": 2, "ucsc": [1, 3, 4, 5], "ucsc_db_dir": 1, "ucsc_refseq": 1, "uint8": [1, 5], "ultim": 5, "unambigu": [0, 5], "under": [1, 2, 5], "undergo": 1, "understand": [0, 1], "understood": 0, "unintend": 0, "uniqu": [1, 5], "unit": 1, "unittest": [0, 2], "unittestbuild": 2, "univers": 5, "unless": [1, 5], "unlik": 1, "unpatch": 4, "unpredict": 2, "unreach": 1, "unrecogn": 1, "unspecifi": 1, "unstrand": 1, "unsupport": 1, "until": 1, "up": [0, 1, 4, 5], "updat": 2, "upload": [1, 5], "upload_fil": [1, 5], "upon": [1, 2], "upper": 1, "upstream": [1, 5], "upstream_of": [1, 5], "url": 1, "us": [0, 1, 2, 3, 5], "usag": 1, "user": [1, 2, 4, 5], "user_data_dir": [3, 5], "userguid": 5, "usual": [0, 5], "utr": 1, "utr3": 1, "utr3tabl": 1, "utr5": 1, "utr5tabl": 1, "utrtabl": 1, "v": [1, 2, 5], "v101": [1, 4], "v105": [1, 4], "v109": [1, 4], "v110": [1, 4], "v19": [1, 2, 4, 5], "v22": 1, "v29": [1, 4], "v29lift37": 1, "v41": [1, 4], "v41lift37": 1, "valid": [1, 5], "valu": [0, 1, 5], "valueerror": 1, "var": 5, "var1": 5, "var2": 5, "var3": 5, "var_seq": 0, "variabl": [2, 5], "variablestep": 1, "variant": [0, 4], "variant_dna": 1, "variant_genom": [0, 1, 5], "variant_str": 1, "variantgenom": [0, 4, 5], "vcf": [2, 4], "vcfbin": 1, "vcfbin_vers": 1, "vcfbinpath": 1, "vcfpath": 1, "vcftabl": 5, "vcfv4": 5, "vcfvariant": 5, "vcproj": 2, "vcvars64": 2, "ve": 5, "verbos": 2, "veri": 1, "verifi": 5, "version": [1, 2, 5], "vertic": 0, "via": [1, 2], "view": 1, "visibl": 1, "visual": 5, "vm19": [1, 4], "vm30": [1, 4], "vm31": [1, 4], "w": 5, "wa": [0, 1], "wai": [1, 4, 5], "walk": 5, "walk_annot": 5, "want": [0, 1, 2, 5], "warn": 1, "we": [0, 1, 5], "weak": 1, "well": [0, 5], "were": [0, 1, 5], "wget": 1, "what": [0, 1, 5], "when": [0, 1, 2, 5], "where": [0, 1, 5], "wherea": [0, 1, 5], "whether": [0, 1], "which": [0, 1, 2, 5], "while": [1, 2, 5], "wide": 1, "wig": 1, "wiki": 1, "wild": 0, "window": [0, 1, 2, 5], "wish": [1, 5], "with_anchor": [1, 5], "within": [0, 1, 5], "without": [1, 2, 4, 5], "word": 1, "work": [0, 2, 4], "would": [0, 1, 5], "wrapper": 1, "write": 1, "write2bit": 1, "written": 1, "wt": 0, "x": [1, 5], "x64": 2, "xcode": 2, "xm": 1, "xxxxxyyyyi": 5, "y": [1, 5], "yet": 1, "yml": 2, "you": [0, 1, 2, 3, 4, 5], "your": [1, 2, 3, 5], "yourself": 2, "z": [1, 5], "zero": [0, 1, 5], "zlib": 2, "zzzz": 5}, "titles": ["Anchors", "API Documentation", "Developers", "Genomes", "GenomeKit", "Quickstart"], "titleterms": {"_apply_vari": 1, "_cxx_util": 1, "_twobitutil": 1, "access": [1, 5], "ad": 2, "align": 1, "alignmentmatch": 1, "alignmentmatcht": 1, "alignmentt": 1, "an": 0, "anchor": 0, "annot": [1, 3, 5], "api": 1, "assembli": 3, "avail": 4, "basic": 5, "build": 2, "c": 2, "cd": 1, "cdstabl": 1, "chang": [2, 5], "clion": 2, "conda": 5, "data": [1, 2, 5], "data_manag": 1, "datamanag": 1, "debug": 2, "delet": 0, "develop": 2, "document": 1, "exampl": 3, "exon": 1, "exont": 1, "extract": 5, "featur": [4, 5], "file": [2, 5], "find": 5, "from": 5, "full": 2, "gene": 1, "genet": 1, "genom": [1, 3, 5], "genome_kit": 1, "genomeannot": 1, "genomedna": 1, "genomekit": [2, 4], "genometrack": 1, "genometrackbuild": 1, "gk_data": 1, "i": 4, "indic": 4, "initi": 5, "insert": 0, "instal": 5, "intern": 1, "interv": [0, 1, 5], "intron": 1, "intront": 1, "jetbrain": 2, "jreadalign": 1, "junction": 1, "junctionreadalign": 1, "junctionreadalignmentst": 1, "junctionreadcount": 1, "junctionreaddistribut": 1, "junctionreaddistributiont": 1, "junctiont": 1, "length": 5, "log": 5, "make": 2, "motif": 5, "new": 5, "next": 4, "outsid": 0, "quickstart": 5, "read": 1, "readalign": 1, "readdistribut": 1, "releas": 2, "resourc": 4, "set": 2, "setup": 2, "share": 5, "specifi": 0, "step": 4, "tabl": [1, 4], "test": 2, "todo": 0, "track": [1, 5], "transcript": 1, "transcriptt": 1, "troubleshoot": 5, "up": 2, "us": 4, "usag": 5, "variant": [1, 5], "variantgenom": 1, "variantt": 1, "vcf": [1, 5], "vcftabl": 1, "vcfvariant": 1, "via": 5, "what": 4, "without": 0, "work": 5}})