Search.setIndex({"docnames": ["anchors", "api", "develop", "genomes", "index", "quickstart"], "filenames": ["anchors.rst", "api.rst", "develop.rst", "genomes.rst", "index.rst", "quickstart.rst"], "titles": ["Anchors", "API Documentation", "Developers", "Genomes", "GenomeKit", "Quickstart"], "terms": {"when": [0, 1, 2, 5], "appli": [0, 1, 5], "variant": [0, 4], "genom": [0, 2, 4, 5], "extract": [0, 1, 4], "featur": [0, 2], "most": [0, 1, 5], "importantli": [0, 5], "dna": [0, 1, 4, 5], "sequenc": [0, 1, 4, 5], "from": [0, 1, 2, 3, 4], "result": [0, 1, 3, 5], "variantgenom": [0, 4, 5], "need": [0, 1, 2, 3, 5], "some": [0, 1, 4, 5], "notion": 0, "align": [0, 2, 4, 5], "between": [0, 1, 5], "wild": 0, "type": [0, 1, 2, 5], "mutant": [0, 5], "chang": [0, 1, 4, 5], "length": [0, 1, 5], "furthermor": 0, "mani": [0, 2, 5], "machin": [0, 4], "learn": [0, 4, 5], "method": [0, 1, 2, 5], "expect": [0, 1, 5], "fix": [0, 1], "input": [0, 1], "so": [0, 1, 2, 5], "predict": 0, "unambigu": [0, 5], "mechan": [0, 5], "how": [0, 1, 2, 4, 5], "affect": [0, 1], "genomekit": [0, 1, 3, 5], "solv": 0, "thi": [0, 1, 2, 5], "problem": 0, "through": [0, 1], "call": [0, 1, 4, 5], "rather": [0, 5], "than": [0, 1, 5], "allow": [0, 1, 5], "arbitrari": [0, 1], "i": [0, 1, 2, 3, 5], "singl": [0, 1, 5], "coordin": [0, 1, 5], "usual": [0, 5], "posit": [0, 1, 5], "interest": [0, 5], "splice": [0, 1, 5], "site": [0, 1, 2, 5], "It": [0, 1, 2, 5], "import": [0, 1, 2, 3, 5], "understand": [0, 1], "concept": [0, 5], "object": [0, 1, 4, 5], "return": [0, 1, 5], "practic": 0, "rare": 0, "necessari": [0, 1], "set": [0, 1, 4, 5], "manual": 0, "function": [0, 1, 5], "typic": 0, "appropri": 0, "conceptu": [0, 1], "gap": [0, 1], "two": [0, 1, 2, 5], "base": [0, 1, 2, 5], "alwai": [0, 1, 5], "consid": [0, 1, 5], "chr7": [0, 5], "117231977": 0, "117231997": 0, "hg19": [0, 1, 4, 5], "which": [0, 1, 2, 5], "short": [0, 4, 5], "window": [0, 1, 2, 5], "size": [0, 1, 2, 3], "20": [0, 1, 2, 5], "around": [0, 5], "acceptor": [0, 1, 5], "exon": [0, 5], "13": [0, 4, 5], "cftr": [0, 5], "117": [0, 5], "231": [0, 5], "977": 0, "987": 0, "996": 0, "0": [0, 1, 4, 5], "1": [0, 1, 2, 4, 5], "2": [0, 1, 5], "3": [0, 1, 5], "4": [0, 1, 5], "5": [0, 1, 5], "6": [0, 1, 4, 5], "7": [0, 1, 5], "8": [0, 1, 5], "9": [0, 1, 5], "10": [0, 1, 5], "11": [0, 1, 5], "12": [0, 1, 5], "14": [0, 5], "15": [0, 1, 5], "16": [0, 1, 5], "17": [0, 5], "18": [0, 5], "19": [0, 5], "t": [0, 1, 5], "A": [0, 1, 5], "c": [0, 1, 3, 5], "g": [0, 1, 5], "The": [0, 1, 2, 3, 4, 5], "first": [0, 1, 5], "line": [0, 1, 5], "show": [0, 1, 5], "last": [0, 1], "element": [0, 1, 5], "well": [0, 5], "middl": [0, 1], "In": [0, 1, 2, 4, 5], "chapter": 0, "we": [0, 1, 5], "stick": 0, "convent": [0, 1, 4, 5], "zero": [0, 1, 5], "howev": [0, 1, 5], "ani": [0, 1, 2, 3, 5], "string": [0, 1, 5], "like": [0, 1, 4, 5], "993": [0, 5], "tct": [0, 5], "ar": [0, 1, 2, 4, 5], "one": [0, 1, 5], "clinvar": [0, 5], "format": [0, 1, 2, 3, 5], "now": [0, 1, 2, 3, 5], "go": 0, "ct": [0, 1, 5], "dinucleotid": 0, "default": [0, 1, 2, 5], "behaviour": [0, 1, 2, 5], "genome_kit": [0, 2, 3, 5], "na\u00efv": 0, "variant_genom": [0, 1, 5], "var_seq": 0, "tatcttaaagctgtgtgt": 0, "len": [0, 1, 5], "As": [0, 1], "you": [0, 1, 2, 3, 4, 5], "can": [0, 1, 2, 4, 5], "see": [0, 1, 2, 5], "were": [0, 1, 5], "remov": [0, 1], "shorter": 0, "If": [0, 1, 2, 3, 5], "keep": [0, 1], "expand": [0, 1, 5], "its": [0, 1, 5], "constant": [0, 1], "117231987": [0, 5], "tatcttaaagctgtgtgtaa": 0, "what": [0, 1, 5], "happen": 0, "here": [0, 1, 5], "best": [0, 1, 5], "understood": 0, "look": 0, "wt": 0, "mt": [0, 4], "vertic": 0, "bar": [0, 1], "indic": [0, 1, 5], "sinc": [0, 1, 5], "occur": [0, 1], "right": [0, 1], "side": [0, 1], "extend": [0, 1], "includ": [0, 1, 2, 4, 5], "extra": [0, 1], "equal": [0, 1, 5], "integ": [0, 1, 5], "chromosom": [0, 1, 5], "start": [0, 1, 2, 3, 4, 5], "end": [0, 1, 5], "anywher": 0, "insid": [0, 1], "even": [0, 1], "ll": [0, 1, 2], "case": [0, 1, 5], "pass": [0, 1, 5], "5p": [0, 1, 5], "3p": [0, 1, 5], "center": [0, 1, 5], "more": [0, 1, 4, 5], "conveni": [0, 1, 4, 5], "left": [0, 1], "wherea": [0, 1, 5], "pull": [0, 2], "push": 0, "out": [0, 1, 2], "ggg": 0, "982": 0, "tggg": 0, "ctgggtaaagctgtgtctgt": 0, "easili": [0, 1, 4], "seen": 0, "exampl": [0, 1, 2, 4, 5], "stai": [0, 5], "grow": 0, "shrink": [0, 5], "accommod": 0, "ideal": 0, "still": [0, 1], "work": [0, 2, 4], "For": [0, 1, 2, 3, 5], "five": [0, 5], "overlap": [0, 1, 5], "three": 0, "come": [0, 1, 5], "back": [0, 5], "our": [0, 1], "previou": [0, 1, 5], "984": 0, "aaagct": 0, "ttatatcttagtgtctgtaa": 0, "let": [0, 4, 5], "": [0, 1, 2, 5], "examin": 0, "concret": 0, "again": 0, "explain": [0, 4, 5], "abov": [0, 1, 5], "split": [0, 1], "wa": [0, 1], "accordingli": [0, 1], "slightli": 0, "complic": 0, "directli": [0, 1, 2, 5], "clear": 0, "whether": [0, 1], "should": [0, 1, 5], "ride": 0, "There": [0, 1], "also": [0, 1, 2, 4, 5], "special": [0, 1, 5], "would": [0, 1, 5], "within": [0, 1, 5], "actual": [0, 1], "distinguish": [0, 1], "e": [0, 1, 2, 5], "contain": [0, 1, 2, 3, 5], "entir": [0, 1, 5], "core": [0, 1, 5], "motif": [0, 1, 4], "de": 0, "novo": 0, "refer": [0, 1, 4, 5], "To": [0, 1, 2, 5], "handl": [0, 1, 5], "addit": [0, 1], "argument": [0, 1, 5], "anchor_offset": [0, 1, 5], "where": [0, 1, 5], "do": [0, 1, 5], "ttt": 0, "impli": [0, 1, 5], "gttt": 0, "tatcttaaagtttctgtgtc": 0, "studi": [0, 4], "opposit": [0, 1, 5], "want": [0, 1, 2, 5], "instead": [0, 1], "achiev": 0, "us": [0, 1, 2, 3, 5], "same": [0, 1, 5], "befor": [0, 1, 2, 5], "get": [0, 1, 2, 3, 4, 5], "differ": [0, 1, 2, 4, 5], "offset": [0, 1], "cttaaagtttctgtgtctgt": 0, "By": [0, 1, 5], "Of": 0, "cours": 0, "other": [0, 1, 2, 4, 5], "valu": [0, 1, 5], "atcttaaagtttctgtgtct": 0, "confirm": 0, "gener": [0, 1, 3, 5], "specif": [0, 1, 5], "particular": [0, 1], "often": [0, 1], "principl": [0, 1], "independ": [0, 1], "unintend": 0, "have": [0, 1, 2, 5], "therefor": [0, 1, 5], "follow": [0, 1, 2, 3, 5], "kept": [0, 1], "mind": 0, "onli": [0, 1, 2, 5], "effect": [0, 5], "cannot": [0, 1], "greater": [0, 1], "ten": 0, "given": [0, 1, 5], "done": 0, "could": [0, 1, 5], "max_int": 0, "unittest": [0, 2], "code": [0, 1, 2, 5], "correct": [0, 1], "possibl": [0, 1], "defin": [0, 1, 2, 5], "primarili": 0, "complet": [0, 1], "modifi": [0, 1], "equival": [0, 1, 5], "up": [0, 1, 4, 5], "trim": [0, 1, 5], "amount": [0, 1], "fall": [0, 1, 5], "indel": [0, 1, 5], "967": 0, "Then": 0, "117231967": 0, "117231970": 0, "tgt": 0, "tcttaaagctgtgtctgtaa": 0, "check": [0, 2, 5], "denot": [0, 5], "parenthes": 0, "had": [0, 1], "off": [0, 2, 5], "class": [1, 5], "strand": [1, 5], "reference_genom": [1, 5], "anchor": [1, 4, 5], "none": [1, 5], "sourc": [1, 2, 5], "an": [1, 3, 4, 5], "repres": [1, 5], "contigu": 1, "identifi": 1, "__init__": [1, 5], "initi": 1, "dna0": [1, 5], "paramet": [1, 5], "str": [1, 5], "name": [1, 5], "chr1": [1, 5], "chr": [1, 5], "int": 1, "inclus": [1, 5], "exclus": [1, 5], "option": [1, 2, 3, 5], "relev": 1, "specifi": [1, 3, 4, 5], "point": [1, 5], "todo": 1, "link": [1, 2], "detail": [1, 4, 5], "respect": [1, 5], "sens": [1, 5], "forward": [1, 5], "ha": [1, 2, 5], "odd": 1, "round": 1, "toward": 1, "empti": [1, 5], "shift": 1, "purpos": 1, "insert": [1, 4, 5], "itself": 1, "begin": 1, "properti": 1, "shorthand": 1, "chrom": [1, 3, 5], "avail": [1, 5], "end5": [1, 5], "Not": 1, "confus": 1, "attribut": [1, 5], "preced": [1, 5], "upstream": [1, 5], "end3": [1, 5], "succeed": 1, "downstream": [1, 5], "assembli": [1, 2, 5], "ucsc": [1, 3, 4, 5], "refg": [1, 3], "associ": [1, 5], "accord": [1, 5], "current": [1, 2, 5], "neg": [1, 5], "dnstream": 1, "take": [1, 5], "number": [1, 5], "add": 1, "symmetr": 1, "intersect": 1, "both": [1, 2, 5], "possibli": 1, "disjoint": 1, "subtract": 1, "list": [1, 4, 5], "anoth": 1, "after": 1, "self": [1, 5], "is_positive_strand": 1, "test": [1, 4, 5], "true": [1, 5], "bool": 1, "as_positive_strand": 1, "as_negative_strand": 1, "as_opposite_strand": [1, 5], "upstream_of": [1, 5], "strictli": 1, "subclass": [1, 5], "dnstream_of": [1, 5], "li": 1, "extent": 1, "express": 1, "static": 1, "from_dna0_coord": [1, 5], "creat": [1, 2, 3, 5], "span": [1, 5], "from_dna0": 1, "from_dna1": 1, "dna1": [1, 5], "automat": [1, 2, 5], "convert": [1, 5], "from_rna1": 1, "rna1": 1, "min": 1, "ab": 1, "max": 1, "sign": 1, "interval1": 1, "interval2": 1, "smallest": 1, "pair": 1, "exist": 1, "rais": [1, 5], "except": [1, 5], "requir": [1, 2, 3, 5], "mai": [1, 2, 5], "100": [1, 5], "150": [1, 5], "b": [1, 2, 5], "200": 1, "250": 1, "note": [1, 5], "awar": 1, "context": 1, "midpoint": 1, "125": 1, "151": 1, "126": 1, "distanc": 1, "interval_a": 1, "interval_b": 1, "300": [1, 5], "interval_c": 1, "400": 1, "500": 1, "325": 1, "determin": 1, "comput": 1, "must": [1, 2, 3], "float": [1, 5], "otherwis": 1, "as_dna0": 1, "tupl": 1, "as_dna1": 1, "as_rna1": 1, "as_ucsc": [1, 5], "browser": [1, 5], "chr5": 1, "500201": 1, "5002010": 1, "with_anchor": [1, 5], "version": [1, 2, 5], "ignor": [1, 5], "__repr__": 1, "repr": 1, "__str__": 1, "ref": [1, 5], "alt": [1, 5], "allel": [1, 5], "altern": [1, 5], "inherit": 1, "synonym": 1, "provid": [1, 2, 5], "histor": 1, "reason": [1, 2, 5], "consist": 1, "prefer": 1, "alter": 1, "as_variant_str": 1, "represent": 1, "from_str": [1, 5], "diverg": 1, "x": [1, 5], "y": [1, 5], "v": [1, 2, 5], "n": [1, 2, 3], "260": 1, "config": 1, "correspond": [1, 5], "resourc": [1, 5], "etc": [1, 5], "gencod": [1, 2, 3, 4, 5], "v19": [1, 2, 4, 5], "1st": [1, 5], "v29": [1, 4], "hg38": [1, 3, 4, 5], "recogn": 1, "shown": [], "below": [1, 4, 5], "p13": [1, 4], "plusmt": [1, 4], "p12": [1, 4], "p14": [1, 4], "mm10": [1, 4], "p6": [1, 4], "mm39": [1, 4], "rn6": [1, 4], "macfas5": [1, 4], "susscr11": [1, 4], "basic": [1, 4], "v29lift37": 1, "v41": [1, 4], "v41lift37": 1, "ucsc_refseq": 1, "2017": 1, "06": [1, 5], "25": [1, 5], "ncbi_refseq": 1, "v105": [1, 4], "20190906": [1, 4], "v109": [1, 4], "v110": [1, 4], "vm19": [1, 4], "vm30": [1, 4], "vm31": [1, 4], "m39": [1, 4], "ensembl": [1, 3, 4, 5], "rnor_6": [1, 4], "88": [1, 4], "macfas_5": [1, 4], "v101": [1, 4], "95": [1, 4], "sscrofa11": [1, 4], "98": [1, 4], "request": [1, 5], "creation": 1, "disambigu": [1, 4], "serv": [1, 5], "__reduce__": 1, "helper": 1, "pickl": 1, "__new__": 1, "cl": 1, "variant_str": 1, "revers": [1, 5], "complement": [1, 5], "outsid": [1, 4], "rang": [1, 5], "indexerror": 1, "queri": [1, 4, 5], "variant_dna": 1, "rule": 1, "intend": [1, 5], "either": [1, 2, 5], "anno": [1, 2], "support": [1, 2, 3, 5], "loop": 1, "over": [1, 5], "tran": [1, 5], "cdss": [1, 5], "utr5": 1, "utr5tabl": 1, "utr3": 1, "utr3tabl": 1, "find_motif": [1, 5], "match_posit": [1, 5], "find_overlapping_motif": [1, 5], "fals": [1, 5], "find": [1, 2, 4], "search": [1, 4, 5], "veri": 1, "ag": [1, 5], "match": [1, 5], "put": 1, "non": [1, 5], "all": [1, 2, 4, 5], "found": [1, 5], "genome37": 1, "40000": 1, "40080": 1, "aa": 1, "40031": 1, "40061": 1, "map": [1, 5], "valid": [1, 5], "chromosome_s": 1, "bp": 1, "basepair": 1, "configur": [1, 5], "data_dir": [1, 5], "open": [1, 2, 5], "path": 1, "file": [1, 3, 4], "being": 1, "appris_transcript": 1, "order": [1, 5], "custom": [1, 5], "appri": [1, 2], "index": [1, 4, 5], "decreas": 1, "prioriti": 1, "restrict": 1, "geneid": 1, "score": 1, "least": [1, 5], "canon": 1, "appris_princip": 1, "as_str": 1, "princip": 1, "isoform": 1, "prinicp": 1, "origin": 1, "databas": [1, 5], "level": [1, 5], "report": 1, "appris_index_nam": 1, "2017_05": 1, "v22": 1, "_build_appri": 1, "remote_fil": 1, "rest": 1, "toggl": 1, "desir": 1, "__weakref__": 1, "weak": 1, "behav": 1, "quantiti": 1, "depend": [1, 2, 4, 5], "alreadi": 1, "system": [1, 2, 5], "explan": [1, 5], "normal": [1, 2], "yet": 1, "implement": [1, 5], "user": [1, 2, 4, 5], "respons": 1, "longer": 1, "pleas": [1, 2], "inform": 1, "assum": [1, 5], "never": 1, "mean": [1, 5], "polici": 1, "occurr": 1, "kind": 1, "complex": [1, 5], "py": [1, 2, 5], "40033": 1, "tag": 1, "40030": 1, "50000": 1, "50010": 1, "taaaccacat": 1, "__call__": 1, "filenam": [1, 5], "2bit": [1, 2, 3], "infil": 1, "view": 1, "gtrack": [1, 5], "chr17": 1, "41246881": 1, "41246886": 1, "phastcons_mamm": 1, "f4": 1, "arrai": [1, 5], "857": 1, "571": 1, "357": 1, "286": 1, "dtype": [1, 5], "float16": 1, "scope": 1, "foo": 1, "encod": 1, "etyp": 1, "trade": 1, "power": [1, 4], "storag": [1, 5], "speed": [1, 2], "decod": 1, "m0": 1, "32": 1, "uint8": [1, 5], "bit": 1, "mask": [1, 5], "u1": 1, "u2": 1, "u3": [1, 5], "u4": 1, "u5": 1, "31": 1, "u6": 1, "63": 1, "u8": 1, "255": 1, "i8": 1, "int8": [1, 5], "128": 1, "127": 1, "f2": 1, "65504": 1, "dictionari": 1, "f3": 1, "f5": 1, "f6": 1, "f8": 1, "f16": 1, "f32": 1, "2e": 1, "38": 1, "4e38": 1, "np": [1, 5], "linspac": 1, "num_bit": 1, "inde": 1, "numpi": [1, 5], "bool_": 1, "ieee": 1, "754": 1, "half": 1, "precis": 1, "float32": [1, 5], "new": [1, 2, 4], "sensit": 1, "unstrand": 1, "row": [1, 5], "dim": 1, "column": [1, 5], "ndarrai": 1, "dimension": 1, "each": [1, 5], "resolut": 1, "datum": 1, "stranded": [1, 5], "store": [1, 2, 5], "uniqu": [1, 5], "phastcon": 1, "histogram": 1, "region": 1, "separate_dim": 1, "kei": 1, "appear": 1, "onc": [1, 2], "count": 1, "time": 1, "nan": 1, "omit": 1, "wide": 1, "dimens": 1, "upon": [1, 2], "separ": [1, 5], "dict": [1, 5], "close": 1, "statement": 1, "gtrack_vers": 1, "build": [1, 3, 4, 5], "attempt": [1, 5], "mismatch": 1, "ioerror": 1, "outfil": 1, "involv": 1, "four": [1, 5], "step": [1, 2], "write": 1, "further": 1, "disk": 1, "final": [1, 2, 5], "fill": [1, 5], "them": [1, 2, 5], "single_strand": [1, 5], "set_default_valu": 1, "entri": [1, 5], "set_data": [1, 5], "finish": 1, "coarser": 1, "1bp": 1, "load": 1, "wig": 1, "5bp": 1, "fixedstep": 1, "57": 1, "23": [1, 5], "share": [1, 4], "strand_unawar": [1, 5], "strand_awar": [1, 5], "doe": [1, 5], "fine": 1, "set_spars": 1, "min_run": 1, "48": 1, "min_delta": 1, "enabl": [1, 5], "sparsif": 1, "reduc": 1, "domin": 1, "default_valu": 1, "spars": 1, "highli": 1, "recommend": 1, "variablestep": 1, "analyz": 1, "avoid": 1, "run": [1, 2, 5], "treat": 1, "explicitli": 1, "exclud": [1, 5], "byte": 1, "addd": 1, "small": [1, 2], "increas": 1, "definit": 1, "minimum": 1, "delta": 1, "distinct": 1, "set_clamp": 1, "clamp": 1, "trigger": 1, "error": 1, "notifi": 1, "hi": 1, "her": 1, "properli": 1, "intent": 1, "f": [1, 2], "set_dict": 1, "numer": 1, "wish": [1, 5], "abl": [1, 2], "compris": 1, "set_transform": 1, "d": [1, 5], "transform": 1, "about": [1, 4, 5], "undergo": 1, "affin": 1, "target": [1, 2], "set_restrict": 1, "act": 1, "crop": 1, "awai": 1, "make": [1, 4, 5], "full": 1, "pipelin": [1, 2], "sake": 1, "unit": 1, "iter": 1, "faster": 1, "dure": 1, "develop": [1, 4], "until": 1, "previous": 1, "compat": [1, 2], "built": [1, 2, 5], "nxm": 1, "m": [1, 2], "r": [1, 5], "xm": 1, "word": 1, "suppli": [1, 5], "coarsen": 1, "builder": [1, 5], "constructor": [1, 5], "set_data_from_wig": 1, "pos_strand_fil": 1, "neg_strand_fil": 1, "gz": [1, 2, 5], "second": [1, 5], "adjac": 1, "merg": [1, 2], "set_data_from_bedgraph": 1, "bedgraph": 1, "direct": 1, "analogu": 1, "set_data_from_b": 1, "bedfil": 1, "categori": 1, "bed": 1, "incorpor": 1, "repeat": [1, 5], "unlik": 1, "per": [1, 5], "larg": 1, "especi": 1, "manag": [1, 5], "flush": 1, "accumul": 1, "free": 1, "memori": 1, "while": [1, 2, 5], "immedi": [1, 5], "becom": 1, "written": 1, "data_s": 1, "sparsiti": 1, "disabl": 1, "via": [1, 2], "coars": 1, "occupi": 1, "approxim": 1, "bits_per_datum": 1, "exact": 1, "trivial": 1, "portion": 1, "dedic": 1, "higher": 1, "lower": 1, "rough": 1, "estim": 1, "index_s": 1, "expos": 1, "shortcut": 1, "refseq": [1, 4, 5], "multipl": 1, "These": [1, 5], "tend": 1, "ambigu": [1, 5], "thei": [1, 2, 3, 5], "re": [1, 2], "retain": 1, "ident": [1, 5], "smn1": 1, "locat": [1, 2, 5], "70220767": 1, "70249769": 1, "69345349": 1, "69374349": 1, "duplic": 1, "70248838": 1, "69373418": 1, "own": [1, 5], "nm_000344": 1, "nm_001297715": 1, "nm_022874": 1, "similarli": 1, "smn2": 1, "70248842": 1, "69373422": 1, "utrtabl": 1, "retriev": 1, "dganno": [1, 2, 3], "binary_vers": 1, "binari": [1, 2, 5], "build_gencod": [1, 3], "gff3": [1, 2, 3], "convers": 1, "cfg": [1, 3], "destin": 1, "build_ncbi_refseq": 1, "ncbi": [1, 3, 4], "build_ucsc_refseq": 1, "ucsc_db_dir": 1, "directori": [1, 2, 3, 5], "refgen": 1, "txt": [1, 2, 3], "reflink": 1, "pars": 1, "id": [1, 5], "entrez": 1, "ensg00000115275": 1, "7841": 1, "brca1": 1, "gene_typ": 1, "field": [1, 5], "standard": [1, 4, 5], "biotyp": 1, "protein_cod": 1, "protein": 1, "non_cod": 1, "evid": 1, "copi": [1, 2, 3], "find_5p_align": [1, 5], "brute": 1, "forc": 1, "filter": [1, 4, 5], "sort": 1, "find_3p_align": [1, 5], "find_5p_within": [1, 5], "find_3p_within": [1, 5], "find_within": [1, 5], "find_overlap": [1, 5], "find_exact": [1, 5], "exactli": [1, 5], "signific": 1, "find_x": 1, "__getitem__": 1, "lookup": 1, "linear": 1, "mog": 1, "suffix": 1, "One": 1, "first_by_nam": 1, "enst00000233616": 1, "nm_006302": 1, "tsl": 1, "tsl1": 1, "suspect": 1, "mrna": 1, "tsl2": 1, "flag": [1, 2], "est": 1, "tsl3": 1, "tsl4": 1, "tsl5": 1, "model": 1, "structur": [1, 4, 5], "tslna": 1, "transcript_typ": 1, "parent": 1, "ccds_id": 1, "ccdsid": 1, "applic": [1, 5], "ccds53759": 1, "protein_id": 1, "ensp00000233616": 1, "np_006293": 1, "product": 1, "descript": [1, 5], "breast": 1, "cancer": 1, "suscept": 1, "segment": 1, "utr": 1, "exonic_interv": 1, "ense00000846737": 1, "next_exon": [1, 5], "next": [1, 5], "prev_exon": 1, "prev_intron": 1, "next_intron": 1, "intronic_interv": 1, "intr": 1, "phase": 1, "reach": 1, "codon": 1, "frame": 1, "http": [1, 2, 3, 5], "gmod": 1, "org": [1, 5], "wiki": 1, "next_cd": 1, "prev_cd": 1, "ralign": 1, "incompat": 1, "infer": 1, "optim": 1, "bam": [1, 4], "observ": 1, "mnt": 1, "sample000132": 1, "my_fil": 1, "ralign_vers": 1, "build_ralign": 1, "include_dupl": 1, "library_format": 1, "u": 1, "sam": [1, 4], "pool": 1, "cigar": 1, "1m1d1m2n1m": 1, "delet": [1, 4, 5], "1m1i1m2n1m": 1, "rdist": 1, "sy": 1, "stdin": 1, "stream": 1, "samtool": 1, "h": 1, "Be": 1, "sure": [1, 5], "header": 1, "against": [1, 5], "long": 1, "github": [1, 2], "com": [1, 2], "ht": 1, "spec": 1, "blob": 1, "master": 1, "samv1": 1, "pdf": 1, "subsect": 1, "librari": [1, 4], "sf": 1, "sr": 1, "salmon": 1, "readthedoc": 1, "io": [1, 5], "en": [1, 5], "stabl": 1, "library_typ": 1, "html": [1, 3, 5], "pileup": 1, "int32": [1, 5], "bodi": 1, "acgtacgt": 1, "substitut": 1, "shape": [1, 5], "pileup_a": 1, "pileup_c": 1, "pileup_g": 1, "pileup_t": 1, "pileup_del": 1, "compar": 1, "across": 1, "junc": 1, "jralign": 1, "distribut": [1, 2], "child": 1, "been": 1, "invalid": 1, "jralign_vers": 1, "build_jralign": 1, "min_read": 1, "min_overhang": 1, "include_vari": 1, "overhang_error": 1, "process": [1, 2], "5m8d5m": 1, "5m100n10m50n5m": 1, "multi": 1, "block": 1, "contribut": 1, "1s2m2n1m1n1m": 1, "overhang": 1, "sum": [1, 5], "2nd": [1, 5], "scenario": 1, "scheme": 1, "facilit": 1, "rel": [1, 5], "usag": 1, "bootstrap": 1, "correctli": 1, "without": [1, 2, 4, 5], "modif": 1, "20m100n10m100n3m": 1, "becaus": [1, 5], "clip": 1, "qualiti": 1, "doesn": 1, "privat": 1, "output": [1, 5], "stop": 1, "limit": 1, "maximum": 1, "your": [1, 2, 3, 5], "num_read": 1, "meaning": 1, "ad": [1, 4, 5], "num_vari": 1, "rdist_vers": 1, "build_rdist": 1, "pass_filt": 1, "callabl": 1, "junction_index": 1, "read_index": 1, "downsampl": 1, "lambda": 1, "_": 1, "random": 1, "binomi": 1, "from_jralign": 1, "cach": 1, "brca1_interv": 1, "content": [1, 4], "regener": 1, "preserv": 1, "part": [1, 5], "num_count": 1, "divid": 1, "rtype": 1, "far": 1, "vcfbin": 1, "section": [1, 5], "quickstart": [1, 4], "demo": [1, 2, 5], "query_vcf": [1, 5], "build_vcfbin": 1, "from_vcf": [1, 5], "vcfpath": 1, "arg": 1, "typing_extens": 1, "liter": 1, "ebi": 1, "vcfbinpath": 1, "kwarg": 1, "fast": [1, 2, 4, 5], "bin": 1, "def": [1, 5], "open_clinvar": 1, "info_id": [1, 5], "af_exac": 1, "clnsig": 1, "clinvar_20180128": 1, "bcftool": 1, "instal": [1, 2, 4], "eg": 1, "gtex": 1, "intermedi": 1, "__len__": 1, "boolean": 1, "select": [1, 2, 5], "astyp": 1, "Or": 1, "subset": 1, "abca4": 1, "abca4_vcf": 1, "index_of": [1, 5], "instanc": [1, 5], "gt_homozygous_ref": 1, "gt": [1, 5], "gt_heterozyg": 1, "gt_homozygous_alt": 1, "gt_unknown": 1, "svtype_na": 1, "svtype": 1, "miss": 1, "num_sampl": 1, "sampl": [1, 5], "samples_nam": [], "info": [1, 5], "dp": 1, "keyerror": 1, "unrecogn": 1, "typeerror": 1, "format_id": 1, "matrix": 1, "sample1": [1, 5], "sample2": [1, 5], "sample3": [1, 5], "fmt_id": [1, 5], "ancestr": 1, "norm": 1, "cat": 1, "fail": 1, "ok": 1, "ca": [1, 5], "int16": 1, "genotyp": [1, 4, 5], "del": 1, "svtype_del": 1, "INS": 1, "svtype_in": 1, "dup": 1, "svtype_dup": 1, "inv": 1, "svtype_inv": 1, "cnv": 1, "svtype_cnv": 1, "bnd": 1, "svtype_bnd": 1, "mateid": 1, "diploid": 1, "interpret": [1, 5], "overrid": 1, "bitwidth": 1, "perform": 1, "save": [1, 5], "space": [1, 5], "reserv": [1, 2], "caveat": 1, "ascii": 1, "3a": 1, "unless": [1, 5], "sole": 1, "depth": [1, 5], "multidimension": 1, "bewar": 1, "incorrectli": 1, "caller": 1, "ensur": [1, 2], "b37": 1, "fa": 1, "ncbi37": 1, "control": [1, 2, 5], "encount": 1, "warn": 1, "valueerror": 1, "under": [1, 2, 5], "condit": 1, "malform": 1, "everi": [1, 5], "unsupport": 1, "nucleotid": [1, 5], "vcfbin_vers": 1, "sequence_vari": 1, "overestim": 1, "due": 1, "spuriou": 1, "variant_combin": 1, "heterozygous_alt_af_threshold": 1, "difficult": 1, "consensu": 1, "af": [1, 5], "unobserv": 1, "primari": 1, "multiallel": 1, "unknown": 1, "combin": 1, "threshold": 1, "matric": 1, "scalar": 1, "dynam": 1, "accept": [1, 5], "param": 1, "local": [1, 2, 5], "get_fil": [1, 5], "download": [1, 2, 5], "upload_fil": [1, 5], "filepath": [1, 5], "metadata": [1, 3, 5], "upload": [1, 5], "remot": [1, 2, 5], "attach": 1, "modul": [1, 2, 4], "fetch": 1, "demand": [1, 2], "wrapper": 1, "wget": 1, "url": 1, "dst": 1, "timestamp": 1, "progress": 1, "unspecifi": 1, "temp": 1, "assign": 1, "date": 1, "newer": [1, 2], "server": 1, "httperror": 1, "socket": 1, "timeout": 1, "connect": 1, "interrupt": 1, "progresspercentag": 1, "files": 1, "progress_msg": 1, "displai": 1, "text": 1, "termin": 1, "transfer": 1, "percentag": 1, "throughput": 1, "myfil": 1, "eta": 1, "00": [1, 5], "147": 1, "truncat": 1, "messag": 1, "mypackag": 1, "bytes_amount": 1, "write2bit": 1, "seq": 1, "mix": 1, "upper": 1, "acgtn": 1, "charact": 1, "aaaannnnccaaaatttt": 1, "chr2": 1, "nnnnnnngggggggg": 1, "mydna": 1, "apply_vari": 1, "reference_align": 1, "mock": 1, "mark": 1, "python": [1, 2, 3, 4, 5], "partial": 1, "visibl": 1, "autocomplet": 1, "doc": 1, "sphinx": 1, "_cxx": [1, 2], "regist": [1, 2, 5], "mock_result": 1, "autodoc": 1, "decor": 1, "face": 1, "analysi": 1, "propag": 1, "backend": 1, "those": [1, 4, 5], "pydeletemockattr": 1, "strip": 1, "symbol": 1, "result_typ": 1, "give": [1, 4], "hint": 1, "mere": 1, "mock_unreach": 1, "print": [1, 3, 5], "try": 1, "instanti": 1, "unreach": 1, "ever": 1, "caus": 1, "headach": 1, "pylint": 1, "pycharm": 1, "noth": 1, "strip_mock_bas": 1, "subtyp": 1, "runtim": 1, "hierarchi": 1, "z": [1, 5], "know": [1, 5], "member": 1, "hacki": 1, "wai": [1, 4, 5], "trick": 1, "simpli": 1, "declar": 1, "think": 1, "though": 1, "compil": [2, 3, 5], "conda": [2, 3], "gxx_linux": 2, "64": 2, "clangxx_osx": 2, "osx": 2, "sdk": 2, "xcode": 2, "appl": 2, "clone": 2, "tree": 2, "git": 2, "deepgenom": 2, "environ": [2, 5], "mamba": [], "env": [2, 5], "genomekit_dev": 2, "yml": 2, "activ": [2, 3], "On": 2, "m1": 2, "mac": 2, "might": 2, "cxx": 2, "zlib": 2, "fmt": 2, "forg": [2, 5], "bioconda": 2, "dep": 2, "packag": [2, 5], "mode": [2, 5], "pip": 2, "extens": 2, "prerequisit": 2, "vcvars64": 2, "bat": 2, "command": 2, "prompt": 2, "setup": 4, "vcproj": 2, "sln": 2, "discov": 2, "switch": 2, "branch": [2, 5], "edit": 2, "yourself": 2, "forget": 2, "lead": [2, 5], "unpredict": 2, "ubuntu": 2, "ccach": 2, "repeatedli": 2, "just": [2, 5], "testinterv": 2, "test_seri": 2, "test_interv": 2, "src": 2, "main": 2, "cpp": 2, "cmake": 2, "dcmake_build_typ": 2, "unittestbuild": 2, "ng": [], "envvar": 2, "gk_debugbreak": 2, "break": 2, "gk_check": 2, "failur": 2, "debugg": 2, "reli": 2, "pre": [2, 5], "annot": [2, 4], "updat": 2, "tini": 2, "excerpt": 2, "enough": 2, "continu": 2, "integr": 2, "help": [2, 3], "instruct": [2, 3], "resid": 2, "track": [2, 4], "subcommand": 2, "circleci": [], "dev": [], "channel": [], "commit": [], "ie": [], "v0": [], "versionrc": [], "json": [], "regular": [], "bump": [], "npx": [], "minor": [], "patch": 4, "major": [], "futur": 4, "brew": [], "openssl": [], "xz": [], "curl": [], "o": [], "www": [], "ftp": [], "tar": [], "xjf": [], "cd": 5, "as_ptr": [], "tp_basics": [], "1788c1788": [], "assert": [], "t_size": [], "b_size": [], "els": [], "smaller": [], "cppflag": [], "prefix": [], "ldflag": [], "l": [], "lib": [], "macosx_deployment_target": [], "pydebug": [], "echo": [], "j": [], "abi": [], "rebuilt": [], "pypa": 5, "pythonhost": [], "2d": [], "80": [], "1809de155bad674b494248bcfca0e49eb4c5d8bee58f26fe7a0dd45029e2": [], "zip": [], "saniti": [], "fsanit": [], "address": [], "dyld_insert_librari": [], "app": [], "toolchain": [], "xcodedefault": [], "xctoolchain": [], "usr": [], "clang": [], "darwin": [], "libclang_rt": [], "asan_osx_dynam": [], "dylib": [], "schema": 3, "chromalia": 3, "hgdownload": 3, "soe": 3, "edu": 3, "goldenpath": 3, "bigzip": 3, "appdir": [3, 5], "user_data_dir": [3, 5], "fasta": 3, "tool": [2, 3], "fatotwobit": 3, "twobitinfo": 3, "twobit": 3, "contig": 3, "alias": 3, "gk": [3, 5], "genomeannot": [3, 5], "my_anno": 3, "my_assembli": 3, "deep": 4, "easi": [4, 5], "access": 4, "data": 4, "goal": 4, "research": 4, "creativ": 4, "design": 4, "secondari": 4, "conserv": 4, "read": [2, 4, 5], "vcf": [2, 4], "interv": 4, "lightweight": 4, "scan": 4, "unpatch": 4, "grch37": [4, 5], "rcr": 4, "grch38": 4, "grcm38": 4, "grcm39": 4, "rgsc": 4, "macaca_fascicularis_5": 4, "sgsc": 4, "comprehens": 4, "snapshot": 4, "tutori": [4, 5], "api": [2, 4, 5], "document": [2, 4], "junction": [2, 4], "intern": [4, 5], "releas": 4, "debug": [], "garbag": [], "collect": [], "issu": [], "page": 4, "quick": 5, "cover": 5, "anaconda": 5, "replac": [], "project": [2, 5], "breviti": 5, "ve": 5, "sever": 5, "namespac": 5, "117120016": 5, "117120201": 5, "irrespect": 5, "185": 5, "describ": 5, "117120017": 5, "aattggaagcaaa": 5, "aactttttttcag": 5, "gene": 5, "ensg00000001626": 5, "transcript": 5, "earlier": 5, "plu": 5, "enst00000003084": 5, "27": 5, "genometrackbuild": 5, "arang": 5, "genometrack": 5, "walk": 5, "straight": 5, "walk_annot": 5, "ensg00000223972": 5, "ddx11l1": 5, "enst00000456328": 5, "intron": 5, "thought": 5, "tabl": [2, 5], "exont": 5, "26": 5, "enst00000454343": 5, "enst00000426809": 5, "top": 5, "ultim": 5, "slice": 5, "oper": 5, "__________": 5, "few": [3, 5], "0123456789": 5, "aaaaabbbbb": 5, "cccc": 5, "live": 5, "univers": 5, "xxxxxyyyyi": 5, "zzzz": 5, "w": 5, "notic": 5, "consecut": 5, "visual": 5, "_____________": 5, "ctgaaaaaaagtt": 5, "tttgcttccaatt": 5, "meet": 5, "certain": 5, "criteria": 5, "has_coding_seq": 5, "has_good_level": 5, "not_first_exon": 5, "1196293": 5, "724078": 5, "remain": 5, "605573": 5, "558099": 5, "198992": 5, "10nt": 5, "surround": 5, "tgcagggaac": 5, "ttcagctgct": 5, "tgtaggaaac": 5, "tccaggctat": 5, "gccagaggac": 5, "gacagaacca": 5, "cccagattgg": 5, "nearest": 5, "100nt": 5, "coord": 5, "91661": 5, "169295": 5, "320861": 5, "91662": 5, "enst00000466430": 5, "169296": 5, "enst00000466557": 5, "320862": 5, "enst00000432964": 5, "candid": 5, "enst00000601486": 5, "enst00000599771": 5, "3rd": 5, "individu": 5, "117120148": 5, "AT": 5, "117120150": 5, "120": 5, "149": 5, "style": 5, "compact": 5, "vcftabl": 5, "suppos": 5, "fileformat": 5, "vcfv4": 5, "frequenc": 5, "po": 5, "qual": 5, "949523": 5, "949608": 5, "01": 5, "949696": 5, "02": 5, "949739": 5, "tc": 5, "03": 5, "977028": 5, "04": 5, "977330": 5, "05": 5, "977516": 5, "21": 5, "977570": 5, "07": 5, "22": 5, "24": 5, "978604": 5, "08": 5, "978628": 5, "09": 5, "28": 5, "29": 5, "30": 5, "carri": 5, "vcfvariant": 5, "949522": 5, "0500000007451": 5, "977030": 5, "977630": 5, "977329": 5, "977515": 5, "977569": 5, "10l": 5, "3l": 5, "fastest": 5, "axi": 5, "978603": 5, "978627": 5, "besid": 5, "idea": 5, "transpar": 5, "mixtur": 5, "toi": 5, "flank": 5, "extract_featur": 5, "7nt": 5, "117120149": 5, "rs397508328": 5, "117120151": 5, "rs397508657": 5, "var": 5, "ccatgca": 5, "ccgttca": 5, "plan": 5, "snv": 5, "made": [1, 4, 5], "var1": 5, "117120188": 5, "rs397508673": 5, "var2": 5, "117120190": 5, "rs397508710": 5, "dela": 5, "var3": 5, "togeth": 5, "technic": 5, "pad": 5, "994": 5, "comma": 5, "nuclear": 5, "117120185": 5, "117120192": 5, "ccaaact": 5, "cctaact": 5, "ccaact": 5, "cctact": 5, "manner": 5, "lift": 5, "suit": 5, "anchored_5p": 5, "anchored_3p": 5, "ccaactt": 5, "tccaact": 5, "pwm": 5, "117231957": 5, "117232030": 5, "ttgatatttatatgtttttatatcttaaagctgtgtctgtaaactgatggctaacaaaactaggattttggtc": 5, "aacaa": 5, "117232009": 5, "117232004": 5, "117232014": 5, "But": 5, "potenti": 5, "117232020": 5, "verifi": 5, "find_overlapping_match": 5, "tt": 5, "ttgatatttatatgttttta": 5, "117231963": 5, "117231971": 5, "117231973": 5, "117231981": 5, "117232022": 5, "117232024": 5, "117231964": 5, "117231972": 5, "117231974": 5, "117232023": 5, "117231980": 5, "ttagtt": 5, "117231979": 5, "third": 5, "fourth": 5, "variabl": [2, 5], "genomekit_data_dir": 5, "public": 5, "googl": 5, "cloud": 5, "bucket": 5, "gc": 5, "genomekit_gcs_bucket": 5, "data_manag": [4, 5], "datamanag": 5, "mydatamanag": 5, "gk_data": [4, 5], "plugin": 5, "advertis": 5, "endpoint": 5, "group": 5, "setuptool": 5, "latest": 5, "userguid": 5, "entry_point": 5, "linux": 2, "flow": [], "goe": [], "offici": [], "prior": [], "allow_outside_chromosom": 1, "lack": 1, "sample_nam": 1, "comment": 2, "2019": 2, "shell": 2, "x64": 2, "nativ": 2, "menu": 2, "preconfigur": 2, "migrat": [], "ci": 2, "apprisnotavailableerror": 1, "apprisnotavailableerro": [], "err": [], "common": 1, "ones": [1, 4], "list_available_genom": [1, 4], "in_clion": 2, "conda_prefix": 2, "home": 2, "parallel": 2, "verbos": 2, "commonli": 4, "jetbrain": 4, "clion": 4, "pai": 5, "gcloud": 5, "credenti": 5, "beforehand": [], "genomekit_gcs_billing_project": 5, "repo": 2, "bot": 2, "pr": 2, "feedstock": 2, "regro": 2, "cf": 2, "autotick": 2, "kick": 2, "publish": 2, "starter": [3, 5], "sh": [3, 5], "bash": [3, 5], "script": [3, 5], "place": 5, "good": 5, "larger": 5, "auth": 5, "login": 5, "export": 5, "author": [], "__": []}, "objects": {"genome_kit": [[1, 0, 1, "", "Alignment"], [1, 0, 1, "", "AlignmentMatch"], [1, 0, 1, "", "AlignmentMatchTable"], [1, 0, 1, "", "AlignmentTable"], [1, 0, 1, "", "Cds"], [1, 0, 1, "", "CdsTable"], [1, 0, 1, "", "DataManager"], [1, 0, 1, "", "Exon"], [1, 0, 1, "", "ExonTable"], [1, 0, 1, "", "Gene"], [1, 0, 1, "", "GeneTable"], [1, 0, 1, "", "Genome"], [1, 0, 1, "", "GenomeAnnotation"], [1, 0, 1, "", "GenomeDNA"], [1, 0, 1, "", "GenomeTrack"], [1, 0, 1, "", "GenomeTrackBuilder"], [1, 0, 1, "", "Interval"], [1, 0, 1, "", "Intron"], [1, 0, 1, "", "IntronTable"], [1, 0, 1, "", "JReadAlignments"], [1, 0, 1, "", "Junction"], [1, 0, 1, "", "JunctionReadAlignment"], [1, 0, 1, "", "JunctionReadAlignments"], [1, 0, 1, "", "JunctionReadAlignmentsTable"], [1, 0, 1, "", "JunctionReadCount"], [1, 0, 1, "", "JunctionReadDistribution"], [1, 0, 1, "", "JunctionReadDistributionTable"], [1, 0, 1, "", "JunctionTable"], [1, 0, 1, "", "ReadAlignments"], [1, 0, 1, "", "ReadDistributions"], [1, 0, 1, "", "Transcript"], [1, 0, 1, "", "TranscriptTable"], [1, 0, 1, "", "VCFTable"], [1, 0, 1, "", "VCFVariant"], [1, 0, 1, "", "Variant"], [1, 0, 1, "", "VariantGenome"], [1, 0, 1, "", "VariantTable"], [1, 5, 0, "-", "gk_data"]], "genome_kit.Alignment": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "matches"]], "genome_kit.AlignmentMatch": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "variants"]], "genome_kit.AlignmentMatchTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.AlignmentTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.Cds": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "exon"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_cds"], [1, 2, 1, "", "phase"], [1, 2, 1, "", "prev_cds"], [1, 2, 1, "", "transcript"]], "genome_kit.CdsTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.DataManager": [[1, 1, 1, "", "__init__"], [1, 3, 1, "", "__weakref__"], [1, 1, 1, "", "get_file"], [1, 1, 1, "", "list_available_genomes"], [1, 1, 1, "", "upload_file"]], "genome_kit.Exon": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "cds"], [1, 2, 1, "", "id"], [1, 2, 1, "", "index"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_exon"], [1, 2, 1, "", "next_intron"], [1, 2, 1, "", "prev_exon"], [1, 2, 1, "", "prev_intron"], [1, 2, 1, "", "transcript"], [1, 2, 1, "", "utr3"], [1, 2, 1, "", "utr5"]], "genome_kit.ExonTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.Gene": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "id"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "level"], [1, 2, 1, "", "name"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "type"]], "genome_kit.GeneTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 1, 1, "", "first_by_name"], [1, 2, 1, "", "stranded"]], "genome_kit.Genome": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__new__"], [1, 1, 1, "", "__reduce__"], [1, 1, 1, "", "__repr__"], [1, 3, 1, "", "__weakref__"], [1, 2, 1, "", "annotation"], [1, 1, 1, "", "appris_principality"], [1, 1, 1, "", "appris_transcripts"], [1, 2, 1, "", "cdss"], [1, 1, 1, "", "chromosome_size"], [1, 2, 1, "", "chromosomes"], [1, 2, 1, "", "config"], [1, 2, 1, "", "data_dir"], [1, 1, 1, "", "dna"], [1, 2, 1, "", "exons"], [1, 1, 1, "", "find_motif"], [1, 2, 1, "", "genes"], [1, 1, 1, "", "interval"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"], [1, 1, 1, "", "variant"], [1, 1, 1, "", "variant_dna"]], "genome_kit.GenomeAnnotation": [[1, 1, 1, "", "__repr__"], [1, 1, 1, "", "binary_version"], [1, 1, 1, "", "build_gencode"], [1, 1, 1, "", "build_ncbi_refseq"], [1, 1, 1, "", "build_ucsc_refseq"], [1, 2, 1, "", "cdss"], [1, 2, 1, "", "exons"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "genes"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"]], "genome_kit.GenomeDNA": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "reference_genome"]], "genome_kit.GenomeTrack": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "close"], [1, 2, 1, "", "dim"], [1, 2, 1, "", "dtype"], [1, 2, 1, "", "etype"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "gtrack_version"], [1, 1, 1, "", "histogram"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "resolution"], [1, 2, 1, "", "stranded"]], "genome_kit.GenomeTrackBuilder": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "data_size"], [1, 2, 1, "", "dim"], [1, 2, 1, "", "dtype"], [1, 2, 1, "", "etype"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "finalize"], [1, 1, 1, "", "flush"], [1, 2, 1, "", "index_size"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "resolution"], [1, 1, 1, "", "set_clamping"], [1, 1, 1, "", "set_data"], [1, 1, 1, "", "set_data_from_bed"], [1, 1, 1, "", "set_data_from_bedgraph"], [1, 1, 1, "", "set_data_from_wig"], [1, 1, 1, "", "set_default_value"], [1, 1, 1, "", "set_dict"], [1, 1, 1, "", "set_restriction"], [1, 1, 1, "", "set_sparsity"], [1, 1, 1, "", "set_transform"], [1, 2, 1, "", "stranded"]], "genome_kit.Interval": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 2, 1, "", "anchor"], [1, 2, 1, "", "anchor_offset"], [1, 1, 1, "", "as_dna0"], [1, 1, 1, "", "as_dna1"], [1, 1, 1, "", "as_negative_strand"], [1, 1, 1, "", "as_opposite_strand"], [1, 1, 1, "", "as_positive_strand"], [1, 1, 1, "", "as_rna1"], [1, 1, 1, "", "as_ucsc"], [1, 2, 1, "", "chromosome"], [1, 1, 1, "", "contains"], [1, 1, 1, "", "distance"], [1, 1, 1, "", "dnstream_of"], [1, 2, 1, "", "end"], [1, 2, 1, "", "end3"], [1, 2, 1, "", "end5"], [1, 1, 1, "", "expand"], [1, 1, 1, "", "from_dna0"], [1, 1, 1, "", "from_dna0_coord"], [1, 1, 1, "", "from_dna1"], [1, 1, 1, "", "from_rna1"], [1, 1, 1, "", "intersect"], [1, 1, 1, "", "is_positive_strand"], [1, 2, 1, "", "midpoint"], [1, 1, 1, "", "overlaps"], [1, 2, 1, "", "reference_genome"], [1, 1, 1, "", "shift"], [1, 1, 1, "", "spanning"], [1, 2, 1, "", "start"], [1, 2, 1, "", "strand"], [1, 1, 1, "", "subtract"], [1, 1, 1, "", "upstream_of"], [1, 1, 1, "", "with_anchor"], [1, 1, 1, "", "within"]], "genome_kit.Intron": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "index"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_exon"], [1, 2, 1, "", "next_intron"], [1, 2, 1, "", "prev_exon"], [1, 2, 1, "", "prev_intron"], [1, 2, 1, "", "transcript"]], "genome_kit.IntronTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JReadAlignments": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_jralign"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "jralign_version"], [1, 2, 1, "", "junctions"], [1, 2, 1, "", "variants"]], "genome_kit.Junction": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "alignments"], [1, 2, 1, "", "interval"]], "genome_kit.JunctionReadAlignment": [[1, 1, 1, "", "__repr__"], [1, 2, 1, "", "left"], [1, 2, 1, "", "num_variants"], [1, 2, 1, "", "right"], [1, 2, 1, "", "strand"], [1, 1, 1, "", "variants"]], "genome_kit.JunctionReadAlignments": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "num_reads"]], "genome_kit.JunctionReadAlignmentsTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JunctionReadCount": [[1, 1, 1, "", "__repr__"], [1, 2, 1, "", "count"], [1, 2, 1, "", "shift"], [1, 2, 1, "", "strand"]], "genome_kit.JunctionReadDistribution": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "num_counts"], [1, 2, 1, "", "num_reads"]], "genome_kit.JunctionReadDistributionTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JunctionTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.ReadAlignments": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "alignments"], [1, 1, 1, "", "build_ralign"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "junctions"], [1, 2, 1, "", "matches"], [1, 1, 1, "", "pileup"], [1, 1, 1, "", "ralign_version"], [1, 2, 1, "", "variants"]], "genome_kit.ReadDistributions": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_rdist"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "from_jraligns"], [1, 2, 1, "", "junctions"], [1, 1, 1, "", "rdist_version"]], "genome_kit.Transcript": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "ccds_id"], [1, 2, 1, "", "cdss"], [1, 2, 1, "", "exons"], [1, 2, 1, "", "gene"], [1, 2, 1, "", "id"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "level"], [1, 2, 1, "", "product"], [1, 2, 1, "", "protein_id"], [1, 2, 1, "", "tsl"], [1, 2, 1, "", "type"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"]], "genome_kit.TranscriptTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.VCFTable": [[1, 3, 1, "", "GT_HETEROZYGOUS"], [1, 3, 1, "", "GT_HOMOZYGOUS_ALT"], [1, 3, 1, "", "GT_HOMOZYGOUS_REF"], [1, 3, 1, "", "GT_UNKNOWN"], [1, 3, 1, "", "SVTYPE_NA"], [1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__len__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_vcfbin"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 1, 1, "", "format"], [1, 2, 1, "", "format_ids"], [1, 1, 1, "", "from_vcf"], [1, 1, 1, "", "info"], [1, 2, 1, "", "info_ids"], [1, 1, 1, "", "masked"], [1, 2, 1, "", "num_samples"], [1, 2, 1, "", "sample_names"], [1, 1, 1, "", "sequence_variations"], [1, 2, 1, "", "stranded"], [1, 1, 1, "", "variant_combinations"], [1, 1, 1, "", "vcfbin_version"], [1, 1, 1, "", "where"]], "genome_kit.VCFVariant": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 1, 1, "", "as_variant_string"]], "genome_kit.Variant": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 2, 1, "", "alt"], [1, 1, 1, "", "as_variant_string"], [1, 1, 1, "", "from_string"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "position"], [1, 2, 1, "", "ref"], [1, 1, 1, "", "spanning"]], "genome_kit.VariantGenome": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "dna"], [1, 1, 1, "", "find_motif"]], "genome_kit.VariantTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"], [1, 1, 1, "", "where"]], "genome_kit._apply_variants": [[1, 4, 1, "", "apply_variants"]], "genome_kit._cxx_util": [[1, 4, 1, "", "mock"], [1, 4, 1, "", "mock_result"], [1, 4, 1, "", "mock_unreachable"], [1, 4, 1, "", "strip_mock_bases"]], "genome_kit._twobitutils": [[1, 4, 1, "", "write2bit"]], "genome_kit.data_manager": [[1, 0, 1, "", "ProgressPercentage"]], "genome_kit.data_manager.ProgressPercentage": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__init__"], [1, 3, 1, "", "__weakref__"]], "genome_kit.gk_data": [[1, 4, 1, "", "get_file"], [1, 4, 1, "", "upload_file"], [1, 4, 1, "", "wget"]]}, "objtypes": {"0": "py:class", "1": "py:method", "2": "py:property", "3": "py:attribute", "4": "py:function", "5": "py:module"}, "objnames": {"0": ["py", "class", "Python class"], "1": ["py", "method", "Python method"], "2": ["py", "property", "Python property"], "3": ["py", "attribute", "Python attribute"], "4": ["py", "function", "Python function"], "5": ["py", "module", "Python module"]}, "titleterms": {"anchor": 0, "interv": [0, 1, 5], "without": 0, "delet": 0, "an": 0, "specifi": 0, "insert": 0, "todo": 0, "outsid": 0, "api": 1, "document": 1, "variant": [1, 5], "genom": [1, 3], "variantgenom": 1, "track": [1, 5], "genomedna": 1, "genometrack": 1, "genometrackbuild": 1, "annot": [1, 3, 5], "genomeannot": 1, "gene": 1, "genet": 1, "transcript": 1, "transcriptt": 1, "exon": 1, "exont": 1, "intron": 1, "intront": 1, "cd": 1, "cdstabl": 1, "read": 1, "align": 1, "readalign": 1, "junction": 1, "junctiont": 1, "alignmentt": 1, "alignmentmatch": 1, "alignmentmatcht": 1, "jreadalign": 1, "junctionreadalignmentst": 1, "junctionreadalign": 1, "readdistribut": 1, "junctionreaddistributiont": 1, "junctionreaddistribut": 1, "junctionreadcount": 1, "vcf": [1, 5], "tabl": [1, 4], "vcftabl": 1, "vcfvariant": 1, "variantt": 1, "data": [1, 2, 5], "access": [1, 5], "marshal": [], "unmarsh": [], "as_proto": [], "from_proto": [], "genome_kit": 1, "datamanag": 1, "intern": 1, "gk_data": 1, "data_manag": 1, "_twobitutil": 1, "_apply_vari": 1, "_cxx_util": 1, "develop": 2, "set": 2, "up": 2, "make": 2, "chang": 2, "ad": 2, "test": 2, "c": 2, "debug": 2, "build": 2, "file": [2, 5], "full": 2, "releas": 2, "genomekit": [2, 4], "garbag": [], "collect": [], "issu": [], "maco": [], "exampl": 3, "assembli": 3, "what": 4, "i": 4, "us": 4, "featur": [4, 5], "resourc": 4, "avail": 4, "next": 4, "step": 4, "indic": 4, "quickstart": 5, "instal": 5, "via": 5, "conda": 5, "basic": 5, "work": 5, "extract": 5, "from": 5, "motif": 5, "find": 5, "share": 5, "new": 5, "jetbrain": 2, "clion": 2, "setup": 2, "initi": 5}, "envversion": {"sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.todo": 2, "sphinx.ext.viewcode": 1, "sphinx": 58}, "alltitles": {"Anchors": [[0, "anchors"]], "Intervals without anchors": [[0, "intervals-without-anchors"]], "Deletions with an anchored interval": [[0, "deletions-with-an-anchored-interval"]], "Specifying anchors": [[0, "specifying-anchors"]], "Insertions with an anchored interval": [[0, "insertions-with-an-anchored-interval"]], "Deletions at the anchor": [[0, "deletions-at-the-anchor"]], "Insertions at the anchor": [[0, "insertions-at-the-anchor"]], "Todo": [[0, "id3"]], "Anchors outside an interval": [[0, "anchors-outside-an-interval"]], "API Documentation": [[1, "api-documentation"]], "Interval": [[1, "interval"]], "Variant": [[1, "variant"]], "Genome": [[1, "genome"]], "VariantGenome": [[1, "variantgenome"]], "Tracks": [[1, "tracks"], [5, "tracks"]], "GenomeDNA": [[1, "genomedna"]], "GenomeTrack": [[1, "genometrack"]], "GenomeTrackBuilder": [[1, "genometrackbuilder"]], "Annotations": [[1, "annotations"], [3, "annotations"], [5, "annotations"]], "GenomeAnnotation": [[1, "genomeannotation"]], "Gene": [[1, "gene"]], "GeneTable": [[1, "genetable"]], "Transcript": [[1, "transcript"]], "TranscriptTable": [[1, "transcripttable"]], "Exon": [[1, "exon"]], "ExonTable": [[1, "exontable"]], "Intron": [[1, "intron"]], "IntronTable": [[1, "introntable"]], "Cds": [[1, "cds"]], "CdsTable": [[1, "cdstable"]], "Read Alignments": [[1, "read-alignments"]], "ReadAlignments": [[1, "readalignments"]], "Junction": [[1, "junction"]], "JunctionTable": [[1, "junctiontable"]], "Alignment": [[1, "alignment"]], "AlignmentTable": [[1, "alignmenttable"]], "AlignmentMatch": [[1, "alignmentmatch"]], "AlignmentMatchTable": [[1, "alignmentmatchtable"]], "Junction Read Alignments": [[1, "junction-read-alignments"]], "JReadAlignments": [[1, "jreadalignments"]], "JunctionReadAlignmentsTable": [[1, "junctionreadalignmentstable"]], "JunctionReadAlignments": [[1, "junctionreadalignments"]], "JunctionReadAlignment": [[1, "junctionreadalignment"]], "ReadDistributions": [[1, "readdistributions"]], "JunctionReadDistributionTable": [[1, "junctionreaddistributiontable"]], "JunctionReadDistribution": [[1, "junctionreaddistribution"]], "JunctionReadCount": [[1, "junctionreadcount"]], "VCF Tables": [[1, "vcf-tables"]], "VCFTable": [[1, "vcftable"]], "VCFVariant": [[1, "vcfvariant"]], "VariantTable": [[1, "varianttable"]], "Data Access": [[1, "data-access"]], "genome_kit.DataManager": [[1, "genome-kit-datamanager"]], "Internals": [[1, "internals"]], "genome_kit.gk_data": [[1, "module-genome_kit.gk_data"]], "genome_kit.data_manager": [[1, "genome-kit-data-manager"]], "genome_kit._twobitutils": [[1, "genome-kit-twobitutils"]], "genome_kit._apply_variants": [[1, "genome-kit-apply-variants"]], "genome_kit._cxx_util": [[1, "genome-kit-cxx-util"]], "GenomeKit": [[4, "genomekit"]], "What is GenomeKit?": [[4, "what-is-genomekit"]], "Useful Features": [[4, "useful-features"]], "Resources Available": [[4, "resources-available"]], "Next steps": [[4, "next-steps"]], "Indices and tables": [[4, "indices-and-tables"]], "Developers": [[2, "developers"]], "Setting up": [[2, "setting-up"]], "Jetbrains CLion setup": [[2, "jetbrains-clion-setup"]], "Making changes": [[2, "making-changes"]], "Adding tests": [[2, "adding-tests"]], "C++ tests": [[2, "c-tests"]], "Debugging tests": [[2, "debugging-tests"]], "Building data files": [[2, "building-data-files"]], "Building full data files": [[2, "building-full-data-files"]], "Building test data files": [[2, "building-test-data-files"]], "Releasing GenomeKit": [[2, "releasing-genomekit"]], "Genomes": [[3, "genomes"]], "Examples": [[3, "examples"]], "Assemblies": [[3, "assemblies"]], "Quickstart": [[5, "quickstart"]], "Installation": [[5, "installation"]], "Install via conda": [[5, "install-via-conda"]], "Basics": [[5, "basics"]], "Initial Data Access": [[5, "initial-data-access"]], "Working with Intervals": [[5, "working-with-intervals"]], "Feature extraction basics": [[5, "feature-extraction-basics"]], "Variants": [[5, "variants"]], "Variants from VCF Files": [[5, "variants-from-vcf-files"]], "Feature Extraction with Variants": [[5, "feature-extraction-with-variants"]], "Motif finding": [[5, "motif-finding"]], "Sharing new data files": [[5, "sharing-new-data-files"]]}, "indexentries": {}})