

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>genome_kit.variant_genome &mdash; GenomeKit 6.5.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
    <link rel="shortcut icon" href="../../_static/favicon.ico"/>
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=db1c604c"></script>
      <script src="../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            GenomeKit
              <img src="../../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../anchors.html">Anchors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../genomes.html">Genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../develop.html">Developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data_org.html">Remote Data Organization</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">GenomeKit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">genome_kit.variant_genome</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for genome_kit.variant_genome</h1><div class="highlight"><pre>
<span></span><span class="c1"># Copyright (C) 2016-2023 Deep Genomics Inc. All Rights Reserved.</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">__future__</span><span class="w"> </span><span class="kn">import</span> <span class="n">absolute_import</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">functools</span><span class="w"> </span><span class="kn">import</span> <span class="n">partial</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">.</span><span class="w"> </span><span class="kn">import</span> <span class="n">_util</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.</span><span class="w"> </span><span class="kn">import</span> <span class="n">interval</span> <span class="k">as</span> <span class="n">_interval</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._apply_variants</span><span class="w"> </span><span class="kn">import</span> <span class="n">apply_variants</span><span class="p">,</span> <span class="n">check_variants_list</span>


<div class="viewcode-block" id="VariantGenome">
<a class="viewcode-back" href="../../api.html#genome_kit.VariantGenome">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">VariantGenome</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A reference genome with a number of variants applied to it.</span>

<span class="sd">    A variant genome behaves like a :py:class:`~genome_kit.Genome` object,</span>
<span class="sd">    object, except any requests for DNA sequence (or quantities that depend</span>
<span class="sd">    on DNA sequence) will be returned with the variants already applied.</span>
<span class="sd">    If the variants change the length of the sequence, then the</span>
<span class="sd">    interval&#39;s `anchor` and `anchor_offset` determine how the query interval</span>
<span class="sd">    will be aligned to the variant genome&#39;s coordinate system.</span>

<span class="sd">    TODO: see explanation of anchors.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;genome&#39;</span><span class="p">,</span> <span class="s1">&#39;variants&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="VariantGenome.__init__">
<a class="viewcode-back" href="../../api.html#genome_kit.VariantGenome.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference_genome</span><span class="p">,</span> <span class="n">variants</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initialize a variant genome.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        reference_genome : :py:class:`~genome_kit.Genome`</span>
<span class="sd">            The reference genome to which variants are applied.</span>

<span class="sd">        variants : :py:class:`~genome_kit.Variant` | :py:class:`list` of (:py:class:`~genome_kit.Variant`)</span>
<span class="sd">            Either a single variant or a list of variants. Aligned-normalization of</span>
<span class="sd">            variants is not yet implemented, so the user is currently</span>
<span class="sd">            responsible for not passing variants that overlap. The variants must be provided</span>
<span class="sd">            as a :py:class:`~genome_kit.Variant` object. String variants are no longer allowed,</span>
<span class="sd">            please see :py:class:`~genome_kit.Variant` for more information. Since we assume that</span>
<span class="sd">            variants never overlap one another this means that all variant</span>
<span class="sd">            positions are with respect to the reference genome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">reference_genome</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variants</span> <span class="o">=</span> <span class="n">check_variants_list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">variants</span><span class="p">)</span></div>


<div class="viewcode-block" id="VariantGenome.dna">
<a class="viewcode-back" href="../../api.html#genome_kit.VariantGenome.dna">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">dna</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interval</span><span class="p">,</span> <span class="n">allow_outside_chromosome</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract variant DNA for an anchored interval or coordinate.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        interval : :py:class:`~genome_kit.Interval`</span>
<span class="sd">            If `interval` is on the negative strand, the reverse complement</span>
<span class="sd">            sequence is returned. The `interval.anchor` and `interval.anchor_offset`</span>
<span class="sd">            attributes determine the policy of how length-changing variants are applied.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        :py:class:`str`</span>
<span class="sd">            The variant DNA sequence.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">apply_variants</span><span class="p">(</span>
            <span class="n">partial</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">dna</span><span class="p">,</span> <span class="n">allow_outside_chromosome</span><span class="o">=</span><span class="n">allow_outside_chromosome</span><span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">,</span>
            <span class="n">interval</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="VariantGenome.find_motif">
<a class="viewcode-back" href="../../api.html#genome_kit.VariantGenome.find_motif">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">find_motif</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interval</span><span class="p">,</span> <span class="n">motif</span><span class="p">,</span> <span class="n">match_position</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">find_overlapping_motifs</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find a genomic motif in an interval on a :py:class:`~genome_kit.VariantGenome`.</span>

<span class="sd">        This method finds all occurrences of a genomic motif on a VariantGenome. It can</span>
<span class="sd">        handle all kinds of variants including indels and complex variants. When a motif</span>
<span class="sd">        is found within an insertion, it will set the ``anchor_offset`` attribute on</span>
<span class="sd">        the returned :py:class:`~genome_kit.Interval`.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        interval : :py:class:`~genome_kit.Interval`</span>
<span class="sd">            A genomic interval to search for the motif.</span>
<span class="sd">        motif : :py:class:`str`</span>
<span class="sd">            A genomic motif to search for.</span>
<span class="sd">        match_position : :py:class:`int` | :py:class:`str`</span>
<span class="sd">            Very often you want to find a position after a motif or in the middle</span>
<span class="sd">            of a motif. For example, when you are searching for a ``AG`` core</span>
<span class="sd">            acceptor site motif you want to match the 3p end of the ``AG`` motif to</span>
<span class="sd">            return the position of the putative splice site.</span>
<span class="sd">            To do this you can specify a match position as string with the values</span>
<span class="sd">            ``&quot;3p&quot;`` or ``&quot;5p&quot;``, or you can pass an integer that indicates a position</span>
<span class="sd">            in the motif with ``0 &lt;= match_position &lt;= len(motif)``.</span>
<span class="sd">        find_overlapping_motifs : :py:class:`bool`</span>
<span class="sd">            Whether to return matches for overlapping motifs. The default is to only</span>
<span class="sd">            find non-overlapping motifs.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        :py:class`list` of :py:class:`~genome_kit.Interval`</span>
<span class="sd">            A list of all matches (length-0 intervals) found in the query interval.</span>

<span class="sd">        Example</span>
<span class="sd">        -------::</span>
<span class="sd">            &gt;&gt;&gt; from genome_kit import Genome, Variant</span>
<span class="sd">            &gt;&gt;&gt; genome = Genome(&#39;hg19&#39;)</span>
<span class="sd">            &gt;&gt;&gt; variant = Variant.from_string(&quot;chr1:40033:A:G&quot;, genome)</span>
<span class="sd">            &gt;&gt;&gt; variant_genome = VariantGenome(genome, variant)</span>
<span class="sd">            &gt;&gt;&gt; interval = genome.interval(&#39;chr1&#39;, &#39;+&#39;, 40000, 40080)</span>
<span class="sd">            &gt;&gt;&gt; motif = &#39;TAG&#39;</span>
<span class="sd">            &gt;&gt;&gt; variant_genome.find_motif(interval, motif)</span>
<span class="sd">            [Interval(&quot;chr1&quot;, &quot;+&quot;, 40030, 40030, &quot;hg19&quot;, 40030)]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: Longterm we&#39;ll want to move towards a generator function to be more efficient</span>
        <span class="c1"># TODO: on long intervals.</span>

        <span class="k">if</span> <span class="n">match_position</span> <span class="o">==</span> <span class="s1">&#39;5p&#39;</span><span class="p">:</span>
            <span class="n">match_position</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">elif</span> <span class="n">match_position</span> <span class="o">==</span> <span class="s1">&#39;3p&#39;</span><span class="p">:</span>
            <span class="n">match_position</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">match_position</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">motif</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Require 0 &lt;= match_position &lt;= len(motif) [match_position=</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">match_position</span><span class="p">))</span>

        <span class="n">strand</span> <span class="o">=</span> <span class="n">interval</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span>

        <span class="c1"># All motif finding is done on the reference strand. Therefore the motif</span>
        <span class="c1"># is reverse complemented when the interval is on the reverse strand. As</span>
        <span class="c1"># well, the match_position is recomputed to be applied from the end of the</span>
        <span class="c1"># motif (such at it is correct on the reverse strand), e.g. &#39;5p&#39; means</span>
        <span class="c1"># match_position = 0 on the forward strand and match_position = len(motif) on</span>
        <span class="c1"># the reverse strand.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">strand</span><span class="p">:</span>
            <span class="n">match_position</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span> <span class="o">-</span> <span class="n">match_position</span>

        <span class="n">motif</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="n">dna_start</span> <span class="o">=</span> <span class="n">interval</span><span class="o">.</span><span class="n">start</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">strand</span><span class="p">:</span>
            <span class="n">motif</span> <span class="o">=</span> <span class="n">_util</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span>

        <span class="c1"># We extend the interval by one to to handle the case when</span>
        <span class="c1"># match_position=len(motif) and the motif occurs at the very</span>
        <span class="c1"># end of the interval. In that case we would get</span>
        <span class="c1"># pos+match_position == len(reference_alignment) which results in</span>
        <span class="c1"># an IndexError.</span>
        <span class="n">forward_interval</span> <span class="o">=</span> <span class="n">_interval</span><span class="o">.</span><span class="n">Interval</span><span class="p">(</span><span class="n">interval</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">end</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                              <span class="n">interval</span><span class="o">.</span><span class="n">reference_genome</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">anchor</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">anchor_offset</span><span class="p">)</span>

        <span class="c1"># We get both the DNA itself and the reference alignment. The reference alignment</span>
        <span class="c1"># is necessary for constructing a Coordinate that is in terms of the reference</span>
        <span class="c1"># genome.</span>
        <span class="n">dna</span><span class="p">,</span> <span class="n">reference_alignment</span> <span class="o">=</span> <span class="n">apply_variants</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">dna</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">,</span> <span class="n">forward_interval</span><span class="p">,</span> <span class="n">reference_alignment</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># We truncate only the DNA by one so we don&#39;t get any matches past the interval</span>
        <span class="n">dna</span> <span class="o">=</span> <span class="n">dna</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">motif_hits</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">if</span> <span class="n">find_overlapping_motifs</span><span class="p">:</span>
            <span class="n">offset_step</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">offset_step</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span>

        <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">motif</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">break</span>

            <span class="n">pos_ref</span> <span class="o">=</span> <span class="n">reference_alignment</span><span class="p">[</span><span class="n">pos</span> <span class="o">+</span> <span class="n">match_position</span><span class="p">]</span>

            <span class="n">anchor_offset</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pos_ref</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">):</span>
                <span class="n">pos_ref</span><span class="p">,</span> <span class="n">anchor_offset</span> <span class="o">=</span> <span class="n">pos_ref</span>
            <span class="k">elif</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">match_position</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># if this is immediately after an insertion, we need to set</span>
                <span class="c1"># the anchor offset past the insertion so expanding upstream</span>
                <span class="c1"># still works</span>
                <span class="n">prev_ref</span> <span class="o">=</span> <span class="n">reference_alignment</span><span class="p">[</span><span class="n">pos</span> <span class="o">+</span> <span class="n">match_position</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prev_ref</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">):</span>
                    <span class="n">pos_ref</span><span class="p">,</span> <span class="n">anchor_offset</span> <span class="o">=</span> <span class="n">prev_ref</span>
                    <span class="n">anchor_offset</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">anchor</span> <span class="o">=</span> <span class="n">dna_start</span> <span class="o">+</span> <span class="n">pos_ref</span>

            <span class="n">hit</span> <span class="o">=</span> <span class="n">_interval</span><span class="o">.</span><span class="n">Interval</span><span class="p">(</span>
                <span class="n">interval</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span>
                <span class="n">interval</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                <span class="n">anchor</span><span class="p">,</span>
                <span class="n">anchor</span><span class="p">,</span>
                <span class="n">interval</span><span class="o">.</span><span class="n">refg</span><span class="p">,</span>
                <span class="n">anchor</span><span class="o">=</span><span class="n">anchor</span><span class="p">,</span>
                <span class="n">anchor_offset</span><span class="o">=</span><span class="n">anchor_offset</span><span class="p">)</span>
            <span class="n">motif_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hit</span><span class="p">)</span>
            <span class="n">offset</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">offset_step</span>

        <span class="k">return</span> <span class="n">motif_hits</span></div>


<div class="viewcode-block" id="VariantGenome.__repr__">
<a class="viewcode-back" href="../../api.html#genome_kit.VariantGenome.__repr__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;&lt;VariantGenome </span><span class="si">{}</span><span class="s1"> with </span><span class="si">{}</span><span class="s1"> variants&gt;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_genome</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">))</span></div>


    <span class="k">def</span><span class="w"> </span><span class="fm">__getattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="c1"># Intercept any attribute requests that weren&#39;t found on the VariantGenome object itself,</span>
        <span class="c1"># and forward those requests to the Genome object we&#39;re wrapping.</span>

        <span class="c1"># TODO: This is problematic: See DGENGINE-1401</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="fm">__getattribute__</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Deep Genomics.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>