

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Remote Data Organization &mdash; GenomeKit 6.1.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=c7384f69"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Developers" href="develop.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            GenomeKit
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="anchors.html">Anchors</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="genomes.html">Genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="develop.html">Developers</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Remote Data Organization</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#available-file-formats">Available File Formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="#data-access">Data Access</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GenomeKit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Remote Data Organization</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/data_org.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="remote-data-organization">
<span id="data-org"></span><h1>Remote Data Organization<a class="headerlink" href="#remote-data-organization" title="Link to this heading"></a></h1>
<p>In order to facilitate fast operations and queries over genomic data, GenomeKit
uses an internal optimized binary data format. GenomeKit provides APIs for converting
existing data formats such as gff3, vcf, sam, bam, bed, wig, etc into its own format.</p>
<p>In addition, various data and metadata support files are used. Out of the various formats,
only files related to genomic assemblies and annotations can be downloaded from freely
provided remote storage.
Other files (genomic tracks, indexed VCFs, etc) are maintained by users.</p>
<section id="available-file-formats">
<h2>Available File Formats<a class="headerlink" href="#available-file-formats" title="Link to this heading"></a></h2>
<p>Formats not specific to GenomeKit:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">2bit</span></code> (raw assembly data)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">chrom.sizes</span></code> / <code class="docutils literal notranslate"><span class="pre">chromAlias.txt</span></code> (text format)</p></li>
</ul>
<p>GenomeKit-specific data files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">dganno</span></code>: genomic annotation data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cfg</span></code>: annotation metadata, such as associated assembly name (text format)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">appris.*.pkl</span></code>: Python-pickled appris metadata</p></li>
</ul>
</section>
<section id="data-access">
<h2>Data Access<a class="headerlink" href="#data-access" title="Link to this heading"></a></h2>
<p>Assembly and annotation files are accessed transparently by GenomeKit when users
create Genome objects. For example, when users call
<code class="docutils literal notranslate"><span class="pre">gk.Genome(&quot;gencode.v29&quot;).dna(gk.Genome(&quot;gencode.v29&quot;).exons[0].interval)</span></code>,
GenomeKit downloads (if not already locally available) the files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gencode.v29.cfg</span></code> (annotation configuration)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">hg38.p12.2bit</span></code> (associated assembly)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">hg38.p12.chrom.sizes</span></code> (list of chromosomes/contigs and aliases)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gencode.v29.v22.dganno</span></code> (indexed annotation features ; “v22” specifies the format version)</p></li>
</ul>
<p>In the source bucket, files are organized in a flat structure and are easily
found by their name. The binary version is matched via library code.</p>
<p>Files are also directly accessible by using GenomeKit’s data_manager API.
This API also allows users to configure their own buckets (to also allow uploads)
or a different implementation.</p>
<p>Downloads are automatically cached locally and so are only downloaded once.
They are lazily downloaded, so that only the required files are retrieved.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="develop.html" class="btn btn-neutral float-left" title="Developers" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Deep Genomics.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>