Search.setIndex({"alltitles": {"API Documentation": [[1, null]], "Adding tests": [[3, "adding-tests"]], "Alignment": [[1, "alignment"]], "AlignmentMatch": [[1, "alignmentmatch"]], "AlignmentMatchTable": [[1, "alignmentmatchtable"]], "AlignmentTable": [[1, "alignmenttable"]], "Anchors": [[0, null]], "Anchors outside an interval": [[0, "anchors-outside-an-interval"]], "Annotations": [[1, "annotations"], [4, "annotations"], [6, "annotations"]], "Assemblies": [[4, "assemblies"]], "Available File Formats": [[2, "available-file-formats"]], "Basics": [[6, "basics"]], "Building data files": [[3, "building-data-files"]], "Building full data files": [[3, "building-full-data-files"]], "Building test data files": [[3, "building-test-data-files"]], "C++ tests": [[3, "c-tests"]], "Cds": [[1, "cds"]], "CdsTable": [[1, "cdstable"]], "Data Access": [[1, "data-access"], [2, "data-access"]], "Debugging tests": [[3, "debugging-tests"]], "Deletions at the anchor": [[0, "deletions-at-the-anchor"]], "Deletions with an anchored interval": [[0, "deletions-with-an-anchored-interval"]], "Developers": [[3, null]], "Examples": [[4, "examples"]], "Exon": [[1, "exon"]], "ExonTable": [[1, "exontable"]], "Feature Extraction with Variants": [[6, "feature-extraction-with-variants"]], "Feature extraction basics": [[6, "feature-extraction-basics"]], "Gene": [[1, "gene"]], "GeneTable": [[1, "genetable"]], "Generating data with Docker": [[6, "generating-data-with-docker"]], "Genome": [[1, "genome"]], "GenomeAnnotation": [[1, "genomeannotation"]], "GenomeDNA": [[1, "genomedna"]], "GenomeKit": [[5, null]], "GenomeTrack": [[1, "genometrack"]], "GenomeTrackBuilder": [[1, "genometrackbuilder"]], "Genomes": [[4, null]], "Indices and tables": [[5, "indices-and-tables"]], "Initial Data Access": [[6, "initial-data-access"]], "Insertions at the anchor": [[0, "insertions-at-the-anchor"]], "Insertions with an anchored interval": [[0, "insertions-with-an-anchored-interval"]], "Install via conda": [[6, "install-via-conda"]], "Installation": [[6, "installation"]], "Internals": [[1, "internals"]], "Interval": [[1, "interval"]], "Intervals without anchors": [[0, "intervals-without-anchors"]], "Intron": [[1, "intron"]], "IntronTable": [[1, "introntable"]], "JReadAlignments": [[1, "jreadalignments"]], "Jetbrains CLion setup": [[3, "jetbrains-clion-setup"]], "Junction": [[1, "junction"]], "Junction Read Alignments": [[1, "junction-read-alignments"]], "JunctionReadAlignment": [[1, "junctionreadalignment"]], "JunctionReadAlignments": [[1, "junctionreadalignments"]], "JunctionReadAlignmentsTable": [[1, "junctionreadalignmentstable"]], "JunctionReadCount": [[1, "junctionreadcount"]], "JunctionReadDistribution": [[1, "junctionreaddistribution"]], "JunctionReadDistributionTable": [[1, "junctionreaddistributiontable"]], "JunctionTable": [[1, "junctiontable"]], "Length-changing variants": [[6, "length-changing-variants"]], "Logging and troubleshooting": [[6, "logging-and-troubleshooting"]], "Making changes": [[3, "making-changes"]], "Motif finding": [[6, "motif-finding"]], "Next steps": [[5, "next-steps"]], "Quickstart": [[6, null]], "Read Alignments": [[1, "read-alignments"]], "ReadAlignments": [[1, "readalignments"]], "ReadDistributions": [[1, "readdistributions"]], "Releasing GenomeKit": [[3, "releasing-genomekit"]], "Remote Data Organization": [[2, null]], "Resources Available": [[5, "resources-available"]], "Setting up": [[3, "setting-up"]], "Sharing new data files": [[6, "sharing-new-data-files"]], "Specifying anchors": [[0, "specifying-anchors"]], "Todo": [[0, "id3"]], "Tracks": [[1, "tracks"], [6, "tracks"]], "Transcript": [[1, "transcript"]], "TranscriptTable": [[1, "transcripttable"]], "Usage": [[6, "usage"]], "Use with docker": [[6, "use-with-docker"]], "Useful Features": [[5, "useful-features"]], "VCF Tables": [[1, "vcf-tables"]], "VCFTable": [[1, "vcftable"]], "VCFVariant": [[1, "vcfvariant"]], "Variant": [[1, "variant"]], "Variant genomes": [[6, "variant-genomes"]], "VariantGenome": [[1, "variantgenome"]], "VariantTable": [[1, "varianttable"]], "Variants": [[6, "variants"]], "Variants from VCF Files": [[6, "variants-from-vcf-files"]], "What is GenomeKit?": [[5, "what-is-genomekit"]], "Working with Intervals": [[6, "working-with-intervals"]], "genome_kit.DataManager": [[1, "genome-kit-datamanager"]], "genome_kit._apply_variants": [[1, "genome-kit-apply-variants"]], "genome_kit._cxx_util": [[1, "genome-kit-cxx-util"]], "genome_kit._twobitutils": [[1, "genome-kit-twobitutils"]], "genome_kit.data_manager": [[1, "genome-kit-data-manager"]], "genome_kit.gk_data": [[1, "module-genome_kit.gk_data"]]}, "docnames": ["anchors", "api", "data_org", "develop", "genomes", "index", "quickstart"], "envversion": {"sphinx": 62, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.todo": 2, "sphinx.ext.viewcode": 1}, "filenames": ["anchors.rst", "api.rst", "data_org.rst", "develop.rst", "genomes.rst", "index.rst", "quickstart.rst"], "indexentries": {"__call__() (genomedna method)": [[1, "genome_kit.GenomeDNA.__call__", false]], "__call__() (genometrack method)": [[1, "genome_kit.GenomeTrack.__call__", false]], "__call__() (progresspercentage method)": [[1, "genome_kit.data_manager.ProgressPercentage.__call__", false]], "__getitem__() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.__getitem__", false]], "__getitem__() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.__getitem__", false]], "__getitem__() (cdstable method)": [[1, "genome_kit.CdsTable.__getitem__", false]], "__getitem__() (exontable method)": [[1, "genome_kit.ExonTable.__getitem__", false]], "__getitem__() (genetable method)": [[1, "genome_kit.GeneTable.__getitem__", false]], "__getitem__() (introntable method)": [[1, "genome_kit.IntronTable.__getitem__", false]], "__getitem__() (junctionreadalignments method)": [[1, "genome_kit.JunctionReadAlignments.__getitem__", false]], "__getitem__() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.__getitem__", false]], "__getitem__() (junctionreaddistribution method)": [[1, "genome_kit.JunctionReadDistribution.__getitem__", false]], "__getitem__() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.__getitem__", false]], "__getitem__() (junctiontable method)": [[1, "genome_kit.JunctionTable.__getitem__", false]], "__getitem__() (transcripttable method)": [[1, "genome_kit.TranscriptTable.__getitem__", false]], "__getitem__() (varianttable method)": [[1, "genome_kit.VariantTable.__getitem__", false]], "__getitem__() (vcftable method)": [[1, "genome_kit.VCFTable.__getitem__", false]], "__getstate__() (cds method)": [[1, "genome_kit.Cds.__getstate__", false]], "__getstate__() (exon method)": [[1, "genome_kit.Exon.__getstate__", false]], "__getstate__() (gene method)": [[1, "genome_kit.Gene.__getstate__", false]], "__getstate__() (intron method)": [[1, "genome_kit.Intron.__getstate__", false]], "__getstate__() (transcript method)": [[1, "genome_kit.Transcript.__getstate__", false]], "__getstate__() (variant method)": [[1, "genome_kit.Variant.__getstate__", false]], "__init__() (alignment method)": [[1, "genome_kit.Alignment.__init__", false]], "__init__() (alignmentmatch method)": [[1, "genome_kit.AlignmentMatch.__init__", false]], "__init__() (cds method)": [[1, "genome_kit.Cds.__init__", false]], "__init__() (datamanager method)": [[1, "genome_kit.DataManager.__init__", false]], "__init__() (exon method)": [[1, "genome_kit.Exon.__init__", false]], "__init__() (gene method)": [[1, "genome_kit.Gene.__init__", false]], "__init__() (genome method)": [[1, "genome_kit.Genome.__init__", false]], "__init__() (genometrack method)": [[1, "genome_kit.GenomeTrack.__init__", false]], "__init__() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.__init__", false]], "__init__() (interval method)": [[1, "genome_kit.Interval.__init__", false]], "__init__() (intron method)": [[1, "genome_kit.Intron.__init__", false]], "__init__() (jreadalignments method)": [[1, "genome_kit.JReadAlignments.__init__", false]], "__init__() (junction method)": [[1, "genome_kit.Junction.__init__", false]], "__init__() (junctionreadalignments method)": [[1, "genome_kit.JunctionReadAlignments.__init__", false]], "__init__() (junctionreaddistribution method)": [[1, "genome_kit.JunctionReadDistribution.__init__", false]], "__init__() (progresspercentage method)": [[1, "genome_kit.data_manager.ProgressPercentage.__init__", false]], "__init__() (readalignments method)": [[1, "genome_kit.ReadAlignments.__init__", false]], "__init__() (readdistributions method)": [[1, "genome_kit.ReadDistributions.__init__", false]], "__init__() (transcript method)": [[1, "genome_kit.Transcript.__init__", false]], "__init__() (variant method)": [[1, "genome_kit.Variant.__init__", false]], "__init__() (variantgenome method)": [[1, "genome_kit.VariantGenome.__init__", false]], "__init__() (vcftable method)": [[1, "genome_kit.VCFTable.__init__", false]], "__init__() (vcfvariant method)": [[1, "genome_kit.VCFVariant.__init__", false]], "__len__() (vcftable method)": [[1, "genome_kit.VCFTable.__len__", false]], "__new__() (genome static method)": [[1, "genome_kit.Genome.__new__", false]], "__reduce__() (genome method)": [[1, "genome_kit.Genome.__reduce__", false]], "__repr__() (alignment method)": [[1, "genome_kit.Alignment.__repr__", false]], "__repr__() (alignmentmatch method)": [[1, "genome_kit.AlignmentMatch.__repr__", false]], "__repr__() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.__repr__", false]], "__repr__() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.__repr__", false]], "__repr__() (cds method)": [[1, "genome_kit.Cds.__repr__", false]], "__repr__() (cdstable method)": [[1, "genome_kit.CdsTable.__repr__", false]], "__repr__() (exon method)": [[1, "genome_kit.Exon.__repr__", false]], "__repr__() (exontable method)": [[1, "genome_kit.ExonTable.__repr__", false]], "__repr__() (gene method)": [[1, "genome_kit.Gene.__repr__", false]], "__repr__() (genetable method)": [[1, "genome_kit.GeneTable.__repr__", false]], "__repr__() (genome method)": [[1, "genome_kit.Genome.__repr__", false]], "__repr__() (genomeannotation method)": [[1, "genome_kit.GenomeAnnotation.__repr__", false]], "__repr__() (genomedna method)": [[1, "genome_kit.GenomeDNA.__repr__", false]], "__repr__() (genometrack method)": [[1, "genome_kit.GenomeTrack.__repr__", false]], "__repr__() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.__repr__", false]], "__repr__() (interval method)": [[1, "genome_kit.Interval.__repr__", false]], "__repr__() (intron method)": [[1, "genome_kit.Intron.__repr__", false]], "__repr__() (introntable method)": [[1, "genome_kit.IntronTable.__repr__", false]], "__repr__() (jreadalignments method)": [[1, "genome_kit.JReadAlignments.__repr__", false]], "__repr__() (junction method)": [[1, "genome_kit.Junction.__repr__", false]], "__repr__() (junctionreadalignment method)": [[1, "genome_kit.JunctionReadAlignment.__repr__", false]], "__repr__() (junctionreadalignments method)": [[1, "genome_kit.JunctionReadAlignments.__repr__", false]], "__repr__() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.__repr__", false]], "__repr__() (junctionreadcount method)": [[1, "genome_kit.JunctionReadCount.__repr__", false]], "__repr__() (junctionreaddistribution method)": [[1, "genome_kit.JunctionReadDistribution.__repr__", false]], "__repr__() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.__repr__", false]], "__repr__() (junctiontable method)": [[1, "genome_kit.JunctionTable.__repr__", false]], "__repr__() (readalignments method)": [[1, "genome_kit.ReadAlignments.__repr__", false]], "__repr__() (readdistributions method)": [[1, "genome_kit.ReadDistributions.__repr__", false]], "__repr__() (transcript method)": [[1, "genome_kit.Transcript.__repr__", false]], "__repr__() (transcripttable method)": [[1, "genome_kit.TranscriptTable.__repr__", false]], "__repr__() (variant method)": [[1, "genome_kit.Variant.__repr__", false]], "__repr__() (variantgenome method)": [[1, "genome_kit.VariantGenome.__repr__", false]], "__repr__() (varianttable method)": [[1, "genome_kit.VariantTable.__repr__", false]], "__repr__() (vcftable method)": [[1, "genome_kit.VCFTable.__repr__", false]], "__repr__() (vcfvariant method)": [[1, "genome_kit.VCFVariant.__repr__", false]], "__str__() (interval method)": [[1, "genome_kit.Interval.__str__", false]], "__str__() (variant method)": [[1, "genome_kit.Variant.__str__", false]], "__str__() (vcfvariant method)": [[1, "genome_kit.VCFVariant.__str__", false]], "__weakref__ (datamanager attribute)": [[1, "genome_kit.DataManager.__weakref__", false]], "__weakref__ (genome attribute)": [[1, "genome_kit.Genome.__weakref__", false]], "__weakref__ (progresspercentage attribute)": [[1, "genome_kit.data_manager.ProgressPercentage.__weakref__", false]], "alignment (class in genome_kit)": [[1, "genome_kit.Alignment", false]], "alignmentmatch (class in genome_kit)": [[1, "genome_kit.AlignmentMatch", false]], "alignmentmatchtable (class in genome_kit)": [[1, "genome_kit.AlignmentMatchTable", false]], "alignments (junction property)": [[1, "genome_kit.Junction.alignments", false]], "alignments (readalignments property)": [[1, "genome_kit.ReadAlignments.alignments", false]], "alignmenttable (class in genome_kit)": [[1, "genome_kit.AlignmentTable", false]], "alt (variant property)": [[1, "genome_kit.Variant.alt", false]], "anchor (interval property)": [[1, "genome_kit.Interval.anchor", false]], "anchor_offset (interval property)": [[1, "genome_kit.Interval.anchor_offset", false]], "annotation (genome property)": [[1, "genome_kit.Genome.annotation", false]], "apply_variants() (in module genome_kit._apply_variants)": [[1, "genome_kit._apply_variants.apply_variants", false]], "appris_principality() (genome method)": [[1, "genome_kit.Genome.appris_principality", false]], "appris_transcripts() (genome method)": [[1, "genome_kit.Genome.appris_transcripts", false]], "as_dna0() (interval method)": [[1, "genome_kit.Interval.as_dna0", false]], "as_dna1() (interval method)": [[1, "genome_kit.Interval.as_dna1", false]], "as_negative_strand() (interval method)": [[1, "genome_kit.Interval.as_negative_strand", false]], "as_opposite_strand() (interval method)": [[1, "genome_kit.Interval.as_opposite_strand", false]], "as_positive_strand() (interval method)": [[1, "genome_kit.Interval.as_positive_strand", false]], "as_rna1() (interval method)": [[1, "genome_kit.Interval.as_rna1", false]], "as_ucsc() (interval method)": [[1, "genome_kit.Interval.as_ucsc", false]], "as_variant_string() (variant method)": [[1, "genome_kit.Variant.as_variant_string", false]], "as_variant_string() (vcfvariant method)": [[1, "genome_kit.VCFVariant.as_variant_string", false]], "binary_version() (genomeannotation static method)": [[1, "genome_kit.GenomeAnnotation.binary_version", false]], "build_gencode() (genomeannotation static method)": [[1, "genome_kit.GenomeAnnotation.build_gencode", false]], "build_jralign() (jreadalignments static method)": [[1, "genome_kit.JReadAlignments.build_jralign", false]], "build_ncbi_refseq() (genomeannotation static method)": [[1, "genome_kit.GenomeAnnotation.build_ncbi_refseq", false]], "build_ralign() (readalignments static method)": [[1, "genome_kit.ReadAlignments.build_ralign", false]], "build_rdist() (readdistributions static method)": [[1, "genome_kit.ReadDistributions.build_rdist", false]], "build_ucsc_refseq() (genomeannotation static method)": [[1, "genome_kit.GenomeAnnotation.build_ucsc_refseq", false]], "build_vcfbin() (vcftable static method)": [[1, "genome_kit.VCFTable.build_vcfbin", false]], "ccds_id (transcript property)": [[1, "genome_kit.Transcript.ccds_id", false]], "cds (class in genome_kit)": [[1, "genome_kit.Cds", false]], "cds (exon property)": [[1, "genome_kit.Exon.cds", false]], "cdss (genome property)": [[1, "genome_kit.Genome.cdss", false]], "cdss (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.cdss", false]], "cdss (transcript property)": [[1, "genome_kit.Transcript.cdss", false]], "cdstable (class in genome_kit)": [[1, "genome_kit.CdsTable", false]], "chromosome (interval property)": [[1, "genome_kit.Interval.chromosome", false]], "chromosome_size() (genome method)": [[1, "genome_kit.Genome.chromosome_size", false]], "chromosomes (genome property)": [[1, "genome_kit.Genome.chromosomes", false]], "close() (genometrack method)": [[1, "genome_kit.GenomeTrack.close", false]], "close() (jreadalignments method)": [[1, "genome_kit.JReadAlignments.close", false]], "close() (readalignments method)": [[1, "genome_kit.ReadAlignments.close", false]], "close() (readdistributions method)": [[1, "genome_kit.ReadDistributions.close", false]], "close() (vcftable method)": [[1, "genome_kit.VCFTable.close", false]], "config (genome property)": [[1, "genome_kit.Genome.config", false]], "contains() (interval method)": [[1, "genome_kit.Interval.contains", false]], "count (junctionreadcount property)": [[1, "genome_kit.JunctionReadCount.count", false]], "data_dir (genome property)": [[1, "genome_kit.Genome.data_dir", false]], "data_size (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.data_size", false]], "datamanager (class in genome_kit)": [[1, "genome_kit.DataManager", false]], "dim (genometrack property)": [[1, "genome_kit.GenomeTrack.dim", false]], "dim (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.dim", false]], "distance() (interval method)": [[1, "genome_kit.Interval.distance", false]], "dna() (genome method)": [[1, "genome_kit.Genome.dna", false]], "dna() (variantgenome method)": [[1, "genome_kit.VariantGenome.dna", false]], "dnstream_of() (interval method)": [[1, "genome_kit.Interval.dnstream_of", false]], "dtype (genometrack property)": [[1, "genome_kit.GenomeTrack.dtype", false]], "dtype (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.dtype", false]], "end (interval property)": [[1, "genome_kit.Interval.end", false]], "end3 (interval property)": [[1, "genome_kit.Interval.end3", false]], "end5 (interval property)": [[1, "genome_kit.Interval.end5", false]], "etype (genometrack property)": [[1, "genome_kit.GenomeTrack.etype", false]], "etype (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.etype", false]], "exon (cds property)": [[1, "genome_kit.Cds.exon", false]], "exon (class in genome_kit)": [[1, "genome_kit.Exon", false]], "exons (genome property)": [[1, "genome_kit.Genome.exons", false]], "exons (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.exons", false]], "exons (transcript property)": [[1, "genome_kit.Transcript.exons", false]], "exontable (class in genome_kit)": [[1, "genome_kit.ExonTable", false]], "expand() (interval method)": [[1, "genome_kit.Interval.expand", false]], "filename (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.filename", false]], "filename (genomedna property)": [[1, "genome_kit.GenomeDNA.filename", false]], "filename (genometrack property)": [[1, "genome_kit.GenomeTrack.filename", false]], "filename (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.filename", false]], "filename (jreadalignments property)": [[1, "genome_kit.JReadAlignments.filename", false]], "filename (readalignments property)": [[1, "genome_kit.ReadAlignments.filename", false]], "filename (readdistributions property)": [[1, "genome_kit.ReadDistributions.filename", false]], "filename (vcftable property)": [[1, "genome_kit.VCFTable.filename", false]], "finalize() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.finalize", false]], "find_3p_aligned() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_3p_aligned", false]], "find_3p_aligned() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_3p_aligned", false]], "find_3p_aligned() (cdstable method)": [[1, "genome_kit.CdsTable.find_3p_aligned", false]], "find_3p_aligned() (exontable method)": [[1, "genome_kit.ExonTable.find_3p_aligned", false]], "find_3p_aligned() (genetable method)": [[1, "genome_kit.GeneTable.find_3p_aligned", false]], "find_3p_aligned() (introntable method)": [[1, "genome_kit.IntronTable.find_3p_aligned", false]], "find_3p_aligned() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_3p_aligned", false]], "find_3p_aligned() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_3p_aligned", false]], "find_3p_aligned() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_3p_aligned", false]], "find_3p_aligned() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_3p_aligned", false]], "find_3p_aligned() (varianttable method)": [[1, "genome_kit.VariantTable.find_3p_aligned", false]], "find_3p_aligned() (vcftable method)": [[1, "genome_kit.VCFTable.find_3p_aligned", false]], "find_3p_within() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_3p_within", false]], "find_3p_within() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_3p_within", false]], "find_3p_within() (cdstable method)": [[1, "genome_kit.CdsTable.find_3p_within", false]], "find_3p_within() (exontable method)": [[1, "genome_kit.ExonTable.find_3p_within", false]], "find_3p_within() (genetable method)": [[1, "genome_kit.GeneTable.find_3p_within", false]], "find_3p_within() (introntable method)": [[1, "genome_kit.IntronTable.find_3p_within", false]], "find_3p_within() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_3p_within", false]], "find_3p_within() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_3p_within", false]], "find_3p_within() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_3p_within", false]], "find_3p_within() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_3p_within", false]], "find_3p_within() (varianttable method)": [[1, "genome_kit.VariantTable.find_3p_within", false]], "find_3p_within() (vcftable method)": [[1, "genome_kit.VCFTable.find_3p_within", false]], "find_5p_aligned() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_5p_aligned", false]], "find_5p_aligned() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_5p_aligned", false]], "find_5p_aligned() (cdstable method)": [[1, "genome_kit.CdsTable.find_5p_aligned", false]], "find_5p_aligned() (exontable method)": [[1, "genome_kit.ExonTable.find_5p_aligned", false]], "find_5p_aligned() (genetable method)": [[1, "genome_kit.GeneTable.find_5p_aligned", false]], "find_5p_aligned() (introntable method)": [[1, "genome_kit.IntronTable.find_5p_aligned", false]], "find_5p_aligned() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_5p_aligned", false]], "find_5p_aligned() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_5p_aligned", false]], "find_5p_aligned() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_5p_aligned", false]], "find_5p_aligned() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_5p_aligned", false]], "find_5p_aligned() (varianttable method)": [[1, "genome_kit.VariantTable.find_5p_aligned", false]], "find_5p_aligned() (vcftable method)": [[1, "genome_kit.VCFTable.find_5p_aligned", false]], "find_5p_within() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_5p_within", false]], "find_5p_within() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_5p_within", false]], "find_5p_within() (cdstable method)": [[1, "genome_kit.CdsTable.find_5p_within", false]], "find_5p_within() (exontable method)": [[1, "genome_kit.ExonTable.find_5p_within", false]], "find_5p_within() (genetable method)": [[1, "genome_kit.GeneTable.find_5p_within", false]], "find_5p_within() (introntable method)": [[1, "genome_kit.IntronTable.find_5p_within", false]], "find_5p_within() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_5p_within", false]], "find_5p_within() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_5p_within", false]], "find_5p_within() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_5p_within", false]], "find_5p_within() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_5p_within", false]], "find_5p_within() (varianttable method)": [[1, "genome_kit.VariantTable.find_5p_within", false]], "find_5p_within() (vcftable method)": [[1, "genome_kit.VCFTable.find_5p_within", false]], "find_exact() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_exact", false]], "find_exact() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_exact", false]], "find_exact() (cdstable method)": [[1, "genome_kit.CdsTable.find_exact", false]], "find_exact() (exontable method)": [[1, "genome_kit.ExonTable.find_exact", false]], "find_exact() (genetable method)": [[1, "genome_kit.GeneTable.find_exact", false]], "find_exact() (introntable method)": [[1, "genome_kit.IntronTable.find_exact", false]], "find_exact() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_exact", false]], "find_exact() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_exact", false]], "find_exact() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_exact", false]], "find_exact() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_exact", false]], "find_exact() (varianttable method)": [[1, "genome_kit.VariantTable.find_exact", false]], "find_exact() (vcftable method)": [[1, "genome_kit.VCFTable.find_exact", false]], "find_motif() (genome method)": [[1, "genome_kit.Genome.find_motif", false]], "find_motif() (variantgenome method)": [[1, "genome_kit.VariantGenome.find_motif", false]], "find_overlapping() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_overlapping", false]], "find_overlapping() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_overlapping", false]], "find_overlapping() (cdstable method)": [[1, "genome_kit.CdsTable.find_overlapping", false]], "find_overlapping() (exontable method)": [[1, "genome_kit.ExonTable.find_overlapping", false]], "find_overlapping() (genetable method)": [[1, "genome_kit.GeneTable.find_overlapping", false]], "find_overlapping() (introntable method)": [[1, "genome_kit.IntronTable.find_overlapping", false]], "find_overlapping() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_overlapping", false]], "find_overlapping() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_overlapping", false]], "find_overlapping() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_overlapping", false]], "find_overlapping() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_overlapping", false]], "find_overlapping() (varianttable method)": [[1, "genome_kit.VariantTable.find_overlapping", false]], "find_overlapping() (vcftable method)": [[1, "genome_kit.VCFTable.find_overlapping", false]], "find_within() (alignmentmatchtable method)": [[1, "genome_kit.AlignmentMatchTable.find_within", false]], "find_within() (alignmenttable method)": [[1, "genome_kit.AlignmentTable.find_within", false]], "find_within() (cdstable method)": [[1, "genome_kit.CdsTable.find_within", false]], "find_within() (exontable method)": [[1, "genome_kit.ExonTable.find_within", false]], "find_within() (genetable method)": [[1, "genome_kit.GeneTable.find_within", false]], "find_within() (introntable method)": [[1, "genome_kit.IntronTable.find_within", false]], "find_within() (junctionreadalignmentstable method)": [[1, "genome_kit.JunctionReadAlignmentsTable.find_within", false]], "find_within() (junctionreaddistributiontable method)": [[1, "genome_kit.JunctionReadDistributionTable.find_within", false]], "find_within() (junctiontable method)": [[1, "genome_kit.JunctionTable.find_within", false]], "find_within() (transcripttable method)": [[1, "genome_kit.TranscriptTable.find_within", false]], "find_within() (varianttable method)": [[1, "genome_kit.VariantTable.find_within", false]], "find_within() (vcftable method)": [[1, "genome_kit.VCFTable.find_within", false]], "first_by_name() (genetable method)": [[1, "genome_kit.GeneTable.first_by_name", false]], "flush() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.flush", false]], "format() (vcftable method)": [[1, "genome_kit.VCFTable.format", false]], "format_ids (vcftable property)": [[1, "genome_kit.VCFTable.format_ids", false]], "from_dna0() (interval static method)": [[1, "genome_kit.Interval.from_dna0", false]], "from_dna0_coord() (interval static method)": [[1, "genome_kit.Interval.from_dna0_coord", false]], "from_dna1() (interval static method)": [[1, "genome_kit.Interval.from_dna1", false]], "from_jraligns() (readdistributions static method)": [[1, "genome_kit.ReadDistributions.from_jraligns", false]], "from_rna1() (interval static method)": [[1, "genome_kit.Interval.from_rna1", false]], "from_string() (variant static method)": [[1, "genome_kit.Variant.from_string", false]], "from_vcf() (vcftable static method)": [[1, "genome_kit.VCFTable.from_vcf", false]], "gene (class in genome_kit)": [[1, "genome_kit.Gene", false]], "gene (transcript property)": [[1, "genome_kit.Transcript.gene", false]], "genes (genome property)": [[1, "genome_kit.Genome.genes", false]], "genes (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.genes", false]], "genetable (class in genome_kit)": [[1, "genome_kit.GeneTable", false]], "genome (class in genome_kit)": [[1, "genome_kit.Genome", false]], "genome_kit.gk_data": [[1, "module-genome_kit.gk_data", false]], "genomeannotation (class in genome_kit)": [[1, "genome_kit.GenomeAnnotation", false]], "genomedna (class in genome_kit)": [[1, "genome_kit.GenomeDNA", false]], "genometrack (class in genome_kit)": [[1, "genome_kit.GenomeTrack", false]], "genometrackbuilder (class in genome_kit)": [[1, "genome_kit.GenomeTrackBuilder", false]], "get_file() (datamanager method)": [[1, "genome_kit.DataManager.get_file", false]], "get_file() (in module genome_kit.gk_data)": [[1, "genome_kit.gk_data.get_file", false]], "gt_heterozygous_phased_0_1 (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_0_1", false]], "gt_heterozygous_phased_1_0 (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HETEROZYGOUS_PHASED_1_0", false]], "gt_heterozygous_unphased (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HETEROZYGOUS_UNPHASED", false]], "gt_homozygous_alt (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HOMOZYGOUS_ALT", false]], "gt_homozygous_ref (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_HOMOZYGOUS_REF", false]], "gt_unknown (vcftable attribute)": [[1, "genome_kit.VCFTable.GT_UNKNOWN", false]], "gtrack_version() (genometrack static method)": [[1, "genome_kit.GenomeTrack.gtrack_version", false]], "histogram() (genometrack method)": [[1, "genome_kit.GenomeTrack.histogram", false]], "id (exon property)": [[1, "genome_kit.Exon.id", false]], "id (gene property)": [[1, "genome_kit.Gene.id", false]], "id (transcript property)": [[1, "genome_kit.Transcript.id", false]], "index (exon property)": [[1, "genome_kit.Exon.index", false]], "index (intron property)": [[1, "genome_kit.Intron.index", false]], "index_size (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.index_size", false]], "info() (vcftable method)": [[1, "genome_kit.VCFTable.info", false]], "info_ids (vcftable property)": [[1, "genome_kit.VCFTable.info_ids", false]], "intersect() (interval method)": [[1, "genome_kit.Interval.intersect", false]], "interval (alignment property)": [[1, "genome_kit.Alignment.interval", false]], "interval (alignmentmatch property)": [[1, "genome_kit.AlignmentMatch.interval", false]], "interval (cds property)": [[1, "genome_kit.Cds.interval", false]], "interval (class in genome_kit)": [[1, "genome_kit.Interval", false]], "interval (exon property)": [[1, "genome_kit.Exon.interval", false]], "interval (gene property)": [[1, "genome_kit.Gene.interval", false]], "interval (intron property)": [[1, "genome_kit.Intron.interval", false]], "interval (junction property)": [[1, "genome_kit.Junction.interval", false]], "interval (junctionreadalignments property)": [[1, "genome_kit.JunctionReadAlignments.interval", false]], "interval (junctionreaddistribution property)": [[1, "genome_kit.JunctionReadDistribution.interval", false]], "interval (transcript property)": [[1, "genome_kit.Transcript.interval", false]], "interval (variant property)": [[1, "genome_kit.Variant.interval", false]], "interval() (genome method)": [[1, "genome_kit.Genome.interval", false]], "intervals() (genometrack method)": [[1, "genome_kit.GenomeTrack.intervals", false]], "intron (class in genome_kit)": [[1, "genome_kit.Intron", false]], "introns (genome property)": [[1, "genome_kit.Genome.introns", false]], "introns (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.introns", false]], "introns (transcript property)": [[1, "genome_kit.Transcript.introns", false]], "introntable (class in genome_kit)": [[1, "genome_kit.IntronTable", false]], "is_positive_strand() (interval method)": [[1, "genome_kit.Interval.is_positive_strand", false]], "jralign_version() (jreadalignments static method)": [[1, "genome_kit.JReadAlignments.jralign_version", false]], "jreadalignments (class in genome_kit)": [[1, "genome_kit.JReadAlignments", false]], "junction (class in genome_kit)": [[1, "genome_kit.Junction", false]], "junctionreadalignment (class in genome_kit)": [[1, "genome_kit.JunctionReadAlignment", false]], "junctionreadalignments (class in genome_kit)": [[1, "genome_kit.JunctionReadAlignments", false]], "junctionreadalignmentstable (class in genome_kit)": [[1, "genome_kit.JunctionReadAlignmentsTable", false]], "junctionreadcount (class in genome_kit)": [[1, "genome_kit.JunctionReadCount", false]], "junctionreaddistribution (class in genome_kit)": [[1, "genome_kit.JunctionReadDistribution", false]], "junctionreaddistributiontable (class in genome_kit)": [[1, "genome_kit.JunctionReadDistributionTable", false]], "junctions (jreadalignments property)": [[1, "genome_kit.JReadAlignments.junctions", false]], "junctions (readalignments property)": [[1, "genome_kit.ReadAlignments.junctions", false]], "junctions (readdistributions property)": [[1, "genome_kit.ReadDistributions.junctions", false]], "junctiontable (class in genome_kit)": [[1, "genome_kit.JunctionTable", false]], "left (junctionreadalignment property)": [[1, "genome_kit.JunctionReadAlignment.left", false]], "level (gene property)": [[1, "genome_kit.Gene.level", false]], "level (transcript property)": [[1, "genome_kit.Transcript.level", false]], "list_available_genomes() (datamanager method)": [[1, "genome_kit.DataManager.list_available_genomes", false]], "masked() (vcftable method)": [[1, "genome_kit.VCFTable.masked", false]], "matches (alignment property)": [[1, "genome_kit.Alignment.matches", false]], "matches (readalignments property)": [[1, "genome_kit.ReadAlignments.matches", false]], "midpoint (interval property)": [[1, "genome_kit.Interval.midpoint", false]], "mock() (in module genome_kit._cxx_util)": [[1, "genome_kit._cxx_util.mock", false]], "mock_result() (in module genome_kit._cxx_util)": [[1, "genome_kit._cxx_util.mock_result", false]], "mock_unreachable() (in module genome_kit._cxx_util)": [[1, "genome_kit._cxx_util.mock_unreachable", false]], "module": [[1, "module-genome_kit.gk_data", false]], "name (gene property)": [[1, "genome_kit.Gene.name", false]], "next_cds (cds property)": [[1, "genome_kit.Cds.next_cds", false]], "next_exon (exon property)": [[1, "genome_kit.Exon.next_exon", false]], "next_exon (intron property)": [[1, "genome_kit.Intron.next_exon", false]], "next_intron (exon property)": [[1, "genome_kit.Exon.next_intron", false]], "next_intron (intron property)": [[1, "genome_kit.Intron.next_intron", false]], "num_counts (junctionreaddistribution property)": [[1, "genome_kit.JunctionReadDistribution.num_counts", false]], "num_reads (junctionreadalignments property)": [[1, "genome_kit.JunctionReadAlignments.num_reads", false]], "num_reads (junctionreaddistribution property)": [[1, "genome_kit.JunctionReadDistribution.num_reads", false]], "num_samples (vcftable property)": [[1, "genome_kit.VCFTable.num_samples", false]], "num_variants (junctionreadalignment property)": [[1, "genome_kit.JunctionReadAlignment.num_variants", false]], "overlaps() (interval method)": [[1, "genome_kit.Interval.overlaps", false]], "phase (cds property)": [[1, "genome_kit.Cds.phase", false]], "pileup() (readalignments method)": [[1, "genome_kit.ReadAlignments.pileup", false]], "position (variant property)": [[1, "genome_kit.Variant.position", false]], "prev_cds (cds property)": [[1, "genome_kit.Cds.prev_cds", false]], "prev_exon (exon property)": [[1, "genome_kit.Exon.prev_exon", false]], "prev_exon (intron property)": [[1, "genome_kit.Intron.prev_exon", false]], "prev_intron (exon property)": [[1, "genome_kit.Exon.prev_intron", false]], "prev_intron (intron property)": [[1, "genome_kit.Intron.prev_intron", false]], "product (transcript property)": [[1, "genome_kit.Transcript.product", false]], "progresspercentage (class in genome_kit.data_manager)": [[1, "genome_kit.data_manager.ProgressPercentage", false]], "protein_id (transcript property)": [[1, "genome_kit.Transcript.protein_id", false]], "ralign_version() (readalignments static method)": [[1, "genome_kit.ReadAlignments.ralign_version", false]], "rdist_version() (readdistributions static method)": [[1, "genome_kit.ReadDistributions.rdist_version", false]], "readalignments (class in genome_kit)": [[1, "genome_kit.ReadAlignments", false]], "readdistributions (class in genome_kit)": [[1, "genome_kit.ReadDistributions", false]], "ref (variant property)": [[1, "genome_kit.Variant.ref", false]], "reference_genome (genome property)": [[1, "genome_kit.Genome.reference_genome", false]], "reference_genome (genomedna property)": [[1, "genome_kit.GenomeDNA.reference_genome", false]], "reference_genome (genometrack property)": [[1, "genome_kit.GenomeTrack.reference_genome", false]], "reference_genome (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.reference_genome", false]], "reference_genome (interval property)": [[1, "genome_kit.Interval.reference_genome", false]], "resolution (genometrack property)": [[1, "genome_kit.GenomeTrack.resolution", false]], "resolution (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.resolution", false]], "right (junctionreadalignment property)": [[1, "genome_kit.JunctionReadAlignment.right", false]], "sample_names (vcftable property)": [[1, "genome_kit.VCFTable.sample_names", false]], "sequence_variations() (vcftable method)": [[1, "genome_kit.VCFTable.sequence_variations", false]], "set_clamping() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_clamping", false]], "set_data() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_data", false]], "set_data_from_bed() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_data_from_bed", false]], "set_data_from_bedgraph() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_data_from_bedgraph", false]], "set_data_from_wig() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_data_from_wig", false]], "set_default_value() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_default_value", false]], "set_dict() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_dict", false]], "set_restriction() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_restriction", false]], "set_sparsity() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_sparsity", false]], "set_transform() (genometrackbuilder method)": [[1, "genome_kit.GenomeTrackBuilder.set_transform", false]], "shift (junctionreadcount property)": [[1, "genome_kit.JunctionReadCount.shift", false]], "shift() (interval method)": [[1, "genome_kit.Interval.shift", false]], "spanning() (interval static method)": [[1, "genome_kit.Interval.spanning", false]], "spanning() (variant static method)": [[1, "genome_kit.Variant.spanning", false]], "start (interval property)": [[1, "genome_kit.Interval.start", false]], "strand (interval property)": [[1, "genome_kit.Interval.strand", false]], "strand (junctionreadalignment property)": [[1, "genome_kit.JunctionReadAlignment.strand", false]], "strand (junctionreadcount property)": [[1, "genome_kit.JunctionReadCount.strand", false]], "stranded (alignmentmatchtable property)": [[1, "genome_kit.AlignmentMatchTable.stranded", false]], "stranded (alignmenttable property)": [[1, "genome_kit.AlignmentTable.stranded", false]], "stranded (cdstable property)": [[1, "genome_kit.CdsTable.stranded", false]], "stranded (exontable property)": [[1, "genome_kit.ExonTable.stranded", false]], "stranded (genetable property)": [[1, "genome_kit.GeneTable.stranded", false]], "stranded (genometrack property)": [[1, "genome_kit.GenomeTrack.stranded", false]], "stranded (genometrackbuilder property)": [[1, "genome_kit.GenomeTrackBuilder.stranded", false]], "stranded (introntable property)": [[1, "genome_kit.IntronTable.stranded", false]], "stranded (junctionreadalignmentstable property)": [[1, "genome_kit.JunctionReadAlignmentsTable.stranded", false]], "stranded (junctionreaddistributiontable property)": [[1, "genome_kit.JunctionReadDistributionTable.stranded", false]], "stranded (junctiontable property)": [[1, "genome_kit.JunctionTable.stranded", false]], "stranded (transcripttable property)": [[1, "genome_kit.TranscriptTable.stranded", false]], "stranded (varianttable property)": [[1, "genome_kit.VariantTable.stranded", false]], "stranded (vcftable property)": [[1, "genome_kit.VCFTable.stranded", false]], "strip_mock_bases() (in module genome_kit._cxx_util)": [[1, "genome_kit._cxx_util.strip_mock_bases", false]], "subtract() (interval method)": [[1, "genome_kit.Interval.subtract", false]], "svtype_na (vcftable attribute)": [[1, "genome_kit.VCFTable.SVTYPE_NA", false]], "transcript (cds property)": [[1, "genome_kit.Cds.transcript", false]], "transcript (class in genome_kit)": [[1, "genome_kit.Transcript", false]], "transcript (exon property)": [[1, "genome_kit.Exon.transcript", false]], "transcript (intron property)": [[1, "genome_kit.Intron.transcript", false]], "transcripts (gene property)": [[1, "genome_kit.Gene.transcripts", false]], "transcripts (genome property)": [[1, "genome_kit.Genome.transcripts", false]], "transcripts (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.transcripts", false]], "transcripttable (class in genome_kit)": [[1, "genome_kit.TranscriptTable", false]], "tsl (transcript property)": [[1, "genome_kit.Transcript.tsl", false]], "type (gene property)": [[1, "genome_kit.Gene.type", false]], "type (transcript property)": [[1, "genome_kit.Transcript.type", false]], "upload_file() (datamanager method)": [[1, "genome_kit.DataManager.upload_file", false]], "upload_file() (in module genome_kit.gk_data)": [[1, "genome_kit.gk_data.upload_file", false]], "upstream_of() (interval method)": [[1, "genome_kit.Interval.upstream_of", false]], "utr3 (exon property)": [[1, "genome_kit.Exon.utr3", false]], "utr3s (genome property)": [[1, "genome_kit.Genome.utr3s", false]], "utr3s (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.utr3s", false]], "utr3s (transcript property)": [[1, "genome_kit.Transcript.utr3s", false]], "utr5 (exon property)": [[1, "genome_kit.Exon.utr5", false]], "utr5s (genome property)": [[1, "genome_kit.Genome.utr5s", false]], "utr5s (genomeannotation property)": [[1, "genome_kit.GenomeAnnotation.utr5s", false]], "utr5s (transcript property)": [[1, "genome_kit.Transcript.utr5s", false]], "variant (class in genome_kit)": [[1, "genome_kit.Variant", false]], "variant() (genome method)": [[1, "genome_kit.Genome.variant", false]], "variant_dna() (genome method)": [[1, "genome_kit.Genome.variant_dna", false]], "variantgenome (class in genome_kit)": [[1, "genome_kit.VariantGenome", false]], "variants (alignmentmatch property)": [[1, "genome_kit.AlignmentMatch.variants", false]], "variants (jreadalignments property)": [[1, "genome_kit.JReadAlignments.variants", false]], "variants (readalignments property)": [[1, "genome_kit.ReadAlignments.variants", false]], "variants() (junctionreadalignment method)": [[1, "genome_kit.JunctionReadAlignment.variants", false]], "varianttable (class in genome_kit)": [[1, "genome_kit.VariantTable", false]], "vcfbin_version() (vcftable static method)": [[1, "genome_kit.VCFTable.vcfbin_version", false]], "vcftable (class in genome_kit)": [[1, "genome_kit.VCFTable", false]], "vcfvariant (class in genome_kit)": [[1, "genome_kit.VCFVariant", false]], "wget() (in module genome_kit.gk_data)": [[1, "genome_kit.gk_data.wget", false]], "where() (varianttable method)": [[1, "genome_kit.VariantTable.where", false]], "where() (vcftable method)": [[1, "genome_kit.VCFTable.where", false]], "with_anchor() (interval method)": [[1, "genome_kit.Interval.with_anchor", false]], "within() (interval method)": [[1, "genome_kit.Interval.within", false]], "write2bit() (in module genome_kit._twobitutils)": [[1, "genome_kit._twobitutils.write2bit", false]]}, "objects": {"genome_kit": [[1, 0, 1, "", "Alignment"], [1, 0, 1, "", "AlignmentMatch"], [1, 0, 1, "", "AlignmentMatchTable"], [1, 0, 1, "", "AlignmentTable"], [1, 0, 1, "", "Cds"], [1, 0, 1, "", "CdsTable"], [1, 0, 1, "", "DataManager"], [1, 0, 1, "", "Exon"], [1, 0, 1, "", "ExonTable"], [1, 0, 1, "", "Gene"], [1, 0, 1, "", "GeneTable"], [1, 0, 1, "", "Genome"], [1, 0, 1, "", "GenomeAnnotation"], [1, 0, 1, "", "GenomeDNA"], [1, 0, 1, "", "GenomeTrack"], [1, 0, 1, "", "GenomeTrackBuilder"], [1, 0, 1, "", "Interval"], [1, 0, 1, "", "Intron"], [1, 0, 1, "", "IntronTable"], [1, 0, 1, "", "JReadAlignments"], [1, 0, 1, "", "Junction"], [1, 0, 1, "", "JunctionReadAlignment"], [1, 0, 1, "", "JunctionReadAlignments"], [1, 0, 1, "", "JunctionReadAlignmentsTable"], [1, 0, 1, "", "JunctionReadCount"], [1, 0, 1, "", "JunctionReadDistribution"], [1, 0, 1, "", "JunctionReadDistributionTable"], [1, 0, 1, "", "JunctionTable"], [1, 0, 1, "", "ReadAlignments"], [1, 0, 1, "", "ReadDistributions"], [1, 0, 1, "", "Transcript"], [1, 0, 1, "", "TranscriptTable"], [1, 0, 1, "", "VCFTable"], [1, 0, 1, "", "VCFVariant"], [1, 0, 1, "", "Variant"], [1, 0, 1, "", "VariantGenome"], [1, 0, 1, "", "VariantTable"], [1, 5, 0, "-", "gk_data"]], "genome_kit.Alignment": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "matches"]], "genome_kit.AlignmentMatch": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "variants"]], "genome_kit.AlignmentMatchTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.AlignmentTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.Cds": [[1, 1, 1, "", "__getstate__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "exon"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_cds"], [1, 2, 1, "", "phase"], [1, 2, 1, "", "prev_cds"], [1, 2, 1, "", "transcript"]], "genome_kit.CdsTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.DataManager": [[1, 1, 1, "", "__init__"], [1, 3, 1, "", "__weakref__"], [1, 1, 1, "", "get_file"], [1, 1, 1, "", "list_available_genomes"], [1, 1, 1, "", "upload_file"]], "genome_kit.Exon": [[1, 1, 1, "", "__getstate__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "cds"], [1, 2, 1, "", "id"], [1, 2, 1, "", "index"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_exon"], [1, 2, 1, "", "next_intron"], [1, 2, 1, "", "prev_exon"], [1, 2, 1, "", "prev_intron"], [1, 2, 1, "", "transcript"], [1, 2, 1, "", "utr3"], [1, 2, 1, "", "utr5"]], "genome_kit.ExonTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.Gene": [[1, 1, 1, "", "__getstate__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "id"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "level"], [1, 2, 1, "", "name"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "type"]], "genome_kit.GeneTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 1, 1, "", "first_by_name"], [1, 2, 1, "", "stranded"]], "genome_kit.Genome": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__new__"], [1, 1, 1, "", "__reduce__"], [1, 1, 1, "", "__repr__"], [1, 3, 1, "", "__weakref__"], [1, 2, 1, "", "annotation"], [1, 1, 1, "", "appris_principality"], [1, 1, 1, "", "appris_transcripts"], [1, 2, 1, "", "cdss"], [1, 1, 1, "", "chromosome_size"], [1, 2, 1, "", "chromosomes"], [1, 2, 1, "", "config"], [1, 2, 1, "", "data_dir"], [1, 1, 1, "", "dna"], [1, 2, 1, "", "exons"], [1, 1, 1, "", "find_motif"], [1, 2, 1, "", "genes"], [1, 1, 1, "", "interval"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"], [1, 1, 1, "", "variant"], [1, 1, 1, "", "variant_dna"]], "genome_kit.GenomeAnnotation": [[1, 1, 1, "", "__repr__"], [1, 1, 1, "", "binary_version"], [1, 1, 1, "", "build_gencode"], [1, 1, 1, "", "build_ncbi_refseq"], [1, 1, 1, "", "build_ucsc_refseq"], [1, 2, 1, "", "cdss"], [1, 2, 1, "", "exons"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "genes"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "transcripts"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"]], "genome_kit.GenomeDNA": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "reference_genome"]], "genome_kit.GenomeTrack": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "close"], [1, 2, 1, "", "dim"], [1, 2, 1, "", "dtype"], [1, 2, 1, "", "etype"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "gtrack_version"], [1, 1, 1, "", "histogram"], [1, 1, 1, "", "intervals"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "resolution"], [1, 2, 1, "", "stranded"]], "genome_kit.GenomeTrackBuilder": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "data_size"], [1, 2, 1, "", "dim"], [1, 2, 1, "", "dtype"], [1, 2, 1, "", "etype"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "finalize"], [1, 1, 1, "", "flush"], [1, 2, 1, "", "index_size"], [1, 2, 1, "", "reference_genome"], [1, 2, 1, "", "resolution"], [1, 1, 1, "", "set_clamping"], [1, 1, 1, "", "set_data"], [1, 1, 1, "", "set_data_from_bed"], [1, 1, 1, "", "set_data_from_bedgraph"], [1, 1, 1, "", "set_data_from_wig"], [1, 1, 1, "", "set_default_value"], [1, 1, 1, "", "set_dict"], [1, 1, 1, "", "set_restriction"], [1, 1, 1, "", "set_sparsity"], [1, 1, 1, "", "set_transform"], [1, 2, 1, "", "stranded"]], "genome_kit.Interval": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 2, 1, "", "anchor"], [1, 2, 1, "", "anchor_offset"], [1, 1, 1, "", "as_dna0"], [1, 1, 1, "", "as_dna1"], [1, 1, 1, "", "as_negative_strand"], [1, 1, 1, "", "as_opposite_strand"], [1, 1, 1, "", "as_positive_strand"], [1, 1, 1, "", "as_rna1"], [1, 1, 1, "", "as_ucsc"], [1, 2, 1, "", "chromosome"], [1, 1, 1, "", "contains"], [1, 1, 1, "", "distance"], [1, 1, 1, "", "dnstream_of"], [1, 2, 1, "", "end"], [1, 2, 1, "", "end3"], [1, 2, 1, "", "end5"], [1, 1, 1, "", "expand"], [1, 1, 1, "", "from_dna0"], [1, 1, 1, "", "from_dna0_coord"], [1, 1, 1, "", "from_dna1"], [1, 1, 1, "", "from_rna1"], [1, 1, 1, "", "intersect"], [1, 1, 1, "", "is_positive_strand"], [1, 2, 1, "", "midpoint"], [1, 1, 1, "", "overlaps"], [1, 2, 1, "", "reference_genome"], [1, 1, 1, "", "shift"], [1, 1, 1, "", "spanning"], [1, 2, 1, "", "start"], [1, 2, 1, "", "strand"], [1, 1, 1, "", "subtract"], [1, 1, 1, "", "upstream_of"], [1, 1, 1, "", "with_anchor"], [1, 1, 1, "", "within"]], "genome_kit.Intron": [[1, 1, 1, "", "__getstate__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "index"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "next_exon"], [1, 2, 1, "", "next_intron"], [1, 2, 1, "", "prev_exon"], [1, 2, 1, "", "prev_intron"], [1, 2, 1, "", "transcript"]], "genome_kit.IntronTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JReadAlignments": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_jralign"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "jralign_version"], [1, 2, 1, "", "junctions"], [1, 2, 1, "", "variants"]], "genome_kit.Junction": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "alignments"], [1, 2, 1, "", "interval"]], "genome_kit.JunctionReadAlignment": [[1, 1, 1, "", "__repr__"], [1, 2, 1, "", "left"], [1, 2, 1, "", "num_variants"], [1, 2, 1, "", "right"], [1, 2, 1, "", "strand"], [1, 1, 1, "", "variants"]], "genome_kit.JunctionReadAlignments": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "num_reads"]], "genome_kit.JunctionReadAlignmentsTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JunctionReadCount": [[1, 1, 1, "", "__repr__"], [1, 2, 1, "", "count"], [1, 2, 1, "", "shift"], [1, 2, 1, "", "strand"]], "genome_kit.JunctionReadDistribution": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "num_counts"], [1, 2, 1, "", "num_reads"]], "genome_kit.JunctionReadDistributionTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.JunctionTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.ReadAlignments": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "alignments"], [1, 1, 1, "", "build_ralign"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 2, 1, "", "junctions"], [1, 2, 1, "", "matches"], [1, 1, 1, "", "pileup"], [1, 1, 1, "", "ralign_version"], [1, 2, 1, "", "variants"]], "genome_kit.ReadDistributions": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_rdist"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "from_jraligns"], [1, 2, 1, "", "junctions"], [1, 1, 1, "", "rdist_version"]], "genome_kit.Transcript": [[1, 1, 1, "", "__getstate__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 2, 1, "", "ccds_id"], [1, 2, 1, "", "cdss"], [1, 2, 1, "", "exons"], [1, 2, 1, "", "gene"], [1, 2, 1, "", "id"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "introns"], [1, 2, 1, "", "level"], [1, 2, 1, "", "product"], [1, 2, 1, "", "protein_id"], [1, 2, 1, "", "tsl"], [1, 2, 1, "", "type"], [1, 2, 1, "", "utr3s"], [1, 2, 1, "", "utr5s"]], "genome_kit.TranscriptTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"]], "genome_kit.VCFTable": [[1, 3, 1, "", "GT_HETEROZYGOUS_PHASED_0_1"], [1, 3, 1, "", "GT_HETEROZYGOUS_PHASED_1_0"], [1, 3, 1, "", "GT_HETEROZYGOUS_UNPHASED"], [1, 3, 1, "", "GT_HOMOZYGOUS_ALT"], [1, 3, 1, "", "GT_HOMOZYGOUS_REF"], [1, 3, 1, "", "GT_UNKNOWN"], [1, 3, 1, "", "SVTYPE_NA"], [1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__len__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "build_vcfbin"], [1, 1, 1, "", "close"], [1, 2, 1, "", "filename"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 1, 1, "", "format"], [1, 2, 1, "", "format_ids"], [1, 1, 1, "", "from_vcf"], [1, 1, 1, "", "info"], [1, 2, 1, "", "info_ids"], [1, 1, 1, "", "masked"], [1, 2, 1, "", "num_samples"], [1, 2, 1, "", "sample_names"], [1, 1, 1, "", "sequence_variations"], [1, 2, 1, "", "stranded"], [1, 1, 1, "", "vcfbin_version"], [1, 1, 1, "", "where"]], "genome_kit.VCFVariant": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 1, 1, "", "as_variant_string"]], "genome_kit.Variant": [[1, 1, 1, "", "__getstate__"], [1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "__str__"], [1, 2, 1, "", "alt"], [1, 1, 1, "", "as_variant_string"], [1, 1, 1, "", "from_string"], [1, 2, 1, "", "interval"], [1, 2, 1, "", "position"], [1, 2, 1, "", "ref"], [1, 1, 1, "", "spanning"]], "genome_kit.VariantGenome": [[1, 1, 1, "", "__init__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "dna"], [1, 1, 1, "", "find_motif"]], "genome_kit.VariantTable": [[1, 1, 1, "", "__getitem__"], [1, 1, 1, "", "__repr__"], [1, 1, 1, "", "find_3p_aligned"], [1, 1, 1, "", "find_3p_within"], [1, 1, 1, "", "find_5p_aligned"], [1, 1, 1, "", "find_5p_within"], [1, 1, 1, "", "find_exact"], [1, 1, 1, "", "find_overlapping"], [1, 1, 1, "", "find_within"], [1, 2, 1, "", "stranded"], [1, 1, 1, "", "where"]], "genome_kit._apply_variants": [[1, 4, 1, "", "apply_variants"]], "genome_kit._cxx_util": [[1, 4, 1, "", "mock"], [1, 4, 1, "", "mock_result"], [1, 4, 1, "", "mock_unreachable"], [1, 4, 1, "", "strip_mock_bases"]], "genome_kit._twobitutils": [[1, 4, 1, "", "write2bit"]], "genome_kit.data_manager": [[1, 0, 1, "", "ProgressPercentage"]], "genome_kit.data_manager.ProgressPercentage": [[1, 1, 1, "", "__call__"], [1, 1, 1, "", "__init__"], [1, 3, 1, "", "__weakref__"]], "genome_kit.gk_data": [[1, 4, 1, "", "get_file"], [1, 4, 1, "", "upload_file"], [1, 4, 1, "", "wget"]]}, "objnames": {"0": ["py", "class", "Python class"], "1": ["py", "method", "Python method"], "2": ["py", "property", "Python property"], "3": ["py", "attribute", "Python attribute"], "4": ["py", "function", "Python function"], "5": ["py", "module", "Python module"]}, "objtypes": {"0": "py:class", "1": "py:method", "2": "py:property", "3": "py:attribute", "4": "py:function", "5": "py:module"}, "terms": {"": [0, 1, 2, 3, 6], "0": [0, 1, 2, 5, 6], "00": [1, 6], "01": 6, "0123456789": 6, "02": 6, "03": 6, "04": 6, "05": 6, "0500000007451": 6, "06": [1, 6], "07": 6, "08": 6, "09": 6, "1": [0, 1, 3, 5, 6], "10": [0, 1, 6], "100": [1, 6], "100nt": 6, "10l": 6, "10nt": 6, "11": [0, 1, 6], "117": [0, 6], "117120016": 6, "117120017": 6, "117120148": 6, "117120149": 6, "117120150": 6, "117120151": 6, "117120185": 6, "117120188": 6, "117120190": 6, "117120192": 6, "117120201": 6, "117231957": 6, "117231963": 6, "117231964": 6, "117231967": 0, "117231970": 0, "117231971": 6, "117231972": 6, "117231973": 6, "117231974": 6, "117231977": 0, "117231979": 6, "117231980": 6, "117231981": 6, "117231987": [0, 6], "117231997": 0, "117232004": 6, "117232009": 6, "117232014": 6, "117232020": 6, "117232022": 6, "117232023": 6, "117232024": 6, "117232030": 6, "1196293": 6, "12": [0, 1, 6], "120": 6, "125": 1, "126": 1, "127": 1, "128": 1, "13": [0, 5, 6], "14": [0, 6], "147": 1, "149": 6, "15": [0, 1, 6], "150": [1, 6], "151": 1, "16": [0, 1, 6], "169295": 6, "169296": 6, "17": [0, 6], "18": [0, 6], "185": 6, "19": [0, 6], "198992": 6, "1bp": 1, "1m1d1m2n1m": 1, "1m1i1m2n1m": 1, "1s2m2n1m1n1m": 1, "1st": [1, 6], "2": [0, 1, 6], "20": [0, 1, 3, 6], "200": 1, "2017": 1, "2017_05": 1, "2019": 3, "20190906": [1, 5, 6], "20m100n10m100n3m": 1, "21": 6, "22": 6, "23": [1, 6], "231": [0, 6], "24": 6, "25": [1, 6], "250": 1, "255": 1, "26": 6, "260": 1, "27": 6, "28": 6, "286": 1, "29": 6, "2bit": [1, 2, 3, 4], "2e": 1, "2nd": [1, 6], "3": [0, 1, 6], "30": 6, "300": [1, 6], "31": 1, "32": 1, "320861": 6, "320862": 6, "325": 1, "357": 1, "38": 1, "3a": 1, "3l": 6, "3p": [0, 1, 6], "3rd": 6, "4": [0, 1, 6], "400": 1, "40000": 1, "40030": 1, "40031": 1, "40033": 1, "40061": 1, "40080": 1, "41246881": 1, "41246886": 1, "48": 1, "4e38": 1, "5": [0, 1, 6], "500": 1, "50000": 1, "50010": 1, "500201": 1, "5002010": 1, "558099": 6, "57": 1, "571": 1, "5bp": 1, "5m100n10m50n5m": 1, "5m8d5m": 1, "5p": [0, 1, 6], "6": [0, 1, 5, 6], "605573": 6, "63": 1, "64": 3, "65504": 1, "69345349": 1, "69373418": 1, "69373422": 1, "69374349": 1, "7": [0, 1, 6], "70220767": 1, "70248838": 1, "70248842": 1, "70249769": 1, "724078": 6, "754": 1, "7841": 1, "7nt": 6, "8": [0, 1, 6], "857": 1, "88": [1, 5], "9": [0, 1, 6], "91661": 6, "91662": 6, "949522": 6, "949523": 6, "949608": 6, "949696": 6, "949739": 6, "95": [1, 5], "967": 0, "977": 0, "977028": 6, "977030": 6, "977329": 6, "977330": 6, "977515": 6, "977516": 6, "977569": 6, "977570": 6, "977630": 6, "978603": 6, "978604": 6, "978627": 6, "978628": 6, "98": [1, 5], "982": 0, "984": 0, "987": 0, "993": [0, 6], "994": 6, "996": 0, "A": [0, 1, 6], "AT": 6, "As": [0, 1], "Be": 1, "But": 6, "By": [0, 1, 6], "For": [0, 1, 2, 3, 4, 6], "INS": 1, "If": [0, 1, 3, 4, 6], "In": [0, 1, 2, 3, 5, 6], "It": [0, 1, 3], "Not": 1, "Of": 0, "On": 3, "One": 1, "Or": 1, "The": [0, 1, 2, 3, 4, 5, 6], "Then": 0, "There": [0, 1], "These": [1, 6], "To": [0, 1, 3, 6], "_": 1, "__________": 6, "_____________": 6, "__call__": 1, "__getitem__": 1, "__getstate__": 1, "__init__": [1, 6], "__len__": 1, "__new__": 1, "__reduce__": 1, "__repr__": 1, "__str__": 1, "__weakref__": 1, "_build_appri": 1, "_cxx": [1, 3], "_refg_hash": 4, "aa": 1, "aaaaabbbbb": 6, "aaaannnnccaaaatttt": 1, "aaagct": 0, "aacaa": 6, "aactttttttcag": 6, "aattggaagcaaa": 6, "ab": 1, "abca4": 1, "abca4_vcf": 1, "abl": [1, 3], "about": [1, 5, 6], "abov": [0, 1, 6], "accept": [1, 6], "acceptor": [0, 1, 6], "access": 5, "accommod": 0, "accord": [1, 6], "accordingli": [0, 1], "accumul": 1, "acgtacgt": 1, "acgtn": 1, "achiev": 0, "across": 1, "act": 1, "activ": [3, 4], "actual": [0, 1], "ad": [1, 5, 6], "add": 1, "addd": 1, "addit": [0, 1, 2], "adjac": 1, "advertis": 6, "af": 6, "af_exac": 1, "affect": [0, 1], "affin": 1, "after": 1, "ag": [1, 6], "again": 0, "against": [1, 6], "alias": [2, 4], "align": [0, 3, 5, 6], "all": [1, 3, 5, 6], "allel": [1, 6], "alloc": 1, "allow": [0, 1, 2, 6], "allow_outside_chromosom": 1, "alreadi": [1, 2], "also": [0, 1, 2, 3, 5, 6], "alt": [1, 6], "alter": 1, "altern": [1, 6], "alwai": [0, 1, 6], "ambigu": [1, 6], "amd64": 6, "amount": [0, 1], "an": [1, 2, 4, 5, 6], "anaconda": 6, "analogu": 1, "analysi": 1, "analyz": 1, "ancestr": 1, "anchor": [1, 5, 6], "anchor_offset": [0, 1, 6], "anchored_3p": 6, "anchored_5p": 6, "ani": [0, 1, 3, 4, 6], "anno": [1, 3], "annot": [2, 3, 5], "anoth": 1, "anywher": 0, "api": [2, 3, 5, 6], "appdir": [4, 6], "appear": 1, "appl": 3, "appli": [0, 1, 6], "applic": [1, 6], "apply_vari": 1, "appri": [1, 2, 3], "appris_index_nam": 1, "appris_princip": 1, "appris_transcript": 1, "apprisnotavailableerror": 1, "appropri": 0, "approxim": 1, "ar": [0, 1, 2, 3, 4, 5, 6], "arang": 6, "arbitrari": [0, 1], "arg": 1, "argument": [0, 1, 6], "around": [0, 6], "arrai": [1, 6], "as_dna0": 1, "as_dna1": 1, "as_negative_strand": 1, "as_opposite_strand": [1, 6], "as_positive_strand": 1, "as_rna1": 1, "as_str": 1, "as_ucsc": [1, 6], "as_variant_str": 1, "ascii": 1, "assembli": [1, 2, 3, 6], "assign": 1, "associ": [1, 2, 6], "assum": [1, 6], "astyp": 1, "atcttaaagtttctgtgtct": 0, "attach": 1, "attempt": [1, 6], "attribut": [1, 6], "auth": 6, "autocomplet": 1, "autodoc": 1, "automat": [1, 2, 3, 6], "autotick": 3, "avail": [1, 6], "avoid": 1, "awai": 1, "awar": 1, "axi": 6, "b": [1, 3, 6], "b37": 1, "back": [0, 6], "backend": 1, "bam": [1, 2, 5], "bar": [0, 1], "base": [0, 1, 3, 6], "basepair": 1, "basic": [1, 5], "bat": 3, "bcftool": 1, "becaus": [1, 6], "becom": 1, "bed": [1, 2], "bedfil": 1, "bedgraph": 1, "been": 1, "befor": [0, 1, 3, 6], "begin": 1, "behav": 1, "behaviour": [0, 1, 3, 6], "being": 1, "below": [1, 5, 6], "besid": 6, "best": [0, 1, 6], "between": [0, 1, 6], "bewar": 1, "bigzip": 4, "bin": 1, "binari": [1, 2, 3, 6], "binary_vers": 1, "binomi": 1, "bioconda": 3, "biotyp": 1, "bit": 1, "bits_per_datum": 1, "bitwidth": 1, "blob": 1, "block": 1, "bnd": 1, "bodi": 1, "bool": 1, "bool_": 1, "boolean": 1, "bootstrap": 1, "bot": 3, "both": [1, 3, 6], "bp": 1, "branch": [3, 6], "brca1": 1, "brca1_interv": 1, "break": 3, "breast": 1, "breviti": 6, "broadcast": 1, "browser": [1, 6], "brute": 1, "bucket": [2, 6], "build": [1, 4, 5, 6], "build_gencod": [1, 4], "build_jralign": 1, "build_ncbi_refseq": 1, "build_ralign": 1, "build_rdist": 1, "build_ucsc_refseq": 1, "build_vcfbin": 1, "builder": [1, 6], "built": [1, 3, 6], "byte": 1, "bytes_amount": 1, "c": [0, 1, 4, 6], "ca": [1, 6], "cach": [1, 2], "call": [0, 1, 2, 5, 6], "callabl": 1, "caller": 1, "can": [0, 1, 2, 3, 5, 6], "cancer": 1, "candid": 6, "cannot": [0, 1], "canon": 1, "carri": 6, "case": [0, 1, 6], "cat": 1, "categori": 1, "caus": 1, "caveat": 1, "ccaaact": 6, "ccaact": 6, "ccaactt": 6, "ccach": 3, "ccatgca": 6, "cccagattgg": 6, "cccc": 6, "ccds53759": 1, "ccds_id": 1, "ccdsid": 1, "ccgttca": 6, "cctaact": 6, "cctact": 6, "cd": 6, "cdss": [1, 6], "center": [0, 1, 6], "certain": 6, "cf": 3, "cfg": [1, 2, 4], "cftr": [0, 6], "chang": [0, 1, 5], "chapter": 0, "charact": 1, "check": [0, 3, 6], "child": 1, "chr": [1, 6], "chr1": [1, 6], "chr17": 1, "chr2": 1, "chr5": 1, "chr7": [0, 6], "chrom": [1, 2, 4, 6], "chromalia": [2, 4], "chromosom": [0, 1, 2, 6], "chromosome_s": 1, "ci": 3, "cigar": 1, "cl": 1, "clamp": 1, "clangxx_osx": 3, "class": [1, 6], "clear": 0, "clinvar": [0, 6], "clinvar_20180128": 1, "clion": 5, "clip": 1, "clnsig": 1, "clone": [3, 4, 6], "close": 1, "cloud": 6, "cmake": 3, "cnv": 1, "coars": 1, "coarsen": 1, "coarser": 1, "code": [0, 1, 2, 3, 6], "codon": 1, "column": [1, 6], "com": [1, 3, 4, 6], "come": [0, 6], "comma": 6, "command": 3, "comment": 3, "common": 1, "commonli": 5, "compact": 6, "compar": 1, "compat": [1, 3], "compil": [3, 4, 6], "complement": [1, 6], "complet": [0, 1], "complex": [1, 6], "complic": 0, "comprehens": 5, "compris": 1, "comput": 1, "concept": [0, 6], "conceptu": [0, 1], "concret": 0, "conda": [3, 4], "conda_prefix": 3, "condit": 1, "config": 1, "configur": [1, 2, 6], "confirm": 0, "confus": 1, "connect": 1, "consecut": 6, "conserv": 5, "consid": [0, 1, 6], "consist": 1, "constant": [0, 1], "constructor": [1, 6], "contain": [0, 1, 3, 4, 6], "content": [1, 5], "context": 1, "contig": [2, 4], "contigu": 1, "continu": 3, "contribut": 1, "control": [1, 3, 6], "conveni": [0, 1, 5, 6], "convent": [0, 1, 5, 6], "convers": 1, "convert": [1, 2, 6], "coord": 6, "coordin": [0, 1, 6], "copi": [1, 3, 4], "core": [0, 1, 6], "correct": [0, 1], "correctli": 1, "correspond": [1, 6], "could": [0, 1, 6], "count": 1, "cours": 0, "cover": [1, 6], "cpp": 3, "creat": [1, 2, 3, 4, 6], "creation": 1, "creativ": 5, "credenti": 6, "criteria": 6, "crop": 1, "ct": [0, 1, 6], "ctgaaaaaaagtt": 6, "ctgggtaaagctgtgtctgt": 0, "cttaaagtttctgtgtctgt": 0, "current": [1, 3, 6], "custom": [1, 6], "cxx": 3, "d": [1, 6], "data": [4, 5], "data_dir": [1, 6], "data_manag": [2, 5, 6], "data_s": 1, "databas": [1, 6], "datamanag": 6, "date": 1, "datum": 1, "dcmake_build_typ": 3, "ddx11l1": 6, "de": 0, "debugg": 3, "declar": 1, "decod": 1, "decor": 1, "decreas": 1, "dedic": 1, "deep": 5, "deepgenom": [3, 4, 6], "deepgenomicsinc": 6, "def": [1, 6], "default": [0, 1, 3, 6], "default_valu": 1, "defin": [0, 1, 3, 6], "definit": 1, "del": 1, "dela": 6, "delet": [1, 5, 6], "delta": 1, "demand": [1, 3], "demo": [1, 3, 6], "denot": [0, 6], "dep": 3, "depend": [1, 3, 5, 6], "depth": [1, 6], "describ": 6, "descript": [1, 6], "design": 5, "desir": [1, 4], "destin": 1, "detail": [1, 5, 6], "determin": 1, "develop": [1, 5], "dganno": [1, 2, 3, 4], "dict": [1, 6], "dictionari": 1, "differ": [0, 1, 2, 3, 5, 6], "dim": 1, "dimens": 1, "dimension": 1, "dinucleotid": 0, "diploid": 1, "direct": 1, "directli": [0, 1, 2, 3, 6], "directori": [1, 3, 4, 6], "disabl": 1, "disambigu": [1, 5], "discov": 3, "disjoint": 1, "disk": 1, "displai": 1, "distanc": 1, "distinct": 1, "distinguish": [0, 1], "distribut": [1, 3], "diverg": 1, "divid": 1, "dna": [0, 1, 2, 5, 6], "dna0": [1, 6], "dna1": [1, 6], "dnstream": 1, "dnstream_of": [1, 6], "do": [0, 1, 6], "doc": 1, "document": [3, 5], "doe": [1, 6], "doesn": 1, "domin": 1, "don": 6, "done": 0, "download": [1, 2, 3, 6], "downsampl": 1, "downstream": [1, 6], "dp": 1, "dst": 1, "dtype": [1, 6], "due": 1, "dup": 1, "duplic": 1, "dure": 1, "dynam": 1, "e": [0, 1, 3, 4, 6], "each": [1, 6], "earlier": 6, "easi": [5, 6], "easili": [0, 1, 2, 5], "ebi": 1, "echo": 4, "edit": 3, "edu": 4, "effect": [0, 6], "eg": 1, "either": [1, 3, 6], "element": [0, 1, 6], "empti": [1, 6], "en": [1, 6], "enabl": [1, 6], "encod": 1, "encount": 1, "end": [0, 1, 6], "end3": [1, 6], "end5": [1, 6], "endpoint": 6, "enough": 3, "ense00000846737": 1, "ensembl": [1, 4, 5, 6], "ensg00000001626": 6, "ensg00000115275": 1, "ensg00000223972": 6, "ensp00000233616": 1, "enst00000003084": 6, "enst00000233616": 1, "enst00000426809": 6, "enst00000432964": 6, "enst00000454343": 6, "enst00000456328": 6, "enst00000466430": 6, "enst00000466557": 6, "enst00000599771": 6, "enst00000601486": 6, "ensur": [1, 3], "entir": [0, 1, 6], "entrez": 1, "entri": [1, 6], "entry_point": 6, "env": [3, 6], "environ": [3, 4, 6], "envvar": 3, "equal": [0, 1, 6], "equival": [0, 1, 6], "error": 1, "especi": 1, "est": 1, "estim": 1, "eta": 1, "etc": [1, 2, 6], "etyp": 1, "even": [0, 1], "ever": 1, "everi": [1, 6], "evid": 1, "exact": 1, "exactli": [1, 6], "examin": 0, "exampl": [0, 1, 2, 3, 5, 6], "except": [1, 6], "excerpt": 3, "exclud": [1, 6], "exclus": [1, 6], "exist": [1, 2], "exon": [0, 2, 6], "exonic_interv": 1, "exont": 6, "expand": [0, 1, 6], "expect": [0, 1, 6], "explain": [0, 5, 6], "explan": [1, 6], "explicitli": 1, "export": 6, "expos": 1, "express": 1, "extend": [0, 1], "extens": 3, "extent": 1, "extra": [0, 1], "extract": [0, 1, 5], "extract_featur": 6, "f": [1, 3], "f16": 1, "f2": 1, "f3": 1, "f32": 1, "f4": 1, "f5": 1, "f6": 1, "f8": 1, "fa": 1, "face": 1, "facilit": [1, 2], "fail": 1, "failur": 3, "fall": [0, 1, 6], "fals": [1, 6], "far": 1, "fast": [1, 2, 3, 5, 6], "fasta": 4, "faster": 1, "fastest": 6, "fatotwobit": 4, "featur": [0, 2, 3], "feedstock": 3, "fetch": 1, "few": 6, "field": [1, 6], "file": [1, 4, 5], "fileformat": 6, "filenam": [1, 6], "filepath": [1, 6], "files": 1, "fill": [1, 6], "filter": [1, 5, 6], "final": [1, 3, 6], "find": [1, 3, 5], "find_3p_align": [1, 6], "find_3p_within": [1, 6], "find_5p_align": [1, 6], "find_5p_within": [1, 6], "find_exact": [1, 6], "find_motif": [1, 6], "find_overlap": [1, 6], "find_overlapping_match": 6, "find_overlapping_motif": [1, 6], "find_within": [1, 6], "find_x": 1, "fine": 1, "finish": 1, "first": [0, 1, 6], "first_by_nam": 1, "five": [0, 6], "fix": [0, 1], "fixedstep": 1, "flag": [1, 3], "flank": 6, "flat": 2, "float": [1, 6], "float16": 1, "float32": [1, 6], "flush": 1, "fmt": 3, "fmt_id": [1, 6], "follow": [0, 1, 3, 4, 6], "foo": 1, "forc": 1, "forg": [3, 6], "forget": 3, "format": [0, 1, 3, 4, 5, 6], "format_id": 1, "forward": [1, 6], "found": [1, 2, 6], "four": [1, 6], "fourth": 6, "frame": 1, "free": 1, "freeli": 2, "frequenc": 6, "freshli": 1, "from": [0, 1, 2, 3, 4, 5], "from_dna0": 1, "from_dna0_coord": [1, 6], "from_dna1": 1, "from_jralign": 1, "from_rna1": 1, "from_str": [1, 6], "from_vcf": [1, 6], "full": 1, "function": [0, 1, 6], "further": 1, "furthermor": 0, "futur": 5, "g": [0, 1, 4, 6], "gacagaacca": 6, "gap": [0, 1], "gc": 6, "gccagaggac": 6, "gcloud": 6, "gencod": [1, 2, 3, 4, 5, 6], "gene": 6, "gene_typ": 1, "geneid": 1, "gener": [0, 1, 4], "genom": [0, 2, 3, 5], "genome37": 1, "genome_kit": [0, 3, 4, 6], "genomeannot": [4, 6], "genomekit": [0, 1, 2, 4, 6], "genomekit_data_dir": [4, 6], "genomekit_dev": 3, "genomekit_gcs_billing_project": 6, "genomekit_gcs_bucket": 6, "genomekit_quiet": 6, "genomekit_trac": 6, "genometrack": 6, "genometrackbuild": 6, "genotyp": [1, 5, 6], "get": [0, 1, 3, 4, 5, 6], "get_fil": [1, 6], "gff3": [1, 2, 3, 4], "ggg": 0, "git": [3, 4, 6], "github": [1, 3, 4, 6], "give": [1, 5], "given": [0, 1, 6], "gk": [2, 4, 6], "gk_check": 3, "gk_data": [5, 6], "gk_debugbreak": 3, "gmod": 1, "go": 0, "goal": 5, "goldenpath": 4, "googl": 6, "google_cloud_project": 6, "grch37": [5, 6], "grch38": 5, "grcm38": 5, "grcm39": 5, "greater": [0, 1], "group": 6, "grow": 0, "gt": [1, 6], "gt_heterozygous_phased_0_1": 1, "gt_heterozygous_phased_1_0": 1, "gt_heterozygous_unphas": 1, "gt_homozygous_alt": 1, "gt_homozygous_ref": 1, "gt_unknown": 1, "gtex": 1, "gtrack": [1, 6], "gtrack_vers": 1, "gttt": 0, "gxx_linux": 3, "gz": [1, 3, 6], "h": 1, "ha": [1, 3, 6], "hacki": 1, "had": [0, 1], "half": 1, "handl": [0, 1, 6], "happen": 0, "has_coding_seq": 6, "has_good_level": 6, "hash": 4, "have": [0, 1, 3, 6], "headach": 1, "header": 1, "help": [3, 4], "helper": 1, "her": 1, "here": [0, 1, 6], "hg19": [0, 1, 4, 5, 6], "hg38": [1, 2, 4, 5, 6], "hgdownload": 4, "hi": 1, "hierarchi": 1, "higher": 1, "highli": 1, "hint": 1, "histogram": 1, "histor": 1, "hold": 1, "home": 3, "how": [0, 1, 3, 4, 5, 6], "howev": [0, 1, 6], "ht": 1, "html": [1, 4, 6], "http": [1, 3, 4, 6], "httperror": 1, "i": [0, 1, 2, 3, 6], "i8": 1, "id": [1, 6], "idea": 6, "ideal": 0, "ident": [1, 6], "identifi": 1, "ieee": 1, "ignor": [1, 6], "immedi": [1, 6], "implement": [1, 2, 6], "impli": [0, 1, 6], "import": [0, 1, 3, 4, 6], "importantli": [0, 6], "in_clion": 3, "includ": [0, 1, 3, 5, 6], "include_dupl": 1, "include_vari": 1, "inclus": [1, 6], "incompat": 1, "incorpor": 1, "incorrectli": 1, "increas": 1, "inde": 1, "indel": [0, 1, 6], "independ": [0, 1], "index": [1, 2, 5, 6], "index_of": [1, 6], "index_s": 1, "indexerror": 1, "indic": [0, 1, 6], "individu": 6, "infer": 1, "infil": 1, "info": [1, 6], "info_id": [1, 6], "inform": 1, "inherit": 1, "initi": 1, "input": [0, 1], "insert": [1, 5, 6], "insid": [0, 1], "instal": [1, 3, 5], "instanc": [1, 6], "instanti": 1, "instead": [0, 1], "instruct": [3, 4], "int": 1, "int16": 1, "int32": [1, 6], "int8": [1, 6], "integ": [0, 1, 6], "integr": 3, "intend": [1, 6], "intent": 1, "interest": [0, 6], "intermedi": 1, "intern": [2, 5, 6], "interpret": [1, 6], "interrupt": 1, "intersect": 1, "interv": [2, 5], "interval1": 1, "interval2": 1, "interval_a": 1, "interval_b": 1, "interval_c": 1, "intr": 1, "intron": 6, "intronic_interv": 1, "inv": 1, "invalid": 1, "involv": 1, "io": [1, 6], "ioerror": 1, "irrespect": 6, "is_positive_strand": 1, "isoform": 1, "iter": 1, "its": [0, 1, 2, 6], "itself": 1, "jetbrain": 5, "jralign": 1, "jralign_vers": 1, "junc": 1, "junction": [3, 5], "junction_index": 1, "just": [3, 6], "keep": [0, 1], "kei": 1, "kept": [0, 1], "keyerror": 1, "kick": 3, "kind": 1, "know": [1, 6], "kwarg": 1, "lack": 1, "lambda": 1, "larg": 1, "last": [0, 1], "latest": 6, "lazili": 2, "lead": [3, 6], "learn": [0, 5, 6], "least": [1, 6], "left": [0, 1], "len": [0, 1, 6], "length": [0, 1, 5], "let": [0, 5, 6], "level": [1, 6], "li": 1, "librari": [1, 2, 5], "library_format": 1, "library_typ": 1, "lift": 6, "lightweight": 5, "like": [0, 1, 5, 6], "limit": 1, "line": [0, 1, 6], "linear": 1, "link": [1, 3], "linspac": 1, "linux": [3, 6], "list": [1, 2, 5, 6], "list_available_genom": [1, 5], "liter": 1, "live": 6, "ll": [0, 1, 3], "load": 1, "local": [1, 2, 3, 6], "locat": [1, 3, 6], "log": 5, "login": 6, "long": 1, "longer": 1, "look": 0, "lookup": 1, "loop": 1, "lower": 1, "m": [1, 3], "m0": 1, "m1": 3, "m39": [1, 5], "mac": 3, "macaca_fascicularis_5": 5, "macfas5": [1, 5], "macfas_5": [1, 5], "machin": [0, 5], "made": [1, 5, 6], "mai": [1, 3, 6], "main": 3, "maintain": 2, "make": [1, 5, 6], "malform": 1, "manag": [1, 6], "mani": [0, 3, 6], "manner": 6, "manual": 0, "map": [1, 6], "mark": 1, "mask": [1, 6], "master": 1, "match": [1, 2, 6], "match_posit": [1, 6], "mateid": 1, "matric": 1, "matrix": 1, "max": 1, "max_int": 0, "maximum": 1, "mean": [1, 6], "meaning": 1, "mechan": [0, 6], "meet": 6, "member": 1, "memori": 1, "menu": 3, "mere": 1, "merg": [1, 3], "messag": 1, "metadata": [1, 2, 4, 6], "method": [0, 1, 3, 6], "middl": [0, 1], "midpoint": 1, "might": 3, "min": 1, "min_delta": 1, "min_overhang": 1, "min_read": 1, "min_run": 1, "mind": 0, "minimum": 1, "mismatch": 1, "miss": 1, "mix": 1, "mixtur": 6, "mm10": [1, 5], "mm39": [1, 5], "mnt": 1, "mock": 1, "mock_result": 1, "mock_unreach": 1, "mode": [3, 6], "model": 1, "modif": 1, "modifi": [0, 1], "modul": [1, 3, 5], "mog": 1, "more": [0, 1, 5, 6], "most": [0, 1, 6], "motif": [0, 1, 5], "mrna": 1, "mt": [0, 5], "multi": 1, "multidimension": 1, "multipl": 1, "must": [1, 3, 4], "mutant": [0, 6], "my_anno": 4, "my_assembli": 4, "my_fil": 1, "mydatamanag": 6, "mydna": 1, "myfil": 1, "mypackag": 1, "n": [1, 3, 4], "name": [1, 2, 4, 6], "namespac": 6, "nan": 1, "nativ": 3, "na\u00efv": 0, "ncbi": [1, 4, 5, 6], "ncbi37": 1, "ncbi_refseq": 1, "ndarrai": 1, "nearest": 6, "necessari": [0, 1], "need": [0, 1, 3, 4, 6], "neg": [1, 6], "neg_strand_fil": 1, "never": 1, "new": [1, 3, 5], "newer": [1, 3], "next": [1, 6], "next_cd": 1, "next_exon": [1, 6], "next_intron": 1, "nm_000344": 1, "nm_001297715": 1, "nm_006302": 1, "nm_022874": 1, "nnnnnnngggggggg": 1, "non": [1, 6], "non_cod": 1, "none": [1, 6], "norm": 1, "normal": [1, 3], "not_first_exon": 6, "note": [1, 6], "noth": 1, "notic": 6, "notifi": 1, "notion": 0, "novo": 0, "now": [0, 1, 3, 4, 6], "np": [1, 6], "np_006293": 1, "nuclear": 6, "nucleotid": [1, 6], "num_bit": 1, "num_count": 1, "num_read": 1, "num_sampl": 1, "num_vari": 1, "number": [1, 6], "numer": 1, "numpi": [1, 6], "nxm": 1, "o": [4, 6], "object": [0, 1, 2, 5, 6], "observ": 1, "obtain": [4, 6], "occupi": 1, "occur": [0, 1], "occurr": 1, "odd": 1, "off": [0, 3, 6], "offset": [0, 1], "often": [0, 1], "ok": 1, "omit": 1, "onc": [1, 2, 3], "one": [0, 1, 6], "ones": [1, 5], "onli": [0, 1, 2, 3, 6], "open": [1, 3, 6], "open_clinvar": 1, "oper": [2, 6], "opposit": [0, 1, 6], "optim": [1, 2], "option": [1, 3, 4, 6], "order": [1, 2, 6], "org": [1, 6], "organ": 5, "origin": 1, "osx": 3, "other": [0, 1, 2, 3, 5, 6], "otherwis": 1, "our": [0, 1], "out": [0, 1, 2, 3], "outfil": 1, "output": [1, 6], "outsid": [1, 5], "over": [1, 2, 6], "overestim": 1, "overhang": 1, "overhang_error": 1, "overlap": [0, 1, 6], "overrid": 1, "own": [1, 2, 6], "p12": [1, 2, 5], "p13": [1, 5, 6], "p14": [1, 5], "p6": [1, 5], "packag": [3, 6], "pad": 6, "page": 5, "pai": 6, "pair": 1, "parallel": 3, "param": 1, "paramet": [1, 6], "parent": 1, "parenthes": 0, "pars": 1, "part": [1, 6], "partial": 1, "particular": [0, 1], "pass": [0, 1, 6], "pass_filt": 1, "patch": 5, "path": 1, "pdf": 1, "per": [1, 6], "percentag": 1, "perform": 1, "phase": 1, "phastcon": 1, "phastcons_mamm": 1, "pickl": [1, 2], "pileup": 1, "pileup_a": 1, "pileup_c": 1, "pileup_del": 1, "pileup_g": 1, "pileup_t": 1, "pip": 3, "pipelin": [1, 3], "pkl": 2, "plan": 6, "platform": 6, "pleas": [1, 3], "plu": 6, "plugin": 6, "plusmt": [1, 5, 6], "po": 6, "point": [1, 6], "polici": 1, "pool": 1, "portion": 1, "pos_strand_fil": 1, "posit": [0, 1, 6], "possibl": [0, 1], "possibli": 1, "potenti": 6, "power": [1, 5], "pr": 3, "practic": 0, "pre": [3, 6], "preced": [1, 6], "precis": 1, "preconfigur": 3, "predict": 0, "prefer": 1, "prerequisit": 3, "preserv": 1, "prev_cd": 1, "prev_exon": 1, "prev_intron": 1, "previou": [0, 1, 6], "previous": 1, "primarili": 0, "princip": 1, "principl": [0, 1], "prinicp": 1, "print": [1, 4, 6], "prioriti": 1, "privat": 1, "problem": 0, "process": [1, 3], "product": 1, "progress": 1, "progress_msg": 1, "progresspercentag": 1, "project": [3, 6], "prompt": 3, "propag": 1, "properli": 1, "properti": 1, "protein": 1, "protein_cod": 1, "protein_id": 1, "provid": [1, 2, 3, 6], "public": 6, "publish": 3, "pull": [0, 3], "purpos": 1, "push": 0, "pushd": [4, 6], "put": 1, "pwm": 6, "py": [1, 3, 6], "pycharm": 1, "pydeletemockattr": 1, "pylint": 1, "pypa": 6, "python": [1, 2, 3, 4, 5, 6], "qual": 6, "qualiti": 1, "quantiti": 1, "queri": [1, 2, 5, 6], "query_vcf": [1, 6], "quick": 6, "quickstart": [1, 5], "r": [1, 6], "rais": [1, 6], "ralign": 1, "ralign_vers": 1, "random": 1, "rang": [1, 6], "rare": 0, "rather": [0, 6], "raw": 2, "rcr": 5, "rdist": 1, "rdist_vers": 1, "re": [1, 3], "reach": 1, "read": [3, 5, 6], "read_index": 1, "readi": 6, "readthedoc": 1, "reason": [1, 3, 6], "recogn": 1, "recommend": 1, "reduc": 1, "ref": [1, 6], "refer": [0, 1, 5, 6], "reference_align": 1, "reference_genom": [1, 6], "refg": [1, 4], "refgen": 1, "reflink": 1, "refseq": [1, 5, 6], "regener": 1, "region": 1, "regist": [1, 3, 6], "regro": 3, "rel": [1, 6], "relat": [2, 6], "releas": 5, "relev": 1, "reli": 3, "remain": 6, "remot": [1, 3, 5, 6], "remote_fil": 1, "remov": [0, 1], "repeat": [1, 6], "repeatedli": 3, "replac": 4, "repo": [3, 4, 6], "report": 1, "repr": 1, "repres": [1, 6], "represent": 1, "request": [1, 6], "requir": [1, 2, 3, 4, 6], "research": 5, "reserv": [1, 3], "resid": 3, "resolut": 1, "resourc": [1, 6], "respect": [1, 6], "respons": 1, "rest": 1, "restrict": 1, "result": [0, 1, 4, 6], "result_typ": 1, "retain": 1, "retriev": [1, 2], "return": [0, 1, 6], "revers": [1, 6], "rgsc": 5, "ride": 0, "right": [0, 1], "rm": 6, "rn6": [1, 5], "rna1": 1, "rnor_6": [1, 5], "rough": 1, "round": 1, "row": [1, 6], "rs397508328": 6, "rs397508657": 6, "rs397508673": 6, "rs397508710": 6, "rtype": 1, "rule": 1, "run": [1, 3, 6], "runtim": 1, "sake": 1, "salmon": 1, "sam": [1, 2, 5], "same": [0, 1, 6], "sampl": [1, 6], "sample000132": 1, "sample1": [1, 6], "sample2": [1, 6], "sample3": [1, 6], "sample_nam": 1, "samtool": 1, "samv1": 1, "save": [1, 6], "scalar": 1, "scan": 5, "scenario": 1, "schema": 4, "scheme": 1, "scope": 1, "score": 1, "script": [4, 6], "sdk": 3, "search": [1, 5, 6], "second": [1, 6], "secondari": 5, "section": [1, 6], "see": [0, 1, 3, 4, 6], "seen": 0, "segment": 1, "select": [1, 3, 6], "self": [1, 6], "sens": [1, 6], "sensit": 1, "separ": [1, 6], "separate_dim": 1, "seq": 1, "sequenc": [0, 1, 5, 6], "sequence_vari": 1, "serv": 1, "server": 1, "set": [0, 1, 5, 6], "set_clamp": 1, "set_data": [1, 6], "set_data_from_b": 1, "set_data_from_bedgraph": 1, "set_data_from_wig": 1, "set_default_valu": 1, "set_dict": 1, "set_restrict": 1, "set_spars": 1, "set_transform": 1, "setup": 5, "setuptool": 6, "sever": 6, "sf": 1, "sgsc": 5, "shape": [1, 6], "share": [1, 5], "shell": 3, "shift": 1, "short": [0, 5, 6], "shortcut": 1, "shorter": 0, "shorthand": 1, "should": [0, 1, 6], "show": [0, 1, 6], "shrink": [0, 6], "side": [0, 1], "sign": 1, "signific": 1, "similarli": 1, "simpli": 1, "sinc": [0, 1, 6], "singl": [0, 1, 6], "single_strand": [1, 6], "site": [0, 1, 3, 6], "size": [0, 1, 2, 3, 4], "slice": 6, "slightli": 0, "sln": 3, "small": [1, 3], "smallest": 1, "smn1": 1, "smn2": 1, "snapshot": 5, "snv": 6, "so": [0, 1, 2, 3, 6], "socket": 1, "soe": 4, "sole": 1, "solv": 0, "some": [0, 1, 5, 6], "sort": 1, "sourc": [1, 2, 3, 4, 6], "space": [1, 6], "span": [1, 6], "spars": 1, "sparsif": 1, "sparsiti": 1, "spec": 1, "special": [0, 1, 6], "specif": [0, 1, 2, 6], "specifi": [1, 2, 4, 5, 6], "speed": [1, 3], "sphinx": 1, "splice": [0, 1, 6], "split": [0, 1], "spuriou": 1, "sr": 1, "src": [3, 4, 6], "sscrofa11": [1, 5], "stabl": 1, "stai": [0, 6], "standard": [1, 5, 6], "start": [0, 1, 3, 5, 6], "statement": 1, "static": 1, "stdin": 1, "step": [1, 3], "stick": 0, "still": [0, 1], "stop": 1, "storag": [1, 2, 6], "store": [1, 3, 6], "str": [1, 6], "straight": 6, "strand": [1, 6], "strand_awar": [1, 6], "strand_unawar": [1, 6], "stranded": [1, 6], "stream": 1, "strictli": 1, "string": [0, 1, 6], "strip": 1, "strip_mock_bas": 1, "structur": [1, 2, 5, 6], "studi": [0, 5], "style": 6, "subclass": [1, 6], "subcommand": 3, "subsect": 1, "subset": 1, "substitut": 1, "subtract": 1, "subtyp": 1, "succeed": 1, "suffix": 1, "suit": 6, "sum": [1, 6], "suppli": [1, 6], "support": [1, 2, 3, 4, 6], "suppos": 6, "suppress": 6, "sure": [1, 6], "surround": 6, "suscept": 1, "suspect": 1, "susscr11": [1, 5], "svtype": 1, "svtype_bnd": 1, "svtype_cnv": 1, "svtype_del": 1, "svtype_dup": 1, "svtype_in": 1, "svtype_inv": 1, "svtype_na": 1, "switch": 3, "sy": 1, "symbol": 1, "symmetr": 1, "synonym": 1, "system": [1, 3, 6], "t": [0, 1, 6], "taaaccacat": 1, "tabl": [3, 6], "tag": 1, "take": [1, 6], "target": [1, 3], "tatcttaaagctgtgtgt": 0, "tatcttaaagctgtgtgtaa": 0, "tatcttaaagtttctgtgtc": 0, "tc": 6, "tccaact": 6, "tccaggctat": 6, "tct": [0, 6], "tcttaaagctgtgtctgtaa": 0, "technic": 6, "temp": 1, "ten": 0, "tend": 1, "termin": 1, "test": [1, 5, 6], "test_interv": 3, "test_seri": 3, "testinterv": 3, "text": [1, 2], "tgcagggaac": 6, "tggg": 0, "tgt": 0, "tgtaggaaac": 6, "than": [0, 1, 6], "thei": [1, 2, 3, 4, 6], "them": [1, 3], "therefor": [0, 1, 6], "thi": [0, 1, 2, 3, 6], "think": 1, "third": 6, "those": [1, 5, 6], "though": 1, "thought": 6, "three": 0, "through": [0, 1], "throughput": 1, "time": 1, "timeout": 1, "timestamp": 1, "tini": 3, "todo": 1, "togeth": 6, "toggl": 1, "toi": 6, "tool": [3, 4], "top": 6, "toward": 1, "trace": 6, "track": [2, 3, 5], "trade": 1, "tran": [1, 6], "transcript": 6, "transcript_typ": 1, "transfer": 1, "transform": 1, "transpar": [2, 6], "treat": 1, "tree": 3, "trick": 1, "trigger": 1, "trim": [0, 1, 6], "trivial": 1, "troubleshoot": 5, "true": [1, 6], "truncat": 1, "try": 1, "tsl": 1, "tsl1": 1, "tsl2": 1, "tsl3": 1, "tsl4": 1, "tsl5": 1, "tslna": 1, "tt": 6, "ttagtt": 6, "ttatatcttagtgtctgtaa": 0, "ttcagctgct": 6, "ttgatatttatatgttttta": 6, "ttgatatttatatgtttttatatcttaaagctgtgtctgtaaactgatggctaacaaaactaggattttggtc": 6, "ttt": 0, "tttgcttccaatt": 6, "tupl": 1, "tutori": [5, 6], "two": [0, 1, 3, 6], "twobit": 4, "twobitinfo": 4, "txt": [1, 2, 3, 4], "type": [0, 1, 3, 6], "typeerror": 1, "typic": 0, "u": 1, "u1": 1, "u2": 1, "u3": [1, 6], "u4": 1, "u5": 1, "u6": 1, "u8": 1, "ubuntu": 3, "ucsc": [1, 4, 5, 6], "ucsc_db_dir": 1, "ucsc_refseq": 1, "uint8": [1, 6], "ultim": 6, "unambigu": [0, 6], "under": [1, 3, 4, 6], "undergo": 1, "understand": [0, 1], "understood": 0, "unintend": 0, "uniqu": [1, 6], "unit": 1, "unittest": [0, 3], "unittestbuild": 3, "univers": 6, "unless": [1, 6], "unlik": 1, "unpatch": 5, "unpredict": 3, "unreach": 1, "unrecogn": 1, "unspecifi": 1, "unstrand": 1, "unsupport": 1, "until": 1, "up": [0, 1, 5, 6], "updat": 3, "upload": [1, 2, 6], "upload_fil": [1, 6], "upon": [1, 3], "upper": 1, "upstream": [1, 6], "upstream_of": [1, 6], "url": 1, "us": [0, 1, 2, 3, 4], "usag": 1, "user": [1, 2, 3, 5, 6], "user_data_dir": [4, 6], "userguid": 6, "usual": [0, 6], "utr": 1, "utr3": 1, "utr3tabl": 1, "utr5": 1, "utr5tabl": 1, "utrtabl": 1, "v": [1, 3, 6], "v101": [1, 5], "v105": [1, 5, 6], "v109": [1, 5], "v110": [1, 5], "v19": [1, 3, 5, 6], "v22": [1, 2], "v26lift37": [4, 6], "v29": [1, 2, 5], "v29lift37": 1, "v41": [1, 5], "v41lift37": 1, "valid": [1, 6], "valu": [0, 1, 6], "valueerror": 1, "var": 6, "var1": 6, "var2": 6, "var3": 6, "var_seq": 0, "variabl": [3, 6], "variablestep": 1, "variant": [0, 5], "variant_dna": 1, "variant_genom": [0, 1, 6], "variant_str": 1, "variantgenom": [0, 5, 6], "variou": 2, "vcf": [2, 3, 5], "vcfbin": 1, "vcfbin_vers": 1, "vcfbinpath": 1, "vcfpath": 1, "vcftabl": 6, "vcfv4": 6, "vcfvariant": 6, "vcproj": 3, "vcvars64": 3, "ve": 6, "verbos": 3, "veri": 1, "verifi": 6, "version": [1, 2, 3, 6], "vertic": 0, "via": [1, 2, 3], "view": 1, "visibl": 1, "visual": 6, "vm19": [1, 5], "vm30": [1, 5], "vm31": [1, 5], "w": 6, "wa": [0, 1], "wai": [1, 5, 6], "walk": 6, "walk_annot": 6, "want": [0, 1, 3, 6], "warn": 1, "we": [0, 1, 6], "weak": 1, "well": [0, 6], "were": [0, 1, 6], "wget": 1, "what": [0, 1, 6], "when": [0, 1, 2, 3, 6], "where": [0, 1, 6], "wherea": [0, 1, 6], "whether": [0, 1], "which": [0, 1, 3, 6], "while": [1, 3, 6], "wide": 1, "wig": [1, 2], "wiki": 1, "wild": 0, "window": [0, 1, 3, 6], "wish": [1, 6], "with_anchor": [1, 6], "within": [0, 1, 6], "without": [1, 3, 5, 6], "word": 1, "work": [0, 3, 5], "would": [0, 1, 6], "wrapper": 1, "write": 1, "write2bit": 1, "written": 1, "wt": 0, "x": [1, 6], "x64": 3, "xcode": 3, "xm": 1, "xxxxxyyyyi": 6, "y": [1, 6], "yet": 1, "yml": 3, "you": [0, 1, 3, 4, 5, 6], "your": [1, 3, 4, 6], "yourself": 3, "z": [1, 6], "zero": [0, 1, 6], "zlib": 3, "zzzz": 6}, "titles": ["Anchors", "API Documentation", "Remote Data Organization", "Developers", "Genomes", "GenomeKit", "Quickstart"], "titleterms": {"_apply_vari": 1, "_cxx_util": 1, "_twobitutil": 1, "access": [1, 2, 6], "ad": 3, "align": 1, "alignmentmatch": 1, "alignmentmatcht": 1, "alignmentt": 1, "an": 0, "anchor": 0, "annot": [1, 4, 6], "api": 1, "assembli": 4, "avail": [2, 5], "basic": 6, "build": 3, "c": 3, "cd": 1, "cdstabl": 1, "chang": [3, 6], "clion": 3, "conda": 6, "data": [1, 2, 3, 6], "data_manag": 1, "datamanag": 1, "debug": 3, "delet": 0, "develop": 3, "docker": 6, "document": 1, "exampl": 4, "exon": 1, "exont": 1, "extract": 6, "featur": [5, 6], "file": [2, 3, 6], "find": 6, "format": 2, "from": 6, "full": 3, "gene": 1, "gener": 6, "genet": 1, "genom": [1, 4, 6], "genome_kit": 1, "genomeannot": 1, "genomedna": 1, "genomekit": [3, 5], "genometrack": 1, "genometrackbuild": 1, "gk_data": 1, "i": 5, "indic": 5, "initi": 6, "insert": 0, "instal": 6, "intern": 1, "interv": [0, 1, 6], "intron": 1, "intront": 1, "jetbrain": 3, "jreadalign": 1, "junction": 1, "junctionreadalign": 1, "junctionreadalignmentst": 1, "junctionreadcount": 1, "junctionreaddistribut": 1, "junctionreaddistributiont": 1, "junctiont": 1, "length": 6, "log": 6, "make": 3, "motif": 6, "new": 6, "next": 5, "organ": 2, "outsid": 0, "quickstart": 6, "read": 1, "readalign": 1, "readdistribut": 1, "releas": 3, "remot": 2, "resourc": 5, "set": 3, "setup": 3, "share": 6, "specifi": 0, "step": 5, "tabl": [1, 5], "test": 3, "todo": 0, "track": [1, 6], "transcript": 1, "transcriptt": 1, "troubleshoot": 6, "up": 3, "us": [5, 6], "usag": 6, "variant": [1, 6], "variantgenom": 1, "variantt": 1, "vcf": [1, 6], "vcftabl": 1, "vcfvariant": 1, "via": 6, "what": 5, "without": 0, "work": 6}})