

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Developers &mdash; GenomeKit 6.0.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=256d0b43"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Genomes" href="genomes.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            GenomeKit
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="anchors.html">Anchors</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="genomes.html">Genomes</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Developers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#setting-up">Setting up</a></li>
<li class="toctree-l2"><a class="reference internal" href="#jetbrains-clion-setup">Jetbrains CLion setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="#making-changes">Making changes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#adding-tests">Adding tests</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#c-tests">C++ tests</a></li>
<li class="toctree-l3"><a class="reference internal" href="#debugging-tests">Debugging tests</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#building-data-files">Building data files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#building-full-data-files">Building full data files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#building-test-data-files">Building test data files</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#releasing-genomekit">Releasing GenomeKit</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GenomeKit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Developers</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/develop.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="developers">
<span id="develop"></span><h1>Developers<a class="headerlink" href="#developers" title="Link to this heading"></a></h1>
<p>Developing for GenomeKit currently requires a C++ compiler that support C++20 features. Both can be installed via conda (<code class="docutils literal notranslate"><span class="pre">gxx_linux-64</span></code> and <code class="docutils literal notranslate"><span class="pre">`clangxx_osx-64</span></code>). For OSX, you will also need the SDK via <code class="docutils literal notranslate"><span class="pre">xcode-select</span> <span class="pre">--install</span></code> or from <a class="reference external" href="https://developer.apple.com/download/all/">https://developer.apple.com/download/all/</a>.</p>
<section id="setting-up">
<h2>Setting up<a class="headerlink" href="#setting-up" title="Link to this heading"></a></h2>
<p>Clone the source tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">git</span><span class="nd">@github</span><span class="o">.</span><span class="n">com</span><span class="p">:</span><span class="n">deepgenomics</span><span class="o">/</span><span class="n">GenomeKit</span><span class="o">.</span><span class="n">git</span>
</pre></div>
</div>
<p>From the <code class="docutils literal notranslate"><span class="pre">GenomeKit</span></code> directory, install the provided conda environment which
contains all dependencies:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">env</span> <span class="n">create</span> <span class="o">-</span><span class="n">f</span> <span class="n">genomekit_dev</span><span class="o">.</span><span class="n">yml</span>
<span class="n">conda</span> <span class="n">activate</span> <span class="n">genomekit_dev</span>
</pre></div>
</div>
<p>On Windows, you’ll need to comment out the mac/linux only test dependencies from genomekit_dev.yml.</p>
<p>On M1 macs, you might need to set up the environment differently:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">create</span> <span class="o">-</span><span class="n">n</span> <span class="n">cxx</span> <span class="n">cxx</span><span class="o">-</span><span class="n">compiler</span> <span class="n">zlib</span> <span class="n">fmt</span>
<span class="n">conda</span> <span class="n">activate</span> <span class="n">cxx</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="o">--</span><span class="n">file</span> <span class="n">a</span><span class="o">-</span><span class="n">file</span><span class="o">-</span><span class="k">with</span><span class="o">-</span><span class="n">the</span><span class="o">-</span><span class="n">deps</span><span class="o">-</span><span class="n">from</span><span class="o">-</span><span class="n">genomekit_dev</span><span class="o">-</span><span class="n">yml</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Build the package in development mode:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="o">-</span><span class="n">e</span> <span class="o">.</span>
</pre></div>
</div>
<p>This builds the C++ extension and copies it into
your source tree (<code class="docutils literal notranslate"><span class="pre">genome_kit/_cxx.so</span></code>).
It also ensures that <code class="docutils literal notranslate"><span class="pre">import</span> <span class="pre">genome_kit</span></code> works from any directory
by linking your source tree from python’s <code class="docutils literal notranslate"><span class="pre">site-packages</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Windows Prerequisites</p>
<p>You will need VS 2019 or newer installed. To get a compatible shell, either locate
and run <code class="docutils literal notranslate"><span class="pre">vcvars64.bat</span></code>, or start the x64 Native Tools Command Prompt from the
Start menu.
To open VS with a preconfigured project, directly run in that command prompt:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">.</span><span class="n">vcproj</span>\<span class="n">genome_kit</span><span class="o">.</span><span class="n">sln</span>
</pre></div>
</div>
</div>
<p>Finally, run the all the tests:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">unittest</span> <span class="n">discover</span>
</pre></div>
</div>
<p>You can also run examples from the <code class="docutils literal notranslate"><span class="pre">demos</span></code> directory.</p>
</section>
<section id="jetbrains-clion-setup">
<h2>Jetbrains CLion setup<a class="headerlink" href="#jetbrains-clion-setup" title="Link to this heading"></a></h2>
<p>In the CMake settings, set the following environment variables:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>IN_CLION=1;CONDA_PREFIX=$HOME/conda/envs/genomekit_dev
</pre></div>
</div>
</section>
<section id="making-changes">
<h2>Making changes<a class="headerlink" href="#making-changes" title="Link to this heading"></a></h2>
<p>If the C/C++ code changed, you must re-run the <code class="docutils literal notranslate"><span class="pre">develop</span></code> command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="o">-</span><span class="n">e</span> <span class="o">.</span>
</pre></div>
</div>
<p>This includes switching branches, merging changes, or editing the C/C++ code
yourself. <em>Forgetting this step may lead to unpredictable behaviour.</em></p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>To speed up compilation on Ubuntu or Mac, install <code class="docutils literal notranslate"><span class="pre">ccache</span></code>.</p>
</div>
<p>Before checking in any changes, run all tests locally:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">unittest</span> <span class="n">discover</span>
</pre></div>
</div>
</section>
<section id="adding-tests">
<h2>Adding tests<a class="headerlink" href="#adding-tests" title="Link to this heading"></a></h2>
<p>Tests are located in the <code class="docutils literal notranslate"><span class="pre">tests</span></code> directory, and any data they need
is located in the <code class="docutils literal notranslate"><span class="pre">tests/data</span></code> directory.</p>
<p>While developing a test, you may want to run it repeatedly, without
all other tests.
For example, to run just the <code class="docutils literal notranslate"><span class="pre">TestInterval.test_serialize</span></code> method in
<code class="docutils literal notranslate"><span class="pre">tests/test_interval.py</span></code> use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">unittest</span> <span class="n">tests</span><span class="o">.</span><span class="n">test_interval</span><span class="o">.</span><span class="n">TestInterval</span><span class="o">.</span><span class="n">test_serialize</span>
</pre></div>
</div>
<section id="c-tests">
<h3>C++ tests<a class="headerlink" href="#c-tests" title="Link to this heading"></a></h3>
<p>To test C++ code directly, you can compile and run src/main.cpp:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cmake</span> <span class="o">-</span><span class="n">DCMAKE_BUILD_TYPE</span><span class="o">=</span><span class="n">Debug</span> <span class="o">-</span><span class="n">B</span> <span class="n">unittestbuild</span>
<span class="n">cmake</span> <span class="o">--</span><span class="n">build</span> <span class="n">unittestbuild</span> <span class="o">--</span><span class="n">parallel</span> <span class="o">--</span><span class="n">verbose</span> <span class="o">--</span><span class="n">target</span> <span class="n">main</span> <span class="n">test</span>
</pre></div>
</div>
</section>
<section id="debugging-tests">
<h3>Debugging tests<a class="headerlink" href="#debugging-tests" title="Link to this heading"></a></h3>
<p>Define envvar <cite>GK_DEBUGBREAK</cite> to break upon GK_CHECK failures when running
under a debugger.</p>
</section>
</section>
<section id="building-data-files">
<h2>Building data files<a class="headerlink" href="#building-data-files" title="Link to this heading"></a></h2>
<p>GenomeKit relies on many pre-built files.
For example, the binary annotation <code class="docutils literal notranslate"><span class="pre">gencode.v19.annotation.dganno</span></code>
is built from <code class="docutils literal notranslate"><span class="pre">gencode.v19.annotation.gff3.gz</span></code>.
Reasons to re-build these files include:</p>
<ul class="simple">
<li><p>Changes to the binary file format.</p></li>
<li><p>Updates to the source data.</p></li>
<li><p>Changes to the processing of source data.</p></li>
</ul>
<p>GenomeKit has two sets of data files:</p>
<ul class="simple">
<li><p><em>Full data files</em> are for normal use.
They are stored remotely in the GenomeKit store
and pulled to the user’s local file system on-demand.</p></li>
<li><p><em>Test data files</em> are for testing.
They are tiny excerpts of the full files, small enough
to check in to source control, fast enough to run in
continuous integration testing.
They are stored in the source tree under <code class="docutils literal notranslate"><span class="pre">tests/data</span></code>.</p></li>
</ul>
<p>The <code class="docutils literal notranslate"><span class="pre">genome_kit</span></code> module’s <code class="docutils literal notranslate"><span class="pre">build</span></code> command can be used to build full
Appris data files, and Appris/dganno/2bit test data files.</p>
<p>For a full set of options, run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">genome_kit</span> <span class="n">build</span> <span class="o">--</span><span class="n">help</span>
</pre></div>
</div>
<section id="building-full-data-files">
<h3>Building full data files<a class="headerlink" href="#building-full-data-files" title="Link to this heading"></a></h3>
<p>For instructions on how to build annotation (dganno) files and assembly
(2bit) files, see <a class="reference external" href="genomes.html">Genomes</a>.</p>
<p>Full-sized data files reside in a local user directory reserved
for GenomeKit, downloaded from the data store on-demand.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See the API Documentation for instructions on how to build
<a class="reference external" href="api.html#genometrackbuilder">data tracks</a>,
<a class="reference external" href="api.html#genome_kit.ReadAlignments.build_ralign">read alignments</a>,
<a class="reference external" href="api.html#genome_kit.ReadDistributions.build_rdist">read distributions</a>,
<a class="reference external" href="api.html#genome_kit.JReadAlignments.build_jralign">junction read alignments</a>,
and <a class="reference external" href="api.html#genome_kit.VCFTable.build_vcfbin">VCF tables</a>.</p>
</div>
</section>
<section id="building-test-data-files">
<h3>Building test data files<a class="headerlink" href="#building-test-data-files" title="Link to this heading"></a></h3>
<p>Test data files reside in the source tree under <code class="docutils literal notranslate"><span class="pre">tests/data</span></code>.
To build them, you must have registered your source tree in
develop mode:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="o">-</span><span class="n">e</span> <span class="o">.</span>
</pre></div>
</div>
<p>Now that your source tree is the default <cite>genome_kit</cite> import,
the <code class="docutils literal notranslate"><span class="pre">build</span></code> subcommand will be able to find
your test data directory.</p>
<p>To build test annotation, 2bit, and Appris files, use <cite>–test-&lt;type&gt;</cite>
flags on the <code class="docutils literal notranslate"><span class="pre">build</span></code> subcommand:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">genome_kit</span> <span class="n">build</span> <span class="o">--</span><span class="n">test</span><span class="o">-</span><span class="n">anno</span> <span class="o">--</span><span class="n">test</span><span class="o">-</span><span class="mi">2</span><span class="n">bit</span> <span class="o">--</span><span class="n">test</span><span class="o">-</span><span class="n">appris</span>
</pre></div>
</div>
</section>
</section>
<section id="releasing-genomekit">
<h2>Releasing GenomeKit<a class="headerlink" href="#releasing-genomekit" title="Link to this heading"></a></h2>
<p>The <a class="reference external" href="https://github.com/deepgenomics/GenomeKit">GenomeKit repo</a> uses
the <a class="reference external" href="https://github.com/googleapis/release-please">Release Please bot</a>
to create Github releases based on PRs. When the bot creates a PR, you can
merge it to create a release.</p>
<p>Once a Github release is created, a PR will automatically be created in
the <a class="reference external" href="https://github.com/conda-forge/genomekit-feedstock">GenomeKit conda-forge feedstock repo</a>
by regro-cf-autotick-bot. Once that PR is merged, conda-forge’s CI
pipeline is kicked off and the new version of GenomeKit is built and published
to conda-forge.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="genomes.html" class="btn btn-neutral float-left" title="Genomes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Deep Genomics.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>